// @spartacus/core/esm5/src/cart/store/reducers/multi-cart.reducer.js
import { CartActions } from '../actions/index';
export var activeCartInitialState = '';
export var wishListInitialState = '';
export function activeCartReducer(state, action) {
    if (state === void 0) { state = activeCartInitialState; }
    var _a, _b, _c;
    switch (action.type) {
        case CartActions.LOAD_CART_SUCCESS:
        case CartActions.CREATE_CART_SUCCESS:
        // point to `temp-${uuid}` cart when we are creating/merging cart
        case CartActions.CREATE_CART:
            if ((_b = (_a = action === null || action === void 0 ? void 0 : action.payload) === null || _a === void 0 ? void 0 : _a.extraData) === null || _b === void 0 ? void 0 : _b.active) {
                return action.meta.entityId;
            }
            else {
                return state;
            }
        case CartActions.SET_ACTIVE_CART_ID:
            return action.payload;
        case CartActions.REMOVE_CART:
        case CartActions.DELETE_CART_SUCCESS:
            if (((_c = action.payload) === null || _c === void 0 ? void 0 : _c.cartId) === state) {
                return activeCartInitialState;
            }
            return state;
        case CartActions.CLEAR_CART_STATE:
            return activeCartInitialState;
    }
    return state;
}
export var cartEntitiesInitialState = undefined;
export function cartEntitiesReducer(state, action) {
    if (state === void 0) { state = cartEntitiesInitialState; }
    switch (action.type) {
        case CartActions.LOAD_CART_SUCCESS:
        case CartActions.CREATE_CART_SUCCESS:
        case CartActions.CREATE_WISH_LIST_SUCCESS:
        case CartActions.LOAD_WISH_LIST_SUCCESS:
        case CartActions.SET_TEMP_CART:
            return action.payload.cart;
    }
    return state;
}
export function wishListReducer(state, action) {
    if (state === void 0) { state = wishListInitialState; }
    switch (action.type) {
        case CartActions.CREATE_WISH_LIST_SUCCESS:
        case CartActions.LOAD_WISH_LIST_SUCCESS:
            return action.meta.entityId;
        case CartActions.CLEAR_CART_STATE:
            return wishListInitialState;
    }
    return state;
}
