// @spartacus/storefront/esm5/shared/components/generic-link/generic-link.component.js
import { __decorate, __read } from "tslib";
import { Component, Input } from '@angular/core';
import { Params, Router } from '@angular/router';
/**
 * This component navigates using [routerLink] attribute when input 'url' is a relative url. Otherwise (when it's absolute), [href] is used.
 */
var GenericLinkComponent = /** @class */ (function () {
    function GenericLinkComponent(router) {
        this.router = router;
        /**
         * Pattern matching string starting with `http://` or `https://`.
         */
        this.PROTOCOL_REGEX = /^https?:\/\//i;
        /**
         * Used to split url into 2 parts:
         * 1. the path
         * 2. query params + hash fragment
         */
        this.URL_SPLIT = /(^[^#?]*)(.*)/;
        /**
         * Parsed parts of the @Input `url`, when it's a local URL.
         * It should not be used when the `url` is external.
         * @see `url`
         */
        this.routeParts = {};
    }
    /**
     * Returns true when the @Input `url` is a string starting with `http://` or `https://`.
     */
    GenericLinkComponent.prototype.isExternalUrl = function () {
        return typeof this.url === 'string' && this.PROTOCOL_REGEX.test(this.url);
    };
    Object.defineProperty(GenericLinkComponent.prototype, "rel", {
        get: function () {
            return this.target === '_blank' ? 'noopener' : null;
        },
        enumerable: true,
        configurable: true
    });
    GenericLinkComponent.prototype.ngOnChanges = function (changes) {
        if (changes['url']) {
            this.setUrlParts(changes['url'].currentValue);
        }
    };
    Object.defineProperty(GenericLinkComponent.prototype, "routerUrl", {
        /**
         * The part with the path of the local url.
         */
        get: function () {
            return this.routeParts.path;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GenericLinkComponent.prototype, "queryParams", {
        /**
         * The part with the query params of the local url.
         */
        get: function () {
            return this.routeParts.queryParams;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GenericLinkComponent.prototype, "fragment", {
        /**
         * The part with the hash fragment of the local url.
         */
        get: function () {
            return this.routeParts.fragment;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Parses the given url and sets the property `urlParts` accordingly.
     */
    GenericLinkComponent.prototype.setUrlParts = function (url) {
        if (typeof url === 'string') {
            url = this.getAbsoluteUrl(url); // string links in CMS sometimes don't have the leading slash, so fix it here
            this.routeParts = this.splitUrl(url);
        }
        else {
            this.routeParts = { path: url };
        }
    };
    /**
     * Parses the given string into 3 parts:
     * - string path (wrapped in an array to be compatible with Angular syntax for the `routerLink`)
     * - query params (as an object)
     * - hash fragment (string)
     */
    GenericLinkComponent.prototype.splitUrl = function (url) {
        if (url === void 0) { url = ''; }
        var _a = this.router.parseUrl(url), queryParams = _a.queryParams, fragment = _a.fragment;
        var _b = __read(url.match(this.URL_SPLIT), 2), path = _b[1];
        // wrap path in an array, to have the Angular-like path format
        return { path: [path], queryParams: queryParams, fragment: fragment };
    };
    /**
     * Prepends a leading slash to the given URL string, in case it doesn't have it.
     */
    GenericLinkComponent.prototype.getAbsoluteUrl = function (url) {
        return url.startsWith('/') ? url : '/' + url;
    };
    GenericLinkComponent.ctorParameters = function () { return [
        { type: Router }
    ]; };
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "url", void 0);
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "target", void 0);
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "class", void 0);
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "style", void 0);
    __decorate([
        Input()
    ], GenericLinkComponent.prototype, "title", void 0);
    GenericLinkComponent = __decorate([
        Component({
            selector: 'cx-generic-link',
            template: "<!-- https://github.com/angular/angular/issues/24567 -->\r\n\r\n<ng-container *ngIf=\"isExternalUrl(); else isLocalUrl\">\r\n  <a\r\n    role=\"link\"\r\n    [href]=\"url\"\r\n    [attr.target]=\"target\"\r\n    [attr.rel]=\"rel\"\r\n    [attr.class]=\"class\"\r\n    [attr.id]=\"id\"\r\n    [attr.style]=\"style\"\r\n    [attr.title]=\"title\"\r\n  >\r\n    <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n  </a>\r\n</ng-container>\r\n\r\n<ng-template #isLocalUrl>\r\n  <a\r\n    role=\"link\"\r\n    [routerLink]=\"routerUrl\"\r\n    [queryParams]=\"queryParams\"\r\n    [fragment]=\"fragment\"\r\n    [attr.target]=\"target\"\r\n    [attr.class]=\"class\"\r\n    [attr.id]=\"id\"\r\n    [attr.style]=\"style\"\r\n    [attr.title]=\"title\"\r\n  >\r\n    <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n  </a>\r\n</ng-template>\r\n\r\n<ng-template #content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n"
        })
    ], GenericLinkComponent);
    return GenericLinkComponent;
}());
export { GenericLinkComponent };
