// @spartacus/storefront/esm5/cms-components/cart/cart-totals/cart-totals.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { CartCouponModule } from '../cart-coupon/cart-coupon.module';
import { CartSharedModule } from '../cart-shared/cart-shared.module';
import { CartTotalsComponent } from './cart-totals.component';
var CartTotalsModule = /** @class */ (function () {
    function CartTotalsModule() {
    }
    CartTotalsModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                UrlModule,
                CartSharedModule,
                I18nModule,
                CartCouponModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CartTotalsComponent: {
                            component: CartTotalsComponent,
                        },
                    },
                }),
            ],
            declarations: [CartTotalsComponent],
            exports: [CartTotalsComponent],
            entryComponents: [CartTotalsComponent],
        })
    ], CartTotalsModule);
    return CartTotalsModule;
}());
export { CartTotalsModule };
