// @spartacus/storefront/esm5/cms-components/cart/cart-totals/cart-totals.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActiveCartService, Cart, OrderEntry } from '@spartacus/core';
import { filter } from 'rxjs/operators';
var CartTotalsComponent = /** @class */ (function () {
    function CartTotalsComponent(activeCartService) {
        this.activeCartService = activeCartService;
    }
    CartTotalsComponent.prototype.ngOnInit = function () {
        this.cart$ = this.activeCartService.getActive();
        this.entries$ = this.activeCartService
            .getEntries()
            .pipe(filter(function (entries) { return entries.length > 0; }));
    };
    CartTotalsComponent.ctorParameters = function () { return [
        { type: ActiveCartService }
    ]; };
    CartTotalsComponent = __decorate([
        Component({
            selector: 'cx-cart-totals',
            template: "<ng-container *ngIf=\"cart$ | async as cart\">\n  <ng-container *ngIf=\"entries$ | async as entries\">\n    <cx-order-summary [cart]=\"cart\"></cx-order-summary>\n    <button\n      [routerLink]=\"{ cxRoute: 'checkout' } | cxUrl\"\n      *ngIf=\"entries.length\"\n      class=\"btn btn-primary btn-block\"\n      type=\"button\"\n    >\n      {{ 'cartDetails.proceedToCheckout' | cxTranslate }}\n    </button>\n  </ng-container>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CartTotalsComponent);
    return CartTotalsComponent;
}());
export { CartTotalsComponent };
