// @spartacus/core/esm2015/src/product/store/reducers/product-references.reducer.js
import { ProductActions } from '../actions/index';
export const initialState = {
    productCode: '',
    list: [],
};
export function reducer(state = initialState, action) {
    switch (action.type) {
        case ProductActions.LOAD_PRODUCT_REFERENCES_SUCCESS: {
            const productCode = action.payload.productCode;
            const list = action.payload.list;
            return Object.assign(Object.assign({}, state), { list: [...state.list, ...(list ? list : [])].reduce((productReferences, productReference) => {
                    if (!productReferences.some((obj) => obj.referenceType === productReference.referenceType &&
                        obj.target.code === productReference.target.code)) {
                        productReferences.push(productReference);
                    }
                    return productReferences;
                }, []), productCode });
        }
        case ProductActions.CLEAN_PRODUCT_REFERENCES: {
            return initialState;
        }
    }
    return state;
}
export const getProductReferenceList = (state) => state.list;
export const getProductReferenceProductCode = (state) => state.productCode;
