// @spartacus/storefront/esm2015/cms-components/checkout/components/checkout-progress/checkout-progress.component.js
import { __decorate } from "tslib";
import { RoutingService, RoutingConfigService } from '@spartacus/core';
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { CheckoutConfig } from '../../config/checkout-config';
import { tap } from 'rxjs/operators';
let CheckoutProgressComponent = class CheckoutProgressComponent {
    constructor(config, routingService, routingConfigService) {
        this.config = config;
        this.routingService = routingService;
        this.routingConfigService = routingConfigService;
    }
    ngOnInit() {
        this.steps = this.config.checkout.steps;
        this.routerState$ = this.routingService.getRouterState().pipe(tap((router) => {
            this.activeStepUrl = router.state.context.id;
            this.steps.forEach((step, index) => {
                const routeUrl = `/${this.routingConfigService.getRouteConfig(step.routeName).paths[0]}`;
                if (routeUrl === this.activeStepUrl) {
                    this.activeStepIndex = index;
                }
            });
        }));
    }
    getTabIndex(stepIndex) {
        return !this.isActive(stepIndex) && !this.isDisabled(stepIndex) ? 0 : -1;
    }
    isActive(index) {
        return index === this.activeStepIndex;
    }
    isDisabled(index) {
        return index > this.activeStepIndex;
    }
};
CheckoutProgressComponent.ctorParameters = () => [
    { type: CheckoutConfig },
    { type: RoutingService },
    { type: RoutingConfigService }
];
CheckoutProgressComponent = __decorate([
    Component({
        selector: 'cx-checkout-progress',
        template: "<section *ngIf=\"routerState$ | async as routerState\">\r\n  <div class=\"cx-nav d-none d-lg-block d-xl-block\">\r\n    <ul class=\"cx-list\">\r\n      <li class=\"cx-item\" *ngFor=\"let step of steps; let i = index\">\r\n        <a\r\n          [routerLink]=\"{ cxRoute: step.routeName } | cxUrl\"\r\n          class=\"cx-link\"\r\n          [class.active]=\"isActive(i)\"\r\n          [class.disabled]=\"isDisabled(i)\"\r\n          [tabindex]=\"getTabIndex(i)\"\r\n        >\r\n          {{ i + 1 }}. {{ step.name | cxTranslate }}\r\n        </a>\r\n      </li>\r\n    </ul>\r\n  </div>\r\n</section>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], CheckoutProgressComponent);
export { CheckoutProgressComponent };
