// @spartacus/core/esm5/src/user/store/actions/customer-coupon.action.js
import { __extends } from "tslib";
import { CUSTOMER_COUPONS, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, CLAIM_CUSTOMER_COUPON_PROCESS_ID, } from '../user-state';
import { LoaderLoadAction, LoaderFailAction, LoaderSuccessAction, LoaderResetAction, } from '../../../state/utils/loader/loader.action';
import { EntityFailAction, EntityLoadAction, EntityLoaderResetAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { PROCESS_FEATURE } from '../../../process/store';
export var LOAD_CUSTOMER_COUPONS = '[User] Load Customer Coupons';
export var LOAD_CUSTOMER_COUPONS_FAIL = '[User] Load Customer Coupons Fail';
export var LOAD_CUSTOMER_COUPONS_SUCCESS = '[User] Load Customer Coupons Success';
export var RESET_LOAD_CUSTOMER_COUPONS = '[User] Reset Load Customer Coupons';
export var SUBSCRIBE_CUSTOMER_COUPON = '[User] Subscribe Customer Notification Coupon';
export var SUBSCRIBE_CUSTOMER_COUPON_FAIL = '[User] Subscribe Customer Coupon Notification Fail';
export var SUBSCRIBE_CUSTOMER_COUPON_SUCCESS = '[User] Subscribe Customer Coupon Notification Success';
export var RESET_SUBSCRIBE_CUSTOMER_COUPON_PROCESS = '[User] Reset Subscribe Customer Coupon Process';
export var UNSUBSCRIBE_CUSTOMER_COUPON = '[User] Unsubscribe Customer Notification Coupon';
export var UNSUBSCRIBE_CUSTOMER_COUPON_FAIL = '[User] Unsubscribe Customer Coupon Notification Fail';
export var UNSUBSCRIBE_CUSTOMER_COUPON_SUCCESS = '[User] Unsubscribe Customer Coupon Notification Success';
export var RESET_UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS = '[User] Reset Unsubscribe Customer Coupon Process';
export var CLAIM_CUSTOMER_COUPON = '[User] Claim Customer';
export var CLAIM_CUSTOMER_COUPON_FAIL = '[User] Claim Customer Fail';
export var CLAIM_CUSTOMER_COUPON_SUCCESS = '[User] Claim Customer Success';
var LoadCustomerCoupons = /** @class */ (function (_super) {
    __extends(LoadCustomerCoupons, _super);
    function LoadCustomerCoupons(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS;
        return _this;
    }
    return LoadCustomerCoupons;
}(LoaderLoadAction));
export { LoadCustomerCoupons };
var LoadCustomerCouponsFail = /** @class */ (function (_super) {
    __extends(LoadCustomerCouponsFail, _super);
    function LoadCustomerCouponsFail(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS_FAIL;
        return _this;
    }
    return LoadCustomerCouponsFail;
}(LoaderFailAction));
export { LoadCustomerCouponsFail };
var LoadCustomerCouponsSuccess = /** @class */ (function (_super) {
    __extends(LoadCustomerCouponsSuccess, _super);
    function LoadCustomerCouponsSuccess(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS_SUCCESS;
        return _this;
    }
    return LoadCustomerCouponsSuccess;
}(LoaderSuccessAction));
export { LoadCustomerCouponsSuccess };
var ResetLoadCustomerCoupons = /** @class */ (function (_super) {
    __extends(ResetLoadCustomerCoupons, _super);
    function ResetLoadCustomerCoupons() {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.type = RESET_LOAD_CUSTOMER_COUPONS;
        return _this;
    }
    return ResetLoadCustomerCoupons;
}(LoaderResetAction));
export { ResetLoadCustomerCoupons };
// Subscribe coupon notification actions
var SubscribeCustomerCoupon = /** @class */ (function (_super) {
    __extends(SubscribeCustomerCoupon, _super);
    function SubscribeCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON;
        return _this;
    }
    return SubscribeCustomerCoupon;
}(EntityLoadAction));
export { SubscribeCustomerCoupon };
var SubscribeCustomerCouponFail = /** @class */ (function (_super) {
    __extends(SubscribeCustomerCouponFail, _super);
    function SubscribeCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return SubscribeCustomerCouponFail;
}(EntityFailAction));
export { SubscribeCustomerCouponFail };
var SubscribeCustomerCouponSuccess = /** @class */ (function (_super) {
    __extends(SubscribeCustomerCouponSuccess, _super);
    function SubscribeCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return SubscribeCustomerCouponSuccess;
}(EntitySuccessAction));
export { SubscribeCustomerCouponSuccess };
var ResetSubscribeCustomerCouponProcess = /** @class */ (function (_super) {
    __extends(ResetSubscribeCustomerCouponProcess, _super);
    function ResetSubscribeCustomerCouponProcess() {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.type = RESET_SUBSCRIBE_CUSTOMER_COUPON_PROCESS;
        return _this;
    }
    return ResetSubscribeCustomerCouponProcess;
}(EntityLoaderResetAction));
export { ResetSubscribeCustomerCouponProcess };
var UnsubscribeCustomerCoupon = /** @class */ (function (_super) {
    __extends(UnsubscribeCustomerCoupon, _super);
    function UnsubscribeCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON;
        return _this;
    }
    return UnsubscribeCustomerCoupon;
}(EntityLoadAction));
export { UnsubscribeCustomerCoupon };
var UnsubscribeCustomerCouponFail = /** @class */ (function (_super) {
    __extends(UnsubscribeCustomerCouponFail, _super);
    function UnsubscribeCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return UnsubscribeCustomerCouponFail;
}(EntityFailAction));
export { UnsubscribeCustomerCouponFail };
var UnsubscribeCustomerCouponSuccess = /** @class */ (function (_super) {
    __extends(UnsubscribeCustomerCouponSuccess, _super);
    function UnsubscribeCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return UnsubscribeCustomerCouponSuccess;
}(EntitySuccessAction));
export { UnsubscribeCustomerCouponSuccess };
var ResetUnsubscribeCustomerCouponProcess = /** @class */ (function (_super) {
    __extends(ResetUnsubscribeCustomerCouponProcess, _super);
    function ResetUnsubscribeCustomerCouponProcess() {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.type = RESET_UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS;
        return _this;
    }
    return ResetUnsubscribeCustomerCouponProcess;
}(EntityLoaderResetAction));
export { ResetUnsubscribeCustomerCouponProcess };
var ClaimCustomerCoupon = /** @class */ (function (_super) {
    __extends(ClaimCustomerCoupon, _super);
    function ClaimCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON;
        return _this;
    }
    return ClaimCustomerCoupon;
}(EntityLoadAction));
export { ClaimCustomerCoupon };
var ClaimCustomerCouponFail = /** @class */ (function (_super) {
    __extends(ClaimCustomerCouponFail, _super);
    function ClaimCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return ClaimCustomerCouponFail;
}(EntityFailAction));
export { ClaimCustomerCouponFail };
var ClaimCustomerCouponSuccess = /** @class */ (function (_super) {
    __extends(ClaimCustomerCouponSuccess, _super);
    function ClaimCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return ClaimCustomerCouponSuccess;
}(EntitySuccessAction));
export { ClaimCustomerCouponSuccess };
