// @spartacus/core/esm5/src/user/store/actions/order-details.action.js
import { __extends } from "tslib";
import { LoaderFailAction, LoaderLoadAction, LoaderSuccessAction, LoaderResetAction, } from '../../../state/utils/loader/loader.action';
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { USER_ORDER_DETAILS, CANCEL_ORDER_PROCESS_ID } from '../user-state';
export var LOAD_ORDER_DETAILS = '[User] Load Order Details';
export var LOAD_ORDER_DETAILS_FAIL = '[User] Load Order Details Fail';
export var LOAD_ORDER_DETAILS_SUCCESS = '[User] Load Order Details Success';
export var CLEAR_ORDER_DETAILS = '[User] Clear Order Details';
export var CANCEL_ORDER = '[User] Cancel Order';
export var CANCEL_ORDER_FAIL = '[User] Cancel Order Fail';
export var CANCEL_ORDER_SUCCESS = '[User] Cancel Order Success';
export var RESET_CANCEL_ORDER_PROCESS = '[User] Reset Cancel Order Process';
var LoadOrderDetails = /** @class */ (function (_super) {
    __extends(LoadOrderDetails, _super);
    function LoadOrderDetails(payload) {
        var _this = _super.call(this, USER_ORDER_DETAILS) || this;
        _this.payload = payload;
        _this.type = LOAD_ORDER_DETAILS;
        return _this;
    }
    return LoadOrderDetails;
}(LoaderLoadAction));
export { LoadOrderDetails };
var LoadOrderDetailsFail = /** @class */ (function (_super) {
    __extends(LoadOrderDetailsFail, _super);
    function LoadOrderDetailsFail(payload) {
        var _this = _super.call(this, USER_ORDER_DETAILS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_ORDER_DETAILS_FAIL;
        return _this;
    }
    return LoadOrderDetailsFail;
}(LoaderFailAction));
export { LoadOrderDetailsFail };
var LoadOrderDetailsSuccess = /** @class */ (function (_super) {
    __extends(LoadOrderDetailsSuccess, _super);
    function LoadOrderDetailsSuccess(payload) {
        var _this = _super.call(this, USER_ORDER_DETAILS) || this;
        _this.payload = payload;
        _this.type = LOAD_ORDER_DETAILS_SUCCESS;
        return _this;
    }
    return LoadOrderDetailsSuccess;
}(LoaderSuccessAction));
export { LoadOrderDetailsSuccess };
var ClearOrderDetails = /** @class */ (function (_super) {
    __extends(ClearOrderDetails, _super);
    function ClearOrderDetails() {
        var _this = _super.call(this, USER_ORDER_DETAILS) || this;
        _this.type = CLEAR_ORDER_DETAILS;
        return _this;
    }
    return ClearOrderDetails;
}(LoaderResetAction));
export { ClearOrderDetails };
var CancelOrder = /** @class */ (function (_super) {
    __extends(CancelOrder, _super);
    function CancelOrder(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CANCEL_ORDER_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CANCEL_ORDER;
        return _this;
    }
    return CancelOrder;
}(StateUtils.EntityLoadAction));
export { CancelOrder };
var CancelOrderFail = /** @class */ (function (_super) {
    __extends(CancelOrderFail, _super);
    function CancelOrderFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CANCEL_ORDER_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CANCEL_ORDER_FAIL;
        return _this;
    }
    return CancelOrderFail;
}(StateUtils.EntityFailAction));
export { CancelOrderFail };
var CancelOrderSuccess = /** @class */ (function (_super) {
    __extends(CancelOrderSuccess, _super);
    function CancelOrderSuccess() {
        var _this = _super.call(this, PROCESS_FEATURE, CANCEL_ORDER_PROCESS_ID) || this;
        _this.type = CANCEL_ORDER_SUCCESS;
        return _this;
    }
    return CancelOrderSuccess;
}(StateUtils.EntitySuccessAction));
export { CancelOrderSuccess };
var ResetCancelOrderProcess = /** @class */ (function (_super) {
    __extends(ResetCancelOrderProcess, _super);
    function ResetCancelOrderProcess() {
        var _this = _super.call(this, PROCESS_FEATURE, CANCEL_ORDER_PROCESS_ID) || this;
        _this.type = RESET_CANCEL_ORDER_PROCESS;
        return _this;
    }
    return ResetCancelOrderProcess;
}(StateUtils.EntityLoaderResetAction));
export { ResetCancelOrderProcess };
