// @spartacus/core/esm5/src/user/store/actions/payment-methods.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { USER_PAYMENT_METHODS } from '../user-state';
export var LOAD_USER_PAYMENT_METHODS = '[User] Load User Payment Methods';
export var LOAD_USER_PAYMENT_METHODS_FAIL = '[User] Load User Payment Methods Fail';
export var LOAD_USER_PAYMENT_METHODS_SUCCESS = '[User] Load User Payment Methods Success';
export var SET_DEFAULT_USER_PAYMENT_METHOD = '[User] Set Default User Payment Method';
export var SET_DEFAULT_USER_PAYMENT_METHOD_FAIL = '[User] Set Default User Payment Method Fail';
export var SET_DEFAULT_USER_PAYMENT_METHOD_SUCCESS = '[User] Set Default User Payment Method Success';
export var DELETE_USER_PAYMENT_METHOD = '[User] Delete User Payment Method';
export var DELETE_USER_PAYMENT_METHOD_FAIL = '[User] Delete User Payment Method Fail';
export var DELETE_USER_PAYMENT_METHOD_SUCCESS = '[User] Delete User  Payment Method Success';
var LoadUserPaymentMethods = /** @class */ (function (_super) {
    __extends(LoadUserPaymentMethods, _super);
    function LoadUserPaymentMethods(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_PAYMENT_METHODS;
        return _this;
    }
    return LoadUserPaymentMethods;
}(StateUtils.LoaderLoadAction));
export { LoadUserPaymentMethods };
var LoadUserPaymentMethodsFail = /** @class */ (function (_super) {
    __extends(LoadUserPaymentMethodsFail, _super);
    function LoadUserPaymentMethodsFail(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_PAYMENT_METHODS_FAIL;
        return _this;
    }
    return LoadUserPaymentMethodsFail;
}(StateUtils.LoaderFailAction));
export { LoadUserPaymentMethodsFail };
var LoadUserPaymentMethodsSuccess = /** @class */ (function (_super) {
    __extends(LoadUserPaymentMethodsSuccess, _super);
    function LoadUserPaymentMethodsSuccess(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_PAYMENT_METHODS_SUCCESS;
        return _this;
    }
    return LoadUserPaymentMethodsSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadUserPaymentMethodsSuccess };
var SetDefaultUserPaymentMethod = /** @class */ (function (_super) {
    __extends(SetDefaultUserPaymentMethod, _super);
    function SetDefaultUserPaymentMethod(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = SET_DEFAULT_USER_PAYMENT_METHOD;
        return _this;
    }
    return SetDefaultUserPaymentMethod;
}(StateUtils.LoaderLoadAction));
export { SetDefaultUserPaymentMethod };
var SetDefaultUserPaymentMethodFail = /** @class */ (function (_super) {
    __extends(SetDefaultUserPaymentMethodFail, _super);
    function SetDefaultUserPaymentMethodFail(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS, payload) || this;
        _this.payload = payload;
        _this.type = SET_DEFAULT_USER_PAYMENT_METHOD_FAIL;
        return _this;
    }
    return SetDefaultUserPaymentMethodFail;
}(StateUtils.LoaderFailAction));
export { SetDefaultUserPaymentMethodFail };
var SetDefaultUserPaymentMethodSuccess = /** @class */ (function (_super) {
    __extends(SetDefaultUserPaymentMethodSuccess, _super);
    function SetDefaultUserPaymentMethodSuccess(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = SET_DEFAULT_USER_PAYMENT_METHOD_SUCCESS;
        return _this;
    }
    return SetDefaultUserPaymentMethodSuccess;
}(StateUtils.LoaderSuccessAction));
export { SetDefaultUserPaymentMethodSuccess };
var DeleteUserPaymentMethod = /** @class */ (function (_super) {
    __extends(DeleteUserPaymentMethod, _super);
    function DeleteUserPaymentMethod(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_PAYMENT_METHOD;
        return _this;
    }
    return DeleteUserPaymentMethod;
}(StateUtils.LoaderLoadAction));
export { DeleteUserPaymentMethod };
var DeleteUserPaymentMethodFail = /** @class */ (function (_super) {
    __extends(DeleteUserPaymentMethodFail, _super);
    function DeleteUserPaymentMethodFail(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS, payload) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_PAYMENT_METHOD_FAIL;
        return _this;
    }
    return DeleteUserPaymentMethodFail;
}(StateUtils.LoaderFailAction));
export { DeleteUserPaymentMethodFail };
var DeleteUserPaymentMethodSuccess = /** @class */ (function (_super) {
    __extends(DeleteUserPaymentMethodSuccess, _super);
    function DeleteUserPaymentMethodSuccess(payload) {
        var _this = _super.call(this, USER_PAYMENT_METHODS) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_PAYMENT_METHOD_SUCCESS;
        return _this;
    }
    return DeleteUserPaymentMethodSuccess;
}(StateUtils.LoaderSuccessAction));
export { DeleteUserPaymentMethodSuccess };
