// @spartacus/core/esm5/src/user/store/actions/regions.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { REGIONS } from '../user-state';
export var LOAD_REGIONS = '[User] Load Regions';
export var LOAD_REGIONS_SUCCESS = '[User] Load Regions Success';
export var LOAD_REGIONS_FAIL = '[User] Load Regions Fail';
export var CLEAR_REGIONS = '[User] Clear Regions';
var LoadRegions = /** @class */ (function (_super) {
    __extends(LoadRegions, _super);
    function LoadRegions(payload) {
        var _this = _super.call(this, REGIONS) || this;
        _this.payload = payload;
        _this.type = LOAD_REGIONS;
        return _this;
    }
    return LoadRegions;
}(StateUtils.LoaderLoadAction));
export { LoadRegions };
var LoadRegionsFail = /** @class */ (function (_super) {
    __extends(LoadRegionsFail, _super);
    function LoadRegionsFail(payload) {
        var _this = _super.call(this, REGIONS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_REGIONS_FAIL;
        return _this;
    }
    return LoadRegionsFail;
}(StateUtils.LoaderFailAction));
export { LoadRegionsFail };
var LoadRegionsSuccess = /** @class */ (function (_super) {
    __extends(LoadRegionsSuccess, _super);
    function LoadRegionsSuccess(payload) {
        var _this = _super.call(this, REGIONS) || this;
        _this.payload = payload;
        _this.type = LOAD_REGIONS_SUCCESS;
        return _this;
    }
    return LoadRegionsSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadRegionsSuccess };
var ClearRegions = /** @class */ (function () {
    function ClearRegions() {
        this.type = CLEAR_REGIONS;
    }
    return ClearRegions;
}());
export { ClearRegions };
