// @spartacus/core/esm5/src/user/store/actions/update-password.action.js
import { __extends } from "tslib";
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { UPDATE_PASSWORD_PROCESS_ID } from '../user-state';
export var UPDATE_PASSWORD = '[User] Update Password';
export var UPDATE_PASSWORD_FAIL = '[User] Update Password Fail';
export var UPDATE_PASSWORD_SUCCESS = '[User] Update Password Success';
export var UPDATE_PASSWORD_RESET = '[User] Reset Update Password Process State';
var UpdatePassword = /** @class */ (function (_super) {
    __extends(UpdatePassword, _super);
    function UpdatePassword(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_PASSWORD_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UPDATE_PASSWORD;
        return _this;
    }
    return UpdatePassword;
}(StateUtils.EntityLoadAction));
export { UpdatePassword };
var UpdatePasswordFail = /** @class */ (function (_super) {
    __extends(UpdatePasswordFail, _super);
    function UpdatePasswordFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_PASSWORD_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UPDATE_PASSWORD_FAIL;
        return _this;
    }
    return UpdatePasswordFail;
}(StateUtils.EntityFailAction));
export { UpdatePasswordFail };
var UpdatePasswordSuccess = /** @class */ (function (_super) {
    __extends(UpdatePasswordSuccess, _super);
    function UpdatePasswordSuccess() {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_PASSWORD_PROCESS_ID) || this;
        _this.type = UPDATE_PASSWORD_SUCCESS;
        return _this;
    }
    return UpdatePasswordSuccess;
}(StateUtils.EntitySuccessAction));
export { UpdatePasswordSuccess };
var UpdatePasswordReset = /** @class */ (function (_super) {
    __extends(UpdatePasswordReset, _super);
    function UpdatePasswordReset() {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_PASSWORD_PROCESS_ID) || this;
        _this.type = UPDATE_PASSWORD_RESET;
        return _this;
    }
    return UpdatePasswordReset;
}(StateUtils.EntityLoaderResetAction));
export { UpdatePasswordReset };
