// @spartacus/core/esm5/src/user/store/actions/update-email.action.js
import { __extends } from "tslib";
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { UPDATE_EMAIL_PROCESS_ID } from '../user-state';
export var UPDATE_EMAIL = '[User] Update Email';
export var UPDATE_EMAIL_ERROR = '[User] Update Email Error';
export var UPDATE_EMAIL_SUCCESS = '[User] Update Email Success';
export var RESET_EMAIL = '[User] Reset Email';
var UpdateEmailAction = /** @class */ (function (_super) {
    __extends(UpdateEmailAction, _super);
    function UpdateEmailAction(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UPDATE_EMAIL;
        return _this;
    }
    return UpdateEmailAction;
}(StateUtils.EntityLoadAction));
export { UpdateEmailAction };
var UpdateEmailSuccessAction = /** @class */ (function (_super) {
    __extends(UpdateEmailSuccessAction, _super);
    function UpdateEmailSuccessAction(newUid) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID) || this;
        _this.newUid = newUid;
        _this.type = UPDATE_EMAIL_SUCCESS;
        return _this;
    }
    return UpdateEmailSuccessAction;
}(StateUtils.EntitySuccessAction));
export { UpdateEmailSuccessAction };
var UpdateEmailErrorAction = /** @class */ (function (_super) {
    __extends(UpdateEmailErrorAction, _super);
    function UpdateEmailErrorAction(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UPDATE_EMAIL_ERROR;
        return _this;
    }
    return UpdateEmailErrorAction;
}(StateUtils.EntityFailAction));
export { UpdateEmailErrorAction };
var ResetUpdateEmailAction = /** @class */ (function (_super) {
    __extends(ResetUpdateEmailAction, _super);
    function ResetUpdateEmailAction() {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID) || this;
        _this.type = RESET_EMAIL;
        return _this;
    }
    return ResetUpdateEmailAction;
}(StateUtils.EntityLoaderResetAction));
export { ResetUpdateEmailAction };
