// @spartacus/core/esm5/src/user/store/actions/user-details.action.js
import { __extends } from "tslib";
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { UPDATE_USER_DETAILS_PROCESS_ID } from '../user-state';
export var LOAD_USER_DETAILS = '[User] Load User Details';
export var LOAD_USER_DETAILS_FAIL = '[User] Load User Details Fail';
export var LOAD_USER_DETAILS_SUCCESS = '[User] Load User Details Success';
export var UPDATE_USER_DETAILS = '[User] Update User Details';
export var UPDATE_USER_DETAILS_FAIL = '[User] Update User Details Fail';
export var UPDATE_USER_DETAILS_SUCCESS = '[User] Update User Details Success';
export var RESET_USER_DETAILS = '[User] Reset User Details';
var LoadUserDetails = /** @class */ (function () {
    function LoadUserDetails(payload) {
        this.payload = payload;
        this.type = LOAD_USER_DETAILS;
    }
    return LoadUserDetails;
}());
export { LoadUserDetails };
var LoadUserDetailsFail = /** @class */ (function () {
    function LoadUserDetailsFail(payload) {
        this.payload = payload;
        this.type = LOAD_USER_DETAILS_FAIL;
    }
    return LoadUserDetailsFail;
}());
export { LoadUserDetailsFail };
var LoadUserDetailsSuccess = /** @class */ (function () {
    function LoadUserDetailsSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_USER_DETAILS_SUCCESS;
    }
    return LoadUserDetailsSuccess;
}());
export { LoadUserDetailsSuccess };
var UpdateUserDetails = /** @class */ (function (_super) {
    __extends(UpdateUserDetails, _super);
    function UpdateUserDetails(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_USER_DETAILS_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UPDATE_USER_DETAILS;
        return _this;
    }
    return UpdateUserDetails;
}(StateUtils.EntityLoadAction));
export { UpdateUserDetails };
var UpdateUserDetailsFail = /** @class */ (function (_super) {
    __extends(UpdateUserDetailsFail, _super);
    function UpdateUserDetailsFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_USER_DETAILS_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UPDATE_USER_DETAILS_FAIL;
        return _this;
    }
    return UpdateUserDetailsFail;
}(StateUtils.EntityFailAction));
export { UpdateUserDetailsFail };
var UpdateUserDetailsSuccess = /** @class */ (function (_super) {
    __extends(UpdateUserDetailsSuccess, _super);
    function UpdateUserDetailsSuccess(userUpdates) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_USER_DETAILS_PROCESS_ID) || this;
        _this.userUpdates = userUpdates;
        _this.type = UPDATE_USER_DETAILS_SUCCESS;
        return _this;
    }
    return UpdateUserDetailsSuccess;
}(StateUtils.EntitySuccessAction));
export { UpdateUserDetailsSuccess };
var ResetUpdateUserDetails = /** @class */ (function (_super) {
    __extends(ResetUpdateUserDetails, _super);
    function ResetUpdateUserDetails() {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_USER_DETAILS_PROCESS_ID) || this;
        _this.type = RESET_USER_DETAILS;
        return _this;
    }
    return ResetUpdateUserDetails;
}(StateUtils.EntityLoaderResetAction));
export { ResetUpdateUserDetails };
