// @spartacus/core/esm5/src/user/store/actions/user-orders.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { USER_ORDERS } from '../user-state';
export var LOAD_USER_ORDERS = '[User] Load User Orders';
export var LOAD_USER_ORDERS_FAIL = '[User] Load User Orders Fail';
export var LOAD_USER_ORDERS_SUCCESS = '[User] Load User Orders Success';
export var CLEAR_USER_ORDERS = '[User] Clear User Orders';
var LoadUserOrders = /** @class */ (function (_super) {
    __extends(LoadUserOrders, _super);
    function LoadUserOrders(payload) {
        var _this = _super.call(this, USER_ORDERS) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ORDERS;
        return _this;
    }
    return LoadUserOrders;
}(StateUtils.LoaderLoadAction));
export { LoadUserOrders };
var LoadUserOrdersFail = /** @class */ (function (_super) {
    __extends(LoadUserOrdersFail, _super);
    function LoadUserOrdersFail(payload) {
        var _this = _super.call(this, USER_ORDERS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ORDERS_FAIL;
        return _this;
    }
    return LoadUserOrdersFail;
}(StateUtils.LoaderFailAction));
export { LoadUserOrdersFail };
var LoadUserOrdersSuccess = /** @class */ (function (_super) {
    __extends(LoadUserOrdersSuccess, _super);
    function LoadUserOrdersSuccess(payload) {
        var _this = _super.call(this, USER_ORDERS) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ORDERS_SUCCESS;
        return _this;
    }
    return LoadUserOrdersSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadUserOrdersSuccess };
var ClearUserOrders = /** @class */ (function (_super) {
    __extends(ClearUserOrders, _super);
    function ClearUserOrders() {
        var _this = _super.call(this, USER_ORDERS) || this;
        _this.type = CLEAR_USER_ORDERS;
        return _this;
    }
    return ClearUserOrders;
}(StateUtils.LoaderResetAction));
export { ClearUserOrders };
