// @spartacus/core/esm5/src/user/store/actions/user-addresses.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { USER_ADDRESSES } from '../user-state';
export var LOAD_USER_ADDRESSES = '[User] Load User Addresses';
export var LOAD_USER_ADDRESSES_FAIL = '[User] Load User Addresses Fail';
export var LOAD_USER_ADDRESSES_SUCCESS = '[User] Load User Addresses Success';
export var ADD_USER_ADDRESS = '[User] Add User Address';
export var ADD_USER_ADDRESS_FAIL = '[User] Add User Address Fail';
export var ADD_USER_ADDRESS_SUCCESS = '[User] Add User Address Success';
export var UPDATE_USER_ADDRESS = '[User] Update User Address';
export var UPDATE_USER_ADDRESS_FAIL = '[User] Update User Address Fail';
export var UPDATE_USER_ADDRESS_SUCCESS = '[User] Update User Address Success';
export var DELETE_USER_ADDRESS = '[User] Delete User Address';
export var DELETE_USER_ADDRESS_FAIL = '[User] Delete User Address Fail';
export var DELETE_USER_ADDRESS_SUCCESS = '[User] Delete User Address Success';
var LoadUserAddresses = /** @class */ (function (_super) {
    __extends(LoadUserAddresses, _super);
    function LoadUserAddresses(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ADDRESSES;
        return _this;
    }
    return LoadUserAddresses;
}(StateUtils.LoaderLoadAction));
export { LoadUserAddresses };
var LoadUserAddressesFail = /** @class */ (function (_super) {
    __extends(LoadUserAddressesFail, _super);
    function LoadUserAddressesFail(payload) {
        var _this = _super.call(this, USER_ADDRESSES, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ADDRESSES_FAIL;
        return _this;
    }
    return LoadUserAddressesFail;
}(StateUtils.LoaderFailAction));
export { LoadUserAddressesFail };
var LoadUserAddressesSuccess = /** @class */ (function (_super) {
    __extends(LoadUserAddressesSuccess, _super);
    function LoadUserAddressesSuccess(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = LOAD_USER_ADDRESSES_SUCCESS;
        return _this;
    }
    return LoadUserAddressesSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadUserAddressesSuccess };
// Adding address actions
var AddUserAddress = /** @class */ (function (_super) {
    __extends(AddUserAddress, _super);
    function AddUserAddress(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = ADD_USER_ADDRESS;
        return _this;
    }
    return AddUserAddress;
}(StateUtils.LoaderLoadAction));
export { AddUserAddress };
var AddUserAddressFail = /** @class */ (function (_super) {
    __extends(AddUserAddressFail, _super);
    function AddUserAddressFail(payload) {
        var _this = _super.call(this, USER_ADDRESSES, payload) || this;
        _this.payload = payload;
        _this.type = ADD_USER_ADDRESS_FAIL;
        return _this;
    }
    return AddUserAddressFail;
}(StateUtils.LoaderFailAction));
export { AddUserAddressFail };
var AddUserAddressSuccess = /** @class */ (function (_super) {
    __extends(AddUserAddressSuccess, _super);
    function AddUserAddressSuccess(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = ADD_USER_ADDRESS_SUCCESS;
        return _this;
    }
    return AddUserAddressSuccess;
}(StateUtils.LoaderSuccessAction));
export { AddUserAddressSuccess };
// Updating address actions
var UpdateUserAddress = /** @class */ (function (_super) {
    __extends(UpdateUserAddress, _super);
    function UpdateUserAddress(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = UPDATE_USER_ADDRESS;
        return _this;
    }
    return UpdateUserAddress;
}(StateUtils.LoaderLoadAction));
export { UpdateUserAddress };
var UpdateUserAddressFail = /** @class */ (function (_super) {
    __extends(UpdateUserAddressFail, _super);
    function UpdateUserAddressFail(payload) {
        var _this = _super.call(this, USER_ADDRESSES, payload) || this;
        _this.payload = payload;
        _this.type = UPDATE_USER_ADDRESS_FAIL;
        return _this;
    }
    return UpdateUserAddressFail;
}(StateUtils.LoaderFailAction));
export { UpdateUserAddressFail };
var UpdateUserAddressSuccess = /** @class */ (function (_super) {
    __extends(UpdateUserAddressSuccess, _super);
    function UpdateUserAddressSuccess(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = UPDATE_USER_ADDRESS_SUCCESS;
        return _this;
    }
    return UpdateUserAddressSuccess;
}(StateUtils.LoaderSuccessAction));
export { UpdateUserAddressSuccess };
// Deleting address actions
var DeleteUserAddress = /** @class */ (function (_super) {
    __extends(DeleteUserAddress, _super);
    function DeleteUserAddress(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_ADDRESS;
        return _this;
    }
    return DeleteUserAddress;
}(StateUtils.LoaderLoadAction));
export { DeleteUserAddress };
var DeleteUserAddressFail = /** @class */ (function (_super) {
    __extends(DeleteUserAddressFail, _super);
    function DeleteUserAddressFail(payload) {
        var _this = _super.call(this, USER_ADDRESSES, payload) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_ADDRESS_FAIL;
        return _this;
    }
    return DeleteUserAddressFail;
}(StateUtils.LoaderFailAction));
export { DeleteUserAddressFail };
var DeleteUserAddressSuccess = /** @class */ (function (_super) {
    __extends(DeleteUserAddressSuccess, _super);
    function DeleteUserAddressSuccess(payload) {
        var _this = _super.call(this, USER_ADDRESSES) || this;
        _this.payload = payload;
        _this.type = DELETE_USER_ADDRESS_SUCCESS;
        return _this;
    }
    return DeleteUserAddressSuccess;
}(StateUtils.LoaderSuccessAction));
export { DeleteUserAddressSuccess };
