// @spartacus/storefront/esm2015/cms-components/asm/csagent-login-form/csagent-login-form.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output, ViewEncapsulation, } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
let CSAgentLoginFormComponent = class CSAgentLoginFormComponent {
    constructor(fb) {
        this.fb = fb;
        this.csAgentTokenLoading = false;
        this.submitEvent = new EventEmitter();
    }
    ngOnInit() {
        this.csAgentLoginForm = this.fb.group({
            userId: ['', [Validators.required]],
            password: ['', [Validators.required]],
        });
    }
    onSubmit() {
        if (this.csAgentLoginForm.valid) {
            this.submitEvent.emit({
                userId: this.csAgentLoginForm.get('userId').value,
                password: this.csAgentLoginForm.get('password').value,
            });
        }
        else {
            this.csAgentLoginForm.markAllAsTouched();
        }
    }
};
CSAgentLoginFormComponent.ctorParameters = () => [
    { type: FormBuilder }
];
__decorate([
    Input()
], CSAgentLoginFormComponent.prototype, "csAgentTokenLoading", void 0);
__decorate([
    Output()
], CSAgentLoginFormComponent.prototype, "submitEvent", void 0);
CSAgentLoginFormComponent = __decorate([
    Component({
        selector: 'cx-csagent-login-form',
        template: "<form\r\n  (ngSubmit)=\"onSubmit()\"\r\n  [formGroup]=\"csAgentLoginForm\"\r\n  *ngIf=\"!csAgentTokenLoading\"\r\n>\r\n  <label>\r\n    <input\r\n      type=\"text\"\r\n      formControlName=\"userId\"\r\n      placeholder=\"{{ 'asm.loginForm.userId.label' | cxTranslate }}\"\r\n    />\r\n    <cx-form-errors [control]=\"csAgentLoginForm.get('userId')\"></cx-form-errors>\r\n  </label>\r\n\r\n  <label>\r\n    <input\r\n      type=\"password\"\r\n      placeholder=\"{{ 'asm.loginForm.password.label' | cxTranslate }}\"\r\n      formControlName=\"password\"\r\n    />\r\n    <cx-form-errors\r\n      [control]=\"csAgentLoginForm.get('password')\"\r\n    ></cx-form-errors>\r\n  </label>\r\n  <button type=\"submit\">\r\n    {{ 'asm.loginForm.submit' | cxTranslate }}\r\n  </button>\r\n</form>\r\n\r\n<div\r\n  *ngIf=\"csAgentTokenLoading\"\r\n  class=\"spinner\"\r\n  aria-hidden=\"false\"\r\n  aria-label=\"Loading\"\r\n>\r\n  <div></div>\r\n  <div></div>\r\n  <div></div>\r\n</div>\r\n",
        encapsulation: ViewEncapsulation.None,
        styles: ["cx-csagent-login-form form{display:flex;width:100%}@media (max-width:575px){cx-csagent-login-form form{flex-direction:column}}cx-csagent-login-form form label{margin:0 0 15px;min-width:auto}@media (min-width:575px){cx-csagent-login-form form label{margin:0 15px 0 0;min-width:15rem}}cx-csagent-login-form form label input{width:100%}cx-csagent-login-form button[type=submit]{color:#fff;border-color:#0a6ed1;background-color:#0a6ed1}cx-csagent-login-form button[type=submit]:hover{background-color:#085caf}"]
    })
], CSAgentLoginFormComponent);
export { CSAgentLoginFormComponent };
