// @spartacus/storefront/esm5/cms-components/product/stock-notification/stock-notification.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StockNotificationComponent } from './stock-notification.component';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { StockNotificationDialogComponent } from './stock-notification-dialog/stock-notification-dialog.component';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { RouterModule } from '@angular/router';
var StockNotificationModule = /** @class */ (function () {
    function StockNotificationModule() {
    }
    StockNotificationModule = __decorate([
        NgModule({
            declarations: [StockNotificationComponent, StockNotificationDialogComponent],
            imports: [CommonModule, RouterModule, I18nModule, SpinnerModule, UrlModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        StockNotificationComponent: {
                            component: StockNotificationComponent,
                        },
                    },
                }),
            ],
            entryComponents: [
                StockNotificationComponent,
                StockNotificationDialogComponent,
            ],
            exports: [StockNotificationComponent, StockNotificationDialogComponent],
        })
    ], StockNotificationModule);
    return StockNotificationModule;
}());
export { StockNotificationModule };
