// @spartacus/core/esm5/src/product/store/effects/product-reviews.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, mergeMap, tap } from 'rxjs/operators';
import { ProductReviewsConnector } from '../../connectors/reviews/product-reviews.connector';
import { ProductActions } from '../actions/index';
import { GlobalMessageService, GlobalMessageType, } from '../../../global-message/index';
var ProductReviewsEffects = /** @class */ (function () {
    function ProductReviewsEffects(actions$, productReviewsConnector, globalMessageService) {
        var _this = this;
        this.actions$ = actions$;
        this.productReviewsConnector = productReviewsConnector;
        this.globalMessageService = globalMessageService;
        this.loadProductReviews$ = this.actions$.pipe(ofType(ProductActions.LOAD_PRODUCT_REVIEWS), map(function (action) { return action.payload; }), mergeMap(function (productCode) {
            return _this.productReviewsConnector.get(productCode).pipe(map(function (data) {
                return new ProductActions.LoadProductReviewsSuccess({
                    productCode: productCode,
                    list: data,
                });
            }), catchError(function (_error) {
                return of(new ProductActions.LoadProductReviewsFail({
                    message: productCode,
                }));
            }));
        }));
        this.postProductReview = this.actions$.pipe(ofType(ProductActions.POST_PRODUCT_REVIEW), map(function (action) { return action.payload; }), mergeMap(function (payload) {
            return _this.productReviewsConnector
                .add(payload.productCode, payload.review)
                .pipe(map(function (reviewResponse) {
                return new ProductActions.PostProductReviewSuccess(reviewResponse);
            }), catchError(function (_error) {
                return of(new ProductActions.PostProductReviewFail(payload.productCode));
            }));
        }));
        this.showGlobalMessageOnPostProductReviewSuccess$ = this.actions$.pipe(ofType(ProductActions.POST_PRODUCT_REVIEW_SUCCESS), tap(function () {
            _this.globalMessageService.add({ key: 'productReview.thankYouForReview' }, GlobalMessageType.MSG_TYPE_CONFIRMATION);
        }));
    }
    ProductReviewsEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: ProductReviewsConnector },
        { type: GlobalMessageService }
    ]; };
    __decorate([
        Effect()
    ], ProductReviewsEffects.prototype, "loadProductReviews$", void 0);
    __decorate([
        Effect()
    ], ProductReviewsEffects.prototype, "postProductReview", void 0);
    __decorate([
        Effect({ dispatch: false })
    ], ProductReviewsEffects.prototype, "showGlobalMessageOnPostProductReviewSuccess$", void 0);
    ProductReviewsEffects = __decorate([
        Injectable()
    ], ProductReviewsEffects);
    return ProductReviewsEffects;
}());
export { ProductReviewsEffects };
