// @spartacus/core/esm5/src/product/store/effects/product.effect.js
import { __assign, __decorate, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { merge, of } from 'rxjs';
import { catchError, map, mergeMap } from 'rxjs/operators';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { ProductConnector } from '../../connectors/product/product.connector';
import { ProductActions } from '../actions/index';
import { SiteContextActions } from '../../../site-context/store/actions/index';
import { bufferDebounceTime } from '../../../util/buffer-debounce-time';
import { withdrawOn } from '../../../util/withdraw-on';
var ProductEffects = /** @class */ (function () {
    function ProductEffects(actions$, productConnector) {
        var _this = this;
        this.actions$ = actions$;
        this.productConnector = productConnector;
        // we want to cancel all ongoing requests when currency or language changes,
        this.contextChange$ = this.actions$.pipe(ofType(SiteContextActions.CURRENCY_CHANGE, SiteContextActions.LANGUAGE_CHANGE));
        this.loadProduct$ = createEffect(function () { return function (_a) {
            var _b = _a === void 0 ? {} : _a, scheduler = _b.scheduler, _c = _b.debounce, debounce = _c === void 0 ? 0 : _c;
            return _this.actions$.pipe(ofType(ProductActions.LOAD_PRODUCT), map(function (action) { return ({
                code: action.payload,
                scope: action.meta.scope,
            }); }), 
            // we are grouping all load actions that happens at the same time
            // to optimize loading and pass them all to productConnector.getMany
            bufferDebounceTime(debounce, scheduler), mergeMap(function (products) {
                return merge.apply(void 0, __spread(_this.productConnector
                    .getMany(products)
                    .map(_this.productLoadEffect)));
            }), withdrawOn(_this.contextChange$));
        }; });
    }
    ProductEffects.prototype.productLoadEffect = function (productLoad) {
        return productLoad.data$.pipe(map(function (data) {
            return new ProductActions.LoadProductSuccess(__assign({ code: productLoad.code }, data), productLoad.scope);
        }), catchError(function (error) {
            return of(new ProductActions.LoadProductFail(productLoad.code, makeErrorSerializable(error), productLoad.scope));
        }));
    };
    ProductEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: ProductConnector }
    ]; };
    ProductEffects = __decorate([
        Injectable()
    ], ProductEffects);
    return ProductEffects;
}());
export { ProductEffects };
