// @spartacus/storefront/esm2015/cms-components/product/product-list/product-facet-navigation/product-facet-navigation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ViewChild, } from '@angular/core';
import { asapScheduler, BehaviorSubject, interval, of } from 'rxjs';
import { delayWhen, observeOn, switchMap } from 'rxjs/operators';
import { ICON_TYPE } from '../../../../cms-components/misc/icon/icon.model';
import { BreakpointService } from '../../../../layout/breakpoint/breakpoint.service';
let ProductFacetNavigationComponent = class ProductFacetNavigationComponent {
    constructor(breakpointService) {
        this.breakpointService = breakpointService;
        this.iconTypes = ICON_TYPE;
        /**
         * We delay the removal of DOM so that animations can finish playing before the
         * DOM is removed. Removing the DOM, as hidding is not enough to stop elements
         * to be focused.
         */
        this.CLOSE_DELAY = 300;
        this.open$ = new BehaviorSubject(false);
        /**
         * Emits the open state that indicates whether the facet list should be rendered.
         * This is either done instantly, or after the user triggers this by using the trigger
         * button. This driven by the visiibility of the trigger, so that the CSS drives
         * the behaviour. This can differ per breakpoint.
         *
         * There's a configurable delay for the closed state, so that the DOM is not removed
         * before some CSS animations are done.
         */
        this.isOpen$ = this.breakpointService.breakpoint$.pipe(
        // deffer emitting a new value to the next micro-task to ensure that the `hasTrigger`
        // method represents the actual UI state.
        observeOn(asapScheduler), switchMap(() => (this.hasTrigger ? this.open$ : of(true))), delayWhen((launched) => interval(launched ? 0 : this.CLOSE_DELAY)));
        /**
         * Emits the active state that indicates whether the facet list is activated. Activation
         * is related to the css, so that a animation or transition can visualize opening/closing
         * the list (i.e. dialog).
         */
        this.isActive$ = this.open$.pipe(
        // deffer emitting a new value to the next micro-task to ensure the active class is
        //  applied after the DOM is created
        observeOn(asapScheduler));
    }
    launch() {
        this.open$.next(true);
    }
    close() {
        this.open$.next(false);
        this.trigger.nativeElement.focus();
    }
    /**
     * Indicates that the facet navigation should be open explicitely by a trigger.
     * This is fully controlled by CSS, where the trigger button can be hidden
     * (display:none) for certain screen sizes.
     */
    get hasTrigger() {
        return this.trigger.nativeElement.offsetParent !== null;
    }
};
ProductFacetNavigationComponent.ctorParameters = () => [
    { type: BreakpointService }
];
__decorate([
    ViewChild('trigger')
], ProductFacetNavigationComponent.prototype, "trigger", void 0);
ProductFacetNavigationComponent = __decorate([
    Component({
        selector: 'cx-product-facet-navigation',
        template: "<button\r\n  #trigger\r\n  class=\"btn btn-action btn-block dialog-trigger\"\r\n  (click)=\"launch()\"\r\n>\r\n  <cx-icon [type]=\"iconTypes.FILTER\"></cx-icon>\r\n  {{ 'productList.filterBy.label' | cxTranslate }}\r\n</button>\r\n\r\n<cx-active-facets></cx-active-facets>\r\n\r\n<cx-facet-list\r\n  *ngIf=\"isOpen$ | async\"\r\n  [isDialog]=\"hasTrigger\"\r\n  (closeList)=\"close()\"\r\n  [class.active]=\"isActive$ | async\"\r\n  [class.dialog]=\"hasTrigger\"\r\n></cx-facet-list>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], ProductFacetNavigationComponent);
export { ProductFacetNavigationComponent };
