// @spartacus/storefront/esm5/cms-components/myaccount/address-book/address-book.component.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Address, CheckoutDeliveryService, UserAddressService, } from '@spartacus/core';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var AddressBookComponentService = /** @class */ (function () {
    function AddressBookComponentService(userAddressService, checkoutDeliveryService) {
        this.userAddressService = userAddressService;
        this.checkoutDeliveryService = checkoutDeliveryService;
    }
    AddressBookComponentService.prototype.getAddresses = function () {
        return this.userAddressService.getAddresses();
    };
    AddressBookComponentService.prototype.getAddressesStateLoading = function () {
        return this.userAddressService.getAddressesLoading();
    };
    AddressBookComponentService.prototype.loadAddresses = function () {
        this.userAddressService.loadAddresses();
    };
    AddressBookComponentService.prototype.addUserAddress = function (address) {
        this.userAddressService.addUserAddress(address);
    };
    AddressBookComponentService.prototype.updateUserAddress = function (addressId, address) {
        this.userAddressService.updateUserAddress(addressId, address);
        this.checkoutDeliveryService.clearCheckoutDeliveryDetails();
    };
    AddressBookComponentService.ctorParameters = function () { return [
        { type: UserAddressService },
        { type: CheckoutDeliveryService }
    ]; };
    AddressBookComponentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AddressBookComponentService_Factory() { return new AddressBookComponentService(i0.ɵɵinject(i1.UserAddressService), i0.ɵɵinject(i1.CheckoutDeliveryService)); }, token: AddressBookComponentService, providedIn: "root" });
    AddressBookComponentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AddressBookComponentService);
    return AddressBookComponentService;
}());
export { AddressBookComponentService };
