// @spartacus/storefront/esm5/cms-components/product/product-variants/variant-style-selector/variant-style-selector.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { OccConfig, BaseOption, VariantQualifier, VariantOptionQualifier, Product, ProductService, ProductScope, RoutingService, } from '@spartacus/core';
import { filter, take } from 'rxjs/operators';
var VariantStyleSelectorComponent = /** @class */ (function () {
    function VariantStyleSelectorComponent(config, productService, routingService) {
        this.config = config;
        this.productService = productService;
        this.routingService = routingService;
        this.variantQualifier = VariantQualifier;
    }
    VariantStyleSelectorComponent.prototype.getVariantOptionValue = function (qualifiers) {
        var obj = qualifiers.find(function (q) { return q.qualifier === VariantQualifier.STYLE; });
        return obj ? obj.value : '';
    };
    VariantStyleSelectorComponent.prototype.getVariantThumbnailUrl = function (variantOptionQualifiers) {
        var qualifier = variantOptionQualifiers.find(function (item) { return item.image; });
        return qualifier
            ? "" + this.config.backend.occ.baseUrl + qualifier.image.url
            : '';
    };
    VariantStyleSelectorComponent.prototype.changeStyle = function (code) {
        var _this = this;
        if (code) {
            this.productService
                .get(code, ProductScope.LIST)
                .pipe(
            // below call might looks redundant but in fact this data is going to be loaded anyways
            // we're just calling it earlier and storing
            filter(Boolean), take(1))
                .subscribe(function (product) {
                _this.routingService.go({
                    cxRoute: 'product',
                    params: product,
                });
            });
        }
        return null;
    };
    VariantStyleSelectorComponent.ctorParameters = function () { return [
        { type: OccConfig },
        { type: ProductService },
        { type: RoutingService }
    ]; };
    __decorate([
        Input()
    ], VariantStyleSelectorComponent.prototype, "variants", void 0);
    VariantStyleSelectorComponent = __decorate([
        Component({
            selector: 'cx-variant-style-selector',
            template: "<ng-container>\n  <div class=\"variant-selector\">\n    <div *ngIf=\"variants.selected\" class=\"variant-name\">\n      {{ 'variant.style' | cxTranslate }}:\n      <span class=\"style-name\">{{\n        getVariantOptionValue(variants?.selected.variantOptionQualifiers)\n      }}</span>\n    </div>\n    <ul class=\"variant-list\">\n      <li\n        *ngFor=\"let v of variants?.options\"\n        [ngClass]=\"{\n          'selected-variant': v.code === variants?.selected.code\n        }\"\n      >\n        <img\n          (click)=\"changeStyle(v.code)\"\n          src=\"{{ getVariantThumbnailUrl(v.variantOptionQualifiers) }}\"\n          title=\"{{ getVariantOptionValue(v.variantOptionQualifiers) }}\"\n          alt=\"{{ getVariantOptionValue(v.variantOptionQualifiers) }}\"\n        />\n      </li>\n    </ul>\n  </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], VariantStyleSelectorComponent);
    return VariantStyleSelectorComponent;
}());
export { VariantStyleSelectorComponent };
