// @spartacus/storefront/esm5/cms-components/cart/cart-coupon/cart-coupon.component.js
import { __decorate, __read } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ActiveCartService, Cart, CartVoucherService, CustomerCoupon, CustomerCouponSearchResult, CustomerCouponService, } from '@spartacus/core';
import { combineLatest, Subscription } from 'rxjs';
import { map, tap } from 'rxjs/operators';
var CartCouponComponent = /** @class */ (function () {
    function CartCouponComponent(cartVoucherService, formBuilder, customerCouponService, activeCartService) {
        this.cartVoucherService = cartVoucherService;
        this.formBuilder = formBuilder;
        this.customerCouponService = customerCouponService;
        this.activeCartService = activeCartService;
        this.MAX_CUSTOMER_COUPON_PAGE = 100;
        this.ignoreCloseEvent = false;
        this.subscription = new Subscription();
        this.couponBoxIsActive = false;
    }
    CartCouponComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (this.customerCouponService) {
            this.customerCouponService.loadCustomerCoupons(this.MAX_CUSTOMER_COUPON_PAGE);
        }
        this.cart$ = combineLatest([
            this.activeCartService.getActive(),
            this.activeCartService.getActiveCartId(),
            this.customerCouponService.getCustomerCoupons(this.MAX_CUSTOMER_COUPON_PAGE),
        ]).pipe(tap(function (_a) {
            var _b = __read(_a, 3), cart = _b[0], activeCardId = _b[1], customerCoupons = _b[2];
            _this.cartId = activeCardId;
            _this.getApplicableCustomerCoupons(cart, customerCoupons.coupons);
        }), map(function (_a) {
            var _b = __read(_a, 1), cart = _b[0];
            return cart;
        }));
        this.cartIsLoading$ = this.activeCartService
            .isStable()
            .pipe(map(function (loaded) { return !loaded; }));
        this.cartVoucherService.resetAddVoucherProcessingState();
        this.couponForm = this.formBuilder.group({
            couponCode: ['', [Validators.required]],
        });
        // TODO(#7241): Replace process subscriptions with event listeners and drop process for ADD_VOUCHER
        this.subscription.add(this.cartVoucherService
            .getAddVoucherResultSuccess()
            .subscribe(function (success) {
            _this.onSuccess(success);
        }));
        // TODO(#7241): Replace process subscriptions with event listeners and drop process for ADD_VOUCHER
        this.subscription.add(this.cartVoucherService.getAddVoucherResultError().subscribe(function (error) {
            _this.onError(error);
        }));
    };
    CartCouponComponent.prototype.onError = function (error) {
        if (error) {
            this.customerCouponService.loadCustomerCoupons(this.MAX_CUSTOMER_COUPON_PAGE);
            this.cartVoucherService.resetAddVoucherProcessingState();
        }
    };
    CartCouponComponent.prototype.onSuccess = function (success) {
        if (success) {
            this.couponForm.reset();
            this.cartVoucherService.resetAddVoucherProcessingState();
        }
    };
    CartCouponComponent.prototype.getApplicableCustomerCoupons = function (cart, coupons) {
        var _this = this;
        this.applicableCoupons = coupons || [];
        if (cart.appliedVouchers) {
            cart.appliedVouchers.forEach(function (appliedVoucher) {
                _this.applicableCoupons = _this.applicableCoupons.filter(function (coupon) { return coupon.couponId !== appliedVoucher.code; });
            });
        }
    };
    CartCouponComponent.prototype.applyVoucher = function () {
        if (this.couponForm.valid) {
            this.cartVoucherService.addVoucher(this.couponForm.value.couponCode, this.cartId);
        }
        else {
            this.couponForm.markAllAsTouched();
        }
    };
    CartCouponComponent.prototype.applyCustomerCoupon = function (couponId) {
        this.cartVoucherService.addVoucher(couponId, this.cartId);
        this.couponBoxIsActive = false;
    };
    CartCouponComponent.prototype.close = function (event) {
        if (!this.ignoreCloseEvent) {
            this.couponBoxIsActive = false;
            if (event && event.target) {
                event.target.blur();
            }
        }
        this.ignoreCloseEvent = false;
    };
    CartCouponComponent.prototype.disableClose = function () {
        this.ignoreCloseEvent = true;
    };
    CartCouponComponent.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        this.cartVoucherService.resetAddVoucherProcessingState();
    };
    CartCouponComponent.ctorParameters = function () { return [
        { type: CartVoucherService },
        { type: FormBuilder },
        { type: CustomerCouponService },
        { type: ActiveCartService }
    ]; };
    CartCouponComponent = __decorate([
        Component({
            selector: 'cx-cart-coupon',
            template: "<ng-container *ngIf=\"cart$ | async as cart\">\r\n  <div class=\"cx-cart-coupon-title\">\r\n    {{ 'voucher.coupon' | cxTranslate }}\r\n  </div>\r\n  <div class=\"form-group\">\r\n    <form (ngSubmit)=\"applyVoucher()\" [formGroup]=\"couponForm\">\r\n      <div class=\"cx-cart-coupon-container\">\r\n        <input\r\n          type=\"text\"\r\n          class=\"form-control input-coupon-code\"\r\n          formControlName=\"couponCode\"\r\n          placeholder=\"{{ 'voucher.placeholder' | cxTranslate }}\"\r\n        />\r\n        <button\r\n          class=\"btn btn-block btn-action apply-coupon-button\"\r\n          type=\"submit\"\r\n          [disabled]=\"cartIsLoading$ | async\"\r\n          [class.disabled]=\"cartIsLoading$ | async\"\r\n        >\r\n          {{ 'voucher.apply' | cxTranslate }}\r\n        </button>\r\n        <cx-form-errors\r\n          [control]=\"couponForm.get('couponCode')\"\r\n        ></cx-form-errors>\r\n      </div>\r\n    </form>\r\n  </div>\r\n\r\n  <cx-applied-coupons\r\n    [vouchers]=\"cart.appliedVouchers\"\r\n    [cartIsLoading]=\"cartIsLoading$ | async\"\r\n    [isReadOnly]=\"false\"\r\n  >\r\n  </cx-applied-coupons>\r\n\r\n  <ng-container *ngIf=\"applicableCoupons && applicableCoupons.length > 0\">\r\n    <div class=\"cx-available-coupon\">\r\n      <div class=\"title cx-cart-coupon-title\">\r\n        {{ 'voucher.availableCoupons' | cxTranslate }}\r\n      </div>\r\n      <div class=\"message\">\r\n        {{ 'voucher.availableCouponsLabel' | cxTranslate }}\r\n      </div>\r\n      <div class=\"scroll\">\r\n        <div class=\"coupons card\" *ngFor=\"let coupon of applicableCoupons\">\r\n          <button\r\n            (click)=\"applyCustomerCoupon(coupon.couponId)\"\r\n            class=\"coupon-id link\"\r\n            [disabled]=\"cartIsLoading$ | async\"\r\n            [class.disabled]=\"cartIsLoading$ | async\"\r\n          >\r\n            {{ coupon.couponId }}\r\n          </button>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </ng-container>\r\n</ng-container>\r\n"
        })
    ], CartCouponComponent);
    return CartCouponComponent;
}());
export { CartCouponComponent };
