// @spartacus/storefront/esm5/layout/a11y/keyboard-focus/trap/trap-focus.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { TabFocusService } from '../tab/tab-focus.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/select-focus.util";
var TrapFocusService = /** @class */ (function (_super) {
    __extends(TrapFocusService, _super);
    function TrapFocusService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Indicates whether any of the child elements of the host are focusable.
     *
     * @param host `HTMLElement` that is used to query the focusable elements.
     */
    TrapFocusService.prototype.hasFocusableChildren = function (host) {
        return this.findFocusable(host).length > 0;
    };
    /**
     * Focus the next or previous element of all available focusable elements.
     * The focus is _trapped_ in case there's no next or previous available element.
     * The focus will automatically move the start or end of the list.
     */
    TrapFocusService.prototype.moveFocus = function (host, config, increment, event) {
        var focusable = this.findFocusable(host);
        var index = focusable.findIndex(function (v) { return v === event.target; }) + increment;
        var shouldMoveFocus = (index >= 0 && index < focusable.length) ||
            (index < 0 && this.getTrapStart(config.trap)) ||
            (index >= focusable.length && this.getTrapEnd(config.trap));
        if (shouldMoveFocus) {
            if (index >= focusable.length) {
                index = 0;
            }
            if (index < 0) {
                index = focusable.length - 1;
            }
            event.preventDefault();
            event.stopPropagation();
            var el = focusable[index];
            el.focus();
        }
    };
    TrapFocusService.prototype.getTrapStart = function (trap) {
        return trap === true || trap === 'start';
    };
    TrapFocusService.prototype.getTrapEnd = function (trap) {
        return trap === true || trap === 'end';
    };
    TrapFocusService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TrapFocusService_Factory() { return new TrapFocusService(i0.ɵɵinject(i1.SelectFocusUtility)); }, token: TrapFocusService, providedIn: "root" });
    TrapFocusService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], TrapFocusService);
    return TrapFocusService;
}(TabFocusService));
export { TrapFocusService };
