// @spartacus/storefront/esm2015/cms-components/storefinder/components/store-finder-grid/store-finder-grid.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { StoreFinderService, GeoPoint } from '@spartacus/core';
let StoreFinderGridComponent = class StoreFinderGridComponent {
    constructor(storeFinderService, route) {
        this.storeFinderService = storeFinderService;
        this.route = route;
    }
    ngOnInit() {
        this.isLoading$ = this.storeFinderService.getViewAllStoresLoading();
        this.locations$ = this.storeFinderService.getViewAllStoresEntities();
        this.defaultLocation = {};
        if (this.route.snapshot.params.country) {
            this.storeFinderService.findStoresAction('', {
                pageSize: -1,
            }, undefined, this.route.snapshot.params.country);
        }
    }
    ngOnDestroy() { }
};
StoreFinderGridComponent.ctorParameters = () => [
    { type: StoreFinderService },
    { type: ActivatedRoute }
];
StoreFinderGridComponent = __decorate([
    Component({
        selector: 'cx-store-finder-grid',
        template: "<ng-container\r\n  *ngIf=\"\r\n    !(isLoading$ | async) && (locations$ | async) as locations;\r\n    else loading\r\n  \"\r\n>\r\n  <div class=\"container\">\r\n    <div class=\"row\">\r\n      <div\r\n        class=\"col-sm-4 col-md-4 col-lg-3 item\"\r\n        *ngFor=\"let location of locations?.stores\"\r\n      >\r\n        <cx-store-finder-list-item\r\n          [location]=\"location\"\r\n        ></cx-store-finder-list-item>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</ng-container>\r\n\r\n<ng-template #loading>\r\n  <div class=\"cx-spinner\"><cx-spinner></cx-spinner></div>\r\n</ng-template>\r\n"
    })
], StoreFinderGridComponent);
export { StoreFinderGridComponent };
