// @spartacus/storefront/esm2015/recipes/storefront.module.js
var StorefrontModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { ExternalRoutesModule, OccModule, PersonalizationModule, provideConfig, SiteContextModule, SmartEditModule, } from '@spartacus/core';
import { AsmModule } from '../cms-components/asm/asm.module';
import { ProductDetailsPageModule } from '../cms-pages/product-details-page/product-details-page.module';
import { ProductListingPageModule } from '../cms-pages/product-listing-page/product-listing-page.module';
import { MainModule } from '../layout/main/main.module';
import { StorefrontFoundationModule } from './storefront-foundation.module';
let StorefrontModule = StorefrontModule_1 = class StorefrontModule {
    static withConfig(config) {
        return {
            ngModule: StorefrontModule_1,
            providers: [provideConfig(config)],
        };
    }
};
StorefrontModule = StorefrontModule_1 = __decorate([
    NgModule({
        imports: [
            RouterModule.forRoot([], {
                scrollPositionRestoration: 'enabled',
                anchorScrolling: 'enabled',
            }),
            StoreModule.forRoot({}),
            EffectsModule.forRoot([]),
            // ASM module must be imported before the `AuthModule (which is imported in `StorefrontFoundationModule`)
            // since we might have conflicting interceptor logic. See #5461.
            AsmModule,
            StorefrontFoundationModule,
            MainModule,
            SiteContextModule.forRoot(),
            SmartEditModule.forRoot(),
            PersonalizationModule.forRoot(),
            // opt-in explicitly
            OccModule.forRoot(),
            ProductDetailsPageModule,
            ProductListingPageModule,
            ExternalRoutesModule.forRoot(),
        ],
        exports: [MainModule, StorefrontFoundationModule],
    })
], StorefrontModule);
export { StorefrontModule };
