// @spartacus/storefront/esm5/layout/launch-dialog/launch-dialog.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { Config } from '@spartacus/core';
import { LayoutConfig } from '../config/layout-config';
import { InlineRenderStrategy, LaunchRenderStrategy, OutletRenderStrategy, RoutingRenderStrategy, } from './services/index';
var LaunchDialogModule = /** @class */ (function () {
    function LaunchDialogModule() {
    }
    LaunchDialogModule_1 = LaunchDialogModule;
    LaunchDialogModule.forRoot = function () {
        return {
            ngModule: LaunchDialogModule_1,
            providers: [{ provide: LayoutConfig, useExisting: Config }],
        };
    };
    var LaunchDialogModule_1;
    LaunchDialogModule = LaunchDialogModule_1 = __decorate([
        NgModule({
            providers: [
                {
                    provide: LaunchRenderStrategy,
                    useExisting: OutletRenderStrategy,
                    multi: true,
                },
                {
                    provide: LaunchRenderStrategy,
                    useExisting: InlineRenderStrategy,
                    multi: true,
                },
                {
                    provide: LaunchRenderStrategy,
                    useExisting: RoutingRenderStrategy,
                    multi: true,
                },
            ],
        })
    ], LaunchDialogModule);
    return LaunchDialogModule;
}());
export { LaunchDialogModule };
