// @spartacus/core/esm5/src/anonymous-consents/store/anonymous-consents-store.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { provideDefaultConfigFactory } from '../../config/config.module';
import { StorageSyncType } from '../../state/index';
import { StateModule } from '../../state/state.module';
import { ANONYMOUS_CONSENTS_STORE_FEATURE } from './anonymous-consents-state';
import { effects } from './effects/index';
import { metaReducers, reducerProvider, reducerToken } from './reducers/index';
export function anonymousConsentsStoreConfigFactory() {
    var _a;
    var config = {
        state: {
            storageSync: {
                keys: (_a = {},
                    _a[ANONYMOUS_CONSENTS_STORE_FEATURE] = StorageSyncType.LOCAL_STORAGE,
                    _a),
            },
        },
    };
    return config;
}
var AnonymousConsentsStoreModule = /** @class */ (function () {
    function AnonymousConsentsStoreModule() {
    }
    AnonymousConsentsStoreModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                StateModule,
                StoreModule.forFeature(ANONYMOUS_CONSENTS_STORE_FEATURE, reducerToken, {
                    metaReducers: metaReducers,
                }),
                EffectsModule.forFeature(effects),
            ],
            providers: [
                provideDefaultConfigFactory(anonymousConsentsStoreConfigFactory),
                reducerProvider,
            ],
        })
    ], AnonymousConsentsStoreModule);
    return AnonymousConsentsStoreModule;
}());
export { AnonymousConsentsStoreModule };
