// @spartacus/storefront/esm2015/cms-components/user/checkout-login/checkout-login.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ActiveCartService, AuthRedirectService } from '@spartacus/core';
import { CustomFormValidators } from '../../../shared/utils/validators/custom-form-validators';
let CheckoutLoginComponent = class CheckoutLoginComponent {
    constructor(formBuilder, authRedirectService, activeCartService) {
        this.formBuilder = formBuilder;
        this.authRedirectService = authRedirectService;
        this.activeCartService = activeCartService;
        this.checkoutLoginForm = this.formBuilder.group({
            email: ['', [Validators.required, CustomFormValidators.emailValidator]],
            emailConfirmation: ['', [Validators.required]],
        }, {
            validators: CustomFormValidators.emailsMustMatch('email', 'emailConfirmation'),
        });
    }
    onSubmit() {
        if (this.checkoutLoginForm.valid) {
            const email = this.checkoutLoginForm.get('email').value;
            this.activeCartService.addEmail(email);
            if (!this.sub) {
                this.sub = this.activeCartService.getAssignedUser().subscribe(() => {
                    if (this.activeCartService.isGuestCart()) {
                        this.authRedirectService.redirect();
                    }
                });
            }
        }
        else {
            this.checkoutLoginForm.markAllAsTouched();
        }
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
};
CheckoutLoginComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AuthRedirectService },
    { type: ActiveCartService }
];
CheckoutLoginComponent = __decorate([
    Component({
        selector: 'cx-checkout-login',
        template: "<form (ngSubmit)=\"onSubmit()\" [formGroup]=\"checkoutLoginForm\">\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'checkoutLogin.emailAddress.label' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        type=\"email\"\r\n        name=\"email\"\r\n        class=\"form-control\"\r\n        formControlName=\"email\"\r\n        placeholder=\"{{\r\n          'checkoutLogin.emailAddress.placeholder' | cxTranslate\r\n        }}\"\r\n      />\r\n      <cx-form-errors\r\n        [control]=\"checkoutLoginForm.get('email')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'checkoutLogin.confirmEmail.label' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        type=\"email\"\r\n        name=\"emailConfirmation\"\r\n        class=\"form-control\"\r\n        formControlName=\"emailConfirmation\"\r\n        placeholder=\"{{\r\n          'checkoutLogin.confirmEmail.placeholder' | cxTranslate\r\n        }}\"\r\n      />\r\n      <cx-form-errors\r\n        [control]=\"checkoutLoginForm.get('emailConfirmation')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <button type=\"submit\" class=\"btn btn-block btn-primary\">\r\n    {{ 'checkoutLogin.continue' | cxTranslate }}\r\n  </button>\r\n</form>\r\n"
    })
], CheckoutLoginComponent);
export { CheckoutLoginComponent };
