// @spartacus/core/esm5/src/routing/services/url-matcher.service.js
import { __decorate } from "tslib";
import { Injectable, isDevMode } from '@angular/core';
import { GlobService } from '../../util/glob.service';
import * as i0 from "@angular/core";
import * as i1 from "../../util/glob.service";
var UrlMatcherService = /** @class */ (function () {
    function UrlMatcherService(globService) {
        this.globService = globService;
    }
    /**
     * Returns a matcher that is always fails
     */
    UrlMatcherService.prototype.getFalsy = function () {
        return function falsyUrlMatcher() {
            return null;
        };
    };
    /**
     * Returns a matcher for given list of paths
     */
    UrlMatcherService.prototype.getFromPaths = function (paths) {
        var _this = this;
        var matchers = paths.map(function (path) { return _this.getFromPath(path); });
        var matcher = this.getCombined(matchers);
        if (isDevMode()) {
            matcher['_paths'] = paths; // property added for easier debugging of routes
        }
        return matcher;
    };
    /**
     * Returns a matcher that combines the given matchers
     * */
    UrlMatcherService.prototype.getCombined = function (matchers) {
        var matcher = function combinedUrlMatchers(segments, segmentGroup, route) {
            for (var i = 0; i < matchers.length; i++) {
                var result = matchers[i](segments, segmentGroup, route);
                if (result) {
                    return result;
                }
            }
            return null;
        };
        if (isDevMode()) {
            matcher['_matchers'] = matchers; // property added for easier debugging of routes
        }
        return matcher;
    };
    /**
     * Similar to Angular's defaultUrlMatcher. Differences:
     * - the `path` comes from function's argument, not from `route.path`
     * - the empty path `''` is handled here, but in Angular is handled one level higher in the match() function
     */
    UrlMatcherService.prototype.getFromPath = function (path) {
        if (path === void 0) { path = ''; }
        var matcher = function pathUrlMatcher(segments, segmentGroup, route) {
            /**
             * @license
             * The MIT License
             * Copyright (c) 2010-2019 Google LLC. http://angular.io/license
             *
             * See:
             * - https://github.com/angular/angular/blob/6f5f481fdae03f1d8db36284b64c7b82d9519d85/packages/router/src/shared.ts#L121
             */
            // use function's argument, not the `route.path`
            if (path === '') {
                if (route.pathMatch === 'full' &&
                    (segmentGroup.hasChildren() || segments.length > 0)) {
                    return null;
                }
                return { consumed: [], posParams: {} };
            }
            var parts = path.split('/'); // use function's argument, not the `route.path`
            if (parts.length > segments.length) {
                // The actual URL is shorter than the config, no match
                return null;
            }
            if (route.pathMatch === 'full' &&
                (segmentGroup.hasChildren() || parts.length < segments.length)) {
                // The config is longer than the actual URL but we are looking for a full match, return null
                return null;
            }
            var posParams = {};
            // Check each config part against the actual URL
            for (var index = 0; index < parts.length; index++) {
                var part = parts[index];
                var segment = segments[index];
                var isParameter = part.startsWith(':');
                if (isParameter) {
                    posParams[part.substring(1)] = segment;
                }
                else if (part !== segment.path) {
                    // The actual URL part does not match the config, no match
                    return null;
                }
            }
            return { consumed: segments.slice(0, parts.length), posParams: posParams };
        };
        if (isDevMode()) {
            matcher['_path'] = path; // property added for easier debugging of routes
        }
        return matcher;
    };
    /**
     * Returns URL matcher that accepts almost everything (like `**` route), but not paths accepted by the given matcher
     */
    UrlMatcherService.prototype.getOpposite = function (originalMatcher) {
        var matcher = function oppositeUrlMatcher(segments, group, route) {
            return originalMatcher(segments, group, route)
                ? null
                : { consumed: segments, posParams: {} };
        };
        if (isDevMode()) {
            matcher['_originalMatcher'] = originalMatcher; // property added for easier debugging of routes
        }
        return matcher;
    };
    /**
     * Returns URL matcher for the given list of glob-like patterns. Each pattern must start with `/` or `!/`.
     */
    UrlMatcherService.prototype.getFromGlob = function (globPatterns) {
        var globValidator = this.globService.getValidator(globPatterns);
        var matcher = function globUrlMatcher(segments) {
            var fullPath = "/" + segments.map(function (s) { return s.path; }).join('/');
            return globValidator(fullPath)
                ? { consumed: segments, posParams: {} }
                : null;
        };
        if (isDevMode()) {
            matcher['_globPatterns'] = globPatterns; // property added for easier debugging of routes
        }
        return matcher;
    };
    UrlMatcherService.ctorParameters = function () { return [
        { type: GlobService }
    ]; };
    UrlMatcherService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UrlMatcherService_Factory() { return new UrlMatcherService(i0.ɵɵinject(i1.GlobService)); }, token: UrlMatcherService, providedIn: "root" });
    UrlMatcherService = __decorate([
        Injectable({ providedIn: 'root' })
    ], UrlMatcherService);
    return UrlMatcherService;
}());
export { UrlMatcherService };
