// @spartacus/storefront/esm5/cms-components/myaccount/order/order-details/order-details.module.js
import { __decorate, __read, __spread } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, FeaturesConfigModule, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ConsignmentTrackingComponent } from '../../../../cms-components/myaccount/order/order-details/order-detail-items/consignment-tracking/consignment-tracking.component';
import { CmsPageGuard } from '../../../../cms-structure/guards/cms-page.guard';
import { PageLayoutComponent } from '../../../../cms-structure/page/page-layout/page-layout.component';
import { CardModule } from '../../../../shared/components/card/card.module';
import { SpinnerModule } from '../../../../shared/components/spinner/spinner.module';
import { CartSharedModule } from '../../../cart/cart-shared/cart-shared.module';
import { PromotionsModule } from '../../../checkout/components/promotions/promotions.module';
import { OrderDetailActionsComponent } from './order-detail-actions/order-detail-actions.component';
import { OrderDetailHeadlineComponent } from './order-detail-headline/order-detail-headline.component';
import { TrackingEventsComponent } from './order-detail-items/consignment-tracking/tracking-events/tracking-events.component';
import { OrderConsignedEntriesComponent } from './order-detail-items/order-consigned-entries/order-consigned-entries.component';
import { OrderDetailItemsComponent } from './order-detail-items/order-detail-items.component';
import { OrderDetailShippingComponent } from './order-detail-shipping/order-detail-shipping.component';
import { OrderDetailTotalsComponent } from './order-detail-totals/order-detail-totals.component';
import { OrderDetailsService } from './order-details.service';
var moduleComponents = [
    OrderDetailActionsComponent,
    OrderDetailHeadlineComponent,
    OrderDetailItemsComponent,
    OrderDetailTotalsComponent,
    OrderDetailShippingComponent,
    TrackingEventsComponent,
    ConsignmentTrackingComponent,
    OrderConsignedEntriesComponent,
];
var ɵ0 = { pageLabel: 'order', cxRoute: 'orderGuest' }, ɵ1 = { cxRoute: 'orderDetails' };
var OrderDetailsModule = /** @class */ (function () {
    function OrderDetailsModule() {
    }
    OrderDetailsModule = __decorate([
        NgModule({
            imports: [
                CartSharedModule,
                CardModule,
                CommonModule,
                I18nModule,
                FeaturesConfigModule,
                PromotionsModule,
                UrlModule,
                RouterModule.forChild([
                    {
                        path: null,
                        canActivate: [AuthGuard, CmsPageGuard],
                        component: PageLayoutComponent,
                        data: ɵ0,
                    },
                    {
                        path: null,
                        canActivate: [AuthGuard, CmsPageGuard],
                        component: PageLayoutComponent,
                        data: ɵ1,
                    },
                ]),
                SpinnerModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        AccountOrderDetailsActionsComponent: {
                            component: OrderDetailActionsComponent,
                        },
                        AccountOrderDetailsHeadlineComponent: {
                            component: OrderDetailHeadlineComponent,
                        },
                        AccountOrderDetailsItemsComponent: {
                            component: OrderDetailItemsComponent,
                        },
                        AccountOrderDetailsTotalsComponent: {
                            component: OrderDetailTotalsComponent,
                        },
                        AccountOrderDetailsShippingComponent: {
                            component: OrderDetailShippingComponent,
                        },
                    },
                    features: {
                        consignmentTracking: '1.2',
                    },
                }),
                OrderDetailsService,
            ],
            declarations: __spread(moduleComponents),
            exports: __spread(moduleComponents),
            entryComponents: __spread(moduleComponents),
        })
    ], OrderDetailsModule);
    return OrderDetailsModule;
}());
export { OrderDetailsModule };
export { ɵ0, ɵ1 };
