// @spartacus/storefront/esm5/layout/launch-dialog/services/launch-dialog.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable, isDevMode, } from '@angular/core';
import { resolveApplicable } from '@spartacus/core';
import { BehaviorSubject } from 'rxjs';
import { LayoutConfig } from '../../config/layout-config';
import { LaunchRenderStrategy } from './launch-render.strategy';
import * as i0 from "@angular/core";
import * as i1 from "./launch-render.strategy";
import * as i2 from "../../config/layout-config";
var LaunchDialogService = /** @class */ (function () {
    function LaunchDialogService(renderStrategies, layoutConfig) {
        this.renderStrategies = renderStrategies;
        this.layoutConfig = layoutConfig;
        this._dialogClose = new BehaviorSubject(undefined);
        this.renderStrategies = this.renderStrategies || [];
    }
    /**
     * Render the element based on the strategy from the launch configuration
     *
     * @param caller LAUNCH_CALLER
     * @param vcr View Container Ref of the container for inline rendering
     */
    LaunchDialogService.prototype.launch = function (caller, vcr) {
        var config = this.findConfiguration(caller);
        if (config) {
            var renderer = this.getStrategy(config);
            // Render if the strategy exists
            if (renderer) {
                this._dialogClose.next(undefined);
                return renderer.render(config, caller, vcr);
            }
        }
        else if (isDevMode()) {
            console.warn('No configuration provided for caller ' + caller);
        }
    };
    /**
     * Util method to remove element from rendered elements list
     *
     * @param caller LAUNCH_CALLER
     */
    LaunchDialogService.prototype.clear = function (caller) {
        var config = this.findConfiguration(caller);
        var renderer = this.getStrategy(config);
        // Render if the strategy exists
        if (renderer) {
            renderer.remove(caller, config);
        }
    };
    Object.defineProperty(LaunchDialogService.prototype, "dialogClose", {
        get: function () {
            return this._dialogClose.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    LaunchDialogService.prototype.closeDialog = function (reason) {
        this._dialogClose.next(reason);
    };
    /**
     * Returns the configuration for the caller
     *
     * @param caller LAUNCH_CALLER
     */
    LaunchDialogService.prototype.findConfiguration = function (caller) {
        var _a;
        if ((_a = this.layoutConfig) === null || _a === void 0 ? void 0 : _a.launch) {
            return this.layoutConfig.launch[caller];
        }
        return undefined;
    };
    /**
     * Returns the render strategy based on the configuration
     *
     * @param config Configuration for launch
     */
    LaunchDialogService.prototype.getStrategy = function (config) {
        return resolveApplicable(this.renderStrategies, [config]);
    };
    LaunchDialogService.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Inject, args: [LaunchRenderStrategy,] }] },
        { type: LayoutConfig }
    ]; };
    LaunchDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function LaunchDialogService_Factory() { return new LaunchDialogService(i0.ɵɵinject(i1.LaunchRenderStrategy), i0.ɵɵinject(i2.LayoutConfig)); }, token: LaunchDialogService, providedIn: "root" });
    LaunchDialogService = __decorate([
        Injectable({ providedIn: 'root' }),
        __param(0, Inject(LaunchRenderStrategy))
    ], LaunchDialogService);
    return LaunchDialogService;
}());
export { LaunchDialogService };
