// @spartacus/storefront/esm5/layout/launch-dialog/services/launch-render.strategy.js
import { __decorate, __param, __values } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Inject, } from '@angular/core';
import { DIALOG_TYPE, } from '../config';
var LaunchRenderStrategy = /** @class */ (function () {
    function LaunchRenderStrategy(document, rendererFactory) {
        this.document = document;
        this.rendererFactory = rendererFactory;
        // List of called references; only used for rendered elements
        this.renderedCallers = [];
        /**
         * Classes to apply to the component when the dialog is a DIALOG
         */
        this.dialogClasses = ['d-block', 'fade', 'modal', 'show'];
        /**
         * Classes to apply to the component when the dialog is a POPOVER
         */
        this.popoverClasses = ['cx-dialog-popover'];
        /**
         * Classes to apply to the component when the dialog is a SIDEBAR_END
         */
        this.sidebarEndClasses = ['cx-sidebar-end'];
        /**
         * Classes to apply to the component when the dialog is a SIDEBAR_START
         */
        this.sidebarStartClasses = ['cx-sidebar-start'];
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * Determines if element should render
     *
     * @param caller
     * @param config
     */
    LaunchRenderStrategy.prototype.shouldRender = function (caller, config) {
        return (Boolean(config.component) &&
            (this.renderedCallers.some(function (el) { return el.caller === caller; })
                ? !!config.multi
                : true));
    };
    LaunchRenderStrategy.prototype.applyClasses = function (component, dialogType) {
        var e_1, _a;
        var classes = [];
        // TODO: make classes configurable
        switch (dialogType) {
            case DIALOG_TYPE.DIALOG:
                classes = this.dialogClasses;
                this.renderer.addClass(this.document.body, 'modal-open');
                break;
            case DIALOG_TYPE.POPOVER:
                classes = this.popoverClasses;
                break;
            case DIALOG_TYPE.SIDEBAR_END:
                classes = this.sidebarEndClasses;
                break;
            case DIALOG_TYPE.SIDEBAR_START:
                classes = this.sidebarStartClasses;
                break;
        }
        try {
            for (var classes_1 = __values(classes), classes_1_1 = classes_1.next(); !classes_1_1.done; classes_1_1 = classes_1.next()) {
                var newClass = classes_1_1.value;
                this.renderer.addClass(component.location.nativeElement, newClass);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (classes_1_1 && !classes_1_1.done && (_a = classes_1.return)) _a.call(classes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * Method to call when rendered element is destroyed
     * The element will be removed from the list of rendered elements
     *
     * @param caller
     * @param _config optional parameters used in children strategies
     */
    LaunchRenderStrategy.prototype.remove = function (caller, config) {
        var _a;
        this.renderedCallers = this.renderedCallers.filter(function (el) { return el.caller !== caller; });
        if (((_a = config) === null || _a === void 0 ? void 0 : _a.dialogType) === DIALOG_TYPE.DIALOG) {
            this.renderer.removeClass(this.document.body, 'modal-open');
        }
    };
    LaunchRenderStrategy.prototype.getPriority = function () {
        return -10 /* LOW */; // low, as it's a default matcher
    };
    LaunchRenderStrategy = __decorate([
        __param(0, Inject(DOCUMENT))
    ], LaunchRenderStrategy);
    return LaunchRenderStrategy;
}());
export { LaunchRenderStrategy };
