// @spartacus/core/esm5/src/occ/adapters/product/converters/occ-product-search-page-normalizer.js
import { __assign, __decorate, __rest } from "tslib";
import { Injectable } from '@angular/core';
import { PRODUCT_NORMALIZER } from '../../../../product/connectors/product/converters';
import { Converter, ConverterService, } from '../../../../util/converter.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../../util/converter.service";
var OccProductSearchPageNormalizer = /** @class */ (function () {
    function OccProductSearchPageNormalizer(converterService) {
        this.converterService = converterService;
        /**
         * Specifies the minimal number of top values in case
         * non have been setup by the business.
         */
        this.DEFAULT_TOP_VALUES = 6;
    }
    OccProductSearchPageNormalizer.prototype.convert = function (source, target) {
        var _this = this;
        if (target === void 0) { target = {}; }
        target = __assign(__assign({}, target), source);
        this.normalizeFacets(target);
        if (source.products) {
            target.products = source.products.map(function (product) {
                return _this.converterService.convert(product, PRODUCT_NORMALIZER);
            });
        }
        return target;
    };
    OccProductSearchPageNormalizer.prototype.normalizeFacets = function (target) {
        this.normalizeFacetValues(target);
        this.normalizeUselessFacets(target);
    };
    /**
     * The (current) backend returns facets with values that do not contribute
     * to the facet navigation much, as the number in the result list will not get
     * behaviour, see https://jira.hybris.com/browse/CS-427.
     *
     * As long as this is not in place, we manually filter the facet from the list;
     * any facet that does not have a count < the total results will be dropped from
     * the facets.
     */
    OccProductSearchPageNormalizer.prototype.normalizeUselessFacets = function (target) {
        target.facets = target.facets.filter(function (facet) {
            return (!target.pagination ||
                !target.pagination.totalResults ||
                ((!facet.hasOwnProperty('visible') || facet.visible) &&
                    facet.values &&
                    facet.values.find(function (value) {
                        return (value.selected || value.count < target.pagination.totalResults);
                    })));
        });
    };
    /*
     * In case there are so-called `topValues` given for the facet values,
     * values are obsolete.
     *
     * `topValues` is a feature in Adaptive Search which can limit a large
     * amount of facet values to a small set (5 by default). As long as the backend
     * provides all facet values AND topValues, we normalize the data to not bother
     * the UI with this specific feature.
     */
    OccProductSearchPageNormalizer.prototype.normalizeFacetValues = function (target) {
        var _this = this;
        if (target.facets) {
            target.facets = target.facets.map(function (facetSource) {
                var topValues = facetSource.topValues, facetTarget = __rest(facetSource, ["topValues"]);
                facetTarget.topValueCount = topValues
                    ? topValues.length
                    : _this.DEFAULT_TOP_VALUES;
                return facetTarget;
            });
        }
    };
    OccProductSearchPageNormalizer.ctorParameters = function () { return [
        { type: ConverterService }
    ]; };
    OccProductSearchPageNormalizer.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccProductSearchPageNormalizer_Factory() { return new OccProductSearchPageNormalizer(i0.ɵɵinject(i1.ConverterService)); }, token: OccProductSearchPageNormalizer, providedIn: "root" });
    OccProductSearchPageNormalizer = __decorate([
        Injectable({ providedIn: 'root' })
    ], OccProductSearchPageNormalizer);
    return OccProductSearchPageNormalizer;
}());
export { OccProductSearchPageNormalizer };
