// @spartacus/storefront/esm5/cms-components/product/product-images/product-images.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig } from '@spartacus/core';
import { OutletModule } from '../../../cms-structure/outlet/index';
import { CarouselModule } from '../../../shared/components/carousel/index';
import { MediaModule } from '../../../shared/components/media/media.module';
import { ProductImagesComponent } from './product-images.component';
var ProductImagesModule = /** @class */ (function () {
    function ProductImagesModule() {
    }
    ProductImagesModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                MediaModule,
                OutletModule,
                CarouselModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ProductImagesComponent: {
                            component: ProductImagesComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductImagesComponent],
            entryComponents: [ProductImagesComponent],
            exports: [ProductImagesComponent],
        })
    ], ProductImagesModule);
    return ProductImagesModule;
}());
export { ProductImagesModule };
