// @spartacus/storefront/esm5/cms-components/product/product-images/product-images.component.js
import { __decorate, __read } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { BehaviorSubject, combineLatest, of } from 'rxjs';
import { distinctUntilChanged, filter, map, tap } from 'rxjs/operators';
import { CurrentProductService } from '../current-product.service';
var ProductImagesComponent = /** @class */ (function () {
    function ProductImagesComponent(currentProductService) {
        var _this = this;
        this.currentProductService = currentProductService;
        this.mainMediaContainer = new BehaviorSubject(null);
        this.product$ = this.currentProductService.getProduct().pipe(filter(Boolean), distinctUntilChanged(), tap(function (p) {
            var _a;
            _this.mainMediaContainer.next(((_a = p.images) === null || _a === void 0 ? void 0 : _a.PRIMARY) ? p.images.PRIMARY : {});
        }));
        this.thumbs$ = this.product$.pipe(map(function (p) { return _this.createThumbs(p); }));
        this.mainImage$ = combineLatest([this.product$, this.mainMediaContainer]).pipe(map(function (_a) {
            var _b = __read(_a, 2), container = _b[1];
            return container;
        }));
    }
    ProductImagesComponent.prototype.openImage = function (item) {
        this.mainMediaContainer.next(item);
    };
    ProductImagesComponent.prototype.isActive = function (thumbnail) {
        return this.mainMediaContainer.pipe(filter(Boolean), map(function (container) {
            return (container.zoom &&
                container.zoom.url &&
                thumbnail.zoom &&
                thumbnail.zoom.url &&
                container.zoom.url === thumbnail.zoom.url);
        }));
    };
    /** find the index of the main media in the list of media */
    ProductImagesComponent.prototype.getActive = function (thumbs) {
        return this.mainMediaContainer.pipe(filter(Boolean), map(function (container) {
            var current = thumbs.find(function (t) {
                return t.media &&
                    container.zoom &&
                    t.media.container &&
                    t.media.container.zoom &&
                    t.media.container.zoom.url === container.zoom.url;
            });
            return thumbs.indexOf(current);
        }));
    };
    /**
     * Return an array of CarouselItems for the product thumbnails.
     * In case there are less then 2 thumbs, we return null.
     */
    ProductImagesComponent.prototype.createThumbs = function (product) {
        if (!product.images ||
            !product.images.GALLERY ||
            product.images.GALLERY.length < 2) {
            return [];
        }
        return product.images.GALLERY.map(function (c) { return of({ container: c }); });
    };
    ProductImagesComponent.ctorParameters = function () { return [
        { type: CurrentProductService }
    ]; };
    ProductImagesComponent = __decorate([
        Component({
            selector: 'cx-product-images',
            template: "<ng-container *ngIf=\"mainImage$ | async as main\">\r\n  <cx-media [container]=\"main\"> </cx-media>\r\n</ng-container>\r\n\r\n<ng-container *ngIf=\"thumbs$ | async as thumbs\">\r\n  <cx-carousel\r\n    *ngIf=\"thumbs.length\"\r\n    class=\"thumbs\"\r\n    [items]=\"thumbs\"\r\n    itemWidth=\"120px\"\r\n    [hideIndicators]=\"false\"\r\n    [template]=\"thumb\"\r\n  ></cx-carousel>\r\n</ng-container>\r\n\r\n<ng-template #thumb let-item=\"item\">\r\n  <cx-media\r\n    [container]=\"item.container\"\r\n    tabindex=\"0\"\r\n    (focus)=\"openImage(item.container)\"\r\n    [class.is-active]=\"isActive(item.container) | async\"\r\n  >\r\n  </cx-media>\r\n</ng-template>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ProductImagesComponent);
    return ProductImagesComponent;
}());
export { ProductImagesComponent };
