// @spartacus/storefront/esm5/cms-components/navigation/footer-navigation/footer-navigation.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { GenericLinkModule } from '../../../shared/components/generic-link/generic-link.module';
import { NavigationModule } from '../navigation/navigation.module';
import { FooterNavigationComponent } from './footer-navigation.component';
var FooterNavigationModule = /** @class */ (function () {
    function FooterNavigationModule() {
    }
    FooterNavigationModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                NavigationModule,
                GenericLinkModule,
                I18nModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        FooterNavigationComponent: {
                            component: FooterNavigationComponent,
                        },
                    },
                }),
            ],
            declarations: [FooterNavigationComponent],
            entryComponents: [FooterNavigationComponent],
            exports: [FooterNavigationComponent],
        })
    ], FooterNavigationModule);
    return FooterNavigationModule;
}());
export { FooterNavigationModule };
