// @spartacus/core/esm5/src/event/utils/merging-subject.js
import { Observable } from 'rxjs';
import { share } from 'rxjs/operators';
// PRIVATE API
/**
 * Allows for dynamic adding and removing source observables
 * and exposes them as one merged observable at a property `output$`.
 *
 * Thanks to the `share()` operator used inside, it subscribes to source observables
 * only when someone subscribes to it. And it unsubscribes from source observables
 * when the counter of consumers drops to 0.
 *
 * **To avoid memory leaks**, all manually added sources should be manually removed
 * when not plan to emit values anymore. In particular closed event sources won't be
 * automatically removed.
 */
var MergingSubject = /** @class */ (function () {
    function MergingSubject() {
        var _this = this;
        /**
         * List of already added sources (but not removed yet)
         */
        this.sources = [];
        /**
         * For each source: it stores a subscription responsible for
         * passing all values from source to the consumer
         */
        this.subscriptionsToSources = new Map();
        /**
         * Observable with all sources merged.
         *
         * Only after subscribing to it, under the hood it subscribes to the source observables.
         * When the number of subscribers drops to 0, it unsubscribes from all source observables.
         * But if later on something subscribes to it again, it subscribes to the source observables again.
         *
         * It multicasts the emissions for each subscriber.
         */
        this.output$ = new Observable(function (consumer) {
            // There can be only 0 or 1 consumer of this observable coming from the `share()` operator
            // that is piped right after this observable.
            // `share()` not only multicasts the results but also  When all end-subscribers unsubscribe from `share()` operator, it will unsubscribe
            // from this observable (by the nature `refCount`-nature of the `share()` operator).
            _this.consumer = consumer;
            _this.bindAllSourcesToConsumer(consumer);
            return function () {
                _this.consumer = null;
                _this.unbindAllSourcesFromConsumer();
            };
        }).pipe(share());
        /**
         * Reference to the subscriber coming from the `share()` operator piped to the `output$` observable.
         * For more, see docs of the `output$` observable;
         */
        this.consumer = null;
    }
    /**
     * Registers the given source to pass its values to the `output$` observable.
     *
     * It does nothing, when the source has been already added (but not removed yet).
     */
    MergingSubject.prototype.add = function (source) {
        if (this.has(source)) {
            return;
        }
        if (this.consumer) {
            this.bindSourceToConsumer(source, this.consumer);
        }
        this.sources.push(source);
    };
    /**
     * Starts passing all values from already added sources to consumer
     */
    MergingSubject.prototype.bindAllSourcesToConsumer = function (consumer) {
        var _this = this;
        this.sources.forEach(function (source) {
            return _this.bindSourceToConsumer(source, consumer);
        });
    };
    /**
     * Stops passing all values from already added sources to consumer
     * (if any consumer is active at the moment)
     */
    MergingSubject.prototype.unbindAllSourcesFromConsumer = function () {
        var _this = this;
        this.sources.forEach(function (source) { return _this.unbindSourceFromConsumer(source); });
    };
    /**
     * Starts passing all values from a single source to consumer
     */
    MergingSubject.prototype.bindSourceToConsumer = function (source, consumer) {
        var subscriptionToSource = source.subscribe(function (val) { return consumer.next(val); }); // passes all emissions from source to consumer
        this.subscriptionsToSources.set(source, subscriptionToSource);
    };
    /**
     * Stops passing all values from a single source to consumer
     * (if any consumer is active at the moment)
     */
    MergingSubject.prototype.unbindSourceFromConsumer = function (source) {
        var subscriptionToSource = this.subscriptionsToSources.get(source);
        if (subscriptionToSource !== undefined) {
            subscriptionToSource.unsubscribe();
            this.subscriptionsToSources.delete(source);
        }
    };
    /**
     * Unregisters the given source so it stops passing its values to `output$` observable.
     *
     * Should be used when a source is no longer maintained **to avoid memory leaks**.
     */
    MergingSubject.prototype.remove = function (source) {
        // clear binding from source to consumer (if any consumer exists at the moment)
        this.unbindSourceFromConsumer(source);
        // remove source from array
        var i;
        if ((i = this.sources.findIndex(function (s) { return s === source; })) !== -1) {
            this.sources.splice(i, 1);
        }
    };
    /**
     * Returns whether the given source has been already addded
     */
    MergingSubject.prototype.has = function (source) {
        return this.sources.includes(source);
    };
    return MergingSubject;
}());
export { MergingSubject };
