// @spartacus/core/esm5/src/cart/facade/cart-voucher.service.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { combineLatest } from 'rxjs';
import { map, take } from 'rxjs/operators';
import { AuthService } from '../../auth/index';
import { getProcessErrorFactory, getProcessLoadingFactory, getProcessSuccessFactory, } from '../../process/store/selectors/process.selectors';
import { CartActions } from '../store/actions/index';
import { ADD_VOUCHER_PROCESS_ID } from '../store/multi-cart-state';
import { ActiveCartService } from './active-cart.service';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
import * as i3 from "./active-cart.service";
var CartVoucherService = /** @class */ (function () {
    function CartVoucherService(store, authService, activeCartService) {
        this.store = store;
        this.authService = authService;
        this.activeCartService = activeCartService;
    }
    CartVoucherService.prototype.addVoucher = function (voucherId, cartId) {
        var _this = this;
        this.combineUserAndCartId(cartId).subscribe(function (_a) {
            var _b = __read(_a, 2), occUserId = _b[0], cartIdentifier = _b[1];
            return _this.store.dispatch(new CartActions.CartAddVoucher({
                userId: occUserId,
                cartId: cartIdentifier,
                voucherId: voucherId,
            }));
        });
    };
    CartVoucherService.prototype.removeVoucher = function (voucherId, cartId) {
        var _this = this;
        this.combineUserAndCartId(cartId).subscribe(function (_a) {
            var _b = __read(_a, 2), occUserId = _b[0], cartIdentifier = _b[1];
            return _this.store.dispatch(new CartActions.CartRemoveVoucher({
                userId: occUserId,
                cartId: cartIdentifier,
                voucherId: voucherId,
            }));
        });
    };
    // TODO(#7241): Remove when switching to event system for add voucher
    /**
     * Get add voucher process error flag
     * @deprecated since 2.0
     */
    CartVoucherService.prototype.getAddVoucherResultError = function () {
        return this.store.pipe(select(getProcessErrorFactory(ADD_VOUCHER_PROCESS_ID)));
    };
    // TODO(#7241): Remove when switching to event system for add voucher
    /**
     * Get add voucher process success flag
     * @deprecated since 2.0
     */
    CartVoucherService.prototype.getAddVoucherResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(ADD_VOUCHER_PROCESS_ID)));
    };
    // TODO(#7241): Remove when switching to event system for add voucher
    /**
     * Get add voucher process loading flag
     * @deprecated since 2.0
     */
    CartVoucherService.prototype.getAddVoucherResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(ADD_VOUCHER_PROCESS_ID)));
    };
    // TODO(#7241): Remove when switching to event system for add voucher
    /**
     * Reset add voucher process
     * @deprecated since 2.0
     */
    CartVoucherService.prototype.resetAddVoucherProcessingState = function () {
        this.store.dispatch(new CartActions.CartResetAddVoucher());
    };
    CartVoucherService.prototype.combineUserAndCartId = function (cartId) {
        if (cartId) {
            return this.authService.getOccUserId().pipe(take(1), map(function (userId) { return [userId, cartId]; }));
        }
        else {
            return combineLatest([
                this.authService.getOccUserId(),
                this.activeCartService.getActiveCartId(),
            ]).pipe(take(1));
        }
    };
    CartVoucherService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService },
        { type: ActiveCartService }
    ]; };
    CartVoucherService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartVoucherService_Factory() { return new CartVoucherService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService), i0.ɵɵinject(i3.ActiveCartService)); }, token: CartVoucherService, providedIn: "root" });
    CartVoucherService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CartVoucherService);
    return CartVoucherService;
}());
export { CartVoucherService };
