// @spartacus/storefront/esm5/cms-components/myaccount/order/return-request-detail/return-request-detail.module.js
import { __decorate, __read, __spread } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, FeaturesConfigModule, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { MediaModule } from '../../../../shared/index';
import { CmsPageGuard } from '../../../../cms-structure/guards/cms-page.guard';
import { PageLayoutComponent } from '../../../../cms-structure/page/page-layout/page-layout.component';
import { ReturnRequestOverviewComponent } from './return-request-overview/return-request-overview.component';
import { ReturnRequestItemsComponent } from './return-request-items/return-request-items.component';
import { ReturnRequestTotalsComponent } from './return-request-totals/return-request-totals.component';
var components = [
    ReturnRequestOverviewComponent,
    ReturnRequestItemsComponent,
    ReturnRequestTotalsComponent,
];
var ɵ0 = { cxRoute: 'returnRequestDetails' };
var ReturnRequestDetailModule = /** @class */ (function () {
    function ReturnRequestDetailModule() {
    }
    ReturnRequestDetailModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule.forChild([
                    {
                        path: null,
                        canActivate: [AuthGuard, CmsPageGuard],
                        component: PageLayoutComponent,
                        data: ɵ0,
                    },
                ]),
                RouterModule,
                UrlModule,
                I18nModule,
                MediaModule,
                FeaturesConfigModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ReturnRequestOverviewComponent: {
                            component: ReturnRequestOverviewComponent,
                        },
                        ReturnRequestItemsComponent: {
                            component: ReturnRequestItemsComponent,
                        },
                        ReturnRequestTotalsComponent: {
                            component: ReturnRequestTotalsComponent,
                        },
                    },
                }),
            ],
            declarations: __spread(components),
            exports: __spread(components),
            entryComponents: __spread(components),
        })
    ], ReturnRequestDetailModule);
    return ReturnRequestDetailModule;
}());
export { ReturnRequestDetailModule };
export { ɵ0 };
