// @spartacus/storefront/esm5/cms-components/content/banner-carousel/banner-carousel.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { PageComponentModule } from '../../../cms-structure/page/component/page-component.module';
import { CarouselModule, MediaModule } from '../../../shared/components/index';
import { BannerCarouselComponent } from './banner-carousel.component';
var BannerCarouselModule = /** @class */ (function () {
    function BannerCarouselModule() {
    }
    BannerCarouselModule = __decorate([
        NgModule({
            imports: [CommonModule, PageComponentModule, CarouselModule, MediaModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        RotatingImagesComponent: {
                            component: BannerCarouselComponent,
                        },
                    },
                }),
            ],
            declarations: [BannerCarouselComponent],
            entryComponents: [BannerCarouselComponent],
            exports: [BannerCarouselComponent],
        })
    ], BannerCarouselModule);
    return BannerCarouselModule;
}());
export { BannerCarouselModule };
