// @spartacus/storefront/esm5/cms-components/content/banner-carousel/banner-carousel.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { CmsBannerCarouselComponent as model, CmsService, ContentSlotComponentData, } from '@spartacus/core';
import { filter, map, tap } from 'rxjs/operators';
import { CmsComponentData } from '../../../cms-structure/index';
/**
 * Generic carousel that renders CMS Components.
 */
var BannerCarouselComponent = /** @class */ (function () {
    function BannerCarouselComponent(componentData, cmsService) {
        var _this = this;
        this.componentData = componentData;
        this.cmsService = cmsService;
        this.componentData$ = this.componentData.data$.pipe(filter(Boolean), tap(function (d) { return (_this.theme = d.effect + "-theme"); }));
        this.items$ = this.componentData$.pipe(map(function (data) { return data.banners.trim().split(' '); }), map(function (codes) { return codes.map(function (code) { return _this.cmsService.getComponentData(code); }); }));
        /**
         * Adds a specific theme for the carousel. The effect can be
         * used in CSS customisations.
         */
        this.theme = '';
    }
    /**
     * Returns an Obervable with an Array of Observables. This is done, so that
     * the component UI could consider to lazy load the UI components when they're
     * in the viewpoint.
     */
    BannerCarouselComponent.prototype.getItems = function () {
        return this.items$;
    };
    BannerCarouselComponent.ctorParameters = function () { return [
        { type: CmsComponentData },
        { type: CmsService }
    ]; };
    __decorate([
        HostBinding('class')
    ], BannerCarouselComponent.prototype, "theme", void 0);
    BannerCarouselComponent = __decorate([
        Component({
            selector: 'cx-banner-carousel',
            template: "<cx-carousel\n  [items]=\"getItems() | async\"\n  [template]=\"template\"\n  itemWidth=\"100%\"\n  class=\"inline-navigation\"\n></cx-carousel>\n\n<ng-template #template let-item=\"item\">\n  <ng-container\n    [cxComponentWrapper]=\"{\n      flexType: item.typeCode,\n      typeCode: item.typeCode,\n      uid: item?.uid\n    }\"\n  >\n  </ng-container>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], BannerCarouselComponent);
    return BannerCarouselComponent;
}());
export { BannerCarouselComponent };
