// @spartacus/storefront/esm5/cms-components/cart/cart-details/cart-details.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { FeaturesConfigModule, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { PromotionsModule } from '../../checkout/components/promotions/promotions.module';
import { CartSharedModule } from '../cart-shared/cart-shared.module';
import { CartCouponModule } from '../cart-coupon/cart-coupon.module';
import { CartDetailsComponent } from './cart-details.component';
var CartDetailsModule = /** @class */ (function () {
    function CartDetailsModule() {
    }
    CartDetailsModule = __decorate([
        NgModule({
            imports: [
                CartSharedModule,
                CommonModule,
                CartCouponModule,
                RouterModule,
                UrlModule,
                PromotionsModule,
                FeaturesConfigModule,
                I18nModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CartComponent: {
                            component: CartDetailsComponent,
                        },
                    },
                }),
            ],
            declarations: [CartDetailsComponent],
            exports: [CartDetailsComponent],
            entryComponents: [CartDetailsComponent],
        })
    ], CartDetailsModule);
    return CartDetailsModule;
}());
export { CartDetailsModule };
