// @spartacus/storefront/esm5/cms-components/cart/cart-details/cart-details.component.js
import { __decorate, __read } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActiveCartService, AuthService, Cart, OrderEntry, PromotionLocation, PromotionResult, RoutingService, SelectiveCartService, } from '@spartacus/core';
import { combineLatest, of } from 'rxjs';
import { filter, map, tap } from 'rxjs/operators';
import { PromotionService } from '../../../shared/services/promotion/promotion.service';
var CartDetailsComponent = /** @class */ (function () {
    function CartDetailsComponent(activeCartService, promotionService, selectiveCartService, authService, routingService) {
        this.activeCartService = activeCartService;
        this.promotionService = promotionService;
        this.selectiveCartService = selectiveCartService;
        this.authService = authService;
        this.routingService = routingService;
        this.loggedIn = false;
        this.promotionLocation = PromotionLocation.ActiveCart;
    }
    CartDetailsComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.cart$ = this.activeCartService.getActive();
        this.promotions$ = this.promotionService.getOrderPromotionsFromCart();
        this.entries$ = this.activeCartService
            .getEntries()
            .pipe(filter(function (entries) { return entries.length > 0; }));
        this.selectiveCartEnabled = this.selectiveCartService.isEnabled();
        this.cartLoaded$ = combineLatest([
            this.activeCartService.isStable(),
            this.selectiveCartEnabled
                ? this.selectiveCartService.getLoaded()
                : of(false),
            this.authService.isUserLoggedIn(),
        ]).pipe(tap(function (_a) {
            var _b = __read(_a, 3), loggedIn = _b[2];
            return (_this.loggedIn = loggedIn);
        }), map(function (_a) {
            var _b = __read(_a, 3), cartLoaded = _b[0], sflLoaded = _b[1], loggedIn = _b[2];
            return loggedIn && _this.selectiveCartEnabled
                ? cartLoaded && sflLoaded
                : cartLoaded;
        }));
        this.orderPromotions$ = this.promotionService.getOrderPromotions(this.promotionLocation);
    };
    CartDetailsComponent.prototype.saveForLater = function (item) {
        if (this.loggedIn) {
            this.activeCartService.removeEntry(item);
            this.selectiveCartService.addEntry(item.product.code, item.quantity);
        }
        else {
            this.routingService.go({ cxRoute: 'login' });
        }
    };
    CartDetailsComponent.ctorParameters = function () { return [
        { type: ActiveCartService },
        { type: PromotionService },
        { type: SelectiveCartService },
        { type: AuthService },
        { type: RoutingService }
    ]; };
    CartDetailsComponent = __decorate([
        Component({
            selector: 'cx-cart-details',
            template: "<ng-container *ngIf=\"cart$ | async as cart\">\r\n  <ng-container *ngIf=\"entries$ | async as entries\">\r\n    <div *ngIf=\"cart.totalItems > 0\" class=\"cart-details-wrapper\">\r\n      <div class=\"cx-total\">\r\n        {{ 'cartDetails.cartName' | cxTranslate: { code: cart.code } }}\r\n      </div>\r\n\r\n      <ng-container *ngIf=\"orderPromotions$ | async as orderPromotions\">\r\n        <cx-promotions [promotions]=\"orderPromotions\"></cx-promotions>\r\n      </ng-container>\r\n\r\n      <cx-cart-item-list\r\n        [items]=\"entries\"\r\n        [cartIsLoading]=\"!(cartLoaded$ | async)\"\r\n        [promotionLocation]=\"promotionLocation\"\r\n        [options]=\"{\r\n          isSaveForLater: false,\r\n          optionalBtn: saveForLaterBtn\r\n        }\"\r\n      ></cx-cart-item-list>\r\n    </div>\r\n  </ng-container>\r\n</ng-container>\r\n\r\n<ng-template let-ctx #saveForLaterBtn>\r\n  <div\r\n    *ngIf=\"selectiveCartEnabled\"\r\n    class=\"col-md-3 col-lg-3 col-xl-3 cx-sfl-btn\"\r\n  >\r\n    <button\r\n      class=\"link cx-action-link\"\r\n      [disabled]=\"ctx.loading\"\r\n      (click)=\"saveForLater(ctx.item)\"\r\n      type=\"button\"\r\n    >\r\n      {{ 'saveForLaterItems.saveForLater' | cxTranslate }}\r\n    </button>\r\n  </div>\r\n</ng-template>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CartDetailsComponent);
    return CartDetailsComponent;
}());
export { CartDetailsComponent };
