// @spartacus/storefront/esm5/cms-components/content/tab-paragraph-container/tab-paragraph-container.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { OutletModule } from '../../../cms-structure/outlet/outlet.module';
import { PageComponentModule } from '../../../cms-structure/page/component/page-component.module';
import { TabParagraphContainerComponent } from './tab-paragraph-container.component';
var TabParagraphContainerModule = /** @class */ (function () {
    function TabParagraphContainerModule() {
    }
    TabParagraphContainerModule = __decorate([
        NgModule({
            imports: [CommonModule, PageComponentModule, OutletModule, I18nModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CMSTabParagraphContainer: {
                            component: TabParagraphContainerComponent,
                        },
                    },
                }),
            ],
            declarations: [TabParagraphContainerComponent],
            entryComponents: [TabParagraphContainerComponent],
            exports: [TabParagraphContainerComponent],
        })
    ], TabParagraphContainerModule);
    return TabParagraphContainerModule;
}());
export { TabParagraphContainerModule };
