// @spartacus/core/esm5/src/user/store/selectors/order-return-request.selectors.js
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
import { StateUtils } from '../../../state/utils/index';
var ɵ0 = function (state) { return state.orderReturn; };
export var getOrderReturnRequestState = createSelector(getUserState, ɵ0);
var ɵ1 = function (state) { return StateUtils.loaderValueSelector(state); };
export var getOrderReturnRequest = createSelector(getOrderReturnRequestState, ɵ1);
var ɵ2 = function (state) { return StateUtils.loaderLoadingSelector(state); };
export var getOrderReturnRequestLoading = createSelector(getOrderReturnRequestState, ɵ2);
var ɵ3 = function (state) {
    return StateUtils.loaderSuccessSelector(state) &&
        !StateUtils.loaderLoadingSelector(state);
};
export var getOrderReturnRequestSuccess = createSelector(getOrderReturnRequestState, ɵ3);
var ɵ4 = function (state) { return state.orderReturnList; };
export var getOrderReturnRequestListState = createSelector(getUserState, ɵ4);
var ɵ5 = function (state) {
    return StateUtils.loaderValueSelector(state);
};
export var getOrderReturnRequestList = createSelector(getOrderReturnRequestListState, ɵ5);
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5 };
