// @spartacus/core/esm5/src/user/store/selectors/regions.selectors.js
import { createSelector } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
var ɵ0 = function (state) { return state.regions; };
export var getRegionsLoaderState = createSelector(getUserState, ɵ0);
var ɵ1 = function (state) {
    return StateUtils.loaderValueSelector(state).entities;
};
export var getAllRegions = createSelector(getRegionsLoaderState, ɵ1);
var ɵ2 = function (state) { return ({
    loaded: StateUtils.loaderSuccessSelector(state),
    loading: StateUtils.loaderLoadingSelector(state),
    regions: StateUtils.loaderValueSelector(state).entities,
    country: StateUtils.loaderValueSelector(state).country,
}); };
export var getRegionsDataAndLoading = createSelector(getRegionsLoaderState, ɵ2);
var ɵ3 = function (state) {
    return StateUtils.loaderValueSelector(state).country;
};
export var getRegionsCountry = createSelector(getRegionsLoaderState, ɵ3);
var ɵ4 = function (state) {
    return StateUtils.loaderLoadingSelector(state);
};
export var getRegionsLoading = createSelector(getRegionsLoaderState, ɵ4);
var ɵ5 = function (state) {
    return StateUtils.loaderSuccessSelector(state);
};
export var getRegionsLoaded = createSelector(getRegionsLoaderState, ɵ5);
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5 };
