// @spartacus/storefront/esm2015/layout/main/storefront.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, HostBinding, HostListener, OnDestroy, OnInit, ViewChild, } from '@angular/core';
import { RoutingService } from '@spartacus/core';
import { FocusConfig, KeyboardFocusService, } from '../a11y/keyboard-focus/index';
import { SkipLinkComponent } from '../a11y/skip-link/index';
import { HamburgerMenuService } from '../header/hamburger-menu/hamburger-menu.service';
let StorefrontComponent = class StorefrontComponent {
    constructor(hamburgerMenuService, routingService, elementRef, keyboardFocusService) {
        this.hamburgerMenuService = hamburgerMenuService;
        this.routingService = routingService;
        this.elementRef = elementRef;
        this.keyboardFocusService = keyboardFocusService;
        this.isExpanded$ = this.hamburgerMenuService.isExpanded;
        // required by esc focus
        this.tabindex = '0';
        this.keyboardFocusConfig = {
            focusOnEscape: true,
            focusOnDoubleEscape: true,
        };
    }
    handleEscape(event) {
        this.keyboardFocusService.handleEscape(this.elementRef.nativeElement, this.keyboardFocusConfig, event);
    }
    ngOnInit() {
        this.navigateSubscription = this.routingService
            .isNavigating()
            .subscribe((val) => {
            this.startNavigating = val === true;
            this.stopNavigating = val === false;
        });
    }
    collapseMenuIfClickOutside(event) {
        if (event.target.className.includes('is-expanded')) {
            this.collapseMenu();
        }
    }
    collapseMenu() {
        this.hamburgerMenuService.toggle(true);
    }
    ngOnDestroy() {
        if (this.navigateSubscription) {
            this.navigateSubscription.unsubscribe();
        }
    }
};
StorefrontComponent.ctorParameters = () => [
    { type: HamburgerMenuService },
    { type: RoutingService },
    { type: ElementRef },
    { type: KeyboardFocusService }
];
__decorate([
    HostBinding('class.start-navigating')
], StorefrontComponent.prototype, "startNavigating", void 0);
__decorate([
    HostBinding('class.stop-navigating')
], StorefrontComponent.prototype, "stopNavigating", void 0);
__decorate([
    HostBinding('tabindex')
], StorefrontComponent.prototype, "tabindex", void 0);
__decorate([
    ViewChild(SkipLinkComponent)
], StorefrontComponent.prototype, "child", void 0);
__decorate([
    HostListener('keydown.escape', ['$event'])
], StorefrontComponent.prototype, "handleEscape", null);
StorefrontComponent = __decorate([
    Component({
        selector: 'cx-storefront',
        template: "<ng-template cxOutlet=\"cx-storefront\">\r\n  <ng-template cxOutlet=\"cx-header\">\r\n    <header\r\n      cxSkipLink=\"cx-header\"\r\n      [cxFocus]=\"{ disableMouseFocus: true }\"\r\n      [class.is-expanded]=\"isExpanded$ | async\"\r\n      (keydown.escape)=\"collapseMenu()\"\r\n      (click)=\"collapseMenuIfClickOutside($event)\"\r\n    >\r\n      <cx-page-layout section=\"header\"></cx-page-layout>\r\n      <cx-page-layout section=\"navigation\"></cx-page-layout>\r\n    </header>\r\n    <cx-page-slot position=\"BottomHeaderSlot\"></cx-page-slot>\r\n    <cx-global-message></cx-global-message>\r\n  </ng-template>\r\n\r\n  <main cxSkipLink=\"cx-main\" [cxFocus]=\"{ disableMouseFocus: true }\">\r\n    <router-outlet></router-outlet>\r\n  </main>\r\n\r\n  <ng-template cxOutlet=\"cx-footer\">\r\n    <footer cxSkipLink=\"cx-footer\" [cxFocus]=\"{ disableMouseFocus: true }\">\r\n      <cx-page-layout section=\"footer\"></cx-page-layout>\r\n    </footer>\r\n  </ng-template>\r\n</ng-template>\r\n"
    })
], StorefrontComponent);
export { StorefrontComponent };
