// @spartacus/storefront/esm5/layout/a11y/skip-link/component/skip-link.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { SkipLinkService } from '../service/skip-link.service';
var SkipLinkComponent = /** @class */ (function () {
    function SkipLinkComponent(skipLinkService) {
        this.skipLinkService = skipLinkService;
        this.skipLinks$ = this.skipLinkService.getSkipLinks();
    }
    SkipLinkComponent.prototype.scrollToTarget = function (skipLink) {
        this.skipLinkService.scrollToTarget(skipLink);
    };
    SkipLinkComponent.ctorParameters = function () { return [
        { type: SkipLinkService }
    ]; };
    SkipLinkComponent = __decorate([
        Component({
            selector: 'cx-skip-link',
            template: "<div [cxFocus]=\"{ tab: true }\" *ngIf=\"skipLinks$ | async as links\">\r\n  <button *ngFor=\"let link of links\" (click)=\"scrollToTarget(link)\">\r\n    {{ 'skipLink.skipTo' | cxTranslate }}\r\n    {{ link.i18nKey | cxTranslate }}\r\n  </button>\r\n</div>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], SkipLinkComponent);
    return SkipLinkComponent;
}());
export { SkipLinkComponent };
