// @spartacus/core/esm5/src/checkout/store/selectors/checkout.selectors.js
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { CHECKOUT_FEATURE, } from '../checkout-state';
var getDeliveryAddressSelector = function (state) { return state.address; };
var ɵ0 = getDeliveryAddressSelector;
var getDeliveryModeSelector = function (state) {
    return state.deliveryMode;
};
var ɵ1 = getDeliveryModeSelector;
var getPaymentDetailsSelector = function (state) {
    return state.paymentDetails;
};
var ɵ2 = getPaymentDetailsSelector;
var getOrderDetailsSelector = function (state) {
    return state.orderDetails;
};
var ɵ3 = getOrderDetailsSelector;
export var getCheckoutState = createFeatureSelector(CHECKOUT_FEATURE);
var ɵ4 = function (checkoutState) { return checkoutState.steps; };
export var getCheckoutStepsState = createSelector(getCheckoutState, ɵ4);
var ɵ5 = function (state) {
    return StateUtils.loaderValueSelector(state);
};
export var getCheckoutSteps = createSelector(getCheckoutStepsState, ɵ5);
export var getDeliveryAddress = createSelector(getCheckoutSteps, getDeliveryAddressSelector);
export var getDeliveryMode = createSelector(getCheckoutSteps, getDeliveryModeSelector);
var ɵ6 = function (deliveryMode) {
    return (deliveryMode &&
        Object.keys(deliveryMode.supported).map(function (code) { return deliveryMode.supported[code]; }));
};
export var getSupportedDeliveryModes = createSelector(getDeliveryMode, ɵ6);
var ɵ7 = function (deliveryMode) {
    return deliveryMode && deliveryMode.selected;
};
export var getSelectedDeliveryModeCode = createSelector(getDeliveryMode, ɵ7);
var ɵ8 = function (deliveryMode) {
    if (deliveryMode.selected !== '') {
        if (Object.keys(deliveryMode.supported).length === 0) {
            return null;
        }
        return deliveryMode.supported[deliveryMode.selected];
    }
};
export var getSelectedDeliveryMode = createSelector(getDeliveryMode, ɵ8);
export var getPaymentDetails = createSelector(getCheckoutSteps, getPaymentDetailsSelector);
export var getCheckoutOrderDetails = createSelector(getCheckoutSteps, getOrderDetailsSelector);
var ɵ9 = function (state) {
    return StateUtils.loaderSuccessSelector(state) &&
        !StateUtils.loaderLoadingSelector(state);
};
export var getCheckoutDetailsLoaded = createSelector(getCheckoutStepsState, ɵ9);
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9 };
