// @spartacus/storefront/esm5/cms-components/myaccount/update-email/update-email.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { UpdateEmailFormComponent } from './update-email-form/update-email-form.component';
import { UpdateEmailComponent } from './update-email.component';
import { FormErrorsModule } from '../../../shared/index';
var UpdateEmailModule = /** @class */ (function () {
    function UpdateEmailModule() {
    }
    UpdateEmailModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
                SpinnerModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        UpdateEmailComponent: {
                            component: UpdateEmailComponent,
                            guards: [AuthGuard],
                        },
                    },
                }),
            ],
            declarations: [UpdateEmailFormComponent, UpdateEmailComponent],
            exports: [UpdateEmailComponent, UpdateEmailFormComponent],
            entryComponents: [UpdateEmailComponent],
        })
    ], UpdateEmailModule);
    return UpdateEmailModule;
}());
export { UpdateEmailModule };
