// @spartacus/storefront/esm5/cms-components/user/login/login.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { AuthService, User, UserService } from '@spartacus/core';
import { of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
var LoginComponent = /** @class */ (function () {
    function LoginComponent(auth, userService) {
        this.auth = auth;
        this.userService = userService;
    }
    LoginComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.user$ = this.auth.isUserLoggedIn().pipe(switchMap(function (isUserLoggedIn) {
            if (isUserLoggedIn) {
                return _this.userService.get();
            }
            else {
                return of(undefined);
            }
        }));
    };
    LoginComponent.ctorParameters = function () { return [
        { type: AuthService },
        { type: UserService }
    ]; };
    LoginComponent = __decorate([
        Component({
            selector: 'cx-login',
            template: "<ng-container *ngIf=\"user$ | async as user; else login\">\n  <div class=\"cx-login-greet\">\n    {{ 'miniLogin.userGreeting' | cxTranslate: { name: user.name } }}\n  </div>\n  <cx-page-slot position=\"HeaderLinks\"></cx-page-slot>\n</ng-container>\n\n<ng-template #login>\n  <a role=\"link\" [routerLink]=\"{ cxRoute: 'login' } | cxUrl\">{{\n    'miniLogin.signInRegister' | cxTranslate\n  }}</a>\n</ng-template>\n"
        })
    ], LoginComponent);
    return LoginComponent;
}());
export { LoginComponent };
