// @spartacus/storefront/esm5/cms-components/user/login/login.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { PageSlotModule } from '../../../cms-structure/page/slot/page-slot.module';
import { LoginComponent } from './login.component';
var LoginModule = /** @class */ (function () {
    function LoginModule() {
    }
    LoginModule = __decorate([
        NgModule({
            imports: [CommonModule, RouterModule, UrlModule, PageSlotModule, I18nModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        LoginComponent: {
                            component: LoginComponent,
                        },
                    },
                }),
            ],
            declarations: [LoginComponent],
            entryComponents: [LoginComponent],
            exports: [LoginComponent],
        })
    ], LoginModule);
    return LoginModule;
}());
export { LoginModule };
