// @spartacus/storefront/esm5/cms-components/myaccount/forgot-password/forgot-password.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ForgotPasswordComponent } from './forgot-password.component';
import { FormErrorsModule } from '../../../shared/index';
var ForgotPasswordModule = /** @class */ (function () {
    function ForgotPasswordModule() {
    }
    ForgotPasswordModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                ReactiveFormsModule,
                RouterModule,
                UrlModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ForgotPasswordComponent: {
                            component: ForgotPasswordComponent,
                            guards: [NotAuthGuard],
                        },
                    },
                }),
            ],
            declarations: [ForgotPasswordComponent],
            exports: [ForgotPasswordComponent],
            entryComponents: [ForgotPasswordComponent],
        })
    ], ForgotPasswordModule);
    return ForgotPasswordModule;
}());
export { ForgotPasswordModule };
