// @spartacus/storefront/esm5/cms-components/myaccount/forgot-password/forgot-password.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { RoutingService, UserService } from '@spartacus/core';
import { CustomFormValidators } from '../../../shared/utils/validators/custom-form-validators';
var ForgotPasswordComponent = /** @class */ (function () {
    function ForgotPasswordComponent(fb, userService, routingService) {
        this.fb = fb;
        this.userService = userService;
        this.routingService = routingService;
    }
    ForgotPasswordComponent.prototype.ngOnInit = function () {
        this.forgotPasswordForm = this.fb.group({
            userEmail: [
                '',
                [Validators.required, CustomFormValidators.emailValidator],
            ],
        });
    };
    ForgotPasswordComponent.prototype.requestForgotPasswordEmail = function () {
        if (this.forgotPasswordForm.valid) {
            this.userService.requestForgotPasswordEmail(this.forgotPasswordForm.value.userEmail);
            this.routingService.go({ cxRoute: 'login' });
        }
        else {
            this.forgotPasswordForm.markAllAsTouched();
        }
    };
    ForgotPasswordComponent.ctorParameters = function () { return [
        { type: FormBuilder },
        { type: UserService },
        { type: RoutingService }
    ]; };
    ForgotPasswordComponent = __decorate([
        Component({
            selector: 'cx-forgot-password',
            template: "<form\r\n  (ngSubmit)=\"requestForgotPasswordEmail()\"\r\n  [formGroup]=\"forgotPasswordForm\"\r\n>\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'forgottenPassword.emailAddress.label' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        type=\"email\"\r\n        class=\"form-control\"\r\n        placeholder=\"{{\r\n          'forgottenPassword.emailAddress.placeholder' | cxTranslate\r\n        }}\"\r\n        formControlName=\"userEmail\"\r\n      />\r\n      <cx-form-errors\r\n        [control]=\"forgotPasswordForm.get('userEmail')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <button type=\"submit\" class=\"btn btn-block btn-primary\">\r\n    {{ 'common.submit' | cxTranslate }}\r\n  </button>\r\n  <a\r\n    class=\"btn btn-block btn-secondary\"\r\n    [routerLink]=\"{ cxRoute: 'login' } | cxUrl\"\r\n    >{{ 'common.cancel' | cxTranslate }}</a\r\n  >\r\n</form>\r\n"
        })
    ], ForgotPasswordComponent);
    return ForgotPasswordComponent;
}());
export { ForgotPasswordComponent };
