// @spartacus/core/esm5/src/auth/store/effects/user-token.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, exhaustMap, map, mergeMap } from 'rxjs/operators';
import { OCC_USER_ID_CURRENT } from '../../../occ/utils/occ-constants';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { UserAuthenticationTokenService } from '../../services/user-authentication/user-authentication-token.service';
import { AuthActions } from '../actions/index';
var UserTokenEffects = /** @class */ (function () {
    function UserTokenEffects(actions$, userTokenService) {
        var _this = this;
        this.actions$ = actions$;
        this.userTokenService = userTokenService;
        this.loadUserToken$ = this.actions$.pipe(ofType(AuthActions.LOAD_USER_TOKEN), map(function (action) { return action.payload; }), mergeMap(function (_a) {
            var userId = _a.userId, password = _a.password;
            return _this.userTokenService.loadToken(userId, password).pipe(map(function (token) {
                var date = new Date();
                date.setSeconds(date.getSeconds() + token.expires_in);
                token.expiration_time = date.toJSON();
                token.userId = OCC_USER_ID_CURRENT;
                return new AuthActions.LoadUserTokenSuccess(token);
            }), catchError(function (error) {
                return of(new AuthActions.LoadUserTokenFail(makeErrorSerializable(error)));
            }));
        }));
        this.login$ = this.actions$.pipe(ofType(AuthActions.LOAD_USER_TOKEN_SUCCESS), map(function () { return new AuthActions.Login(); }));
        this.refreshUserToken$ = this.actions$.pipe(ofType(AuthActions.REFRESH_USER_TOKEN), map(function (action) { return action.payload; }), exhaustMap(function (_a) {
            var refreshToken = _a.refreshToken;
            return _this.userTokenService.refreshToken(refreshToken).pipe(map(function (token) {
                var date = new Date();
                date.setSeconds(date.getSeconds() + token.expires_in);
                token.expiration_time = date.toJSON();
                return new AuthActions.RefreshUserTokenSuccess(token);
            }, catchError(function (error) {
                return of(new AuthActions.RefreshUserTokenFail(makeErrorSerializable(error)));
            })));
        }));
        this.revokeUserToken$ = this.actions$.pipe(ofType(AuthActions.REVOKE_USER_TOKEN), map(function (action) {
            return action.payload;
        }), mergeMap(function (userToken) {
            return _this.userTokenService.revoke(userToken).pipe(map(function () { return new AuthActions.RevokeUserTokenSuccess(userToken); }), catchError(function (error) { return of(new AuthActions.RevokeUserTokenFail(error)); }));
        }));
    }
    UserTokenEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: UserAuthenticationTokenService }
    ]; };
    __decorate([
        Effect()
    ], UserTokenEffects.prototype, "loadUserToken$", void 0);
    __decorate([
        Effect()
    ], UserTokenEffects.prototype, "login$", void 0);
    __decorate([
        Effect()
    ], UserTokenEffects.prototype, "refreshUserToken$", void 0);
    __decorate([
        Effect()
    ], UserTokenEffects.prototype, "revokeUserToken$", void 0);
    UserTokenEffects = __decorate([
        Injectable()
    ], UserTokenEffects);
    return UserTokenEffects;
}());
export { UserTokenEffects };
