// @spartacus/core/esm5/src/site-context/providers/context-service-providers.js
import { APP_INITIALIZER } from '@angular/core';
import { ConfigInitializerService } from '../../config/config-initializer/config-initializer.service';
import { BaseSiteService } from '../facade/base-site.service';
import { CurrencyService } from '../facade/currency.service';
import { LanguageService } from '../facade/language.service';
import { SiteContextRoutesHandler } from '../services/site-context-routes-handler';
export function initializeContext(baseSiteService, langService, currService, configInit, siteContextRoutesHandler) {
    return function () {
        configInit.getStableConfig('context').then(function () {
            siteContextRoutesHandler.init().then(function () {
                baseSiteService.initialize();
                langService.initialize();
                currService.initialize();
            });
        });
    };
}
export var contextServiceProviders = [
    BaseSiteService,
    LanguageService,
    CurrencyService,
    {
        provide: APP_INITIALIZER,
        useFactory: initializeContext,
        deps: [
            BaseSiteService,
            LanguageService,
            CurrencyService,
            ConfigInitializerService,
            SiteContextRoutesHandler,
        ],
        multi: true,
    },
];
