// @spartacus/core/esm5/src/occ/config-loader/occ-config-loader.service.js
import { __decorate, __param, __read, __spread } from "tslib";
import { DOCUMENT, isPlatformBrowser, isPlatformServer } from '@angular/common';
import { Inject, Injectable, isDevMode, Optional, PLATFORM_ID, } from '@angular/core';
import { makeStateKey, StateKey, TransferState, } from '@angular/platform-browser';
import { of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { Config } from '../../config/config.module';
import { deepMerge } from '../../config/utils/deep-merge';
import { SERVER_REQUEST_URL } from '../../ssr/ssr.providers';
import { OccLoadedConfigConverter } from './occ-loaded-config-converter';
import { OccSitesConfigLoader } from './occ-sites-config-loader';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../config/config.module";
import * as i3 from "./occ-sites-config-loader";
import * as i4 from "./occ-loaded-config-converter";
import * as i5 from "@angular/platform-browser";
import * as i6 from "../../ssr/ssr.providers";
export var EXTERNAL_CONFIG_TRANSFER_ID = makeStateKey('cx-external-config');
var OccConfigLoaderService = /** @class */ (function () {
    function OccConfigLoaderService(platform, document, config, sitesConfigLoader, converter, transferState, serverRequestUrl) {
        this.platform = platform;
        this.document = document;
        this.config = config;
        this.sitesConfigLoader = sitesConfigLoader;
        this.converter = converter;
        this.transferState = transferState;
        this.serverRequestUrl = serverRequestUrl;
    }
    Object.defineProperty(OccConfigLoaderService.prototype, "currentUrl", {
        get: function () {
            if (isPlatformBrowser(this.platform)) {
                return this.document.location.href;
            }
            if (this.serverRequestUrl) {
                return this.serverRequestUrl;
            }
            if (isDevMode()) {
                console.error("Please provide token 'SERVER_REQUEST_URL' with the requested URL for SSR");
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initializes the Spartacus config asynchronously basing on the external config
     */
    OccConfigLoaderService.prototype.loadConfig = function () {
        var _this = this;
        return this.get()
            .pipe(tap(function (externalConfig) { return _this.transfer(externalConfig); }), map(function (externalConfig) {
            return deepMerge.apply(void 0, __spread([{}], _this.getConfigChunks(externalConfig)));
        }))
            .toPromise();
    };
    /**
     * Returns the external config
     */
    OccConfigLoaderService.prototype.get = function () {
        var rehydratedExternalConfig = this.rehydrate();
        return rehydratedExternalConfig
            ? of(rehydratedExternalConfig)
            : this.load();
    };
    /**
     * Loads the external config from backend
     */
    OccConfigLoaderService.prototype.load = function () {
        var _this = this;
        return this.sitesConfigLoader
            .load()
            .pipe(map(function (baseSites) {
            return _this.converter.fromOccBaseSites(baseSites, _this.currentUrl);
        }));
    };
    /**
     * Tries to rehydrate external config in the browser from SSR
     */
    OccConfigLoaderService.prototype.rehydrate = function () {
        if (this.transferState && isPlatformBrowser(this.platform)) {
            return this.transferState.get(EXTERNAL_CONFIG_TRANSFER_ID, undefined);
        }
    };
    /**
     * Transfers the given external config in SSR to the browser
     *
     * @param externalConfig
     */
    OccConfigLoaderService.prototype.transfer = function (externalConfig) {
        if (this.transferState &&
            isPlatformServer(this.platform) &&
            externalConfig) {
            this.transferState.set(EXTERNAL_CONFIG_TRANSFER_ID, externalConfig);
        }
    };
    OccConfigLoaderService.prototype.getConfigChunks = function (externalConfig) {
        var chunks = [this.converter.toSiteContextConfig(externalConfig)];
        if (this.shouldReturnI18nChunk()) {
            chunks.push(this.converter.toI18nConfig(externalConfig));
        }
        return chunks;
    };
    OccConfigLoaderService.prototype.shouldReturnI18nChunk = function () {
        var fallbackLangExists = typeof (this.config &&
            this.config.i18n &&
            this.config.i18n.fallbackLang) !== 'undefined';
        if (fallbackLangExists && isDevMode()) {
            console.warn("There is an already provided static config for 'i18n.fallbackLang', so the value from OCC loaded config is ignored.");
        }
        return !fallbackLangExists;
    };
    OccConfigLoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [Config,] }] },
        { type: OccSitesConfigLoader },
        { type: OccLoadedConfigConverter },
        { type: TransferState, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [SERVER_REQUEST_URL,] }] }
    ]; };
    OccConfigLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccConfigLoaderService_Factory() { return new OccConfigLoaderService(i0.ɵɵinject(i0.PLATFORM_ID), i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i2.Config), i0.ɵɵinject(i3.OccSitesConfigLoader), i0.ɵɵinject(i4.OccLoadedConfigConverter), i0.ɵɵinject(i5.TransferState, 8), i0.ɵɵinject(i6.SERVER_REQUEST_URL, 8)); }, token: OccConfigLoaderService, providedIn: "root" });
    OccConfigLoaderService = __decorate([
        Injectable({ providedIn: 'root' }),
        __param(0, Inject(PLATFORM_ID)),
        __param(1, Inject(DOCUMENT)),
        __param(2, Inject(Config)),
        __param(5, Optional()),
        __param(6, Optional()),
        __param(6, Inject(SERVER_REQUEST_URL))
    ], OccConfigLoaderService);
    return OccConfigLoaderService;
}());
export { OccConfigLoaderService };
