// @spartacus/core/esm5/src/occ/config-loader/occ-loaded-config-converter.js
import { __decorate, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { BASE_SITE_CONTEXT_ID, CURRENCY_CONTEXT_ID, LANGUAGE_CONTEXT_ID, } from '../../site-context/providers/context-ids';
import { JavaRegExpConverter } from './java-reg-exp-converter';
import * as i0 from "@angular/core";
import * as i1 from "./java-reg-exp-converter";
var OccLoadedConfigConverter = /** @class */ (function () {
    function OccLoadedConfigConverter(javaRegExpConverter) {
        this.javaRegExpConverter = javaRegExpConverter;
    }
    OccLoadedConfigConverter.prototype.fromOccBaseSites = function (baseSites, currentUrl) {
        var _this = this;
        var baseSite = baseSites.find(function (site) {
            return _this.isCurrentBaseSite(site, currentUrl);
        });
        if (!baseSite) {
            throw this.getError("Current url (" + currentUrl + ") doesn't match with any of url patterns of any base site.");
        }
        // Although `stores` property is an array, typically there is only one store. So we return the first store from the list.
        var baseStore = baseSite.stores && baseSite.stores[0];
        if (!baseStore) {
            throw this.getError("Current base site (" + baseSite.uid + ") doesn't have any base store.");
        }
        return {
            baseSite: baseSite.uid,
            languages: this.getIsoCodes(baseStore.languages, baseSite.defaultLanguage || baseStore.defaultLanguage),
            currencies: this.getIsoCodes(baseStore.currencies, baseStore.defaultCurrency),
            urlParameters: this.getUrlParams(baseSite.urlEncodingAttributes),
        };
    };
    OccLoadedConfigConverter.prototype.toSiteContextConfig = function (_a) {
        var _b;
        var baseSite = _a.baseSite, languages = _a.languages, currencies = _a.currencies, urlEncodingAttributes = _a.urlParameters;
        var result = {
            context: (_b = {
                    urlParameters: urlEncodingAttributes
                },
                _b[BASE_SITE_CONTEXT_ID] = [baseSite],
                _b[LANGUAGE_CONTEXT_ID] = languages,
                _b[CURRENCY_CONTEXT_ID] = currencies,
                _b),
        };
        return result;
    };
    OccLoadedConfigConverter.prototype.toI18nConfig = function (_a) {
        var languages = _a.languages;
        return { i18n: { fallbackLang: languages[0] } };
    };
    OccLoadedConfigConverter.prototype.isCurrentBaseSite = function (site, currentUrl) {
        var _this = this;
        var index = (site.urlPatterns || []).findIndex(function (javaRegexp) {
            var jsRegexp = _this.javaRegExpConverter.toJsRegExp(javaRegexp);
            if (jsRegexp) {
                var result = jsRegexp.test(currentUrl);
                return result;
            }
        });
        return index !== -1;
    };
    /**
     * Returns an array of url encoded site context parameters.
     *
     * It maps the string "storefront" (used in OCC) to the "baseSite" (used in Spartacus)
     */
    OccLoadedConfigConverter.prototype.getUrlParams = function (params) {
        var STOREFRONT_PARAM = 'storefront';
        return (params || []).map(function (param) {
            return param === STOREFRONT_PARAM ? BASE_SITE_CONTEXT_ID : param;
        });
    };
    /**
     * Returns iso codes in a array, where the first element is the default iso code.
     */
    OccLoadedConfigConverter.prototype.getIsoCodes = function (elements, defaultElement) {
        var result = this.moveToFirst(elements, function (el) { return el.isocode === defaultElement.isocode; }).map(function (el) { return el.isocode; });
        return result;
    };
    /**
     * Moves to the start of the array the first element that satisfies the given predicate.
     *
     * @param array array to modify
     * @param predicate function called on elements
     */
    OccLoadedConfigConverter.prototype.moveToFirst = function (array, predicate) {
        array = __spread(array);
        var index = array.findIndex(predicate);
        if (index !== -1) {
            var _a = __read(array.splice(index, 1), 1), el = _a[0];
            array.unshift(el);
        }
        return array;
    };
    OccLoadedConfigConverter.prototype.getError = function (message) {
        return new Error("Error: Cannot get base site config! " + message);
    };
    OccLoadedConfigConverter.ctorParameters = function () { return [
        { type: JavaRegExpConverter }
    ]; };
    OccLoadedConfigConverter.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccLoadedConfigConverter_Factory() { return new OccLoadedConfigConverter(i0.ɵɵinject(i1.JavaRegExpConverter)); }, token: OccLoadedConfigConverter, providedIn: "root" });
    OccLoadedConfigConverter = __decorate([
        Injectable({ providedIn: 'root' })
    ], OccLoadedConfigConverter);
    return OccLoadedConfigConverter;
}());
export { OccLoadedConfigConverter };
