// @spartacus/storefront/esm5/cms-components/product/product-list/product-facet-navigation/active-facets/active-facets.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ICON_TYPE } from '../../../../../cms-components/misc/icon/icon.model';
import { FacetService } from '../services/facet.service';
/**
 * Active facets render the applied facet values as a list of focusable buttons
 * which can be used to remove the applied facet value.
 */
var ActiveFacetsComponent = /** @class */ (function () {
    function ActiveFacetsComponent(facetService) {
        this.facetService = facetService;
        /** Active facets which are applied to the product results. */
        this.facetList$ = this.facetService.facetList$;
        /** Configurable icon which is used for the active facet close button */
        this.closeIcon = ICON_TYPE.CLOSE;
    }
    ActiveFacetsComponent.prototype.getLinkParams = function (facet) {
        var _a, _b;
        return this.facetService.getLinkParams((_b = (_a = facet.removeQuery) === null || _a === void 0 ? void 0 : _a.query) === null || _b === void 0 ? void 0 : _b.value);
    };
    /**
     * The focus key is used to persist the focus on the facet when the DOM is being
     * recreated. We only apply the focus key for the given _active_ facet when there
     * the original facets is not available. This happens for non multi-valued facets.
     *
     * With this approach, the we keep the focus, either at the facet list or on the
     * active facets.
     */
    ActiveFacetsComponent.prototype.getFocusKey = function (facetList, facet) {
        var _a;
        return ((_a = facetList.facets) === null || _a === void 0 ? void 0 : _a.find(function (f) { var _a; return (_a = f.values) === null || _a === void 0 ? void 0 : _a.find(function (val) { return val.name === facet.facetValueName; }); })) ? ''
            : facet.facetValueName;
    };
    ActiveFacetsComponent.ctorParameters = function () { return [
        { type: FacetService }
    ]; };
    __decorate([
        Input()
    ], ActiveFacetsComponent.prototype, "closeIcon", void 0);
    ActiveFacetsComponent = __decorate([
        Component({
            selector: 'cx-active-facets',
            template: "<ng-container *ngIf=\"facetList$ | async as facetList\">\r\n  <h4 *ngIf=\"facetList?.activeFacets?.length > 0\">\r\n    {{ 'productList.appliedFilter' | cxTranslate }}\r\n  </h4>\r\n\r\n  <a\r\n    *ngFor=\"let facet of facetList?.activeFacets\"\r\n    routerLink=\"./\"\r\n    [queryParams]=\"getLinkParams(facet)\"\r\n    [cxFocus]=\"getFocusKey(facetList, facet)\"\r\n  >\r\n    <span>{{ facet.facetValueName }}</span>\r\n    <cx-icon aria-hidden=\"true\" [type]=\"closeIcon\"></cx-icon>\r\n  </a>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.Default
        })
    ], ActiveFacetsComponent);
    return ActiveFacetsComponent;
}());
export { ActiveFacetsComponent };
