// @spartacus/storefront/esm5/cms-structure/page/page-layout/page-layout.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Input, Renderer2, } from '@angular/core';
import { BehaviorSubject, of } from 'rxjs';
import { distinctUntilChanged, switchMap, tap } from 'rxjs/operators';
import { PageLayoutService } from './page-layout.service';
var PageLayoutComponent = /** @class */ (function () {
    function PageLayoutComponent(el, renderer, pageLayoutService) {
        var _this = this;
        this.el = el;
        this.renderer = renderer;
        this.pageLayoutService = pageLayoutService;
        this.section$ = new BehaviorSubject(undefined);
        this.templateName$ = this.pageLayoutService
            .templateName$;
        this.layoutName$ = this.section$.pipe(switchMap(function (section) { return (section ? of(section) : _this.templateName$); }), tap(function (name) {
            _this.styleClass = name;
        }));
        this.slots$ = this.section$.pipe(switchMap(function (section) { return _this.pageLayoutService.getSlots(section); }));
        this.pageFoldSlot$ = this.templateName$.pipe(switchMap(function (templateName) {
            return _this.pageLayoutService.getPageFoldSlot(templateName);
        }), distinctUntilChanged());
    }
    Object.defineProperty(PageLayoutComponent.prototype, "section", {
        set: function (value) {
            this.section$.next(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PageLayoutComponent.prototype, "styleClass", {
        set: function (cls) {
            if (this.currentClass) {
                this.renderer.removeClass(this.el.nativeElement, this.currentClass);
            }
            this.renderer.addClass(this.el.nativeElement, cls);
            this.currentClass = cls;
        },
        enumerable: true,
        configurable: true
    });
    PageLayoutComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: PageLayoutService }
    ]; };
    __decorate([
        Input()
    ], PageLayoutComponent.prototype, "section", null);
    PageLayoutComponent = __decorate([
        Component({
            selector: 'cx-page-layout',
            template: "<ng-template\n  [cxOutlet]=\"layoutName$ | async\"\n  [cxOutletContext]=\"{\n    templateName$: templateName$,\n    slots$: slots$,\n    section$: section$\n  }\"\n>\n  <ng-content></ng-content>\n\n  <cx-page-slot\n    *ngFor=\"let slot of slots$ | async\"\n    [position]=\"slot\"\n    [isPageFold]=\"slot === (pageFoldSlot$ | async)\"\n  ></cx-page-slot>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PageLayoutComponent);
    return PageLayoutComponent;
}());
export { PageLayoutComponent };
