// @spartacus/storefront/esm5/cms-components/checkout/components/checkout-progress/checkout-progress.component.js
import { __decorate } from "tslib";
import { RoutingService, RoutingConfigService } from '@spartacus/core';
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { CheckoutConfig } from '../../config/checkout-config';
import { tap } from 'rxjs/operators';
var CheckoutProgressComponent = /** @class */ (function () {
    function CheckoutProgressComponent(config, routingService, routingConfigService) {
        this.config = config;
        this.routingService = routingService;
        this.routingConfigService = routingConfigService;
    }
    CheckoutProgressComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.steps = this.config.checkout.steps;
        this.routerState$ = this.routingService.getRouterState().pipe(tap(function (router) {
            _this.activeStepUrl = router.state.context.id;
            _this.steps.forEach(function (step, index) {
                var routeUrl = "/" + _this.routingConfigService.getRouteConfig(step.routeName).paths[0];
                if (routeUrl === _this.activeStepUrl) {
                    _this.activeStepIndex = index;
                }
            });
        }));
    };
    CheckoutProgressComponent.prototype.getTabIndex = function (stepIndex) {
        return !this.isActive(stepIndex) && !this.isDisabled(stepIndex) ? 0 : -1;
    };
    CheckoutProgressComponent.prototype.isActive = function (index) {
        return index === this.activeStepIndex;
    };
    CheckoutProgressComponent.prototype.isDisabled = function (index) {
        return index > this.activeStepIndex;
    };
    CheckoutProgressComponent.ctorParameters = function () { return [
        { type: CheckoutConfig },
        { type: RoutingService },
        { type: RoutingConfigService }
    ]; };
    CheckoutProgressComponent = __decorate([
        Component({
            selector: 'cx-checkout-progress',
            template: "<section *ngIf=\"routerState$ | async as routerState\">\r\n  <div class=\"cx-nav d-none d-lg-block d-xl-block\">\r\n    <ul class=\"cx-list\">\r\n      <li class=\"cx-item\" *ngFor=\"let step of steps; let i = index\">\r\n        <a\r\n          [routerLink]=\"{ cxRoute: step.routeName } | cxUrl\"\r\n          class=\"cx-link\"\r\n          [class.active]=\"isActive(i)\"\r\n          [class.disabled]=\"isDisabled(i)\"\r\n          [tabindex]=\"getTabIndex(i)\"\r\n        >\r\n          {{ i + 1 }}. {{ step.name | cxTranslate }}\r\n        </a>\r\n      </li>\r\n    </ul>\r\n  </div>\r\n</section>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CheckoutProgressComponent);
    return CheckoutProgressComponent;
}());
export { CheckoutProgressComponent };
