// @spartacus/storefront/esm2015/cms-components/product/product-list/product-facet-navigation/facet-list/facet-list.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, HostListener, Input, Output, Renderer2, } from '@angular/core';
import { map } from 'rxjs/operators';
import { ICON_TYPE } from '../../../../misc/icon/icon.model';
import { FacetGroupCollapsedState } from '../facet.model';
import { FacetService } from '../services/facet.service';
let FacetListComponent = class FacetListComponent {
    constructor(facetService, elementRef, renderer) {
        this.facetService = facetService;
        this.elementRef = elementRef;
        this.renderer = renderer;
        /** Emits when the list must close */
        this.closeList = new EventEmitter();
        /** The list of all facet and values related to the products in the list */
        this.facetList$ = this.facetService.facetList$;
        this.iconTypes = ICON_TYPE;
        this.dialogFocusConfig = {
            trap: true,
            block: true,
            focusOnEscape: true,
            autofocus: 'cx-facet',
        };
    }
    /**
     * Indicates that the facet navigation is rendered in dialog.
     */
    set isDialog(value) {
        this._isDialog = value;
        if (value) {
            this.renderer.addClass(document.body, 'modal-open');
        }
    }
    get isDialog() {
        return this._isDialog;
    }
    handleClick() {
        this.close();
    }
    /**
     * Toggles the facet group in case it is not expanded.
     */
    expandFacetGroup(facet, ref) {
        if (!ref.isExpanded) {
            this.facetService.toggle(facet, ref.isExpanded);
        }
    }
    /**
     * Indicates that the facet group has been expanded.
     */
    isExpanded(facet) {
        return this.facetService
            .getState(facet)
            .pipe(map((value) => value.toggled === FacetGroupCollapsedState.EXPANDED));
    }
    /**
     * Indicates that the facet group has been collapsed.
     */
    isCollapsed(facet) {
        return this.facetService
            .getState(facet)
            .pipe(map((value) => value.toggled === FacetGroupCollapsedState.COLLAPSED));
    }
    close(event) {
        this.renderer.removeClass(document.body, 'modal-open');
        this.closeList.emit(event);
    }
    block(event) {
        event.stopPropagation();
    }
};
FacetListComponent.ctorParameters = () => [
    { type: FacetService },
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input()
], FacetListComponent.prototype, "isDialog", null);
__decorate([
    Output()
], FacetListComponent.prototype, "closeList", void 0);
__decorate([
    HostListener('click')
], FacetListComponent.prototype, "handleClick", null);
FacetListComponent = __decorate([
    Component({
        selector: 'cx-facet-list',
        template: "<div\r\n  class=\"inner\"\r\n  *ngIf=\"(facetList$ | async)?.facets as facets\"\r\n  [cxFocus]=\"isDialog ? dialogFocusConfig : {}\"\r\n  (esc)=\"close($event)\"\r\n  (click)=\"block($event)\"\r\n>\r\n  <h4>\r\n    {{ 'productList.filterBy.label' | cxTranslate }}\r\n    <button type=\"button\" class=\"close\" aria-label=\"Close\" (click)=\"close()\">\r\n      <cx-icon aria-hidden=\"true\" [type]=\"iconTypes.CLOSE\"></cx-icon>\r\n    </button>\r\n  </h4>\r\n\r\n  <!-- \r\n      Here we'd like to introduce configurable facet components, \r\n      either by using specific configuration or generic sproutlets \r\n  -->\r\n  <cx-facet\r\n    *ngFor=\"let facet of facets\"\r\n    #facetRef\r\n    [facet]=\"facet\"\r\n    [cxFocus]=\"{ lock: true, trap: true, autofocus: 'a' }\"\r\n    (unlock)=\"expandFacetGroup(facet, facetRef)\"\r\n    [class.expanded]=\"isExpanded(facet) | async\"\r\n    [class.collapsed]=\"isCollapsed(facet) | async\"\r\n  ></cx-facet>\r\n</div>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], FacetListComponent);
export { FacetListComponent };
