// @spartacus/storefront/esm5/cms-components/myaccount/my-interests/my-interests.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { ListNavigationModule } from '../../../shared/components/list-navigation/list-navigation.module';
import { MyInterestsComponent } from './my-interests.component';
import { MediaModule } from '../../../shared/components/media/media.module';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { AuthGuard, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
var MyInterestsModule = /** @class */ (function () {
    function MyInterestsModule() {
    }
    MyInterestsModule = __decorate([
        NgModule({
            declarations: [MyInterestsComponent],
            imports: [
                CommonModule,
                I18nModule,
                RouterModule,
                ListNavigationModule,
                I18nModule,
                UrlModule,
                MediaModule,
                SpinnerModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        MyInterestsComponent: {
                            component: MyInterestsComponent,
                            guards: [AuthGuard],
                        },
                    },
                }),
            ],
            exports: [MyInterestsComponent],
            entryComponents: [MyInterestsComponent],
        })
    ], MyInterestsModule);
    return MyInterestsModule;
}());
export { MyInterestsModule };
