// @spartacus/storefront/esm5/cms-components/cart/mini-cart/mini-cart.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { IconModule } from '../../../cms-components/misc/icon/icon.module';
import { MiniCartComponent } from './mini-cart.component';
var MiniCartModule = /** @class */ (function () {
    function MiniCartModule() {
    }
    MiniCartModule = __decorate([
        NgModule({
            imports: [CommonModule, RouterModule, UrlModule, IconModule, I18nModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        MiniCartComponent: {
                            component: MiniCartComponent,
                        },
                    },
                }),
            ],
            declarations: [MiniCartComponent],
            exports: [MiniCartComponent],
            entryComponents: [MiniCartComponent],
        })
    ], MiniCartModule);
    return MiniCartModule;
}());
export { MiniCartModule };
