// @spartacus/storefront/esm5/cms-components/checkout/components/shipping-address/shipping-address.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { CardModule } from '../../../../shared/components/card/card.module';
import { SpinnerModule } from '../../../../shared/components/spinner/spinner.module';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
// tslint:disable-next-line:max-line-length
import { CheckoutProgressMobileBottomModule } from '../checkout-progress/checkout-progress-mobile-bottom/checkout-progress-mobile-bottom.module';
import { CheckoutProgressMobileTopModule } from '../checkout-progress/checkout-progress-mobile-top/checkout-progress-mobile-top.module';
import { CartNotEmptyGuard } from './../../../../cms-components/cart/cart-not-empty.guard';
import { AddressFormModule } from './address-form/address-form.module';
import { CheckoutDetailsLoadedGuard } from '../../guards/checkout-details-loaded.guard';
import { ShippingAddressComponent } from './shipping-address.component';
var ShippingAddressModule = /** @class */ (function () {
    function ShippingAddressModule() {
    }
    ShippingAddressModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                AddressFormModule,
                CardModule,
                SpinnerModule,
                I18nModule,
                CheckoutProgressMobileTopModule,
                CheckoutProgressMobileBottomModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CheckoutShippingAddress: {
                            component: ShippingAddressComponent,
                            guards: [
                                CheckoutAuthGuard,
                                CartNotEmptyGuard,
                                CheckoutDetailsLoadedGuard,
                            ],
                        },
                    },
                }),
            ],
            declarations: [ShippingAddressComponent],
            entryComponents: [ShippingAddressComponent],
            exports: [ShippingAddressComponent],
        })
    ], ShippingAddressModule);
    return ShippingAddressModule;
}());
export { ShippingAddressModule };
