// @spartacus/storefront/esm5/cms-components/checkout/components/review-submit/review-submit.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { FeaturesConfigModule, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { CartSharedModule } from '../../../../cms-components/cart/cart-shared/cart-shared.module';
import { CardModule } from '../../../../shared/components/card/card.module';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { DeliveryModeSetGuard } from '../../guards/delivery-mode-set.guard';
import { PaymentDetailsSetGuard } from '../../guards/payment-details-set.guard';
import { ShippingAddressSetGuard } from '../../guards/shipping-address-set.guard';
import { CartNotEmptyGuard } from './../../../../cms-components/cart/cart-not-empty.guard';
import { ReviewSubmitComponent } from './review-submit.component';
import { PromotionsModule } from '../promotions/promotions.module';
var ReviewSubmitModule = /** @class */ (function () {
    function ReviewSubmitModule() {
    }
    ReviewSubmitModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                CardModule,
                CartSharedModule,
                I18nModule,
                UrlModule,
                RouterModule,
                PromotionsModule,
                FeaturesConfigModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CheckoutReviewOrder: {
                            component: ReviewSubmitComponent,
                            guards: [
                                CheckoutAuthGuard,
                                CartNotEmptyGuard,
                                ShippingAddressSetGuard,
                                DeliveryModeSetGuard,
                                PaymentDetailsSetGuard,
                            ],
                        },
                    },
                }),
            ],
            declarations: [ReviewSubmitComponent],
            entryComponents: [ReviewSubmitComponent],
            exports: [ReviewSubmitComponent],
        })
    ], ReviewSubmitModule);
    return ReviewSubmitModule;
}());
export { ReviewSubmitModule };
