// @spartacus/storefront/esm5/cms-components/user/register/register.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { SpinnerModule, FormErrorsModule } from '../../../shared/index';
import { LoginModule } from '../login/login.module';
import { RegisterComponent } from './register.component';
var RegisterComponentModule = /** @class */ (function () {
    function RegisterComponentModule() {
    }
    RegisterComponentModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                LoginModule,
                ReactiveFormsModule,
                RouterModule,
                UrlModule,
                I18nModule,
                SpinnerModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        RegisterCustomerComponent: {
                            component: RegisterComponent,
                            guards: [NotAuthGuard],
                        },
                    },
                }),
            ],
            declarations: [RegisterComponent],
            exports: [RegisterComponent],
            entryComponents: [RegisterComponent],
        })
    ], RegisterComponentModule);
    return RegisterComponentModule;
}());
export { RegisterComponentModule };
