// @spartacus/storefront/esm2015/cms-components/myaccount/consent-management/consent-management.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { IconModule } from '../../misc/icon/icon.module';
import { ConsentManagementFormComponent } from './components/consent-form/consent-management-form.component';
import { ConsentManagementComponent } from './components/consent-management.component';
let ConsentManagementModule = class ConsentManagementModule {
};
ConsentManagementModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
            SpinnerModule,
            I18nModule,
            IconModule,
        ],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    ConsentManagementComponent: {
                        component: ConsentManagementComponent,
                        guards: [AuthGuard],
                    },
                },
            }),
        ],
        declarations: [ConsentManagementComponent, ConsentManagementFormComponent],
        exports: [ConsentManagementComponent, ConsentManagementFormComponent],
        entryComponents: [ConsentManagementComponent],
    })
], ConsentManagementModule);
export { ConsentManagementModule };
