// @spartacus/core/esm5/src/config/config-initializer/config-initializer.module.js
import { __decorate } from "tslib";
import { APP_INITIALIZER, NgModule, Optional, } from '@angular/core';
import { ConfigInitializerService } from './config-initializer.service';
import { CONFIG_INITIALIZER, CONFIG_INITIALIZER_FORROOT_GUARD, } from './config-initializer';
export function configInitializerFactory(configInitializer, initializers) {
    var isReady = function () { return configInitializer.initialize(initializers); };
    return isReady;
}
var ConfigInitializerModule = /** @class */ (function () {
    function ConfigInitializerModule() {
    }
    ConfigInitializerModule_1 = ConfigInitializerModule;
    ConfigInitializerModule.forRoot = function () {
        return {
            ngModule: ConfigInitializerModule_1,
            providers: [
                {
                    provide: CONFIG_INITIALIZER_FORROOT_GUARD,
                    useValue: true,
                },
                {
                    provide: APP_INITIALIZER,
                    multi: true,
                    useFactory: configInitializerFactory,
                    deps: [
                        ConfigInitializerService,
                        [new Optional(), CONFIG_INITIALIZER],
                    ],
                },
            ],
        };
    };
    var ConfigInitializerModule_1;
    ConfigInitializerModule = ConfigInitializerModule_1 = __decorate([
        NgModule({})
    ], ConfigInitializerModule);
    return ConfigInitializerModule;
}());
export { ConfigInitializerModule };
