// @spartacus/core/esm5/src/site-context/services/site-context-url-serializer.js
import { __decorate, __extends, __read } from "tslib";
import { Injectable } from '@angular/core';
import { DefaultUrlSerializer } from '@angular/router';
import { SiteContextParamsService } from './site-context-params.service';
var UrlSplit = /(^[^#?]*)(.*)/; // used to split url into path and query/fragment parts
var SiteContextUrlSerializer = /** @class */ (function (_super) {
    __extends(SiteContextUrlSerializer, _super);
    function SiteContextUrlSerializer(siteContextParams) {
        var _this = _super.call(this) || this;
        _this.siteContextParams = siteContextParams;
        return _this;
    }
    Object.defineProperty(SiteContextUrlSerializer.prototype, "urlEncodingParameters", {
        get: function () {
            return this.siteContextParams.getUrlEncodingParameters();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SiteContextUrlSerializer.prototype, "hasContextInRoutes", {
        get: function () {
            return this.urlEncodingParameters.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    SiteContextUrlSerializer.prototype.parse = function (url) {
        if (this.hasContextInRoutes) {
            var urlWithParams = this.urlExtractContextParameters(url);
            var parsed = _super.prototype.parse.call(this, urlWithParams.url);
            this.urlTreeIncludeContextParameters(parsed, urlWithParams.params);
            return parsed;
        }
        else {
            return _super.prototype.parse.call(this, url);
        }
    };
    SiteContextUrlSerializer.prototype.urlExtractContextParameters = function (url) {
        var _a = __read(url.match(UrlSplit), 3), urlPart = _a[1], queryPart = _a[2];
        var segments = urlPart.split('/');
        if (segments[0] === '') {
            segments.shift();
        }
        var params = {};
        var paramId = 0;
        var segmentId = 0;
        while (paramId < this.urlEncodingParameters.length &&
            segmentId < segments.length) {
            var paramName = this.urlEncodingParameters[paramId];
            var paramValues = this.siteContextParams.getParamValues(paramName);
            if (paramValues.includes(segments[segmentId])) {
                params[paramName] = segments[segmentId];
                segmentId++;
            }
            paramId++;
        }
        url = segments.slice(Object.keys(params).length).join('/') + queryPart;
        return { url: url, params: params };
    };
    SiteContextUrlSerializer.prototype.urlTreeIncludeContextParameters = function (urlTree, params) {
        urlTree.siteContext = params;
    };
    SiteContextUrlSerializer.prototype.serialize = function (tree) {
        var params = this.urlTreeExtractContextParameters(tree);
        var url = _super.prototype.serialize.call(this, tree);
        var serialized = this.urlIncludeContextParameters(url, params);
        return serialized;
    };
    SiteContextUrlSerializer.prototype.urlTreeExtractContextParameters = function (urlTree) {
        return urlTree.siteContext ? urlTree.siteContext : {};
    };
    SiteContextUrlSerializer.prototype.urlIncludeContextParameters = function (url, params) {
        var _this = this;
        var contextRoutePart = this.urlEncodingParameters
            .map(function (param) {
            return params[param]
                ? params[param]
                : _this.siteContextParams.getValue(param);
        })
            .join('/');
        return contextRoutePart + url;
    };
    SiteContextUrlSerializer.ctorParameters = function () { return [
        { type: SiteContextParamsService }
    ]; };
    SiteContextUrlSerializer = __decorate([
        Injectable()
    ], SiteContextUrlSerializer);
    return SiteContextUrlSerializer;
}(DefaultUrlSerializer));
export { SiteContextUrlSerializer };
