// @spartacus/core/esm5/src/site-context/services/site-context-params.service.js
import { __decorate } from "tslib";
import { Injectable, Injector } from '@angular/core';
import { combineLatest, of } from 'rxjs';
import { distinctUntilChanged, filter } from 'rxjs/operators';
import { getContextParameterDefault, getContextParameterValues, } from '../config/context-config-utils';
import { SiteContextConfig } from '../config/site-context-config';
import { ContextServiceMap } from '../providers/context-service-map';
var SiteContextParamsService = /** @class */ (function () {
    function SiteContextParamsService(config, injector, serviceMap) {
        this.config = config;
        this.injector = injector;
        this.serviceMap = serviceMap;
    }
    SiteContextParamsService.prototype.getContextParameters = function () {
        if (this.config.context) {
            return Object.keys(this.config.context).filter(function (param) { return param !== 'urlParameters'; });
        }
        return [];
    };
    SiteContextParamsService.prototype.getUrlEncodingParameters = function () {
        return (this.config.context && this.config.context.urlParameters) || [];
    };
    SiteContextParamsService.prototype.getParamValues = function (param) {
        return getContextParameterValues(this.config, param);
    };
    SiteContextParamsService.prototype.getParamDefaultValue = function (param) {
        return getContextParameterDefault(this.config, param);
    };
    SiteContextParamsService.prototype.getSiteContextService = function (param) {
        if (this.serviceMap[param]) {
            return this.injector.get(this.serviceMap[param], null);
        }
    };
    SiteContextParamsService.prototype.getValue = function (param) {
        var value;
        var service = this.getSiteContextService(param);
        if (service) {
            service
                .getActive()
                .subscribe(function (val) { return (value = val); })
                .unsubscribe();
        }
        return value !== undefined ? value : this.getParamDefaultValue(param);
    };
    SiteContextParamsService.prototype.setValue = function (param, value) {
        var service = this.getSiteContextService(param);
        if (service) {
            service.setActive(value);
        }
    };
    /**
     * Get active values for all provided context parameters
     *
     * @param params Context parameters
     *
     * @returns Observable emitting array of all passed active context values
     */
    SiteContextParamsService.prototype.getValues = function (params) {
        var _this = this;
        if (params.length === 0) {
            return of([]);
        }
        return combineLatest(params.map(function (param) {
            return _this.getSiteContextService(param)
                .getActive()
                .pipe(distinctUntilChanged());
        })).pipe(filter(function (value) { return value.every(function (param) { return !!param; }); }));
    };
    SiteContextParamsService.ctorParameters = function () { return [
        { type: SiteContextConfig },
        { type: Injector },
        { type: ContextServiceMap }
    ]; };
    SiteContextParamsService = __decorate([
        Injectable()
    ], SiteContextParamsService);
    return SiteContextParamsService;
}());
export { SiteContextParamsService };
