// @spartacus/storefront/esm5/shared/components/media/media.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Input, Output, } from '@angular/core';
import { MediaService } from './media.service';
var MediaComponent = /** @class */ (function () {
    function MediaComponent(mediaService) {
        this.mediaService = mediaService;
        /**
         * Once the media is loaded, we emit an event.
         */
        this.loaded = new EventEmitter();
        /**
         * The `cx-media` component has an `is-initialized` class as long as the
         * media is being initialized.
         */
        this.isInitialized = false;
        /**
         * The `cx-media` component has a `is-loading` class as long as the
         * media is loaded. Wehn the media is loaded, the `is-initialized` class
         * is added.
         */
        this.isLoading = true;
        /**
         * When there's no media provided for the content, or in case an error
         * happened during loading, we add the `is-missing` class. Visual effects
         * can be controlled by CSS.
         */
        this.isMissing = false;
    }
    MediaComponent.prototype.ngOnChanges = function () {
        this.create();
    };
    /**
     * Creates the `Media` object
     */
    MediaComponent.prototype.create = function () {
        var _a;
        this.media = this.mediaService.getMedia(this.container, this.format, this.alt);
        if (!((_a = this.media) === null || _a === void 0 ? void 0 : _a.src)) {
            this.handleMissing();
        }
    };
    /**
     * This handler is called from the UI when the image is loaded.
     */
    MediaComponent.prototype.loadHandler = function () {
        this.isLoading = false;
        this.isInitialized = true;
        this.isMissing = false;
        this.loaded.emit(true);
    };
    /**
     * Whenever an error happens during load, we mark the component
     * with css classes to have a missing media.
     */
    MediaComponent.prototype.errorHandler = function () {
        this.handleMissing();
    };
    MediaComponent.prototype.handleMissing = function () {
        this.isLoading = false;
        this.isInitialized = true;
        this.isMissing = true;
        this.loaded.emit(false);
    };
    MediaComponent.ctorParameters = function () { return [
        { type: MediaService }
    ]; };
    __decorate([
        Input()
    ], MediaComponent.prototype, "container", void 0);
    __decorate([
        Input()
    ], MediaComponent.prototype, "format", void 0);
    __decorate([
        Input()
    ], MediaComponent.prototype, "alt", void 0);
    __decorate([
        Output()
    ], MediaComponent.prototype, "loaded", void 0);
    __decorate([
        HostBinding('class.is-initialized')
    ], MediaComponent.prototype, "isInitialized", void 0);
    __decorate([
        HostBinding('class.is-loading')
    ], MediaComponent.prototype, "isLoading", void 0);
    __decorate([
        HostBinding('class.is-missing')
    ], MediaComponent.prototype, "isMissing", void 0);
    MediaComponent = __decorate([
        Component({
            selector: 'cx-media',
            template: "<img\n  *ngIf=\"media?.src\"\n  [attr.src]=\"media.src\"\n  [attr.srcset]=\"media.srcset\"\n  [attr.alt]=\"media.alt\"\n  (load)=\"loadHandler()\"\n  (error)=\"errorHandler()\"\n/>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], MediaComponent);
    return MediaComponent;
}());
export { MediaComponent };
