// @spartacus/storefront/esm5/cms-structure/services/cms-i18n.service.js
import { __decorate, __values } from "tslib";
import { Injectable } from '@angular/core';
import { TranslationChunkService, TranslationService } from '@spartacus/core';
import { CmsComponentsService } from './cms-components.service';
import * as i0 from "@angular/core";
import * as i1 from "./cms-components.service";
import * as i2 from "@spartacus/core";
var CmsI18nService = /** @class */ (function () {
    function CmsI18nService(cmsComponentsService, translation, translationChunk) {
        this.cmsComponentsService = cmsComponentsService;
        this.translation = translation;
        this.translationChunk = translationChunk;
    }
    CmsI18nService.prototype.loadForComponents = function (componentTypes) {
        var e_1, _a;
        var i18nKeys = this.cmsComponentsService.getI18nKeys(componentTypes);
        var i18nChunks = new Set();
        try {
            for (var i18nKeys_1 = __values(i18nKeys), i18nKeys_1_1 = i18nKeys_1.next(); !i18nKeys_1_1.done; i18nKeys_1_1 = i18nKeys_1.next()) {
                var key = i18nKeys_1_1.value;
                i18nChunks.add(this.translationChunk.getChunkNameForKey(key));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (i18nKeys_1_1 && !i18nKeys_1_1.done && (_a = i18nKeys_1.return)) _a.call(i18nKeys_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.translation.loadChunks(Array.from(i18nChunks));
    };
    CmsI18nService.ctorParameters = function () { return [
        { type: CmsComponentsService },
        { type: TranslationService },
        { type: TranslationChunkService }
    ]; };
    CmsI18nService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CmsI18nService_Factory() { return new CmsI18nService(i0.ɵɵinject(i1.CmsComponentsService), i0.ɵɵinject(i2.TranslationService), i0.ɵɵinject(i2.TranslationChunkService)); }, token: CmsI18nService, providedIn: "root" });
    CmsI18nService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CmsI18nService);
    return CmsI18nService;
}());
export { CmsI18nService };
