// @spartacus/core/esm5/src/cart/store/actions/cart-voucher.action.js
import { __extends } from "tslib";
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { EntityFailAction, EntityLoadAction, EntityLoaderResetAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { EntityProcessesDecrementAction, EntityProcessesIncrementAction, } from '../../../state/utils/entity-processes-loader/entity-processes-loader.action';
import { ADD_VOUCHER_PROCESS_ID, MULTI_CART_DATA } from '../multi-cart-state';
export var CART_ADD_VOUCHER = '[Cart-voucher] Add Cart Vouchers';
export var CART_ADD_VOUCHER_FAIL = '[Cart-voucher] Add Cart Voucher Fail';
export var CART_ADD_VOUCHER_SUCCESS = '[Cart-voucher] Add Cart Voucher Success';
export var CART_RESET_ADD_VOUCHER = '[Cart-voucher] Reset Add Cart Voucher';
export var CART_REMOVE_VOUCHER = '[Cart-voucher] Remove Cart Voucher';
export var CART_REMOVE_VOUCHER_FAIL = '[Cart-voucher] Remove Cart Voucher Fail';
export var CART_REMOVE_VOUCHER_SUCCESS = '[Cart-voucher] Remove Cart Voucher Success';
// Adding cart voucher actions
var CartAddVoucher = /** @class */ (function (_super) {
    __extends(CartAddVoucher, _super);
    function CartAddVoucher(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER;
        return _this;
    }
    return CartAddVoucher;
}(EntityLoadAction));
export { CartAddVoucher };
var CartAddVoucherFail = /** @class */ (function (_super) {
    __extends(CartAddVoucherFail, _super);
    function CartAddVoucherFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID, payload.error) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER_FAIL;
        return _this;
    }
    return CartAddVoucherFail;
}(EntityFailAction));
export { CartAddVoucherFail };
var CartAddVoucherSuccess = /** @class */ (function (_super) {
    __extends(CartAddVoucherSuccess, _super);
    function CartAddVoucherSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER_SUCCESS;
        return _this;
    }
    return CartAddVoucherSuccess;
}(EntitySuccessAction));
export { CartAddVoucherSuccess };
// TODO(#7241): Remove when switching to event system for vouchers
/**
 * Resets add voucher process
 *
 * @deprecated since 2.0
 */
var CartResetAddVoucher = /** @class */ (function (_super) {
    __extends(CartResetAddVoucher, _super);
    function CartResetAddVoucher() {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.type = CART_RESET_ADD_VOUCHER;
        return _this;
    }
    return CartResetAddVoucher;
}(EntityLoaderResetAction));
export { CartResetAddVoucher };
// Deleting cart voucher
var CartRemoveVoucher = /** @class */ (function (_super) {
    __extends(CartRemoveVoucher, _super);
    function CartRemoveVoucher(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER;
        return _this;
    }
    return CartRemoveVoucher;
}(EntityProcessesIncrementAction));
export { CartRemoveVoucher };
var CartRemoveVoucherFail = /** @class */ (function (_super) {
    __extends(CartRemoveVoucherFail, _super);
    function CartRemoveVoucherFail(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER_FAIL;
        return _this;
    }
    return CartRemoveVoucherFail;
}(EntityProcessesDecrementAction));
export { CartRemoveVoucherFail };
var CartRemoveVoucherSuccess = /** @class */ (function (_super) {
    __extends(CartRemoveVoucherSuccess, _super);
    function CartRemoveVoucherSuccess(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER_SUCCESS;
        return _this;
    }
    return CartRemoveVoucherSuccess;
}(EntityProcessesDecrementAction));
export { CartRemoveVoucherSuccess };
