// @spartacus/core/esm5/src/cart/store/actions/wish-list.action.js
import { __extends } from "tslib";
import { EntityFailAction, EntityLoadAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { getCartIdByUserId } from '../../utils/utils';
import { MULTI_CART_DATA } from '../multi-cart-state';
export var CREATE_WISH_LIST = '[Wish List] Create Wish List';
export var CREATE_WISH_LIST_FAIL = '[Wish List] Create Wish List Fail';
export var CREATE_WISH_LIST_SUCCESS = '[Wish List] Create Wish List Success';
export var LOAD_WISH_LIST = '[Wish List] Load Wish List';
export var LOAD_WISH_LIST_SUCCESS = '[Wish List] Load Wish List Success';
export var LOAD_WISH_LIST_FAIL = '[Wish List] Load Wish List Fail';
export var RESET_WISH_LIST_DETAILS = '[Wish List] Reset Wish List';
var CreateWishList = /** @class */ (function () {
    function CreateWishList(payload) {
        this.payload = payload;
        this.type = CREATE_WISH_LIST;
    }
    return CreateWishList;
}());
export { CreateWishList };
var CreateWishListSuccess = /** @class */ (function (_super) {
    __extends(CreateWishListSuccess, _super);
    function CreateWishListSuccess(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, getCartIdByUserId(payload.cart, payload.userId)) || this;
        _this.payload = payload;
        _this.type = CREATE_WISH_LIST_SUCCESS;
        return _this;
    }
    return CreateWishListSuccess;
}(EntitySuccessAction));
export { CreateWishListSuccess };
var CreateWishListFail = /** @class */ (function (_super) {
    __extends(CreateWishListFail, _super);
    function CreateWishListFail(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId, payload.error) || this;
        _this.payload = payload;
        _this.type = CREATE_WISH_LIST_FAIL;
        return _this;
    }
    return CreateWishListFail;
}(EntityFailAction));
export { CreateWishListFail };
var LoadWishList = /** @class */ (function (_super) {
    __extends(LoadWishList, _super);
    function LoadWishList(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.tempCartId) || this;
        _this.payload = payload;
        _this.type = LOAD_WISH_LIST;
        return _this;
    }
    return LoadWishList;
}(EntityLoadAction));
export { LoadWishList };
var LoadWishListSuccess = /** @class */ (function (_super) {
    __extends(LoadWishListSuccess, _super);
    function LoadWishListSuccess(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId) || this;
        _this.payload = payload;
        _this.type = LOAD_WISH_LIST_SUCCESS;
        return _this;
    }
    return LoadWishListSuccess;
}(EntitySuccessAction));
export { LoadWishListSuccess };
var LoadWishListFail = /** @class */ (function (_super) {
    __extends(LoadWishListFail, _super);
    function LoadWishListFail(payload) {
        var _this = _super.call(this, MULTI_CART_DATA, payload.cartId, payload.error) || this;
        _this.payload = payload;
        _this.type = LOAD_WISH_LIST_FAIL;
        return _this;
    }
    return LoadWishListFail;
}(EntityFailAction));
export { LoadWishListFail };
