// @spartacus/core/esm5/src/occ/adapters/cart/converters/occ-cart-normalizer.js
import { __assign, __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { PRODUCT_NORMALIZER } from '../../../../product/connectors/product/converters';
import { Converter, ConverterService, } from '../../../../util/converter.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../../util/converter.service";
var OccCartNormalizer = /** @class */ (function () {
    function OccCartNormalizer(converter) {
        this.converter = converter;
    }
    OccCartNormalizer.prototype.convert = function (source, target) {
        var _this = this;
        if (target === undefined) {
            target = __assign({}, source);
        }
        if (source && source.entries) {
            target.entries = source.entries.map(function (entry) { return (__assign(__assign({}, entry), { product: _this.converter.convert(entry.product, PRODUCT_NORMALIZER) })); });
        }
        this.removeDuplicatePromotions(source, target);
        return target;
    };
    /**
     * Remove all duplicate promotions
     */
    OccCartNormalizer.prototype.removeDuplicatePromotions = function (source, target) {
        if (source && source.potentialOrderPromotions) {
            target.potentialOrderPromotions = this.removeDuplicateItems(source.potentialOrderPromotions);
        }
        if (source && source.potentialProductPromotions) {
            target.potentialProductPromotions = this.removeDuplicateItems(source.potentialProductPromotions);
        }
        if (source && source.appliedOrderPromotions) {
            target.appliedOrderPromotions = this.removeDuplicateItems(source.appliedOrderPromotions);
        }
        if (source && source.appliedProductPromotions) {
            target.appliedProductPromotions = this.removeDuplicateItems(source.appliedProductPromotions);
        }
    };
    OccCartNormalizer.prototype.removeDuplicateItems = function (itemList) {
        return itemList.filter(function (p, i, a) {
            var b = a.map(function (el) { return JSON.stringify(el); });
            return i === b.indexOf(JSON.stringify(p));
        });
    };
    OccCartNormalizer.ctorParameters = function () { return [
        { type: ConverterService }
    ]; };
    OccCartNormalizer.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccCartNormalizer_Factory() { return new OccCartNormalizer(i0.ɵɵinject(i1.ConverterService)); }, token: OccCartNormalizer, providedIn: "root" });
    OccCartNormalizer = __decorate([
        Injectable({ providedIn: 'root' })
    ], OccCartNormalizer);
    return OccCartNormalizer;
}());
export { OccCartNormalizer };
