// @spartacus/core/esm2015/src/global-message/http-interceptors/handlers/unauthorized/unauthorized.handler.js
import { __decorate } from "tslib";
import { Injectable, isDevMode } from '@angular/core';
import { GlobalMessageService } from '../../../facade/global-message.service';
import { GlobalMessageType } from '../../../models/global-message.model';
import { HttpResponseStatus } from '../../../models/response-status.model';
import { HttpErrorHandler } from './../http-error.handler';
import * as i0 from "@angular/core";
import * as i1 from "../../../facade/global-message.service";
/**
 * Handles Oauth client errors when a 401 is returned. This is the case for failing
 * authenticaton requests to OCC.
 */
let UnauthorizedErrorHandler = class UnauthorizedErrorHandler extends HttpErrorHandler {
    constructor(globalMessageService) {
        super(globalMessageService);
        this.globalMessageService = globalMessageService;
        this.responseStatus = HttpResponseStatus.UNAUTHORIZED;
    }
    handleError(_request, response) {
        var _a, _b;
        if (isDevMode()) {
            console.warn(`There's a problem with the "Oauth client" configuration. You must configure a matching Oauth client in the backend and Spartacus.`);
        }
        if (((_a = response.error) === null || _a === void 0 ? void 0 : _a.error) === 'invalid_client') {
            this.globalMessageService.add(((_b = response.error) === null || _b === void 0 ? void 0 : _b.error_description) || {
                key: 'httpHandlers.unauthorized.invalid_client',
            }, GlobalMessageType.MSG_TYPE_ERROR);
        }
        else {
            this.globalMessageService.add({ key: 'httpHandlers.unauthorized.common' }, GlobalMessageType.MSG_TYPE_ERROR);
        }
    }
    getPriority() {
        return -10 /* LOW */;
    }
};
UnauthorizedErrorHandler.ctorParameters = () => [
    { type: GlobalMessageService }
];
UnauthorizedErrorHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function UnauthorizedErrorHandler_Factory() { return new UnauthorizedErrorHandler(i0.ɵɵinject(i1.GlobalMessageService)); }, token: UnauthorizedErrorHandler, providedIn: "root" });
UnauthorizedErrorHandler = __decorate([
    Injectable({
        providedIn: 'root',
    })
], UnauthorizedErrorHandler);
export { UnauthorizedErrorHandler };
