// @spartacus/storefront/esm2015/cms-components/navigation/category-navigation/category-navigation.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { NavigationModule } from '../navigation/navigation.module';
import { CategoryNavigationComponent } from './category-navigation.component';
let CategoryNavigationModule = class CategoryNavigationModule {
};
CategoryNavigationModule = __decorate([
    NgModule({
        imports: [CommonModule, NavigationModule],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    CategoryNavigationComponent: {
                        component: CategoryNavigationComponent,
                    },
                },
            }),
        ],
        declarations: [CategoryNavigationComponent],
        entryComponents: [CategoryNavigationComponent],
        exports: [CategoryNavigationComponent],
    })
], CategoryNavigationModule);
export { CategoryNavigationModule };
