// @spartacus/core/esm5/src/state/services/state-persistence.service.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { of, Subscription } from 'rxjs';
import { map, tap, withLatestFrom } from 'rxjs/operators';
import { StorageSyncType } from '../../state/config/state-config';
import { getStorage, persistToStorage, readFromStorage, } from '../../state/reducers/storage-sync.reducer';
import { WindowRef } from '../../window/window-ref';
import * as i0 from "@angular/core";
import * as i1 from "../../window/window-ref";
var StatePersistenceService = /** @class */ (function () {
    function StatePersistenceService(winRef) {
        this.winRef = winRef;
    }
    /**
     * Helper to synchronize state to more persistent storage (localStorage, sessionStorage).
     * It is context aware, so you can keep different state for te same feature based on specified context.
     *
     * Eg. cart is valid only under the same base site. So you want to synchronize cart only with the same base site.
     * Usage for that case: `syncWithStorage({ key: 'cart', state$: activeCartSelector$, context$: this.siteContextParamsService.getValues([BASE_SITE_CONTEXT_ID]), onRead: (state) => setCorrectStateInStore(state) })`.
     * Active cart for the `electronics` base site will be stored under `spartacus⚿electronics⚿cart` and for apparel under `spartacus⚿apparel⚿cart`.
     *
     * On each context change onRead function will be executed with state from storage provided as a parameter.
     *
     * Omitting context$ will trigger onRead only once at initialization.
     *
     * @param key Key to use in storage for the synchronized state. Should be unique for each feature.
     * @param state$ State to be saved and later restored.
     * @param context$ Context for state
     * @param storageType Storage type to be used to persist state
     * @param onRead Function to be executed on each storage read after context change
     *
     * @returns Subscriptions for reading/writing in storage on context/state change
     */
    StatePersistenceService.prototype.syncWithStorage = function (_a) {
        var _this = this;
        var key = _a.key, state$ = _a.state$, _b = _a.context$, context$ = _b === void 0 ? of('') : _b, _c = _a.storageType, storageType = _c === void 0 ? StorageSyncType.LOCAL_STORAGE : _c, _d = _a.onRead, onRead = _d === void 0 ? function () { } : _d;
        var storage = getStorage(storageType, this.winRef);
        var subscriptions = new Subscription();
        // Do not change order of subscription! Read should happen before write on context change.
        subscriptions.add(context$
            .pipe(map(function (context) {
            return readFromStorage(storage, _this.generateKeyWithContext(context, key));
        }), tap(function (state) { return onRead(state); }))
            .subscribe());
        subscriptions.add(state$.pipe(withLatestFrom(context$)).subscribe(function (_a) {
            var _b = __read(_a, 2), state = _b[0], context = _b[1];
            persistToStorage(_this.generateKeyWithContext(context, key), state, storage);
        }));
        return subscriptions;
    };
    StatePersistenceService.prototype.generateKeyWithContext = function (context, key) {
        return "spartacus\u26BF" + [].concat(context).join('⚿') + "\u26BF" + key;
    };
    StatePersistenceService.ctorParameters = function () { return [
        { type: WindowRef }
    ]; };
    StatePersistenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function StatePersistenceService_Factory() { return new StatePersistenceService(i0.ɵɵinject(i1.WindowRef)); }, token: StatePersistenceService, providedIn: "root" });
    StatePersistenceService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], StatePersistenceService);
    return StatePersistenceService;
}());
export { StatePersistenceService };
