// @spartacus/storefront/esm5/shared/services/promotion/promotion.service.js
import { __decorate, __read, __spread, __values } from "tslib";
import { Injectable } from '@angular/core';
import { ActiveCartService, Cart, CheckoutService, Order, OrderEntry, PromotionLocation, PromotionResult, } from '@spartacus/core';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import { OrderDetailsService } from '../../../cms-components/myaccount/order/order-details/order-details.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../cms-components/myaccount/order/order-details/order-details.service";
import * as i2 from "@spartacus/core";
var PromotionService = /** @class */ (function () {
    function PromotionService(orderDetailsService, checkoutService, activeCartService) {
        this.orderDetailsService = orderDetailsService;
        this.checkoutService = checkoutService;
        this.activeCartService = activeCartService;
    }
    PromotionService.prototype.getOrderPromotions = function (promotionLocation) {
        switch (promotionLocation) {
            case PromotionLocation.ActiveCart:
                return this.getOrderPromotionsFromCart();
            case PromotionLocation.Checkout:
                return this.getOrderPromotionsFromCheckout();
            case PromotionLocation.Order:
                return this.getOrderPromotionsFromOrder();
            default:
                return of([]);
        }
    };
    PromotionService.prototype.getOrderPromotionsFromCart = function () {
        var _this = this;
        return this.activeCartService
            .getActive()
            .pipe(map(function (cart) { return _this.getOrderPromotionsFromCartHelper(cart); }));
    };
    PromotionService.prototype.getOrderPromotionsFromCartHelper = function (cart) {
        var potentialPromotions = [];
        potentialPromotions.push.apply(potentialPromotions, __spread((cart.potentialOrderPromotions || [])));
        var appliedPromotions = [];
        appliedPromotions.push.apply(appliedPromotions, __spread((cart.appliedOrderPromotions || [])));
        return __spread(potentialPromotions, appliedPromotions);
    };
    PromotionService.prototype.getOrderPromotionsFromCheckout = function () {
        var _this = this;
        return this.checkoutService
            .getOrderDetails()
            .pipe(map(function (order) { return _this.getOrderPromotionsFromOrderHelper(order); }));
    };
    PromotionService.prototype.getOrderPromotionsFromOrder = function () {
        var _this = this;
        return this.orderDetailsService
            .getOrderDetails()
            .pipe(map(function (order) { return _this.getOrderPromotionsFromOrderHelper(order); }));
    };
    PromotionService.prototype.getOrderPromotionsFromOrderHelper = function (order) {
        var appliedOrderPromotions = [];
        appliedOrderPromotions.push.apply(appliedOrderPromotions, __spread((order.appliedOrderPromotions || [])));
        return appliedOrderPromotions;
    };
    PromotionService.prototype.getProductPromotionForEntry = function (item, promotionLocation) {
        var _this = this;
        switch (promotionLocation) {
            case PromotionLocation.ActiveCart:
                return this.activeCartService
                    .getActive()
                    .pipe(map(function (cart) {
                    return _this.getProductPromotion(item, cart.appliedProductPromotions || []);
                }));
            case PromotionLocation.Checkout:
                return this.checkoutService
                    .getOrderDetails()
                    .pipe(map(function (order) {
                    return _this.getProductPromotion(item, order.appliedProductPromotions || []);
                }));
            case PromotionLocation.Order:
                return this.orderDetailsService
                    .getOrderDetails()
                    .pipe(map(function (order) {
                    return _this.getProductPromotion(item, order.appliedProductPromotions || []);
                }));
        }
    };
    PromotionService.prototype.getProductPromotion = function (item, promotions) {
        var e_1, _a, e_2, _b;
        var entryPromotions = [];
        if (promotions && promotions.length > 0) {
            try {
                for (var promotions_1 = __values(promotions), promotions_1_1 = promotions_1.next(); !promotions_1_1.done; promotions_1_1 = promotions_1.next()) {
                    var promotion = promotions_1_1.value;
                    if (promotion.description &&
                        promotion.consumedEntries &&
                        promotion.consumedEntries.length > 0) {
                        try {
                            for (var _c = (e_2 = void 0, __values(promotion.consumedEntries)), _d = _c.next(); !_d.done; _d = _c.next()) {
                                var consumedEntry = _d.value;
                                if (this.isConsumedByEntry(consumedEntry, item)) {
                                    entryPromotions.push(promotion);
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (promotions_1_1 && !promotions_1_1.done && (_a = promotions_1.return)) _a.call(promotions_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return entryPromotions;
    };
    PromotionService.prototype.isConsumedByEntry = function (consumedEntry, entry) {
        var e_3, _a;
        var consumedEntryNumber = consumedEntry.orderEntryNumber;
        if (entry.entries && entry.entries.length > 0) {
            try {
                for (var _b = __values(entry.entries), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var subEntry = _c.value;
                    if (subEntry.entryNumber === consumedEntryNumber) {
                        return true;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
            return false;
        }
        else {
            return consumedEntryNumber === entry.entryNumber;
        }
    };
    PromotionService.ctorParameters = function () { return [
        { type: OrderDetailsService },
        { type: CheckoutService },
        { type: ActiveCartService }
    ]; };
    PromotionService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PromotionService_Factory() { return new PromotionService(i0.ɵɵinject(i1.OrderDetailsService), i0.ɵɵinject(i2.CheckoutService), i0.ɵɵinject(i2.ActiveCartService)); }, token: PromotionService, providedIn: "root" });
    PromotionService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], PromotionService);
    return PromotionService;
}());
export { PromotionService };
