// @spartacus/storefront/esm5/cms-components/content/link/link.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig } from '@spartacus/core';
import { LinkComponent } from './link.component';
import { GenericLinkModule } from '../../../shared/components/generic-link/generic-link.module';
var LinkModule = /** @class */ (function () {
    function LinkModule() {
    }
    LinkModule = __decorate([
        NgModule({
            imports: [CommonModule, RouterModule, GenericLinkModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CMSLinkComponent: { component: LinkComponent },
                    },
                }),
            ],
            declarations: [LinkComponent],
            exports: [LinkComponent],
            entryComponents: [LinkComponent],
        })
    ], LinkModule);
    return LinkModule;
}());
export { LinkModule };
