// @spartacus/core/esm5/src/cart/store/selectors/multi-cart.selector.js
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { entityValueSelector } from '../../../state/utils/entity-loader/entity-loader.selectors';
import { entityHasPendingProcessesSelector, entityIsStableSelector, entityProcessesLoaderStateSelector, } from '../../../state/utils/entity-processes-loader/entity-processes-loader.selectors';
import { MULTI_CART_FEATURE, } from '../multi-cart-state';
export var getMultiCartState = createFeatureSelector(MULTI_CART_FEATURE);
var ɵ0 = function (state) { return state.carts; };
export var getMultiCartEntities = createSelector(getMultiCartState, ɵ0);
export var getCartEntitySelectorFactory = function (cartId) {
    return createSelector(getMultiCartEntities, function (state) {
        return entityProcessesLoaderStateSelector(state, cartId);
    });
};
export var getCartSelectorFactory = function (cartId) {
    return createSelector(getMultiCartEntities, function (state) {
        return entityValueSelector(state, cartId);
    });
};
export var getCartIsStableSelectorFactory = function (cartId) {
    return createSelector(getMultiCartEntities, function (state) {
        return entityIsStableSelector(state, cartId);
    });
};
export var getCartHasPendingProcessesSelectorFactory = function (cartId) {
    return createSelector(getMultiCartEntities, function (state) {
        return entityHasPendingProcessesSelector(state, cartId);
    });
};
export var getCartEntriesSelectorFactory = function (cartId) {
    return createSelector(getCartSelectorFactory(cartId), function (state) {
        return state && state.entries ? state.entries : [];
    });
};
export var getCartEntrySelectorFactory = function (cartId, productCode) {
    return createSelector(getCartEntriesSelectorFactory(cartId), function (state) {
        return state
            ? state.find(function (entry) { return entry.product.code === productCode; })
            : undefined;
    });
};
var ɵ1 = function (state) { return state.active; };
export var getActiveCartId = createSelector(getMultiCartState, ɵ1);
var ɵ2 = function (state) { return state.wishList; };
export var getWishListId = createSelector(getMultiCartState, ɵ2);
export { ɵ0, ɵ1, ɵ2 };
