// @spartacus/storefront/esm5/cms-components/product/product-summary/product-summary.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { OutletModule } from '../../../cms-structure/outlet/outlet.module';
import { ProductSummaryComponent } from './product-summary.component';
var ProductSummaryModule = /** @class */ (function () {
    function ProductSummaryModule() {
    }
    ProductSummaryModule = __decorate([
        NgModule({
            imports: [CommonModule, OutletModule, I18nModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ProductSummaryComponent: {
                            component: ProductSummaryComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductSummaryComponent],
            entryComponents: [ProductSummaryComponent],
            exports: [ProductSummaryComponent],
        })
    ], ProductSummaryModule);
    return ProductSummaryModule;
}());
export { ProductSummaryModule };
