// @spartacus/storefront/esm5/shared/components/carousel/carousel.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { WindowRef } from '@spartacus/core';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var CarouselService = /** @class */ (function () {
    function CarouselService(winRef) {
        this.winRef = winRef;
    }
    /**
     * The number of items per slide is calculated by the help of
     * the item width and the available width of the host element.
     * This appoach makes it possible to place the carousel in different
     * layouts. Instead of using the page breakpoints, the host size is
     * taken into account.
     *
     * Since there's no element resize API available, we use the
     * window `resize` event, so that we can adjust the number of items
     * whenever the window got resized.
     */
    CarouselService.prototype.getItemsPerSlide = function (nativeElement, itemWidth) {
        var _this = this;
        return this.winRef.resize$.pipe(map(function () { return nativeElement.clientWidth; }), map(function (totalWidth) { return _this.calculateItems(totalWidth, itemWidth); }));
    };
    /**
     * Calculates the number of items per given hostSize.  calculated based on the given
     * intended size in pixels or percentages. The
     *
     * @param availableWidth The available width in pixels for the carousel items.
     * @param itemWidth The width per carousel item, in px or percentage.
     */
    CarouselService.prototype.calculateItems = function (availableWidth, itemWidth) {
        var calculatedItems = 0;
        if (itemWidth.endsWith('px')) {
            var num = itemWidth.substring(0, itemWidth.length - 2);
            calculatedItems = availableWidth / num;
        }
        if (itemWidth.endsWith('%')) {
            var perc = itemWidth.substring(0, itemWidth.length - 1);
            calculatedItems =
                availableWidth / (availableWidth * (perc / 100));
        }
        return Math.floor(calculatedItems) || 1;
    };
    CarouselService.ctorParameters = function () { return [
        { type: WindowRef }
    ]; };
    CarouselService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CarouselService_Factory() { return new CarouselService(i0.ɵɵinject(i1.WindowRef)); }, token: CarouselService, providedIn: "root" });
    CarouselService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CarouselService);
    return CarouselService;
}());
export { CarouselService };
