// @spartacus/storefront/esm5/shared/components/carousel/carousel.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Input, isDevMode, OnInit, TemplateRef, } from '@angular/core';
import { tap } from 'rxjs/operators';
import { ICON_TYPE } from '../../../cms-components/misc/icon/icon.model';
import { CarouselService } from './carousel.service';
/**
 * Generic carousel component that can be used to render any carousel items,
 * such as products, images, banners, or any component. Carousel items are
 * rendered in so-called carousel slides, and the previous/next buttons as well as
 * the indicator-buttons can used to navigate the slides.
 *
 * The component uses an array of Observables (`items$`) as an input, to allow
 * for lazy loading of items.
 *
 * The number of items per slide is calculated with the `itemWidth`, which can given
 * in pixels or percentage.
 *
 * To allow for flexible rendering of items, the rendering is delegated to the
 * given `template`. This allows for maximum flexibility.
 */
var CarouselComponent = /** @class */ (function () {
    function CarouselComponent(el, service) {
        this.el = el;
        this.service = service;
        /**
         * Specifies the minimum size of the carousel item, either in px or %.
         * This value is used for the calculation of numbers per carousel, so that
         * the number of carousel items is dynamic. The calculation uses the `itemWidth`
         * and the host element `clientWidth`, so that the carousel is reusable in
         * different layouts (for example in a 50% grid).
         */
        this.itemWidth = '300px';
        /**
         * Indicates whether the visual indicators are used.
         */
        this.hideIndicators = false;
        this.indicatorIcon = ICON_TYPE.CIRCLE;
        this.previousIcon = ICON_TYPE.CARET_LEFT;
        this.nextIcon = ICON_TYPE.CARET_RIGHT;
    }
    Object.defineProperty(CarouselComponent.prototype, "setItems", {
        set: function (inputItems) {
            this.items = inputItems;
            //Reset slider when changing products
            this.activeSlide = 0;
        },
        enumerable: true,
        configurable: true
    });
    CarouselComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.template && isDevMode()) {
            console.error('No template reference provided to render the carousel items for the `cx-carousel`');
            return;
        }
        this.size$ = this.service
            .getItemsPerSlide(this.el.nativeElement, this.itemWidth)
            .pipe(tap(function () { return (_this.activeSlide = 0); }));
    };
    CarouselComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: CarouselService }
    ]; };
    __decorate([
        Input()
    ], CarouselComponent.prototype, "title", void 0);
    __decorate([
        Input('items')
    ], CarouselComponent.prototype, "setItems", null);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "template", void 0);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "itemWidth", void 0);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "hideIndicators", void 0);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "indicatorIcon", void 0);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "previousIcon", void 0);
    __decorate([
        Input()
    ], CarouselComponent.prototype, "nextIcon", void 0);
    CarouselComponent = __decorate([
        Component({
            selector: 'cx-carousel',
            template: "<ng-container *ngIf=\"items?.length > 0 && (size$ | async) as size\">\r\n  <h3 *ngIf=\"title\">{{ title }}</h3>\r\n\r\n  <div class=\"carousel-panel\" [ngClass]=\"'size-' + size\">\r\n    <button\r\n      *ngIf=\"size < items.length\"\r\n      class=\"previous\"\r\n      (click)=\"activeSlide = activeSlide - size\"\r\n      [disabled]=\"activeSlide === 0\"\r\n    >\r\n      <cx-icon [type]=\"previousIcon\"></cx-icon>\r\n    </button>\r\n\r\n    <div class=\"slides\">\r\n      <ng-container *ngFor=\"let _ of items; let i = index\">\r\n        <div\r\n          class=\"slide\"\r\n          *ngIf=\"i % size === 0\"\r\n          [class.active]=\"i === activeSlide\"\r\n        >\r\n          <ng-container\r\n            *ngFor=\"let item of items | slice: i:i + size; let j = index\"\r\n          >\r\n            <div\r\n              *ngIf=\"item | async as data\"\r\n              class=\"item\"\r\n              [class.active]=\"i === activeSlide\"\r\n            >\r\n              <ng-container\r\n                *ngTemplateOutlet=\"template; context: { item: data }\"\r\n              ></ng-container>\r\n            </div>\r\n          </ng-container>\r\n        </div>\r\n      </ng-container>\r\n    </div>\r\n\r\n    <button\r\n      *ngIf=\"size < items.length\"\r\n      class=\"next\"\r\n      (click)=\"activeSlide = activeSlide + size\"\r\n      tabindex=\"-1\"\r\n      [disabled]=\"activeSlide > items.length - size - 1\"\r\n    >\r\n      <cx-icon [type]=\"nextIcon\"></cx-icon>\r\n    </button>\r\n  </div>\r\n\r\n  <div *ngIf=\"!hideIndicators && size < items.length\" class=\"indicators\">\r\n    <ng-container *ngFor=\"let _ of items; let i = index\">\r\n      <button\r\n        *ngIf=\"i % size === 0\"\r\n        (focus)=\"activeSlide = i\"\r\n        [disabled]=\"i === activeSlide\"\r\n        tabindex=\"-1\"\r\n      >\r\n        <cx-icon [type]=\"indicatorIcon\"></cx-icon>\r\n      </button>\r\n    </ng-container>\r\n  </div>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CarouselComponent);
    return CarouselComponent;
}());
export { CarouselComponent };
