// @spartacus/storefront/esm5/cms-components/myaccount/consent-management/components/consent-form/consent-management-form.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ANONYMOUS_CONSENT_STATUS, } from '@spartacus/core';
var ConsentManagementFormComponent = /** @class */ (function () {
    function ConsentManagementFormComponent() {
        this.consentGiven = false;
        this.requiredConsents = [];
        this.consentChanged = new EventEmitter();
    }
    ConsentManagementFormComponent.prototype.ngOnInit = function () {
        if (this.consent) {
            this.consentGiven = Boolean(this.consent.consentState === ANONYMOUS_CONSENT_STATUS.GIVEN);
        }
        else {
            if (this.consentTemplate && this.consentTemplate.currentConsent) {
                if (this.consentTemplate.currentConsent.consentWithdrawnDate) {
                    this.consentGiven = false;
                }
                else if (this.consentTemplate.currentConsent.consentGivenDate) {
                    this.consentGiven = true;
                }
            }
        }
    };
    ConsentManagementFormComponent.prototype.onConsentChange = function () {
        this.consentGiven = !this.consentGiven;
        this.consentChanged.emit({
            given: this.consentGiven,
            template: this.consentTemplate,
        });
    };
    ConsentManagementFormComponent.prototype.isRequired = function (templateId) {
        return this.requiredConsents.includes(templateId);
    };
    __decorate([
        Input()
    ], ConsentManagementFormComponent.prototype, "consentTemplate", void 0);
    __decorate([
        Input()
    ], ConsentManagementFormComponent.prototype, "requiredConsents", void 0);
    __decorate([
        Input()
    ], ConsentManagementFormComponent.prototype, "consent", void 0);
    __decorate([
        Output()
    ], ConsentManagementFormComponent.prototype, "consentChanged", void 0);
    ConsentManagementFormComponent = __decorate([
        Component({
            selector: 'cx-consent-management-form',
            template: "<div class=\"form-check\">\r\n  <label>\r\n    <input\r\n      type=\"checkbox\"\r\n      class=\"form-check-input\"\r\n      (change)=\"onConsentChange()\"\r\n      [checked]=\"consentGiven\"\r\n      [disabled]=\"isRequired(consentTemplate?.id)\"\r\n    />\r\n    <span class=\"form-check-label cx-be-bold\">\r\n      {{ consentTemplate?.name }}\r\n    </span>\r\n    <br />\r\n    <span class=\"form-check-label\">\r\n      {{ consentTemplate?.description }}\r\n    </span>\r\n  </label>\r\n</div>\r\n"
        })
    ], ConsentManagementFormComponent);
    return ConsentManagementFormComponent;
}());
export { ConsentManagementFormComponent };
