// @spartacus/storefront/esm5/cms-components/order-confirmation/components/order-confirmation-thank-you-message/order-confirmation-thank-you-message.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, } from '@angular/core';
import { CheckoutService, Order } from '@spartacus/core';
import { tap } from 'rxjs/operators';
var OrderConfirmationThankYouMessageComponent = /** @class */ (function () {
    function OrderConfirmationThankYouMessageComponent(checkoutService) {
        this.checkoutService = checkoutService;
        this.isGuestCustomer = false;
    }
    OrderConfirmationThankYouMessageComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.order$ = this.checkoutService.getOrderDetails().pipe(tap(function (order) {
            _this.isGuestCustomer = order.guestCustomer;
            _this.orderGuid = order.guid;
        }));
    };
    OrderConfirmationThankYouMessageComponent.prototype.ngOnDestroy = function () {
        this.checkoutService.clearCheckoutData();
    };
    OrderConfirmationThankYouMessageComponent.ctorParameters = function () { return [
        { type: CheckoutService }
    ]; };
    OrderConfirmationThankYouMessageComponent = __decorate([
        Component({
            selector: 'cx-order-confirmation-thank-you-message',
            template: "<ng-container *ngIf=\"order$ | async as order\">\r\n  <div class=\"cx-page-header\">\r\n    <h1 class=\"cx-page-title\">\r\n      {{ 'checkoutOrderConfirmation.confirmationOfOrder' | cxTranslate }}\r\n      {{ order.code }}\r\n    </h1>\r\n  </div>\r\n\r\n  <div class=\"cx-order-confirmation-message\">\r\n    <h2>{{ 'checkoutOrderConfirmation.thankYou' | cxTranslate }}</h2>\r\n    <p>\r\n      {{ 'checkoutOrderConfirmation.invoiceHasBeenSentByEmail' | cxTranslate }}\r\n    </p>\r\n    <!-- <a href=\"#\" class=\"btn-link\">Print Page</a> -->\r\n  </div>\r\n\r\n  <div *ngIf=\"isGuestCustomer\">\r\n    <cx-guest-register-form\r\n      [guid]=\"orderGuid\"\r\n      [email]=\"order.paymentInfo.billingAddress.email\"\r\n    ></cx-guest-register-form>\r\n  </div>\r\n\r\n  <cx-add-to-home-screen-banner></cx-add-to-home-screen-banner>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], OrderConfirmationThankYouMessageComponent);
    return OrderConfirmationThankYouMessageComponent;
}());
export { OrderConfirmationThankYouMessageComponent };
