// @spartacus/storefront/esm5/cms-components/navigation/search-box/search-box.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { IconModule } from '../../misc/icon/icon.module';
import { MediaModule } from '../../../shared/components/media/media.module';
import { HighlightPipe } from './highlight.pipe';
import { SearchBoxComponent } from './search-box.component';
var SearchBoxModule = /** @class */ (function () {
    function SearchBoxModule() {
    }
    SearchBoxModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                MediaModule,
                IconModule,
                UrlModule,
                I18nModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        SearchBoxComponent: {
                            component: SearchBoxComponent,
                        },
                    },
                }),
            ],
            declarations: [SearchBoxComponent, HighlightPipe],
            entryComponents: [SearchBoxComponent],
            exports: [SearchBoxComponent],
        })
    ], SearchBoxModule);
    return SearchBoxModule;
}());
export { SearchBoxModule };
