// @spartacus/core/esm5/src/occ/interceptors/with-credentials.interceptor.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
/**
 * Http interceptor to add cookies to all cross-site requests.
 */
var WithCredentialsInterceptor = /** @class */ (function () {
    function WithCredentialsInterceptor(config) {
        this.config = config;
    }
    /**
     * Intercepts each request and adds the `withCredential` flag to it
     * if it hasn't been added already.
     */
    WithCredentialsInterceptor.prototype.intercept = function (request, next) {
        if (this.requiresWithCredentials(request)) {
            request = request.clone({
                withCredentials: true,
            });
        }
        return next.handle(request);
    };
    /**
     * indicates whether the request should use the WithCredentials flag.
     */
    WithCredentialsInterceptor.prototype.requiresWithCredentials = function (request) {
        var _a, _b;
        return (((_a = this.occConfig) === null || _a === void 0 ? void 0 : _a.useWithCredentials) &&
            request.url.indexOf((_b = this.occConfig) === null || _b === void 0 ? void 0 : _b.prefix) > -1);
    };
    Object.defineProperty(WithCredentialsInterceptor.prototype, "occConfig", {
        get: function () {
            return this.config.backend.occ;
        },
        enumerable: true,
        configurable: true
    });
    WithCredentialsInterceptor.ctorParameters = function () { return [
        { type: OccConfig }
    ]; };
    WithCredentialsInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function WithCredentialsInterceptor_Factory() { return new WithCredentialsInterceptor(i0.ɵɵinject(i1.OccConfig)); }, token: WithCredentialsInterceptor, providedIn: "root" });
    WithCredentialsInterceptor = __decorate([
        Injectable({ providedIn: 'root' })
    ], WithCredentialsInterceptor);
    return WithCredentialsInterceptor;
}());
export { WithCredentialsInterceptor };
