// @spartacus/core/esm5/src/state/config/state-config.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Config } from '../../config/config.module';
import * as i0 from "@angular/core";
import * as i1 from "../../config/config.module";
export var StorageSyncType;
(function (StorageSyncType) {
    StorageSyncType["NO_STORAGE"] = "NO_STORAGE";
    StorageSyncType["LOCAL_STORAGE"] = "LOCAL_STORAGE";
    StorageSyncType["SESSION_STORAGE"] = "SESSION_STORAGE";
})(StorageSyncType || (StorageSyncType = {}));
export var StateTransferType;
(function (StateTransferType) {
    StateTransferType["TRANSFER_STATE"] = "SSR";
})(StateTransferType || (StateTransferType = {}));
var StateConfig = /** @class */ (function () {
    function StateConfig() {
    }
    StateConfig.ɵprov = i0.ɵɵdefineInjectable({ factory: function StateConfig_Factory() { return i0.ɵɵinject(i1.Config); }, token: StateConfig, providedIn: "root" });
    StateConfig = __decorate([
        Injectable({
            providedIn: 'root',
            useExisting: Config,
        })
    ], StateConfig);
    return StateConfig;
}());
export { StateConfig };
