// @spartacus/core/esm5/src/product/product.module.js
import { __decorate, __read, __spread } from "tslib";
import { NgModule } from '@angular/core';
import { PageMetaResolver } from '../cms/page/page-meta.resolver';
import { CategoryPageMetaResolver } from './services/category-page-meta.resolver';
import { CouponSearchPageResolver } from './services/coupon-search-page-meta.resolver';
import { ProductPageMetaResolver } from './services/product-page-meta.resolver';
import { SearchPageMetaResolver } from './services/search-page-meta.resolver';
import { ProductStoreModule } from './store/product-store.module';
var pageTitleResolvers = [
    {
        provide: PageMetaResolver,
        useExisting: ProductPageMetaResolver,
        multi: true,
    },
    {
        provide: PageMetaResolver,
        useExisting: CategoryPageMetaResolver,
        multi: true,
    },
    {
        provide: PageMetaResolver,
        useExisting: SearchPageMetaResolver,
        multi: true,
    },
    {
        provide: PageMetaResolver,
        useExisting: CouponSearchPageResolver,
        multi: true,
    },
];
var ProductModule = /** @class */ (function () {
    function ProductModule() {
    }
    ProductModule_1 = ProductModule;
    ProductModule.forRoot = function () {
        return {
            ngModule: ProductModule_1,
            providers: __spread(pageTitleResolvers),
        };
    };
    var ProductModule_1;
    ProductModule = ProductModule_1 = __decorate([
        NgModule({
            imports: [ProductStoreModule],
        })
    ], ProductModule);
    return ProductModule;
}());
export { ProductModule };
