// @spartacus/storefront/esm2015/layout/a11y/keyboard-focus/escape/escape-focus.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, EventEmitter, HostListener, OnInit, Output, } from '@angular/core';
import { PersistFocusDirective } from '../persist/persist-focus.directive';
import { EscapeFocusService } from './escape-focus.service';
/**
 * Directive to focus the host element whenever the `escape` key is captured.
 * UiEvents bubble up by nature, which is why the `cxEscGroup` can be used
 * on a tree of elements. Each time the escape key is used, the focus will
 * move up in the DOM tree.
 *
 */
let EscapeFocusDirective = class EscapeFocusDirective extends PersistFocusDirective {
    constructor(elementRef, service) {
        super(elementRef, service);
        this.elementRef = elementRef;
        this.service = service;
        this.defaultConfig = { focusOnEscape: true };
        this.esc = new EventEmitter();
    }
    /**
     * Handles the escape key event.
     * @param event the native keyboard event which contains the escape keydown event
     */
    handleEscape(event) {
        if (this.service.shouldFocus(this.config)) {
            this.service.handleEscape(this.host, this.config, event);
        }
        this.esc.emit(this.service.shouldFocus(this.config));
    }
    ngOnInit() {
        if (this.service.shouldFocus(this.config)) {
            this.requiredTabindex = -1;
        }
        super.ngOnInit();
    }
};
EscapeFocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: EscapeFocusService }
];
__decorate([
    Output()
], EscapeFocusDirective.prototype, "esc", void 0);
__decorate([
    HostListener('keydown.escape', ['$event'])
], EscapeFocusDirective.prototype, "handleEscape", null);
EscapeFocusDirective = __decorate([
    Directive() // selector: '[cxEscFocus]',
], EscapeFocusDirective);
export { EscapeFocusDirective };
