// @spartacus/storefront/esm2015/layout/a11y/keyboard-focus/escape/escape-focus.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { PersistFocusService } from '../persist/persist-focus.service';
import { SelectFocusUtility } from '../services/select-focus.util';
import * as i0 from "@angular/core";
import * as i1 from "../services/select-focus.util";
let EscapeFocusService = class EscapeFocusService extends PersistFocusService {
    constructor(selectFocusUtil) {
        super();
        this.selectFocusUtil = selectFocusUtil;
    }
    shouldFocus(config) {
        return !!(config === null || config === void 0 ? void 0 : config.focusOnEscape);
    }
    handleEscape(host, config, event) {
        var _a;
        if (this.shouldFocus(config)) {
            if (host !== event.target) {
                host.focus({ preventScroll: true });
                event.preventDefault();
                event.stopPropagation();
            }
            else {
                if (config === null || config === void 0 ? void 0 : config.focusOnDoubleEscape) {
                    (_a = this.selectFocusUtil
                        .findFirstFocusable(host, { autofocus: true })) === null || _a === void 0 ? void 0 : _a.focus();
                }
            }
        }
    }
};
EscapeFocusService.ctorParameters = () => [
    { type: SelectFocusUtility }
];
EscapeFocusService.ɵprov = i0.ɵɵdefineInjectable({ factory: function EscapeFocusService_Factory() { return new EscapeFocusService(i0.ɵɵinject(i1.SelectFocusUtility)); }, token: EscapeFocusService, providedIn: "root" });
EscapeFocusService = __decorate([
    Injectable({
        providedIn: 'root',
    })
], EscapeFocusService);
export { EscapeFocusService };
