// @spartacus/core/esm5/src/auth/store/actions/client-token.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { CLIENT_TOKEN_DATA } from '../auth-state';
export var LOAD_CLIENT_TOKEN = '[Token] Load Client Token';
export var LOAD_CLIENT_TOKEN_FAIL = '[Token] Load Client Token Fail';
export var LOAD_CLIENT_TOKEN_SUCCESS = '[Token] Load Client Token Success';
var LoadClientToken = /** @class */ (function (_super) {
    __extends(LoadClientToken, _super);
    function LoadClientToken() {
        var _this = _super.call(this, CLIENT_TOKEN_DATA) || this;
        _this.type = LOAD_CLIENT_TOKEN;
        return _this;
    }
    return LoadClientToken;
}(StateUtils.LoaderLoadAction));
export { LoadClientToken };
var LoadClientTokenFail = /** @class */ (function (_super) {
    __extends(LoadClientTokenFail, _super);
    function LoadClientTokenFail(payload) {
        var _this = _super.call(this, CLIENT_TOKEN_DATA, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_CLIENT_TOKEN_FAIL;
        return _this;
    }
    return LoadClientTokenFail;
}(StateUtils.LoaderFailAction));
export { LoadClientTokenFail };
var LoadClientTokenSuccess = /** @class */ (function (_super) {
    __extends(LoadClientTokenSuccess, _super);
    function LoadClientTokenSuccess(payload) {
        var _this = _super.call(this, CLIENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CLIENT_TOKEN_SUCCESS;
        return _this;
    }
    return LoadClientTokenSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadClientTokenSuccess };
