// @spartacus/storefront/esm5/cms-components/anonymous-consent-management/banner/anonymous-consent-management-banner.component.js
import { __decorate } from "tslib";
import { Component, OnDestroy, ViewContainerRef } from '@angular/core';
import { AnonymousConsentsService } from '@spartacus/core';
import { Subscription } from 'rxjs';
import { tap } from 'rxjs/operators';
import { AnonymousConsentLaunchDialogService } from '../anonymous-consent-launch-dialog.service';
var AnonymousConsentManagementBannerComponent = /** @class */ (function () {
    function AnonymousConsentManagementBannerComponent(anonymousConsentsService, anonymousConsentLaunchDialogService, vcr) {
        this.anonymousConsentsService = anonymousConsentsService;
        this.anonymousConsentLaunchDialogService = anonymousConsentLaunchDialogService;
        this.vcr = vcr;
        this.subscriptions = new Subscription();
        this.bannerVisible$ = this.anonymousConsentsService.isBannerVisible();
    }
    AnonymousConsentManagementBannerComponent.prototype.viewDetails = function () {
        this.hideBanner();
        var dialog = this.anonymousConsentLaunchDialogService.openDialog(null, this.vcr);
        if (dialog) {
            this.subscriptions.add(dialog.subscribe());
        }
    };
    AnonymousConsentManagementBannerComponent.prototype.allowAll = function () {
        var _this = this;
        this.subscriptions.add(this.anonymousConsentsService
            .giveAllConsents()
            .pipe(tap(function () { return _this.hideBanner(); }))
            .subscribe());
    };
    AnonymousConsentManagementBannerComponent.prototype.hideBanner = function () {
        this.anonymousConsentsService.toggleBannerDismissed(true);
    };
    AnonymousConsentManagementBannerComponent.prototype.ngOnDestroy = function () {
        this.subscriptions.unsubscribe();
    };
    AnonymousConsentManagementBannerComponent.ctorParameters = function () { return [
        { type: AnonymousConsentsService },
        { type: AnonymousConsentLaunchDialogService },
        { type: ViewContainerRef }
    ]; };
    AnonymousConsentManagementBannerComponent = __decorate([
        Component({
            selector: 'cx-anonymous-consent-management-banner',
            template: "<ng-container *ngIf=\"bannerVisible$ | async as bannerVisible\">\n  <div\n    [ngClass]=\"{ 'anonymous-consent-banner-hidden': !bannerVisible }\"\n    class=\"anonymous-consent-banner\"\n  >\n    <div class=\"container\">\n      <div class=\"row\">\n        <div class=\"col-lg-8 col-xs-12\">\n          <div class=\"cx-banner-title\">\n            {{ 'anonymousConsents.banner.title' | cxTranslate }}\n          </div>\n          <div class=\"cx-banner-description\">\n            {{ 'anonymousConsents.banner.description' | cxTranslate }}\n          </div>\n        </div>\n\n        <div class=\"col-lg-4 col-xs-12 cx-banner-buttons\">\n          <button class=\"btn btn-action\" (click)=\"viewDetails()\">\n            {{ 'anonymousConsents.banner.viewDetails' | cxTranslate }}\n          </button>\n          <button class=\"btn btn-primary\" (click)=\"allowAll()\">\n            {{ 'anonymousConsents.banner.allowAll' | cxTranslate }}\n          </button>\n        </div>\n      </div>\n    </div>\n  </div>\n</ng-container>\n"
        })
    ], AnonymousConsentManagementBannerComponent);
    return AnonymousConsentManagementBannerComponent;
}());
export { AnonymousConsentManagementBannerComponent };
