// @spartacus/storefront/esm2015/layout/launch-dialog/launch-dialog.module.js
var LaunchDialogModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { Config } from '@spartacus/core';
import { LayoutConfig } from '../config/layout-config';
import { InlineRenderStrategy, LaunchRenderStrategy, OutletRenderStrategy, RoutingRenderStrategy, } from './services/index';
let LaunchDialogModule = LaunchDialogModule_1 = class LaunchDialogModule {
    static forRoot() {
        return {
            ngModule: LaunchDialogModule_1,
            providers: [{ provide: LayoutConfig, useExisting: Config }],
        };
    }
};
LaunchDialogModule = LaunchDialogModule_1 = __decorate([
    NgModule({
        providers: [
            {
                provide: LaunchRenderStrategy,
                useExisting: OutletRenderStrategy,
                multi: true,
            },
            {
                provide: LaunchRenderStrategy,
                useExisting: InlineRenderStrategy,
                multi: true,
            },
            {
                provide: LaunchRenderStrategy,
                useExisting: RoutingRenderStrategy,
                multi: true,
            },
        ],
    })
], LaunchDialogModule);
export { LaunchDialogModule };
