// @spartacus/storefront/esm5/cms-components/myaccount/update-email/update-email-form/update-email-form.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { CustomFormValidators } from '../../../../shared/utils/validators/custom-form-validators';
var UpdateEmailFormComponent = /** @class */ (function () {
    function UpdateEmailFormComponent(fb) {
        this.fb = fb;
        this.saveEmail = new EventEmitter();
        this.cancelEmail = new EventEmitter();
        this.updateEmailForm = this.fb.group({
            email: ['', [Validators.required, CustomFormValidators.emailValidator]],
            confirmEmail: ['', [Validators.required]],
            password: ['', [Validators.required]],
        }, {
            validators: CustomFormValidators.emailsMustMatch('email', 'confirmEmail'),
        });
    }
    UpdateEmailFormComponent.prototype.onSubmit = function () {
        if (this.updateEmailForm.valid) {
            var newUid = this.updateEmailForm.get('confirmEmail').value;
            var password = this.updateEmailForm.get('password').value;
            this.saveEmail.emit({ newUid: newUid, password: password });
        }
        else {
            this.updateEmailForm.markAllAsTouched();
        }
    };
    UpdateEmailFormComponent.prototype.onCancel = function () {
        this.cancelEmail.emit();
    };
    UpdateEmailFormComponent.ctorParameters = function () { return [
        { type: FormBuilder }
    ]; };
    __decorate([
        Output()
    ], UpdateEmailFormComponent.prototype, "saveEmail", void 0);
    __decorate([
        Output()
    ], UpdateEmailFormComponent.prototype, "cancelEmail", void 0);
    UpdateEmailFormComponent = __decorate([
        Component({
            selector: 'cx-update-email-form',
            template: "<form (ngSubmit)=\"onSubmit()\" [formGroup]=\"updateEmailForm\">\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateEmailForm.newEmailAddress.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          type=\"email\"\r\n          name=\"email\"\r\n          formControlName=\"email\"\r\n          placeholder=\"{{\r\n            'updateEmailForm.newEmailAddress.placeholder' | cxTranslate\r\n          }}\"\r\n          class=\"form-control\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updateEmailForm.get('email')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-sm-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateEmailForm.confirmNewEmailAddress.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          type=\"email\"\r\n          name=\"confirmEmail\"\r\n          formControlName=\"confirmEmail\"\r\n          placeholder=\"{{\r\n            'updateEmailForm.confirmNewEmailAddress.placeholder' | cxTranslate\r\n          }}\"\r\n          class=\"form-control\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updateEmailForm.get('confirmEmail')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-sm-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateEmailForm.password.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          type=\"password\"\r\n          name=\"password\"\r\n          formControlName=\"password\"\r\n          placeholder=\"{{\r\n            'updateEmailForm.password.placeholder' | cxTranslate\r\n          }}\"\r\n          class=\"form-control\"\r\n          autocomplete=\"new-password\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updateEmailForm.get('password')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-lg-6\">\r\n      <button\r\n        class=\"btn btn-block btn-secondary\"\r\n        type=\"button\"\r\n        (click)=\"onCancel()\"\r\n      >\r\n        {{ 'common.cancel' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n    <div class=\"col-lg-6\">\r\n      <button class=\"btn btn-block btn-primary\" type=\"submit\">\r\n        {{ 'common.save' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n  </div>\r\n</form>\r\n"
        })
    ], UpdateEmailFormComponent);
    return UpdateEmailFormComponent;
}());
export { UpdateEmailFormComponent };
