// @spartacus/storefront/esm5/layout/a11y/skip-link/config/skip-link.config.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Config } from '@spartacus/core';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var SkipLinkConfig = /** @class */ (function () {
    function SkipLinkConfig() {
    }
    SkipLinkConfig.ɵprov = i0.ɵɵdefineInjectable({ factory: function SkipLinkConfig_Factory() { return i0.ɵɵinject(i1.Config); }, token: SkipLinkConfig, providedIn: "root" });
    SkipLinkConfig = __decorate([
        Injectable({
            providedIn: 'root',
            useExisting: Config,
        })
    ], SkipLinkConfig);
    return SkipLinkConfig;
}());
export { SkipLinkConfig };
var SkipLink = /** @class */ (function () {
    function SkipLink() {
    }
    return SkipLink;
}());
export { SkipLink };
export var SkipLinkScrollPosition;
(function (SkipLinkScrollPosition) {
    SkipLinkScrollPosition["BEFORE"] = "BEFORE";
    SkipLinkScrollPosition["AFTER"] = "AFTER";
})(SkipLinkScrollPosition || (SkipLinkScrollPosition = {}));
