// @spartacus/core/esm5/src/anonymous-consents/store/selectors/anonymous-consent-templates.selectors.js
import { createSelector } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { getAnonymousConsentState } from './feature.selector';
var ɵ0 = function (state) { return state.templates; };
export var getAnonymousConsentTemplatesState = createSelector(getAnonymousConsentState, ɵ0);
export var getAnonymousConsentTemplatesValue = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderValueSelector);
export var getAnonymousConsentTemplatesLoading = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderLoadingSelector);
export var getAnonymousConsentTemplatesSuccess = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderSuccessSelector);
export var getAnonymousConsentTemplatesError = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderErrorSelector);
export var getAnonymousConsentTemplate = function (templateCode) {
    return createSelector(getAnonymousConsentTemplatesValue, function (templates) {
        return templates
            ? templates.find(function (template) { return template.id === templateCode; })
            : null;
    });
};
export { ɵ0 };
