// @spartacus/storefront/esm5/cms-components/storefinder/components/store-finder-map/store-finder-map.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild, } from '@angular/core';
import { GoogleMapRendererService } from '@spartacus/core';
var StoreFinderMapComponent = /** @class */ (function () {
    function StoreFinderMapComponent(googleMapRendererService) {
        this.googleMapRendererService = googleMapRendererService;
        this.selectedStoreItem = new EventEmitter();
    }
    StoreFinderMapComponent.prototype.ngOnChanges = function (changes) {
        if (changes.locations && this.locations) {
            this.renderMap();
        }
    };
    /**
     * Sets the center of the map to the given location
     * @param latitude latitude of the new center
     * @param longitude longitude of the new center
     */
    StoreFinderMapComponent.prototype.centerMap = function (latitude, longitude) {
        this.googleMapRendererService.centerMap(latitude, longitude);
    };
    StoreFinderMapComponent.prototype.renderMap = function () {
        var _this = this;
        this.googleMapRendererService.renderMap(this.mapElement.nativeElement, this.locations, function (markerIndex) {
            _this.selectStoreItemClickHandle(markerIndex);
        });
    };
    StoreFinderMapComponent.prototype.selectStoreItemClickHandle = function (markerIndex) {
        this.selectedStoreItem.emit(markerIndex);
    };
    StoreFinderMapComponent.ctorParameters = function () { return [
        { type: GoogleMapRendererService }
    ]; };
    __decorate([
        ViewChild('mapElement', { static: true })
    ], StoreFinderMapComponent.prototype, "mapElement", void 0);
    __decorate([
        Input()
    ], StoreFinderMapComponent.prototype, "locations", void 0);
    __decorate([
        Output()
    ], StoreFinderMapComponent.prototype, "selectedStoreItem", void 0);
    StoreFinderMapComponent = __decorate([
        Component({
            selector: 'cx-store-finder-map',
            template: "<div #mapElement class=\"cx-store-map\"></div>\n"
        })
    ], StoreFinderMapComponent);
    return StoreFinderMapComponent;
}());
export { StoreFinderMapComponent };
