// @spartacus/core/esm5/src/user/store/effects/order-details.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { UserOrderConnector } from '../../connectors/order/user-order.connector';
import { UserActions } from '../actions/index';
import { GlobalMessageService, GlobalMessageType, } from '../../../global-message/index';
var OrderDetailsEffect = /** @class */ (function () {
    function OrderDetailsEffect(actions$, orderConnector, globalMessageService) {
        var _this = this;
        this.actions$ = actions$;
        this.orderConnector = orderConnector;
        this.globalMessageService = globalMessageService;
        this.loadOrderDetails$ = this.actions$.pipe(ofType(UserActions.LOAD_ORDER_DETAILS), map(function (action) { return action.payload; }), switchMap(function (payload) {
            return _this.orderConnector.get(payload.userId, payload.orderCode).pipe(map(function (order) {
                return new UserActions.LoadOrderDetailsSuccess(order);
            }), catchError(function (error) {
                return of(new UserActions.LoadOrderDetailsFail(makeErrorSerializable(error)));
            }));
        }));
        this.cancelOrder$ = this.actions$.pipe(ofType(UserActions.CANCEL_ORDER), map(function (action) { return action.payload; }), switchMap(function (payload) {
            return _this.orderConnector
                .cancel(payload.userId, payload.orderCode, payload.cancelRequestInput)
                .pipe(map(function () { return new UserActions.CancelOrderSuccess(); }), catchError(function (error) {
                var _a;
                (_a = error.error) === null || _a === void 0 ? void 0 : _a.errors.forEach(function (err) {
                    return _this.globalMessageService.add(err.message, GlobalMessageType.MSG_TYPE_ERROR);
                });
                return of(new UserActions.CancelOrderFail(makeErrorSerializable(error)));
            }));
        }));
    }
    OrderDetailsEffect.ctorParameters = function () { return [
        { type: Actions },
        { type: UserOrderConnector },
        { type: GlobalMessageService }
    ]; };
    __decorate([
        Effect()
    ], OrderDetailsEffect.prototype, "loadOrderDetails$", void 0);
    __decorate([
        Effect()
    ], OrderDetailsEffect.prototype, "cancelOrder$", void 0);
    OrderDetailsEffect = __decorate([
        Injectable()
    ], OrderDetailsEffect);
    return OrderDetailsEffect;
}());
export { OrderDetailsEffect };
