// @spartacus/core/esm5/src/user/store/effects/payment-methods.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, mergeMap, switchMap } from 'rxjs/operators';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { UserPaymentConnector } from '../../connectors/payment/user-payment.connector';
import { UserActions } from '../actions/index';
var UserPaymentMethodsEffects = /** @class */ (function () {
    function UserPaymentMethodsEffects(actions$, userPaymentMethodConnector) {
        var _this = this;
        this.actions$ = actions$;
        this.userPaymentMethodConnector = userPaymentMethodConnector;
        this.loadUserPaymentMethods$ = this.actions$.pipe(ofType(UserActions.LOAD_USER_PAYMENT_METHODS), map(function (action) { return action.payload; }), mergeMap(function (payload) {
            return _this.userPaymentMethodConnector.getAll(payload).pipe(map(function (payments) {
                return new UserActions.LoadUserPaymentMethodsSuccess(payments);
            }), catchError(function (error) {
                return of(new UserActions.LoadUserPaymentMethodsFail(makeErrorSerializable(error)));
            }));
        }));
        this.setDefaultUserPaymentMethod$ = this.actions$.pipe(ofType(UserActions.SET_DEFAULT_USER_PAYMENT_METHOD), map(function (action) { return action.payload; }), mergeMap(function (payload) {
            return _this.userPaymentMethodConnector
                .setDefault(payload.userId, payload.paymentMethodId)
                .pipe(switchMap(function (data) { return [
                new UserActions.SetDefaultUserPaymentMethodSuccess(data),
                new UserActions.LoadUserPaymentMethods(payload.userId),
            ]; }), catchError(function (error) {
                return of(new UserActions.SetDefaultUserPaymentMethodFail(makeErrorSerializable(error)));
            }));
        }));
        this.deleteUserPaymentMethod$ = this.actions$.pipe(ofType(UserActions.DELETE_USER_PAYMENT_METHOD), map(function (action) { return action.payload; }), mergeMap(function (payload) {
            return _this.userPaymentMethodConnector
                .delete(payload.userId, payload.paymentMethodId)
                .pipe(switchMap(function (data) { return [
                new UserActions.DeleteUserPaymentMethodSuccess(data),
                new UserActions.LoadUserPaymentMethods(payload.userId),
            ]; }), catchError(function (error) {
                return of(new UserActions.DeleteUserPaymentMethodFail(makeErrorSerializable(error)));
            }));
        }));
    }
    UserPaymentMethodsEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: UserPaymentConnector }
    ]; };
    __decorate([
        Effect()
    ], UserPaymentMethodsEffects.prototype, "loadUserPaymentMethods$", void 0);
    __decorate([
        Effect()
    ], UserPaymentMethodsEffects.prototype, "setDefaultUserPaymentMethod$", void 0);
    __decorate([
        Effect()
    ], UserPaymentMethodsEffects.prototype, "deleteUserPaymentMethod$", void 0);
    UserPaymentMethodsEffects = __decorate([
        Injectable()
    ], UserPaymentMethodsEffects);
    return UserPaymentMethodsEffects;
}());
export { UserPaymentMethodsEffects };
