// @spartacus/core/esm5/src/user/store/effects/regions.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { SiteConnector } from '../../../site-context/connectors/site.connector';
import { StateUtils } from '../../../state/utils/index';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { UserActions } from '../actions/index';
import { REGIONS } from '../user-state';
var RegionsEffects = /** @class */ (function () {
    function RegionsEffects(actions$, siteConnector) {
        var _this = this;
        this.actions$ = actions$;
        this.siteConnector = siteConnector;
        this.loadRegions$ = this.actions$.pipe(ofType(UserActions.LOAD_REGIONS), map(function (action) {
            return action.payload;
        }), switchMap(function (countryCode) {
            return _this.siteConnector.getRegions(countryCode).pipe(map(function (regions) {
                return new UserActions.LoadRegionsSuccess({
                    entities: regions,
                    country: countryCode,
                });
            }), catchError(function (error) {
                return of(new UserActions.LoadRegionsFail(makeErrorSerializable(error)));
            }));
        }));
        this.resetRegions$ = this.actions$.pipe(ofType(UserActions.CLEAR_USER_MISCS_DATA, UserActions.CLEAR_REGIONS), map(function () {
            return new StateUtils.LoaderResetAction(REGIONS);
        }));
    }
    RegionsEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: SiteConnector }
    ]; };
    __decorate([
        Effect()
    ], RegionsEffects.prototype, "loadRegions$", void 0);
    __decorate([
        Effect()
    ], RegionsEffects.prototype, "resetRegions$", void 0);
    RegionsEffects = __decorate([
        Injectable()
    ], RegionsEffects);
    return RegionsEffects;
}());
export { RegionsEffects };
