// @spartacus/core/esm5/src/user/store/effects/user-register.effect.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, mergeMap, switchMap } from 'rxjs/operators';
import { AuthActions } from '../../../auth/store/actions/index';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { UserConnector } from '../../connectors/user/user.connector';
import { UserActions } from '../actions/index';
var UserRegisterEffects = /** @class */ (function () {
    function UserRegisterEffects(actions$, userConnector) {
        var _this = this;
        this.actions$ = actions$;
        this.userConnector = userConnector;
        this.registerUser$ = this.actions$.pipe(ofType(UserActions.REGISTER_USER), map(function (action) { return action.payload; }), mergeMap(function (user) {
            return _this.userConnector.register(user).pipe(map(function () { return new UserActions.RegisterUserSuccess(); }), catchError(function (error) {
                return of(new UserActions.RegisterUserFail(makeErrorSerializable(error)));
            }));
        }));
        this.registerGuest$ = this.actions$.pipe(ofType(UserActions.REGISTER_GUEST), map(function (action) { return action.payload; }), mergeMap(function (_a) {
            var guid = _a.guid, password = _a.password;
            return _this.userConnector.registerGuest(guid, password).pipe(switchMap(function (user) { return [
                new AuthActions.LoadUserToken({
                    userId: user.uid,
                    password: password,
                }),
                new UserActions.RegisterGuestSuccess(),
            ]; }), catchError(function (error) {
                return of(new UserActions.RegisterGuestFail(makeErrorSerializable(error)));
            }));
        }));
        this.removeUser$ = this.actions$.pipe(ofType(UserActions.REMOVE_USER), map(function (action) { return action.payload; }), mergeMap(function (userId) {
            return _this.userConnector.remove(userId).pipe(switchMap(function () { return [
                new UserActions.RemoveUserSuccess(),
                new AuthActions.Logout(),
            ]; }), catchError(function (error) {
                return of(new UserActions.RemoveUserFail(makeErrorSerializable(error)));
            }));
        }));
    }
    UserRegisterEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: UserConnector }
    ]; };
    __decorate([
        Effect()
    ], UserRegisterEffects.prototype, "registerUser$", void 0);
    __decorate([
        Effect()
    ], UserRegisterEffects.prototype, "registerGuest$", void 0);
    __decorate([
        Effect()
    ], UserRegisterEffects.prototype, "removeUser$", void 0);
    UserRegisterEffects = __decorate([
        Injectable()
    ], UserRegisterEffects);
    return UserRegisterEffects;
}());
export { UserRegisterEffects };
