// @spartacus/storefront/esm2015/cms-components/cart/cart-not-empty.guard.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ActiveCartService, Cart, RoutingService } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
let CartNotEmptyGuard = class CartNotEmptyGuard {
    constructor(routingService, activeCartService) {
        this.routingService = routingService;
        this.activeCartService = activeCartService;
    }
    canActivate() {
        return combineLatest([
            this.activeCartService.getActive(),
            this.activeCartService.isStable(),
        ]).pipe(filter(([_, loaded]) => loaded), map(([cart]) => {
            if (this.isEmpty(cart)) {
                this.routingService.go({ cxRoute: 'home' });
                return false;
            }
            return true;
        }));
    }
    isEmpty(cart) {
        return cart && !cart.totalItems;
    }
};
CartNotEmptyGuard.ctorParameters = () => [
    { type: RoutingService },
    { type: ActiveCartService }
];
CartNotEmptyGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartNotEmptyGuard_Factory() { return new CartNotEmptyGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.ActiveCartService)); }, token: CartNotEmptyGuard, providedIn: "root" });
CartNotEmptyGuard = __decorate([
    Injectable({
        providedIn: 'root',
    })
], CartNotEmptyGuard);
export { CartNotEmptyGuard };
