// @spartacus/storefront/esm5/cms-structure/page/component/services/cms-injector.service.js
import { __decorate, __read, __spread } from "tslib";
import { Injectable, Injector } from '@angular/core';
import { CmsComponentData } from '../../model';
import { CmsService } from '@spartacus/core';
import { CmsComponentsService } from '../../../services/cms-components.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/cms-components.service";
/**
 * Used to prepare injector for CMS components.
 *
 * Injector will take into account configured providers and provides CmsComponentData
 * for specified component's uid
 */
var CmsInjectorService = /** @class */ (function () {
    function CmsInjectorService(cmsComponentsService, injector) {
        this.cmsComponentsService = cmsComponentsService;
        this.injector = injector;
    }
    CmsInjectorService.prototype.getCmsData = function (uid, parentInjector) {
        return {
            uid: uid,
            data$: (parentInjector !== null && parentInjector !== void 0 ? parentInjector : this.injector)
                .get(CmsService)
                .getComponentData(uid),
        };
    };
    CmsInjectorService.prototype.getInjector = function (type, uid, parentInjector) {
        var _a, _b;
        var configProviders = (_b = (_a = this.cmsComponentsService.getMapping(type)) === null || _a === void 0 ? void 0 : _a.providers) !== null && _b !== void 0 ? _b : [];
        return Injector.create({
            providers: __spread([
                {
                    provide: CmsComponentData,
                    useValue: this.getCmsData(uid),
                }
            ], configProviders),
            parent: parentInjector !== null && parentInjector !== void 0 ? parentInjector : this.injector,
        });
    };
    CmsInjectorService.ctorParameters = function () { return [
        { type: CmsComponentsService },
        { type: Injector }
    ]; };
    CmsInjectorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CmsInjectorService_Factory() { return new CmsInjectorService(i0.ɵɵinject(i1.CmsComponentsService), i0.ɵɵinject(i0.INJECTOR)); }, token: CmsInjectorService, providedIn: "root" });
    CmsInjectorService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CmsInjectorService);
    return CmsInjectorService;
}());
export { CmsInjectorService };
