// @spartacus/storefront/esm5/cms-structure/guards/cms-page.guard.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { CmsActivatedRouteSnapshot, CmsService, ProtectedRoutesGuard, RouteLoadStrategy, RoutingConfigService, RoutingService, } from '@spartacus/core';
import { of } from 'rxjs';
import { first, switchMap } from 'rxjs/operators';
import { CmsPageGuardService } from './cms-page-guard.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "./cms-page-guard.service";
var CmsPageGuard = /** @class */ (function () {
    function CmsPageGuard(routingService, cmsService, protectedRoutesGuard, service, routingConfig) {
        this.routingService = routingService;
        this.cmsService = cmsService;
        this.protectedRoutesGuard = protectedRoutesGuard;
        this.service = service;
        this.routingConfig = routingConfig;
    }
    /**
     * Tries to load the CMS page data for the anticipated route and returns:
     * - `true` - if it can be activated
     * - `false` - if it cannot be activated
     * - `UrlTree` - if user should be redirected to a given `UrlTree`
     *
     * If the route can be activated, it fires additional calculations on the CMS components present on this CMS page,
     * based on their configuration (`cmsComponents` config).
     *
     * For more, see docs of the `CmsPageGuardService.canActivatePage`.
     */
    CmsPageGuard.prototype.canActivate = function (route, state) {
        var _this = this;
        return this.protectedRoutesGuard.canActivate(route).pipe(switchMap(function (canActivate) {
            return canActivate
                ? _this.routingService.getNextPageContext().pipe(switchMap(function (pageContext) {
                    return _this.cmsService.getPage(pageContext, _this.shouldReload()).pipe(first(), switchMap(function (pageData) {
                        return pageData
                            ? _this.service.canActivatePage(pageContext, pageData, route, state)
                            : _this.service.canActivateNotFoundPage(pageContext, route, state);
                    }));
                }))
                : of(false);
        }));
    };
    /**
     * Returns whether we should reload the CMS page data, even when it was loaded before.
     */
    CmsPageGuard.prototype.shouldReload = function () {
        return this.routingConfig.getLoadStrategy() !== "once" /* ONCE */;
    };
    CmsPageGuard.guardName = 'CmsPageGuard';
    CmsPageGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: CmsService },
        { type: ProtectedRoutesGuard },
        { type: CmsPageGuardService },
        { type: RoutingConfigService }
    ]; };
    CmsPageGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function CmsPageGuard_Factory() { return new CmsPageGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.CmsService), i0.ɵɵinject(i1.ProtectedRoutesGuard), i0.ɵɵinject(i2.CmsPageGuardService), i0.ɵɵinject(i1.RoutingConfigService)); }, token: CmsPageGuard, providedIn: "root" });
    CmsPageGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CmsPageGuard);
    return CmsPageGuard;
}());
export { CmsPageGuard };
