// @spartacus/core/esm5/src/auth/facade/auth.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { filter, map, take } from 'rxjs/operators';
import { OCC_USER_ID_ANONYMOUS, OCC_USER_ID_CURRENT, } from '../../occ/utils/occ-constants';
import { AuthActions } from '../store/actions/index';
import { AuthSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var AuthService = /** @class */ (function () {
    function AuthService(store) {
        this.store = store;
    }
    /**
     * Loads a new user token
     * @param userId
     * @param password
     */
    AuthService.prototype.authorize = function (userId, password) {
        this.store.dispatch(new AuthActions.LoadUserToken({
            userId: userId,
            password: password,
        }));
    };
    /**
     * This function provides the userId the OCC calls should use, depending
     * on whether there is an active storefront session or not.
     *
     * It returns the userId of the current storefront user or 'anonymous'
     * in the case there are no signed in user in the storefront.
     *
     * The user id of a regular customer session is 'current'.  In the case of an
     * asm customer emulation session, the userId will be the customerId.
     */
    AuthService.prototype.getOccUserId = function () {
        return this.getUserToken().pipe(map(function (userToken) {
            if (!!userToken && !!userToken.userId) {
                return userToken.userId;
            }
            else {
                return OCC_USER_ID_ANONYMOUS;
            }
        }));
    };
    /**
     * Calls provided callback with current user id.
     *
     * @param cb callback function to invoke
     */
    AuthService.prototype.invokeWithUserId = function (cb) {
        return this.getOccUserId()
            .pipe(take(1))
            .subscribe(function (id) { return cb(id); });
    };
    /**
     * Returns the user's token
     */
    AuthService.prototype.getUserToken = function () {
        return this.store.pipe(select(AuthSelectors.getUserToken));
    };
    /**
     * Refreshes the user token
     * @param token a user token to refresh
     */
    AuthService.prototype.refreshUserToken = function (token) {
        this.store.dispatch(new AuthActions.RefreshUserToken({
            refreshToken: token.refresh_token,
        }));
    };
    /**
     * Store the provided token
     */
    AuthService.prototype.authorizeWithToken = function (token) {
        this.store.dispatch(new AuthActions.LoadUserTokenSuccess(token));
    };
    /**
     * Logout a storefront customer
     */
    AuthService.prototype.logout = function () {
        var _this = this;
        this.getUserToken()
            .pipe(take(1))
            .subscribe(function (userToken) {
            _this.store.dispatch(new AuthActions.Logout());
            if (Boolean(userToken) && userToken.userId === OCC_USER_ID_CURRENT) {
                _this.store.dispatch(new AuthActions.RevokeUserToken(userToken));
            }
        });
    };
    /**
     * Returns a client token.  The client token from the store is returned if there is one.
     * Otherwise, an new token is fetched from the backend and saved in the store.
     */
    AuthService.prototype.getClientToken = function () {
        var _this = this;
        return this.store.pipe(select(AuthSelectors.getClientTokenState), filter(function (state) {
            if (_this.isClientTokenLoaded(state)) {
                return true;
            }
            else {
                if (!state.loading) {
                    _this.store.dispatch(new AuthActions.LoadClientToken());
                }
                return false;
            }
        }), map(function (state) { return state.value; }));
    };
    /**
     * Fetches a clientToken from the backend ans saves it in the store where getClientToken can use it.
     * The new clientToken is returned.
     */
    AuthService.prototype.refreshClientToken = function () {
        var _this = this;
        this.store.dispatch(new AuthActions.LoadClientToken());
        return this.store.pipe(select(AuthSelectors.getClientTokenState), filter(function (state) {
            return _this.isClientTokenLoaded(state);
        }), map(function (state) { return state.value; }));
    };
    AuthService.prototype.isClientTokenLoaded = function (state) {
        return (state.success || state.error) && !state.loading;
    };
    /**
     * Returns `true` if the user is logged in; and `false` if the user is anonymous.
     */
    AuthService.prototype.isUserLoggedIn = function () {
        return this.getUserToken().pipe(map(function (userToken) { return Boolean(userToken) && Boolean(userToken.access_token); }));
    };
    AuthService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    AuthService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthService_Factory() { return new AuthService(i0.ɵɵinject(i1.Store)); }, token: AuthService, providedIn: "root" });
    AuthService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AuthService);
    return AuthService;
}());
export { AuthService };
