// @spartacus/storefront/esm2015/cms-components/myaccount/order/order-details/order-detail-items/order-detail-items.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { PromotionLocation, } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { PromotionService } from '../../../../../shared/services/promotion/promotion.service';
import { OrderDetailsService } from '../order-details.service';
import { cancelledValues, completedValues, } from './order-consigned-entries/order-consigned-entries.model';
let OrderDetailItemsComponent = class OrderDetailItemsComponent {
    constructor(orderDetailsService, promotionService) {
        this.orderDetailsService = orderDetailsService;
        this.promotionService = promotionService;
        this.promotionLocation = PromotionLocation.Order;
        this.order$ = this.orderDetailsService.getOrderDetails();
    }
    ngOnInit() {
        this.orderPromotions$ = this.promotionService.getOrderPromotions(this.promotionLocation);
        this.others$ = this.getOtherStatus(...completedValues, ...cancelledValues);
        this.completed$ = this.getExactStatus(completedValues);
        this.cancel$ = this.getExactStatus(cancelledValues);
    }
    getExactStatus(consignmentStatus) {
        return this.order$.pipe(map((order) => {
            if (Boolean(order.consignments)) {
                return order.consignments.filter((consignment) => consignmentStatus.includes(consignment.status));
            }
        }));
    }
    getOtherStatus(...consignmentStatus) {
        return this.order$.pipe(map((order) => {
            if (Boolean(order.consignments)) {
                return order.consignments.filter((consignment) => !consignmentStatus.includes(consignment.status));
            }
        }));
    }
};
OrderDetailItemsComponent.ctorParameters = () => [
    { type: OrderDetailsService },
    { type: PromotionService }
];
OrderDetailItemsComponent = __decorate([
    Component({
        selector: 'cx-order-details-items',
        template: "<ng-container *ngIf=\"order$ | async as order\">\r\n  <ng-container\r\n    *ngIf=\"order.consignments?.length || order.unconsignedEntries?.length\"\r\n  >\r\n    <ng-container *ngIf=\"orderPromotions$ | async as orderPromotions\">\r\n      <cx-promotions [promotions]=\"orderPromotions\"></cx-promotions>\r\n    </ng-container>\r\n  </ng-container>\r\n\r\n  <!-- consigned entries -->\r\n  <ng-container *ngIf=\"order.consignments?.length\">\r\n    <cx-order-consigned-entries\r\n      *ngIf=\"others$ | async as others\"\r\n      [order]=\"order\"\r\n      [consignments]=\"others\"\r\n    ></cx-order-consigned-entries>\r\n\r\n    <cx-order-consigned-entries\r\n      *ngIf=\"completed$ | async as completed\"\r\n      [order]=\"order\"\r\n      [consignments]=\"completed\"\r\n    ></cx-order-consigned-entries>\r\n\r\n    <cx-order-consigned-entries\r\n      *ngIf=\"cancel$ | async as cancel\"\r\n      [order]=\"order\"\r\n      [consignments]=\"cancel\"\r\n    ></cx-order-consigned-entries>\r\n  </ng-container>\r\n\r\n  <!-- unconsigned entries -->\r\n  <ng-container *ngIf=\"order.unconsignedEntries?.length\">\r\n    <div class=\"cx-list row\">\r\n      <div class=\"cx-list-header col-12\">\r\n        <div class=\"cx-list-status\">\r\n          {{\r\n            'orderDetails.statusDisplay'\r\n              | cxTranslate: { context: order.statusDisplay }\r\n          }}\r\n        </div>\r\n      </div>\r\n      <div class=\"cx-list-item col-12\">\r\n        <cx-cart-item-list\r\n          [items]=\"order?.unconsignedEntries\"\r\n          [readonly]=\"true\"\r\n          [promotionLocation]=\"promotionLocation\"\r\n        ></cx-cart-item-list>\r\n      </div>\r\n    </div>\r\n  </ng-container>\r\n</ng-container>\r\n"
    })
], OrderDetailItemsComponent);
export { OrderDetailItemsComponent };
