// @spartacus/core/esm5/src/occ/adapters/user/occ-customer-coupon.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { CUSTOMER_COUPON_SEARCH_RESULT_NORMALIZER } from '../../../user/connectors/customer-coupon/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
import { OCC_USER_ID_ANONYMOUS } from '../../utils/occ-constants';
var OccCustomerCouponAdapter = /** @class */ (function () {
    function OccCustomerCouponAdapter(http, occEndpoints, converter) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
    }
    OccCustomerCouponAdapter.prototype.getCustomerCoupons = function (userId, pageSize, currentPage, sort) {
        // Currently OCC only supports calls for customer coupons in case of logged users
        if (userId === OCC_USER_ID_ANONYMOUS) {
            return of({});
        }
        var url = this.occEndpoints.getUrl('customerCoupons', { userId: userId });
        var params = new HttpParams().set('sort', sort ? sort : 'startDate:asc');
        if (pageSize) {
            params = params.set('pageSize', pageSize.toString());
        }
        if (currentPage) {
            params = params.set('currentPage', currentPage.toString());
        }
        var headers = this.newHttpHeader();
        return this.http
            .get(url, { headers: headers, params: params })
            .pipe(this.converter.pipeable(CUSTOMER_COUPON_SEARCH_RESULT_NORMALIZER));
    };
    OccCustomerCouponAdapter.prototype.turnOffNotification = function (userId, couponCode) {
        var url = this.occEndpoints.getUrl('couponNotification', {
            userId: userId,
            couponCode: couponCode,
        });
        var headers = this.newHttpHeader();
        return this.http.delete(url, { headers: headers });
    };
    OccCustomerCouponAdapter.prototype.turnOnNotification = function (userId, couponCode) {
        var url = this.occEndpoints.getUrl('couponNotification', {
            userId: userId,
            couponCode: couponCode,
        });
        var headers = this.newHttpHeader();
        return this.http.post(url, { headers: headers });
    };
    OccCustomerCouponAdapter.prototype.claimCustomerCoupon = function (userId, couponCode) {
        var url = this.occEndpoints.getUrl('claimCoupon', {
            userId: userId,
            couponCode: couponCode,
        });
        var headers = this.newHttpHeader();
        return this.http.post(url, { headers: headers });
    };
    OccCustomerCouponAdapter.prototype.newHttpHeader = function () {
        return new HttpHeaders({
            'Content-Type': 'application/json',
        });
    };
    OccCustomerCouponAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService }
    ]; };
    OccCustomerCouponAdapter = __decorate([
        Injectable()
    ], OccCustomerCouponAdapter);
    return OccCustomerCouponAdapter;
}());
export { OccCustomerCouponAdapter };
