// @spartacus/core/esm5/src/occ/adapters/user/occ-user.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { TITLE_NORMALIZER, USER_NORMALIZER, USER_SERIALIZER, USER_SIGN_UP_SERIALIZER, } from '../../../user/connectors/user/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
import { InterceptorUtil, USE_CLIENT_TOKEN, } from '../../utils/interceptor-util';
var OccUserAdapter = /** @class */ (function () {
    function OccUserAdapter(http, occEndpoints, converter) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
    }
    OccUserAdapter.prototype.load = function (userId) {
        var url = this.occEndpoints.getUrl('user', { userId: userId });
        return this.http
            .get(url)
            .pipe(this.converter.pipeable(USER_NORMALIZER));
    };
    OccUserAdapter.prototype.update = function (userId, user) {
        var url = this.occEndpoints.getUrl('user', { userId: userId });
        user = this.converter.convert(user, USER_SERIALIZER);
        return this.http.patch(url, user);
    };
    OccUserAdapter.prototype.register = function (user) {
        var url = this.occEndpoints.getUrl('userRegister');
        var headers = new HttpHeaders({
            'Content-Type': 'application/json',
        });
        headers = InterceptorUtil.createHeader(USE_CLIENT_TOKEN, true, headers);
        user = this.converter.convert(user, USER_SIGN_UP_SERIALIZER);
        return this.http
            .post(url, user, { headers: headers })
            .pipe(this.converter.pipeable(USER_NORMALIZER));
    };
    OccUserAdapter.prototype.registerGuest = function (guid, password) {
        var url = this.occEndpoints.getUrl('userRegister');
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        headers = InterceptorUtil.createHeader(USE_CLIENT_TOKEN, true, headers);
        var httpParams = new HttpParams()
            .set('guid', guid)
            .set('password', password);
        return this.http
            .post(url, httpParams, { headers: headers })
            .pipe(this.converter.pipeable(USER_NORMALIZER));
    };
    OccUserAdapter.prototype.requestForgotPasswordEmail = function (userEmailAddress) {
        var url = this.occEndpoints.getUrl('userForgotPassword');
        var httpParams = new HttpParams().set('userId', userEmailAddress);
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        headers = InterceptorUtil.createHeader(USE_CLIENT_TOKEN, true, headers);
        return this.http.post(url, httpParams, { headers: headers });
    };
    OccUserAdapter.prototype.resetPassword = function (token, newPassword) {
        var url = this.occEndpoints.getUrl('userResetPassword');
        var headers = new HttpHeaders({
            'Content-Type': 'application/json',
        });
        headers = InterceptorUtil.createHeader(USE_CLIENT_TOKEN, true, headers);
        return this.http.post(url, { token: token, newPassword: newPassword }, { headers: headers });
    };
    OccUserAdapter.prototype.updateEmail = function (userId, currentPassword, newUserId) {
        var url = this.occEndpoints.getUrl('userUpdateLoginId', { userId: userId });
        var httpParams = new HttpParams()
            .set('password', currentPassword)
            .set('newLogin', newUserId);
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        return this.http.put(url, httpParams, { headers: headers });
    };
    OccUserAdapter.prototype.updatePassword = function (userId, oldPassword, newPassword) {
        var url = this.occEndpoints.getUrl('userUpdatePassword', { userId: userId });
        var httpParams = new HttpParams()
            .set('old', oldPassword)
            .set('new', newPassword);
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        return this.http.put(url, httpParams, { headers: headers });
    };
    OccUserAdapter.prototype.remove = function (userId) {
        var url = this.occEndpoints.getUrl('user', { userId: userId });
        return this.http.delete(url);
    };
    OccUserAdapter.prototype.loadTitles = function () {
        var url = this.occEndpoints.getUrl('titles');
        return this.http.get(url).pipe(map(function (titleList) { return titleList.titles; }), this.converter.pipeableMany(TITLE_NORMALIZER));
    };
    OccUserAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService }
    ]; };
    OccUserAdapter = __decorate([
        Injectable()
    ], OccUserAdapter);
    return OccUserAdapter;
}());
export { OccUserAdapter };
