// @spartacus/core/esm5/src/occ/adapters/user/occ-user-consent.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { CONSENT_TEMPLATE_NORMALIZER } from '../../../user/connectors/consent/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
var OccUserConsentAdapter = /** @class */ (function () {
    function OccUserConsentAdapter(http, occEndpoints, converter) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
    }
    OccUserConsentAdapter.prototype.loadConsents = function (userId) {
        var url = this.occEndpoints.getUrl('consentTemplates', { userId: userId });
        var headers = new HttpHeaders({ 'Cache-Control': 'no-cache' });
        return this.http
            .get(url, { headers: headers })
            .pipe(catchError(function (error) { return throwError(error); }), map(function (consentList) { return consentList.consentTemplates; }), this.converter.pipeableMany(CONSENT_TEMPLATE_NORMALIZER));
    };
    OccUserConsentAdapter.prototype.giveConsent = function (userId, consentTemplateId, consentTemplateVersion) {
        var url = this.occEndpoints.getUrl('consents', { userId: userId });
        var httpParams = new HttpParams()
            .set('consentTemplateId', consentTemplateId)
            .set('consentTemplateVersion', consentTemplateVersion.toString());
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
            'Cache-Control': 'no-cache',
        });
        return this.http
            .post(url, httpParams, { headers: headers })
            .pipe(catchError(function (error) { return throwError(error); }), this.converter.pipeable(CONSENT_TEMPLATE_NORMALIZER));
    };
    OccUserConsentAdapter.prototype.withdrawConsent = function (userId, consentCode) {
        var headers = new HttpHeaders({
            'Cache-Control': 'no-cache',
        });
        var url = this.occEndpoints.getUrl('consentDetail', {
            userId: userId,
            consentId: consentCode,
        });
        return this.http.delete(url, { headers: headers });
    };
    OccUserConsentAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService }
    ]; };
    OccUserConsentAdapter = __decorate([
        Injectable()
    ], OccUserConsentAdapter);
    return OccUserConsentAdapter;
}());
export { OccUserConsentAdapter };
