// @spartacus/core/esm5/src/occ/adapters/user/occ-user-notification-preference.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { NOTIFICATION_PREFERENCE_NORMALIZER, NOTIFICATION_PREFERENCE_SERIALIZER, } from '../../../user/connectors/notification-preference';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
var headers = new HttpHeaders({
    'Content-Type': 'application/json',
});
var OccUserNotificationPreferenceAdapter = /** @class */ (function () {
    function OccUserNotificationPreferenceAdapter(http, converter, occEndpoints) {
        this.http = http;
        this.converter = converter;
        this.occEndpoints = occEndpoints;
    }
    OccUserNotificationPreferenceAdapter.prototype.loadAll = function (userId) {
        return this.http
            .get(this.occEndpoints.getUrl('notificationPreference', { userId: userId }), {
            headers: headers,
        })
            .pipe(map(function (list) { return list.preferences; }), this.converter.pipeableMany(NOTIFICATION_PREFERENCE_NORMALIZER), catchError(function (error) { return throwError(error); }));
    };
    OccUserNotificationPreferenceAdapter.prototype.update = function (userId, preferences) {
        preferences = this.converter.convert(preferences, NOTIFICATION_PREFERENCE_SERIALIZER);
        return this.http
            .patch(this.occEndpoints.getUrl('notificationPreference', { userId: userId }), { preferences: preferences }, { headers: headers })
            .pipe(catchError(function (error) { return throwError(error); }));
    };
    OccUserNotificationPreferenceAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: ConverterService },
        { type: OccEndpointsService }
    ]; };
    OccUserNotificationPreferenceAdapter = __decorate([
        Injectable()
    ], OccUserNotificationPreferenceAdapter);
    return OccUserNotificationPreferenceAdapter;
}());
export { OccUserNotificationPreferenceAdapter };
