// @spartacus/storefront/esm5/layout/a11y/keyboard-focus/autofocus/auto-focus.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { EscapeFocusService } from '../escape/escape-focus.service';
import { FOCUS_ATTR, } from '../keyboard-focus.model';
import * as i0 from "@angular/core";
import * as i1 from "../services/select-focus.util";
var AutoFocusService = /** @class */ (function (_super) {
    __extends(AutoFocusService, _super);
    function AutoFocusService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Returns the first focusable child element of the host element.
     */
    AutoFocusService.prototype.findFirstFocusable = function (host, config) {
        if (config === void 0) { config = { autofocus: true }; }
        if ((config === null || config === void 0 ? void 0 : config.autofocus) === ':host') {
            return host;
        }
        else if (this.hasPersistedFocus(host, config)) {
            return this.getPersisted(host, this.getPersistenceGroup(host, config));
        }
        else {
            return this.selectFocusUtil.findFirstFocusable(host, config) || host;
        }
    };
    /**
     * Indicates whether any of the focusabe child elements is focused.
     */
    AutoFocusService.prototype.hasPersistedFocus = function (host, config) {
        return !!this.getPersisted(host, this.getPersistenceGroup(host, config));
    };
    /**
     * Returns the element that has a persisted focus state.
     *
     * @param host the `HTMLElement` used to query for focusable children
     * @param group the optional group for the persistent state, to separate different focus
     *   groups and remain the persistence
     */
    AutoFocusService.prototype.getPersisted = function (host, group) {
        if (!this.get(group)) {
            return;
        }
        var focussed = Array.from(host.querySelectorAll("[" + FOCUS_ATTR + "='" + this.get(group) + "']"));
        return focussed.length > 0 ? focussed[0] : null;
    };
    AutoFocusService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AutoFocusService_Factory() { return new AutoFocusService(i0.ɵɵinject(i1.SelectFocusUtility)); }, token: AutoFocusService, providedIn: "root" });
    AutoFocusService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AutoFocusService);
    return AutoFocusService;
}(EscapeFocusService));
export { AutoFocusService };
