// @spartacus/core/esm5/src/asm/facade/asm-auth.service.js
import { __assign, __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { take } from 'rxjs/operators';
import { AuthService } from '../../auth/facade/auth.service';
import { AuthActions } from '../../auth/store/actions';
import { OCC_USER_ID_CURRENT } from '../../occ/utils/occ-constants';
import { AsmActions } from '../store/actions/index';
import { AsmSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var AsmAuthService = /** @class */ (function () {
    function AsmAuthService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Loads a user token for a customer support agent
     * @param userId
     * @param password
     */
    AsmAuthService.prototype.authorizeCustomerSupportAgent = function (userId, password) {
        this.store.dispatch(new AsmActions.LoadCustomerSupportAgentToken({
            userId: userId,
            password: password,
        }));
    };
    /**
     * Starts an ASM customer emulation session.
     * A customer emulation session is stoped by calling logout().
     * @param customerSupportAgentToken
     * @param customerId
     */
    AsmAuthService.prototype.startCustomerEmulationSession = function (customerSupportAgentToken, customerId) {
        this.authService.authorizeWithToken(__assign(__assign({}, customerSupportAgentToken), { userId: customerId }));
    };
    /**
     * Utility function to determine if a given token is a customer emulation session token.
     * @param userToken
     */
    AsmAuthService.prototype.isCustomerEmulationToken = function (userToken) {
        return (Boolean(userToken) &&
            Boolean(userToken.userId) &&
            userToken.userId !== OCC_USER_ID_CURRENT);
    };
    /**
     * Returns the customer support agent's token
     */
    AsmAuthService.prototype.getCustomerSupportAgentToken = function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSupportAgentToken));
    };
    /**
     * Returns the customer support agent's token loading status
     */
    AsmAuthService.prototype.getCustomerSupportAgentTokenLoading = function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSupportAgentTokenLoading));
    };
    /**
     * Logout a customer support agent
     */
    AsmAuthService.prototype.logoutCustomerSupportAgent = function () {
        var _this = this;
        this.getCustomerSupportAgentToken()
            .pipe(take(1))
            .subscribe(function (userToken) {
            _this.store.dispatch(new AsmActions.LogoutCustomerSupportAgent());
            _this.store.dispatch(new AuthActions.RevokeUserToken(userToken));
        });
    };
    AsmAuthService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    AsmAuthService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AsmAuthService_Factory() { return new AsmAuthService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: AsmAuthService, providedIn: "root" });
    AsmAuthService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AsmAuthService);
    return AsmAuthService;
}());
export { AsmAuthService };
