// @spartacus/core/esm5/src/asm/facade/asm.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { AsmActions } from '../store/actions/index';
import { AsmSelectors } from '../store/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var AsmService = /** @class */ (function () {
    function AsmService(store) {
        this.store = store;
    }
    /**
     * Search for customers
     * @param options
     */
    AsmService.prototype.customerSearch = function (options) {
        this.store.dispatch(new AsmActions.CustomerSearch(options));
    };
    /**
     * Reset the customer search result data to the initial state.
     */
    AsmService.prototype.customerSearchReset = function () {
        this.store.dispatch(new AsmActions.CustomerSearchReset());
    };
    /**
     * Returns the customer search result data.
     */
    AsmService.prototype.getCustomerSearchResults = function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSearchResults));
    };
    /**
     * Returns the customer search result loading status.
     */
    AsmService.prototype.getCustomerSearchResultsLoading = function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSearchResultsLoading));
    };
    /**
     * Updates the state of the ASM UI
     */
    AsmService.prototype.updateAsmUiState = function (asmUi) {
        this.store.dispatch(new AsmActions.AsmUiUpdate(asmUi));
    };
    /**
     * Get the state of the ASM UI
     */
    AsmService.prototype.getAsmUiState = function () {
        return this.store.pipe(select(AsmSelectors.getAsmUi));
    };
    AsmService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    AsmService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AsmService_Factory() { return new AsmService(i0.ɵɵinject(i1.Store)); }, token: AsmService, providedIn: "root" });
    AsmService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AsmService);
    return AsmService;
}());
export { AsmService };
