// @spartacus/storefront/esm5/cms-components/myaccount/my-coupons/coupon-claim/coupon-claim.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { RoutingService, CustomerCouponService, GlobalMessageService, GlobalMessageType, } from '@spartacus/core';
var CouponClaimComponent = /** @class */ (function () {
    function CouponClaimComponent(couponService, routingService, messageService) {
        this.couponService = couponService;
        this.routingService = routingService;
        this.messageService = messageService;
    }
    CouponClaimComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.routingService
            .getRouterState()
            .subscribe(function (k) {
            var couponCode = k.state.params.couponCode;
            if (couponCode) {
                _this.couponService.claimCustomerCoupon(couponCode);
                _this.subscription = _this.couponService
                    .getClaimCustomerCouponResultSuccess()
                    .subscribe(function (success) {
                    if (success) {
                        _this.messageService.add({ key: 'myCoupons.claimCustomerCoupon' }, GlobalMessageType.MSG_TYPE_CONFIRMATION);
                    }
                    _this.routingService.go({ cxRoute: 'coupons' });
                });
            }
            else {
                _this.routingService.go({ cxRoute: 'notFound' });
            }
        })
            .unsubscribe();
    };
    CouponClaimComponent.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    CouponClaimComponent.ctorParameters = function () { return [
        { type: CustomerCouponService },
        { type: RoutingService },
        { type: GlobalMessageService }
    ]; };
    CouponClaimComponent = __decorate([
        Component({
            template: "",
            selector: 'cx-coupon-claim'
        })
    ], CouponClaimComponent);
    return CouponClaimComponent;
}());
export { CouponClaimComponent };
