// @spartacus/storefront/esm2015/cms-components/storefinder/components/store-finder-list-item/store-finder-list-item.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { StoreDataService } from '@spartacus/core';
import { AbstractStoreItemComponent } from '../abstract-store-item/abstract-store-item.component';
let StoreFinderListItemComponent = class StoreFinderListItemComponent extends AbstractStoreItemComponent {
    constructor(storeDataService) {
        super(storeDataService);
        this.storeDataService = storeDataService;
        this.locationIndex = null;
        this.storeItemClick = new EventEmitter();
    }
    handleStoreItemClick() {
        if (this.locationIndex !== null) {
            this.storeItemClick.emit(this.locationIndex);
        }
    }
    onKey(event) {
        if (event.key === 'Enter') {
            this.handleStoreItemClick();
        }
    }
};
StoreFinderListItemComponent.ctorParameters = () => [
    { type: StoreDataService }
];
__decorate([
    Input()
], StoreFinderListItemComponent.prototype, "locationIndex", void 0);
__decorate([
    Input()
], StoreFinderListItemComponent.prototype, "listOrderLabel", void 0);
__decorate([
    Input()
], StoreFinderListItemComponent.prototype, "displayDistance", void 0);
__decorate([
    Input()
], StoreFinderListItemComponent.prototype, "useClickEvent", void 0);
__decorate([
    Output()
], StoreFinderListItemComponent.prototype, "storeItemClick", void 0);
StoreFinderListItemComponent = __decorate([
    Component({
        selector: 'cx-store-finder-list-item',
        template: "<ng-container>\r\n  <div>\r\n    <div class=\"cx-store-list-order\">\r\n      {{ listOrderLabel }}\r\n    </div>\r\n    <div class=\"cx-store-name\">\r\n      <button\r\n        *ngIf=\"useClickEvent\"\r\n        (click)=\"handleStoreItemClick()\"\r\n        (keyup)=\"onKey($event)\"\r\n      >\r\n        {{ location.displayName || location.name }}\r\n      </button>\r\n      <a *ngIf=\"!useClickEvent\" [routerLink]=\"[location.name]\">{{\r\n        location.displayName || location.name\r\n      }}</a>\r\n    </div>\r\n    <div class=\"cx-store-address\" *ngIf=\"location.address\">\r\n      <div class=\"cx-store-address-street\">\r\n        {{ location.address.line1 }} {{ location.address.line2 }}\r\n      </div>\r\n      {{\r\n        getFormattedStoreAddress([\r\n          location.address.town,\r\n          location.address.postalCode,\r\n          location.address.country.isocode\r\n        ])\r\n      }}\r\n      <div\r\n        class=\"cx-store-distance\"\r\n        *ngIf=\"location.formattedDistance && displayDistance\"\r\n      >\r\n        {{ location.formattedDistance }}\r\n      </div>\r\n    </div>\r\n    <a\r\n      href=\"{{ getDirections(location) }}\"\r\n      target=\"_blank\"\r\n      class=\"btn btn-sm btn-action btn-block cx-button\"\r\n      (click)=\"$event.stopPropagation()\"\r\n      >{{ 'storeFinder.getDirections' | cxTranslate }}</a\r\n    >\r\n  </div>\r\n</ng-container>\r\n"
    })
], StoreFinderListItemComponent);
export { StoreFinderListItemComponent };
