// @spartacus/storefront/esm2015/cms-components/myaccount/update-password/components/update-password/update-password.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { GlobalMessageService, GlobalMessageType, RoutingService, UserService, } from '@spartacus/core';
import { Subscription } from 'rxjs';
let UpdatePasswordComponent = class UpdatePasswordComponent {
    constructor(routingService, userService, globalMessageService) {
        this.routingService = routingService;
        this.userService = userService;
        this.globalMessageService = globalMessageService;
        this.subscription = new Subscription();
    }
    ngOnInit() {
        this.userService.resetUpdatePasswordProcessState();
        this.loading$ = this.userService.getUpdatePasswordResultLoading();
        this.subscription.add(this.userService
            .getUpdatePasswordResultSuccess()
            .subscribe((success) => this.onSuccess(success)));
    }
    onSuccess(success) {
        if (success) {
            this.globalMessageService.add({ key: 'updatePasswordForm.passwordUpdateSuccess' }, GlobalMessageType.MSG_TYPE_CONFIRMATION);
            this.routingService.go({ cxRoute: 'home' });
        }
    }
    onCancel() {
        this.routingService.go({ cxRoute: 'home' });
    }
    onSubmit({ oldPassword, newPassword, }) {
        this.userService.updatePassword(oldPassword, newPassword);
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
        this.userService.resetUpdatePasswordProcessState();
    }
};
UpdatePasswordComponent.ctorParameters = () => [
    { type: RoutingService },
    { type: UserService },
    { type: GlobalMessageService }
];
UpdatePasswordComponent = __decorate([
    Component({
        selector: 'cx-update-password',
        template: "<ng-container>\r\n  <div *ngIf=\"loading$ | async; else updateForm\">\r\n    <div class=\"cx-spinner\">\r\n      <cx-spinner></cx-spinner>\r\n    </div>\r\n  </div>\r\n\r\n  <ng-template #updateForm>\r\n    <div class=\"container\">\r\n      <div class=\"row d-flex justify-content-center\">\r\n        <cx-update-password-form\r\n          class=\"col-md-6\"\r\n          (cancelled)=\"onCancel()\"\r\n          (submitted)=\"onSubmit($event)\"\r\n        ></cx-update-password-form>\r\n      </div>\r\n    </div>\r\n  </ng-template>\r\n</ng-container>\r\n"
    })
], UpdatePasswordComponent);
export { UpdatePasswordComponent };
