// @spartacus/storefront/esm5/recipes/b2c-storefront.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { provideConfig, provideDefaultConfig, provideDefaultConfigFactory, } from '@spartacus/core';
import { CmsLibModule } from '../cms-components/cms-lib.module';
import { b2cLayoutConfig, mediaConfig } from './config/index';
import { defaultCmsContentConfig } from './config/static-cms-structure/default-cms-content.config';
import { StorefrontModule } from './storefront.module';
var B2cStorefrontModule = /** @class */ (function () {
    function B2cStorefrontModule() {
    }
    B2cStorefrontModule_1 = B2cStorefrontModule;
    B2cStorefrontModule.withConfig = function (config) {
        return {
            ngModule: B2cStorefrontModule_1,
            providers: [provideConfig(config)],
        };
    };
    var B2cStorefrontModule_1;
    B2cStorefrontModule = B2cStorefrontModule_1 = __decorate([
        NgModule({
            imports: [
                StorefrontModule,
                // the cms lib module contains all components that added in the bundle
                CmsLibModule,
            ],
            providers: [
                provideDefaultConfig({
                    pwa: {
                        enabled: true,
                        addToHomeScreen: true,
                    },
                }),
                provideDefaultConfig(b2cLayoutConfig),
                provideDefaultConfig(mediaConfig),
                provideDefaultConfigFactory(defaultCmsContentConfig),
            ],
            exports: [StorefrontModule],
        })
    ], B2cStorefrontModule);
    return B2cStorefrontModule;
}());
export { B2cStorefrontModule };
