// @spartacus/storefront/esm5/cms-components/product/product-list/product-view/product-view.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { ICON_TYPE } from '../../../misc/icon/icon.model';
export var ViewModes;
(function (ViewModes) {
    ViewModes["Grid"] = "grid";
    ViewModes["List"] = "list";
})(ViewModes || (ViewModes = {}));
var ProductViewComponent = /** @class */ (function () {
    function ProductViewComponent() {
        this.iconTypes = ICON_TYPE;
        this.modeChange = new EventEmitter();
    }
    Object.defineProperty(ProductViewComponent.prototype, "buttonClass", {
        get: function () {
            var viewName = this.viewMode.toLowerCase();
            return "cx-product-" + viewName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProductViewComponent.prototype, "viewMode", {
        /**
         *   Display icons inversely to allow users
         *   to see the view they will navigate to
         */
        get: function () {
            if (this.mode === 'list') {
                return this.iconTypes.GRID;
            }
            else if (this.mode === 'grid') {
                return this.iconTypes.LIST;
            }
        },
        enumerable: true,
        configurable: true
    });
    ProductViewComponent.prototype.changeMode = function () {
        var newMode = this.mode === ViewModes.Grid ? ViewModes.List : ViewModes.Grid;
        this.modeChange.emit(newMode);
    };
    __decorate([
        Input()
    ], ProductViewComponent.prototype, "mode", void 0);
    __decorate([
        Output()
    ], ProductViewComponent.prototype, "modeChange", void 0);
    ProductViewComponent = __decorate([
        Component({
            selector: 'cx-product-view',
            template: "<button\r\n  class=\"btn cx-product-layout\"\r\n  [ngClass]=\"buttonClass\"\r\n  (click)=\"changeMode()\"\r\n>\r\n  <cx-icon\r\n    *ngIf=\"viewMode === iconTypes.GRID\"\r\n    [type]=\"iconTypes.GRID\"\r\n  ></cx-icon>\r\n  <cx-icon\r\n    *ngIf=\"viewMode === iconTypes.LIST\"\r\n    [type]=\"iconTypes.LIST\"\r\n  ></cx-icon>\r\n</button>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ProductViewComponent);
    return ProductViewComponent;
}());
export { ProductViewComponent };
