// @spartacus/storefront/esm5/cms-components/product/product-tabs/product-details-tab/product-details-tab.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { provideDefaultConfig } from '@spartacus/core';
import { ProductDetailsTabComponent } from './product-details-tab.component';
var ProductDetailsTabModule = /** @class */ (function () {
    function ProductDetailsTabModule() {
    }
    ProductDetailsTabModule = __decorate([
        NgModule({
            imports: [CommonModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ProductDetailsTabComponent: {
                            component: ProductDetailsTabComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductDetailsTabComponent],
            entryComponents: [ProductDetailsTabComponent],
            exports: [ProductDetailsTabComponent],
        })
    ], ProductDetailsTabModule);
    return ProductDetailsTabModule;
}());
export { ProductDetailsTabModule };
