// @spartacus/core/esm5/src/state/utils/get-state-slice.js
import { __values } from "tslib";
import { deepMerge } from '../../config/utils/deep-merge';
var OBJECT_SEPARATOR = '.';
export function getStateSliceValue(keys, state) {
    return keys
        .split(OBJECT_SEPARATOR)
        .reduce(function (previous, current) { return (previous ? previous[current] : undefined); }, state);
}
export function createShellObject(key, excludeKeys, value) {
    if (!key || !value || Object.keys(value).length === 0) {
        return {};
    }
    var shell = key.split(OBJECT_SEPARATOR).reduceRight(function (acc, previous) {
        var _a;
        return _a = {}, _a[previous] = acc, _a;
    }, value);
    return handleExclusions(key, excludeKeys, shell);
}
export function getStateSlice(keys, excludeKeys, state) {
    var e_1, _a;
    if (keys && keys.length === 0) {
        return {};
    }
    var stateSlices = {};
    try {
        for (var keys_1 = __values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
            var currentKey = keys_1_1.value;
            var stateValue = getStateSliceValue(currentKey, state);
            var shell = createShellObject(currentKey, excludeKeys, stateValue);
            stateSlices = deepMerge(stateSlices, shell);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return stateSlices;
}
export function handleExclusions(key, excludeKeys, value) {
    var e_2, _a;
    var exclusionKeys = getExclusionKeys(key, excludeKeys);
    if (exclusionKeys.length === 0) {
        return value;
    }
    var finalValue = deepMerge({}, value);
    try {
        for (var exclusionKeys_1 = __values(exclusionKeys), exclusionKeys_1_1 = exclusionKeys_1.next(); !exclusionKeys_1_1.done; exclusionKeys_1_1 = exclusionKeys_1.next()) {
            var currentExclusionKey = exclusionKeys_1_1.value;
            var exclusionChunksSplit = currentExclusionKey.split(OBJECT_SEPARATOR);
            var nestedTemp = finalValue;
            for (var i = 0; i < exclusionChunksSplit.length; i++) {
                var currentChunk = exclusionChunksSplit[i];
                // last iteration
                if (i === exclusionChunksSplit.length - 1) {
                    if (nestedTemp && nestedTemp[currentChunk]) {
                        delete nestedTemp[currentChunk];
                    }
                }
                else {
                    nestedTemp = nestedTemp[currentChunk];
                }
            }
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (exclusionKeys_1_1 && !exclusionKeys_1_1.done && (_a = exclusionKeys_1.return)) _a.call(exclusionKeys_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return finalValue;
}
export function getExclusionKeys(key, excludeKeys) {
    var e_3, _a;
    if (!key || !excludeKeys) {
        return [];
    }
    var exclusionKeys = [];
    try {
        for (var excludeKeys_1 = __values(excludeKeys), excludeKeys_1_1 = excludeKeys_1.next(); !excludeKeys_1_1.done; excludeKeys_1_1 = excludeKeys_1.next()) {
            var exclusionKey = excludeKeys_1_1.value;
            if (exclusionKey.includes(key)) {
                exclusionKeys.push(exclusionKey);
            }
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (excludeKeys_1_1 && !excludeKeys_1_1.done && (_a = excludeKeys_1.return)) _a.call(excludeKeys_1);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return exclusionKeys;
}
export function filterKeysByType(keys, type) {
    if (!keys) {
        return [];
    }
    return Object.keys(keys).filter(function (key) { return keys[key] === type; });
}
