// @spartacus/storefront/esm5/cms-components/myaccount/order/amend-order/cancellations/order-cancellation.guard.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { RoutingService } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { OrderCancellationService } from './order-cancellation.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "./order-cancellation.service";
var OrderCancellationGuard = /** @class */ (function () {
    function OrderCancellationGuard(routing, orderAmendService) {
        this.routing = routing;
        this.orderAmendService = orderAmendService;
    }
    OrderCancellationGuard.prototype.canActivate = function () {
        var _this = this;
        return this.orderAmendService.getForm().pipe(map(function (form) {
            if (!form.valid) {
                // the order code is not available in the route
                // as long as we're inside a guard, hence we redirect
                // to the common orders page.
                _this.routing.go({ cxRoute: 'orders' });
                return false;
            }
            else {
                return true;
            }
        }));
    };
    OrderCancellationGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: OrderCancellationService }
    ]; };
    OrderCancellationGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function OrderCancellationGuard_Factory() { return new OrderCancellationGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.OrderCancellationService)); }, token: OrderCancellationGuard, providedIn: "root" });
    OrderCancellationGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], OrderCancellationGuard);
    return OrderCancellationGuard;
}());
export { OrderCancellationGuard };
