// @spartacus/core/esm5/src/user/facade/customer-coupon.service.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { combineLatest } from 'rxjs';
import { filter, map, tap } from 'rxjs/operators';
import { AuthService } from '../../auth/facade/auth.service';
import { getProcessErrorFactory, getProcessLoadingFactory, getProcessSuccessFactory, } from '../../process/store/selectors/process.selectors';
import { UserActions } from '../store/actions/index';
import { UsersSelectors } from '../store/selectors/index';
import { CLAIM_CUSTOMER_COUPON_PROCESS_ID, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, } from '../store/user-state';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var CustomerCouponService = /** @class */ (function () {
    function CustomerCouponService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Retrieves customer's coupons
     * @param pageSize page size
     * @param currentPage current page
     * @param sort sort
     */
    CustomerCouponService.prototype.loadCustomerCoupons = function (pageSize, currentPage, sort) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.LoadCustomerCoupons({
                userId: userId,
                pageSize: pageSize,
                currentPage: currentPage,
                sort: sort,
            }));
        });
    };
    /**
     * Returns customer coupon search result
     * @param pageSize page size
     */
    CustomerCouponService.prototype.getCustomerCoupons = function (pageSize) {
        var _this = this;
        return combineLatest([
            this.store.pipe(select(UsersSelectors.getCustomerCouponsState)),
            this.getClaimCustomerCouponResultLoading(),
        ]).pipe(filter(function (_a) {
            var _b = __read(_a, 2), loading = _b[1];
            return !loading;
        }), tap(function (_a) {
            var _b = __read(_a, 1), customerCouponsState = _b[0];
            var attemptedLoad = customerCouponsState.loading ||
                customerCouponsState.success ||
                customerCouponsState.error;
            if (!attemptedLoad) {
                _this.loadCustomerCoupons(pageSize);
            }
        }), map(function (_a) {
            var _b = __read(_a, 1), customerCouponsState = _b[0];
            return customerCouponsState.value;
        }));
    };
    /**
     * Returns a loaded flag for customer coupons
     */
    CustomerCouponService.prototype.getCustomerCouponsLoaded = function () {
        return this.store.pipe(select(UsersSelectors.getCustomerCouponsLoaded));
    };
    /**
     * Returns a loading flag for customer coupons
     */
    CustomerCouponService.prototype.getCustomerCouponsLoading = function () {
        return this.store.pipe(select(UsersSelectors.getCustomerCouponsLoading));
    };
    /**
     * Subscribe a CustomerCoupon Notification
     * @param couponCode a customer coupon code
     */
    CustomerCouponService.prototype.subscribeCustomerCoupon = function (couponCode) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.SubscribeCustomerCoupon({
                userId: userId,
                couponCode: couponCode,
            }));
        });
    };
    /**
     * Returns the subscribe customer coupon notification process loading flag
     */
    CustomerCouponService.prototype.getSubscribeCustomerCouponResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Returns the subscribe customer coupon notification process success flag
     */
    CustomerCouponService.prototype.getSubscribeCustomerCouponResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Returns the subscribe customer coupon notification process error flag
     */
    CustomerCouponService.prototype.getSubscribeCustomerCouponResultError = function () {
        return this.store.pipe(select(getProcessErrorFactory(SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Unsubscribe a CustomerCoupon Notification
     * @param couponCode a customer coupon code
     */
    CustomerCouponService.prototype.unsubscribeCustomerCoupon = function (couponCode) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.UnsubscribeCustomerCoupon({
                userId: userId,
                couponCode: couponCode,
            }));
        });
    };
    /**
     * Returns the unsubscribe customer coupon notification process loading flag
     */
    CustomerCouponService.prototype.getUnsubscribeCustomerCouponResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Returns the unsubscribe customer coupon notification process success flag
     */
    CustomerCouponService.prototype.getUnsubscribeCustomerCouponResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Returns the unsubscribe customer coupon notification process error flag
     */
    CustomerCouponService.prototype.getUnsubscribeCustomerCouponResultError = function () {
        return this.store.pipe(select(getProcessErrorFactory(UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Claim a CustomerCoupon
     * @param couponCode a customer coupon code
     */
    CustomerCouponService.prototype.claimCustomerCoupon = function (couponCode) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.ClaimCustomerCoupon({
                userId: userId,
                couponCode: couponCode,
            }));
        });
    };
    /**
     * Returns the claim customer coupon notification process success flag
     */
    CustomerCouponService.prototype.getClaimCustomerCouponResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(CLAIM_CUSTOMER_COUPON_PROCESS_ID)));
    };
    /**
     * Returns the claim customer coupon notification process loading flag
     */
    CustomerCouponService.prototype.getClaimCustomerCouponResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(CLAIM_CUSTOMER_COUPON_PROCESS_ID)));
    };
    CustomerCouponService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    CustomerCouponService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CustomerCouponService_Factory() { return new CustomerCouponService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: CustomerCouponService, providedIn: "root" });
    CustomerCouponService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CustomerCouponService);
    return CustomerCouponService;
}());
export { CustomerCouponService };
