// @spartacus/core/esm5/src/user/facade/user-notification-preference.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { AuthService } from '../../auth/facade/auth.service';
import { getProcessLoadingFactory } from '../../process/store/selectors/process.selectors';
import { UserActions } from '../store/actions/index';
import { UsersSelectors } from '../store/selectors/index';
import { UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID, } from '../store/user-state';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var UserNotificationPreferenceService = /** @class */ (function () {
    function UserNotificationPreferenceService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Returns all notification preferences.
     */
    UserNotificationPreferenceService.prototype.getPreferences = function () {
        return this.store.pipe(select(UsersSelectors.getPreferences));
    };
    /**
     * Returns all enabled notification preferences.
     */
    UserNotificationPreferenceService.prototype.getEnabledPreferences = function () {
        return this.store.pipe(select(UsersSelectors.getEnabledPreferences));
    };
    /**
     * Loads all notification preferences.
     */
    UserNotificationPreferenceService.prototype.loadPreferences = function () {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.LoadNotificationPreferences(userId));
        });
    };
    /**
     * Clear all notification preferences.
     */
    UserNotificationPreferenceService.prototype.clearPreferences = function () {
        this.store.dispatch(new UserActions.ClearNotificationPreferences());
    };
    /**
     * Returns a loading flag for notification preferences.
     */
    UserNotificationPreferenceService.prototype.getPreferencesLoading = function () {
        return this.store.pipe(select(UsersSelectors.getPreferencesLoading));
    };
    /**
     * Updating notification preferences.
     * @param preferences a preference list
     */
    UserNotificationPreferenceService.prototype.updatePreferences = function (preferences) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.UpdateNotificationPreferences({
                userId: userId,
                preferences: preferences,
            }));
        });
    };
    /**
     * Returns a loading flag for updating preferences.
     */
    UserNotificationPreferenceService.prototype.getUpdatePreferencesResultLoading = function () {
        return this.store.select(getProcessLoadingFactory(UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID));
    };
    /**
     * Resets the update notification preferences process state. The state needs to be
     * reset after the process concludes, regardless if it's a success or an error.
     */
    UserNotificationPreferenceService.prototype.resetNotificationPreferences = function () {
        this.store.dispatch(new UserActions.ResetNotificationPreferences());
    };
    UserNotificationPreferenceService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    UserNotificationPreferenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserNotificationPreferenceService_Factory() { return new UserNotificationPreferenceService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: UserNotificationPreferenceService, providedIn: "root" });
    UserNotificationPreferenceService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], UserNotificationPreferenceService);
    return UserNotificationPreferenceService;
}());
export { UserNotificationPreferenceService };
