// @spartacus/core/esm5/src/user/facade/user-address.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { map } from 'rxjs/operators';
import { AuthService } from '../../auth/facade/auth.service';
import { UserActions } from '../store/actions/index';
import { UsersSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var UserAddressService = /** @class */ (function () {
    function UserAddressService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Retrieves user's addresses
     */
    UserAddressService.prototype.loadAddresses = function () {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.LoadUserAddresses(userId));
        });
    };
    /**
     * Adds user address
     * @param address a user address
     */
    UserAddressService.prototype.addUserAddress = function (address) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.AddUserAddress({
                userId: userId,
                address: address,
            }));
        });
    };
    /**
     * Sets user address as default
     * @param addressId a user address ID
     */
    UserAddressService.prototype.setAddressAsDefault = function (addressId) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.UpdateUserAddress({
                userId: userId,
                addressId: addressId,
                address: { defaultAddress: true },
            }));
        });
    };
    /**
     * Updates existing user address
     * @param addressId a user address ID
     * @param address a user address
     */
    UserAddressService.prototype.updateUserAddress = function (addressId, address) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.UpdateUserAddress({
                userId: userId,
                addressId: addressId,
                address: address,
            }));
        });
    };
    /**
     * Deletes existing user address
     * @param addressId a user address ID
     */
    UserAddressService.prototype.deleteUserAddress = function (addressId) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.DeleteUserAddress({
                userId: userId,
                addressId: addressId,
            }));
        });
    };
    /**
     * Returns addresses
     */
    UserAddressService.prototype.getAddresses = function () {
        return this.store.pipe(select(UsersSelectors.getAddresses));
    };
    /**
     * Returns a loading flag for addresses
     */
    UserAddressService.prototype.getAddressesLoading = function () {
        return this.store.pipe(select(UsersSelectors.getAddressesLoading));
    };
    UserAddressService.prototype.getAddressesLoadedSuccess = function () {
        return this.store.pipe(select(UsersSelectors.getAddressesLoadedSuccess));
    };
    /**
     * Retrieves delivery countries
     */
    UserAddressService.prototype.loadDeliveryCountries = function () {
        this.store.dispatch(new UserActions.LoadDeliveryCountries());
    };
    /**
     * Returns all delivery countries
     */
    UserAddressService.prototype.getDeliveryCountries = function () {
        return this.store.pipe(select(UsersSelectors.getAllDeliveryCountries));
    };
    /**
     * Returns a country based on the provided `isocode`
     * @param isocode an isocode for a country
     */
    UserAddressService.prototype.getCountry = function (isocode) {
        return this.store.pipe(select(UsersSelectors.countrySelectorFactory(isocode)));
    };
    /**
     * Retrieves regions for specified country by `countryIsoCode`
     * @param countryIsoCode
     */
    UserAddressService.prototype.loadRegions = function (countryIsoCode) {
        this.store.dispatch(new UserActions.LoadRegions(countryIsoCode));
    };
    /**
     * Clear regions in store - useful when changing country
     */
    UserAddressService.prototype.clearRegions = function () {
        this.store.dispatch(new UserActions.ClearRegions());
    };
    /**
     * Returns all regions
     */
    UserAddressService.prototype.getRegions = function (countryIsoCode) {
        var _this = this;
        return this.store.pipe(select(UsersSelectors.getRegionsDataAndLoading), map(function (_a) {
            var regions = _a.regions, country = _a.country, loading = _a.loading, loaded = _a.loaded;
            if (!countryIsoCode && (loading || loaded)) {
                _this.clearRegions();
                return [];
            }
            else if (loading && !loaded) {
                // don't interrupt loading
                return [];
            }
            else if (!loading && countryIsoCode !== country && countryIsoCode) {
                // country changed - clear store and load new regions
                if (country) {
                    _this.clearRegions();
                }
                _this.loadRegions(countryIsoCode);
                return [];
            }
            return regions;
        }));
    };
    UserAddressService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    UserAddressService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserAddressService_Factory() { return new UserAddressService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: UserAddressService, providedIn: "root" });
    UserAddressService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], UserAddressService);
    return UserAddressService;
}());
export { UserAddressService };
