// @spartacus/core/esm5/src/user/facade/user-consent.service.js
import { __decorate, __read, __values } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { iif } from 'rxjs';
import { filter, map, switchMap, tap, withLatestFrom } from 'rxjs/operators';
import { AuthService } from '../../auth/facade/auth.service';
import { getProcessErrorFactory, getProcessLoadingFactory, getProcessSuccessFactory, } from '../../process/store/selectors/process.selectors';
import { UserActions } from '../store/actions/index';
import { UsersSelectors } from '../store/selectors/index';
import { GIVE_CONSENT_PROCESS_ID, WITHDRAW_CONSENT_PROCESS_ID, } from '../store/user-state';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var UserConsentService = /** @class */ (function () {
    function UserConsentService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Retrieves all consents.
     */
    UserConsentService.prototype.loadConsents = function () {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.LoadUserConsents(userId));
        });
    };
    /**
     * Returns all consent templates. If `loadIfMissing` parameter is set to `true`, the method triggers the load if consent templates.
     * @param loadIfMissing is set to `true`, the method will load templates if those are not already present. The default value is `false`.
     */
    UserConsentService.prototype.getConsents = function (loadIfMissing) {
        var _this = this;
        if (loadIfMissing === void 0) { loadIfMissing = false; }
        return iif(function () { return loadIfMissing; }, this.store.pipe(select(UsersSelectors.getConsentsValue), withLatestFrom(this.getConsentsResultLoading(), this.getConsentsResultSuccess()), filter(function (_a) {
            var _b = __read(_a, 3), _templates = _b[0], loading = _b[1], _success = _b[2];
            return !loading;
        }), tap(function (_a) {
            var _b = __read(_a, 3), templates = _b[0], _loading = _b[1], success = _b[2];
            if (!templates || templates.length === 0) {
                // avoid infite loop - if we've already attempted to load templates and we got an empty array as the response
                if (!success) {
                    _this.loadConsents();
                }
            }
        }), filter(function (_a) {
            var _b = __read(_a, 2), templates = _b[0], _loading = _b[1];
            return Boolean(templates);
        }), map(function (_a) {
            var _b = __read(_a, 2), templates = _b[0], _loading = _b[1];
            return templates;
        })), this.store.pipe(select(UsersSelectors.getConsentsValue)));
    };
    /**
     * Returns the consents loading flag
     */
    UserConsentService.prototype.getConsentsResultLoading = function () {
        return this.store.pipe(select(UsersSelectors.getConsentsLoading));
    };
    /**
     * Returns the consents success flag
     */
    UserConsentService.prototype.getConsentsResultSuccess = function () {
        return this.store.pipe(select(UsersSelectors.getConsentsSuccess));
    };
    /**
     * Returns the consents error flag
     */
    UserConsentService.prototype.getConsentsResultError = function () {
        return this.store.pipe(select(UsersSelectors.getConsentsError));
    };
    /**
     * Resets the processing state for consent retrieval
     */
    UserConsentService.prototype.resetConsentsProcessState = function () {
        this.store.dispatch(new UserActions.ResetLoadUserConsents());
    };
    /**
     * Returns the registered consent for the given template ID.
     *
     * As a side-effect, the method will call `getConsents(true)` to load the templates if those are not present.
     *
     * @param templateId a template ID by which to filter the registered templates.
     */
    UserConsentService.prototype.getConsent = function (templateId) {
        var _this = this;
        return this.authService.isUserLoggedIn().pipe(filter(Boolean), tap(function () { return _this.getConsents(true); }), switchMap(function () {
            return _this.store.pipe(select(UsersSelectors.getConsentByTemplateId(templateId)));
        }), filter(function (template) { return Boolean(template); }), map(function (template) { return template.currentConsent; }));
    };
    /**
     * Returns `true` if the consent is truthy and if `consentWithdrawnDate` doesn't exist.
     * Otherwise, `false` is returned.
     *
     * @param consent to check
     */
    UserConsentService.prototype.isConsentGiven = function (consent) {
        return (Boolean(consent) &&
            Boolean(consent.consentGivenDate) &&
            !Boolean(consent.consentWithdrawnDate));
    };
    /**
     * Returns `true` if the consent is either falsy or if `consentWithdrawnDate` is present.
     * Otherwise, `false` is returned.
     *
     * @param consent to check
     */
    UserConsentService.prototype.isConsentWithdrawn = function (consent) {
        if (Boolean(consent)) {
            return Boolean(consent.consentWithdrawnDate);
        }
        return true;
    };
    /**
     * Give consent for specified consent template ID and version.
     * @param consentTemplateId a template ID for which to give a consent
     * @param consentTemplateVersion a template version for which to give a consent
     */
    UserConsentService.prototype.giveConsent = function (consentTemplateId, consentTemplateVersion) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.GiveUserConsent({
                userId: userId,
                consentTemplateId: consentTemplateId,
                consentTemplateVersion: consentTemplateVersion,
            }));
        });
    };
    /**
     * Returns the give consent process loading flag
     */
    UserConsentService.prototype.getGiveConsentResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(GIVE_CONSENT_PROCESS_ID)));
    };
    /**
     * Returns the give consent process success flag
     */
    UserConsentService.prototype.getGiveConsentResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(GIVE_CONSENT_PROCESS_ID)));
    };
    /**
     * Returns the give consent process error flag
     */
    UserConsentService.prototype.getGiveConsentResultError = function () {
        return this.store.pipe(select(getProcessErrorFactory(GIVE_CONSENT_PROCESS_ID)));
    };
    /**
     * Resents the give consent process flags
     */
    UserConsentService.prototype.resetGiveConsentProcessState = function () {
        return this.store.dispatch(new UserActions.ResetGiveUserConsentProcess());
    };
    /**
     * Withdraw consent for the given `consentCode`
     * @param consentCode for which to withdraw the consent
     */
    UserConsentService.prototype.withdrawConsent = function (consentCode) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.WithdrawUserConsent({
                userId: userId,
                consentCode: consentCode,
            }));
        });
    };
    /**
     * Returns the withdraw consent process loading flag
     */
    UserConsentService.prototype.getWithdrawConsentResultLoading = function () {
        return this.store.pipe(select(getProcessLoadingFactory(WITHDRAW_CONSENT_PROCESS_ID)));
    };
    /**
     * Returns the withdraw consent process success flag
     */
    UserConsentService.prototype.getWithdrawConsentResultSuccess = function () {
        return this.store.pipe(select(getProcessSuccessFactory(WITHDRAW_CONSENT_PROCESS_ID)));
    };
    /**
     * Returns the withdraw consent process error flag
     */
    UserConsentService.prototype.getWithdrawConsentResultError = function () {
        return this.store.pipe(select(getProcessErrorFactory(WITHDRAW_CONSENT_PROCESS_ID)));
    };
    /**
     * Resets the process flags for withdraw consent
     */
    UserConsentService.prototype.resetWithdrawConsentProcessState = function () {
        return this.store.dispatch(new UserActions.ResetWithdrawUserConsentProcess());
    };
    /**
     * Filters the provided `templateList`' templates by hiding the template IDs specified in `hideTemplateIds`.
     * If the `hideTemplateIds` is empty, the provided `templateList` is returned.
     *
     * @param templateList a list of consent templates to filter
     * @param hideTemplateIds template IDs to hide
     */
    UserConsentService.prototype.filterConsentTemplates = function (templateList, hideTemplateIds) {
        var e_1, _a;
        if (hideTemplateIds === void 0) { hideTemplateIds = []; }
        if (hideTemplateIds.length === 0) {
            return templateList;
        }
        var updatedTemplateList = [];
        try {
            for (var templateList_1 = __values(templateList), templateList_1_1 = templateList_1.next(); !templateList_1_1.done; templateList_1_1 = templateList_1.next()) {
                var template = templateList_1_1.value;
                var show = !hideTemplateIds.includes(template.id);
                if (show) {
                    updatedTemplateList.push(template);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (templateList_1_1 && !templateList_1_1.done && (_a = templateList_1.return)) _a.call(templateList_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return updatedTemplateList;
    };
    UserConsentService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    UserConsentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserConsentService_Factory() { return new UserConsentService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: UserConsentService, providedIn: "root" });
    UserConsentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], UserConsentService);
    return UserConsentService;
}());
export { UserConsentService };
