// @spartacus/core/esm5/src/user/facade/user-payment.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { AuthService } from '../../auth/facade/auth.service';
import { UserActions } from '../store/actions/index';
import { UsersSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
var UserPaymentService = /** @class */ (function () {
    function UserPaymentService(store, authService) {
        this.store = store;
        this.authService = authService;
    }
    /**
     * Loads all user's payment methods.
     */
    UserPaymentService.prototype.loadPaymentMethods = function () {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.LoadUserPaymentMethods(userId));
        });
    };
    /**
     * Returns all user's payment methods
     */
    UserPaymentService.prototype.getPaymentMethods = function () {
        return this.store.pipe(select(UsersSelectors.getPaymentMethods));
    };
    /**
     * Returns a loading flag for payment methods
     */
    UserPaymentService.prototype.getPaymentMethodsLoading = function () {
        return this.store.pipe(select(UsersSelectors.getPaymentMethodsLoading));
    };
    UserPaymentService.prototype.getPaymentMethodsLoadedSuccess = function () {
        return this.store.pipe(select(UsersSelectors.getPaymentMethodsLoadedSuccess));
    };
    /**
     * Sets the payment as a default one
     * @param paymentMethodId a payment method ID
     */
    UserPaymentService.prototype.setPaymentMethodAsDefault = function (paymentMethodId) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.SetDefaultUserPaymentMethod({
                userId: userId,
                paymentMethodId: paymentMethodId,
            }));
        });
    };
    /**
     * Deletes the payment method
     *
     * @param paymentMethodId a payment method ID
     */
    UserPaymentService.prototype.deletePaymentMethod = function (paymentMethodId) {
        var _this = this;
        this.authService.invokeWithUserId(function (userId) {
            _this.store.dispatch(new UserActions.DeleteUserPaymentMethod({
                userId: userId,
                paymentMethodId: paymentMethodId,
            }));
        });
    };
    /**
     * Returns all billing countries
     */
    UserPaymentService.prototype.getAllBillingCountries = function () {
        return this.store.pipe(select(UsersSelectors.getAllBillingCountries));
    };
    /**
     * Retrieves billing countries
     */
    UserPaymentService.prototype.loadBillingCountries = function () {
        this.store.dispatch(new UserActions.LoadBillingCountries());
    };
    UserPaymentService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService }
    ]; };
    UserPaymentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserPaymentService_Factory() { return new UserPaymentService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService)); }, token: UserPaymentService, providedIn: "root" });
    UserPaymentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], UserPaymentService);
    return UserPaymentService;
}());
export { UserPaymentService };
