// @spartacus/storefront/esm5/cms-structure/seo/structured-data/builders/product/jsonld-product-review.builder.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Product, ProductReviewService, Review } from '@spartacus/core';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
/**
 * Builds the structured data for the product reviews, see https://schema.org/Review.
 * The data includes the aggregated product rating and the individual reviews.
 */
var JsonLdProductReviewBuilder = /** @class */ (function () {
    function JsonLdProductReviewBuilder(reviewService) {
        this.reviewService = reviewService;
    }
    JsonLdProductReviewBuilder.prototype.build = function (product) {
        var _this = this;
        return this.reviewService.getByProductCode(product.code).pipe(filter(Boolean), map(function (reviews) {
            return {
                aggregateRating: _this.buildAggregatedReviews(product, reviews),
                review: reviews.map(function (review) { return _this.buildReviews(review); }),
            };
        }));
    };
    JsonLdProductReviewBuilder.prototype.buildAggregatedReviews = function (product, reviews) {
        var aggregated = {
            '@type': 'AggregateRating',
        };
        if (product.averageRating) {
            aggregated.ratingValue = product.averageRating;
        }
        if (reviews) {
            aggregated.ratingCount = reviews.filter(function (rev) { return !!rev.rating; }).length;
            aggregated.reviewCount = reviews.filter(function (rev) { return !!rev.comment; }).length;
        }
        return aggregated;
    };
    JsonLdProductReviewBuilder.prototype.buildReviews = function (review) {
        var reviewSchema = {
            '@type': 'review',
        };
        if (review.principal && review.principal.name) {
            reviewSchema.author = review.principal.name;
        }
        if (review.date) {
            var date = new Date(review.date);
            reviewSchema.datePublished = date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate();
        }
        if (review.headline) {
            reviewSchema.name = review.headline;
        }
        if (review.comment) {
            reviewSchema.description = review.comment;
        }
        if (review.rating) {
            reviewSchema.reviewRating = {
                '@type': 'Rating',
                ratingValue: review.rating.toString(),
            };
        }
        return reviewSchema;
    };
    JsonLdProductReviewBuilder.ctorParameters = function () { return [
        { type: ProductReviewService }
    ]; };
    JsonLdProductReviewBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function JsonLdProductReviewBuilder_Factory() { return new JsonLdProductReviewBuilder(i0.ɵɵinject(i1.ProductReviewService)); }, token: JsonLdProductReviewBuilder, providedIn: "root" });
    JsonLdProductReviewBuilder = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], JsonLdProductReviewBuilder);
    return JsonLdProductReviewBuilder;
}());
export { JsonLdProductReviewBuilder };
