// @spartacus/storefront/esm5/cms-structure/seo/structured-data/builders/product/product-schema.builder.js
import { __decorate, __param, __read, __spread } from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { combineLatest, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { CurrentProductService } from '../../../../../cms-components/product/current-product.service';
import { JSONLD_PRODUCT_BUILDER } from '../tokens';
import * as i0 from "@angular/core";
import * as i1 from "../../../../../cms-components/product/current-product.service";
import * as i2 from "../tokens";
/**
 * Adds the minimal structured data for the product, see https://schema.org/product.
 * The actual data collection is delegated to `JsonLdBuilder`s, which can be injected
 * using the `JSONLD_PRODUCT_BUILDER` token.
 */
var ProductSchemaBuilder = /** @class */ (function () {
    function ProductSchemaBuilder(currentProduct, builders) {
        this.currentProduct = currentProduct;
        this.builders = builders;
    }
    ProductSchemaBuilder.prototype.build = function () {
        var _this = this;
        return this.currentProduct.getProduct().pipe(switchMap(function (product) {
            if (product) {
                return combineLatest(_this.collect(product)).pipe(map(function (res) { return Object.assign.apply(Object, __spread([{}], res)); }));
            }
            return of({});
        }));
    };
    ProductSchemaBuilder.prototype.collect = function (product) {
        if (!product || !product.code) {
            return [];
        }
        var builders = this.builders
            ? this.builders.map(function (builder) { return builder.build(product); })
            : [];
        return __spread([
            of({
                '@context': 'http://schema.org',
                '@type': 'Product',
            })
        ], builders);
    };
    ProductSchemaBuilder.ctorParameters = function () { return [
        { type: CurrentProductService },
        { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [JSONLD_PRODUCT_BUILDER,] }] }
    ]; };
    ProductSchemaBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductSchemaBuilder_Factory() { return new ProductSchemaBuilder(i0.ɵɵinject(i1.CurrentProductService), i0.ɵɵinject(i2.JSONLD_PRODUCT_BUILDER, 8)); }, token: ProductSchemaBuilder, providedIn: "root" });
    ProductSchemaBuilder = __decorate([
        Injectable({
            providedIn: 'root',
        }),
        __param(1, Optional()),
        __param(1, Inject(JSONLD_PRODUCT_BUILDER))
    ], ProductSchemaBuilder);
    return ProductSchemaBuilder;
}());
export { ProductSchemaBuilder };
