// @spartacus/core/esm5/src/util/converter.service.js
import { __decorate } from "tslib";
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
var ConverterService = /** @class */ (function () {
    function ConverterService(injector) {
        this.injector = injector;
        this.converters = new Map();
    }
    ConverterService.prototype.getConverters = function (injectionToken) {
        if (!this.converters.has(injectionToken)) {
            var converters = this.injector.get(injectionToken, []);
            if (!Array.isArray(converters)) {
                console.warn('Converter must be multi-provided, please use "multi: true" for', injectionToken.toString());
            }
            this.converters.set(injectionToken, converters);
        }
        return this.converters.get(injectionToken);
    };
    /**
     * Will return true if converters for specified token were provided
     */
    ConverterService.prototype.hasConverters = function (injectionToken) {
        var converters = this.getConverters(injectionToken);
        return Array.isArray(converters) && converters.length > 0;
    };
    /**
     * Pipeable operator to apply converter logic in a observable stream
     */
    ConverterService.prototype.pipeable = function (injectionToken) {
        var _this = this;
        if (this.hasConverters(injectionToken)) {
            return map(function (model) { return _this.convertSource(model, injectionToken); });
        }
        else {
            return function (observable) { return observable; };
        }
    };
    /**
     * Pipeable operator to apply converter logic in a observable stream to collection of items
     */
    ConverterService.prototype.pipeableMany = function (injectionToken) {
        var _this = this;
        if (this.hasConverters(injectionToken)) {
            return map(function (model) { return _this.convertMany(model, injectionToken); });
        }
        else {
            return function (observable) { return observable; };
        }
    };
    /**
     * Apply converter logic specified by injection token to source data
     */
    ConverterService.prototype.convert = function (source, injectionToken) {
        if (this.hasConverters(injectionToken)) {
            return this.convertSource(source, injectionToken);
        }
        else {
            return source;
        }
    };
    /**
     * Apply converter logic specified by injection token to a collection
     */
    ConverterService.prototype.convertMany = function (sources, injectionToken) {
        var _this = this;
        if (this.hasConverters(injectionToken) && Array.isArray(sources)) {
            return sources.map(function (source) {
                return _this.convertSource(source, injectionToken);
            });
        }
        else {
            return sources;
        }
    };
    ConverterService.prototype.convertSource = function (source, injectionToken) {
        return this.getConverters(injectionToken).reduce(function (target, converter) {
            return converter.convert(source, target);
        }, undefined);
    };
    ConverterService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    ConverterService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ConverterService_Factory() { return new ConverterService(i0.ɵɵinject(i0.INJECTOR)); }, token: ConverterService, providedIn: "root" });
    ConverterService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ConverterService);
    return ConverterService;
}());
export { ConverterService };
