// @spartacus/storefront/esm5/cms-components/product/product-intro/product-intro.component.js
import { __decorate, __values } from "tslib";
import { ChangeDetectionStrategy, Component, } from '@angular/core';
import { Product, TranslationService, WindowRef } from '@spartacus/core';
import { BehaviorSubject } from 'rxjs';
import { CurrentProductService } from '../current-product.service';
var ProductIntroComponent = /** @class */ (function () {
    function ProductIntroComponent(currentProductService, translationService, winRef) {
        this.currentProductService = currentProductService;
        this.translationService = translationService;
        this.winRef = winRef;
        this.reviewsTabAvailable = new BehaviorSubject(false);
        this.product$ = this.currentProductService.getProduct();
    }
    ProductIntroComponent.prototype.ngAfterContentChecked = function () {
        this.reviewsTabAvailable.next(!!this.getReviewsComponent());
    };
    // Scroll to views component on page and click "Reviews" tab
    ProductIntroComponent.prototype.showReviews = function () {
        var _this = this;
        // Use translated label for Reviews tab reference
        this.translationService
            .translate('TabPanelContainer.tabs.ProductReviewsTabComponent')
            .subscribe(function (reviewsTabLabel) {
            var tabsComponent = _this.getTabsComponent();
            var reviewsTab = _this.getTabByLabel(reviewsTabLabel, tabsComponent);
            var reviewsComponent = _this.getReviewsComponent();
            if (reviewsTab && reviewsComponent) {
                _this.clickTabIfInactive(reviewsTab);
                setTimeout(function () { return reviewsComponent.scrollIntoView({ behavior: 'smooth' }); }, 0);
            }
        })
            .unsubscribe();
    };
    // NOTE: Does not currently exists as its own component
    // but part of tabs component. This is likely to change in refactor.
    ProductIntroComponent.prototype.getReviewsComponent = function () {
        return this.winRef.document.querySelector('cx-product-reviews');
    };
    // Get Tabs Component if exists on page
    ProductIntroComponent.prototype.getTabsComponent = function () {
        return this.winRef.document.querySelector('cx-tab-paragraph-container');
    };
    // Click to activate tab if not already active
    ProductIntroComponent.prototype.clickTabIfInactive = function (tab) {
        if (!tab.classList.contains('active') ||
            tab.classList.contains('toggled')) {
            tab.click();
        }
    };
    // Get Tab by label if exists on page
    ProductIntroComponent.prototype.getTabByLabel = function (label, tabsComponent) {
        var e_1, _a;
        if (tabsComponent) {
            // NOTE: Reads through button tags to click on correct tab
            // There may be a better way of doing this now/after refactor
            var tabElements = tabsComponent.getElementsByTagName('button');
            try {
                // Look through button tab elements until finding tab with label
                for (var _b = __values(Array.from(tabElements)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var buttonElement = _c.value;
                    if (buttonElement.innerHTML.includes(label)) {
                        return buttonElement;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    };
    ProductIntroComponent.ctorParameters = function () { return [
        { type: CurrentProductService },
        { type: TranslationService },
        { type: WindowRef }
    ]; };
    ProductIntroComponent = __decorate([
        Component({
            selector: 'cx-product-intro',
            template: "<ng-container *ngIf=\"product$ | async as product\">\r\n  <div class=\"rating\" *ngIf=\"product.averageRating\">\r\n    <cx-star-rating\r\n      [rating]=\"product?.averageRating\"\r\n      [disabled]=\"true\"\r\n    ></cx-star-rating>\r\n\r\n    <div class=\"count\">({{ product?.numberOfReviews }})</div>\r\n\r\n    <button\r\n      *ngIf=\"reviewsTabAvailable | async\"\r\n      class=\"btn btn-link cx-action-link\"\r\n      (click)=\"showReviews()\"\r\n    >\r\n      {{ 'productSummary.showReviews' | cxTranslate }}\r\n    </button>\r\n  </div>\r\n  <div class=\"rating\" *ngIf=\"!product.averageRating\">\r\n    {{ 'productDetails.noReviews' | cxTranslate }}\r\n  </div>\r\n  <div class=\"code\">\r\n    {{ 'productSummary.id' | cxTranslate }} {{ product?.code }}\r\n  </div>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ProductIntroComponent);
    return ProductIntroComponent;
}());
export { ProductIntroComponent };
