// @spartacus/storefront/esm5/cms-components/order-confirmation/components/guest-register-form/guest-register-form.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { AuthService, RoutingService, UserService } from '@spartacus/core';
import { CustomFormValidators } from '../../../../shared/utils/validators/custom-form-validators';
var GuestRegisterFormComponent = /** @class */ (function () {
    function GuestRegisterFormComponent(userService, routingService, authService, fb) {
        this.userService = userService;
        this.routingService = routingService;
        this.authService = authService;
        this.fb = fb;
        this.guestRegisterForm = this.fb.group({
            password: [
                '',
                [Validators.required, CustomFormValidators.passwordValidator],
            ],
            passwordconf: ['', Validators.required],
        }, {
            validators: CustomFormValidators.passwordsMustMatch('password', 'passwordconf'),
        });
    }
    GuestRegisterFormComponent.prototype.submit = function () {
        var _this = this;
        if (this.guestRegisterForm.valid) {
            this.userService.registerGuest(this.guid, this.guestRegisterForm.value.password);
            if (!this.subscription) {
                this.subscription = this.authService
                    .getUserToken()
                    .subscribe(function (token) {
                    if (token.access_token) {
                        _this.routingService.go({ cxRoute: 'home' });
                    }
                });
            }
        }
        else {
            this.guestRegisterForm.markAllAsTouched();
        }
    };
    GuestRegisterFormComponent.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    GuestRegisterFormComponent.ctorParameters = function () { return [
        { type: UserService },
        { type: RoutingService },
        { type: AuthService },
        { type: FormBuilder }
    ]; };
    __decorate([
        Input()
    ], GuestRegisterFormComponent.prototype, "guid", void 0);
    __decorate([
        Input()
    ], GuestRegisterFormComponent.prototype, "email", void 0);
    GuestRegisterFormComponent = __decorate([
        Component({
            selector: 'cx-guest-register-form',
            template: "<div class=\"register-guest\">\r\n  <div class=\"col-md-6 col-lg-4\">\r\n    <h3>{{ 'checkoutOrderConfirmation.createAccount' | cxTranslate }}</h3>\r\n    <p>\r\n      {{\r\n        'checkoutOrderConfirmation.createAccountForNext'\r\n          | cxTranslate: { email: email }\r\n      }}\r\n    </p>\r\n\r\n    <form (ngSubmit)=\"submit()\" [formGroup]=\"guestRegisterForm\">\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content\">{{\r\n            'register.password.label' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            class=\"form-control\"\r\n            type=\"password\"\r\n            name=\"password\"\r\n            placeholder=\"{{ 'register.password.placeholder' | cxTranslate }}\"\r\n            formControlName=\"password\"\r\n          />\r\n          <cx-form-errors\r\n            [control]=\"guestRegisterForm.get('password')\"\r\n          ></cx-form-errors>\r\n        </label>\r\n      </div>\r\n\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content\">{{\r\n            'register.confirmPassword.label' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            class=\"form-control\"\r\n            type=\"password\"\r\n            name=\"passwordconf\"\r\n            placeholder=\"{{\r\n              'register.confirmPassword.placeholder' | cxTranslate\r\n            }}\"\r\n            formControlName=\"passwordconf\"\r\n          />\r\n          <cx-form-errors\r\n            [control]=\"guestRegisterForm.get('passwordconf')\"\r\n          ></cx-form-errors>\r\n        </label>\r\n      </div>\r\n\r\n      <button type=\"submit\" class=\"btn btn-block btn-primary\">\r\n        {{ 'common.submit' | cxTranslate }}\r\n      </button>\r\n    </form>\r\n  </div>\r\n</div>\r\n"
        })
    ], GuestRegisterFormComponent);
    return GuestRegisterFormComponent;
}());
export { GuestRegisterFormComponent };
