// @spartacus/storefront/esm5/cms-components/order-confirmation/order-confirmation.module.js
import { __decorate, __read, __spread } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FeaturesConfigModule, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { CardModule } from '../../shared/components/card/card.module';
import { CartSharedModule } from '../cart/cart-shared/cart-shared.module';
import { PwaModule } from './../../cms-structure/pwa/pwa.module';
import { OrderConfirmationItemsComponent } from './components/order-confirmation-items/order-confirmation-items.component';
import { OrderConfirmationOverviewComponent } from './components/order-confirmation-overview/order-confirmation-overview.component';
// tslint:disable-next-line
import { OrderConfirmationThankYouMessageComponent } from './components/order-confirmation-thank-you-message/order-confirmation-thank-you-message.component';
import { OrderConfirmationTotalsComponent } from './components/order-confirmation-totals/order-confirmation-totals.component';
import { GuestRegisterFormComponent } from './components/guest-register-form/guest-register-form.component';
import { OrderConfirmationGuard } from './guards/order-confirmation.guard';
import { PromotionsModule } from '../checkout/components/promotions/promotions.module';
import { FormErrorsModule } from '../../shared/index';
var orderConfirmationComponents = [
    OrderConfirmationItemsComponent,
    OrderConfirmationOverviewComponent,
    OrderConfirmationThankYouMessageComponent,
    OrderConfirmationTotalsComponent,
    GuestRegisterFormComponent,
];
var OrderConfirmationModule = /** @class */ (function () {
    function OrderConfirmationModule() {
    }
    OrderConfirmationModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                CartSharedModule,
                CardModule,
                PwaModule,
                PromotionsModule,
                I18nModule,
                ReactiveFormsModule,
                FeaturesConfigModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        OrderConfirmationThankMessageComponent: {
                            component: OrderConfirmationThankYouMessageComponent,
                            guards: [OrderConfirmationGuard],
                        },
                        OrderConfirmationItemsComponent: {
                            component: OrderConfirmationItemsComponent,
                            guards: [OrderConfirmationGuard],
                        },
                        OrderConfirmationTotalsComponent: {
                            component: OrderConfirmationTotalsComponent,
                            guards: [OrderConfirmationGuard],
                        },
                        OrderConfirmationOverviewComponent: {
                            component: OrderConfirmationOverviewComponent,
                            guards: [OrderConfirmationGuard],
                        },
                    },
                }),
            ],
            declarations: __spread(orderConfirmationComponents),
            exports: __spread(orderConfirmationComponents),
            entryComponents: __spread(orderConfirmationComponents),
        })
    ], OrderConfirmationModule);
    return OrderConfirmationModule;
}());
export { OrderConfirmationModule };
