// @spartacus/storefront/esm5/cms-components/misc/qualtrics/qualtrics.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { defaultQualtricsConfig } from './config/default-qualtrics-config';
import { QualtricsComponent } from './qualtrics.component';
var QualtricsModule = /** @class */ (function () {
    function QualtricsModule() {
    }
    QualtricsModule = __decorate([
        NgModule({
            imports: [CommonModule, HttpClientModule],
            declarations: [QualtricsComponent],
            entryComponents: [QualtricsComponent],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        QualtricsComponent: {
                            component: QualtricsComponent,
                        },
                    },
                }),
                provideDefaultConfig(defaultQualtricsConfig),
            ],
        })
    ], QualtricsModule);
    return QualtricsModule;
}());
export { QualtricsModule };
