// @spartacus/core/esm5/src/personalization/services/personalization-context.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { filter, map } from 'rxjs/operators';
import { CmsService } from '../../cms/facade/cms.service';
import { PersonalizationConfig } from '../config/personalization-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/personalization-config";
import * as i2 from "../../cms/facade/cms.service";
var PersonalizationContextService = /** @class */ (function () {
    function PersonalizationContextService(config, cmsService) {
        this.config = config;
        this.cmsService = cmsService;
    }
    PersonalizationContextService.prototype.getPersonalizationContext = function () {
        var _this = this;
        return this.cmsService.getCurrentPage().pipe(filter(Boolean), map(function (page) {
            return page.slots[_this.config.personalization.context.slotPosition];
        }), filter(Boolean), map(function (slot) {
            return slot.components.find(function (i) { return i.uid === _this.config.personalization.context.componentId; });
        }), filter(Boolean), map(function (component) {
            return _this.buildPersonalizationContext(component.properties.script.data);
        }));
    };
    PersonalizationContextService.prototype.buildPersonalizationContext = function (data) {
        var context = JSON.parse(atob(data));
        context.actions.forEach(function (action) {
            Object.keys(action).forEach(function (key) {
                action[key] = atob(action[key]);
            });
        });
        for (var i = 0; i < context.segments.length; i++) {
            context.segments[i] = atob(context.segments[i]);
        }
        return context;
    };
    PersonalizationContextService.ctorParameters = function () { return [
        { type: PersonalizationConfig },
        { type: CmsService }
    ]; };
    PersonalizationContextService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PersonalizationContextService_Factory() { return new PersonalizationContextService(i0.ɵɵinject(i1.PersonalizationConfig), i0.ɵɵinject(i2.CmsService)); }, token: PersonalizationContextService, providedIn: "root" });
    PersonalizationContextService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], PersonalizationContextService);
    return PersonalizationContextService;
}());
export { PersonalizationContextService };
