// @spartacus/core/esm2015/src/user/store/actions/regions.action.js
import { StateUtils } from '../../../state/utils/index';
import { REGIONS } from '../user-state';
export const LOAD_REGIONS = '[User] Load Regions';
export const LOAD_REGIONS_SUCCESS = '[User] Load Regions Success';
export const LOAD_REGIONS_FAIL = '[User] Load Regions Fail';
export const CLEAR_REGIONS = '[User] Clear Regions';
export class LoadRegions extends StateUtils.LoaderLoadAction {
    constructor(payload) {
        super(REGIONS);
        this.payload = payload;
        this.type = LOAD_REGIONS;
    }
}
export class LoadRegionsFail extends StateUtils.LoaderFailAction {
    constructor(payload) {
        super(REGIONS, payload);
        this.payload = payload;
        this.type = LOAD_REGIONS_FAIL;
    }
}
export class LoadRegionsSuccess extends StateUtils.LoaderSuccessAction {
    constructor(payload) {
        super(REGIONS);
        this.payload = payload;
        this.type = LOAD_REGIONS_SUCCESS;
    }
}
export class ClearRegions {
    constructor() {
        this.type = CLEAR_REGIONS;
    }
}
