// @spartacus/core/esm2015/src/user/store/actions/update-email.action.js
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { UPDATE_EMAIL_PROCESS_ID } from '../user-state';
export const UPDATE_EMAIL = '[User] Update Email';
export const UPDATE_EMAIL_ERROR = '[User] Update Email Error';
export const UPDATE_EMAIL_SUCCESS = '[User] Update Email Success';
export const RESET_EMAIL = '[User] Reset Email';
export class UpdateEmailAction extends StateUtils.EntityLoadAction {
    constructor(payload) {
        super(PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID);
        this.payload = payload;
        this.type = UPDATE_EMAIL;
    }
}
export class UpdateEmailSuccessAction extends StateUtils.EntitySuccessAction {
    constructor(newUid) {
        super(PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID);
        this.newUid = newUid;
        this.type = UPDATE_EMAIL_SUCCESS;
    }
}
export class UpdateEmailErrorAction extends StateUtils.EntityFailAction {
    constructor(payload) {
        super(PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID, payload);
        this.payload = payload;
        this.type = UPDATE_EMAIL_ERROR;
    }
}
export class ResetUpdateEmailAction extends StateUtils.EntityLoaderResetAction {
    constructor() {
        super(PROCESS_FEATURE, UPDATE_EMAIL_PROCESS_ID);
        this.type = RESET_EMAIL;
    }
}
