// @spartacus/storefront/esm5/cms-structure/outlet/outlet-renderer.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
var OutletRendererService = /** @class */ (function () {
    function OutletRendererService() {
        this.outletRefs = new BehaviorSubject(new Map());
    }
    /**
     * Dynamically render the templates in the specified array
     *
     * @param outlet
     */
    OutletRendererService.prototype.render = function (outlet) {
        if (this.outletRefs.value.size !== 0) {
            this.outletRefs.value.get(outlet).render();
        }
    };
    /**
     * Register outlet to be available to render dynamically
     *
     * @param cxOutlet
     * @param context
     */
    OutletRendererService.prototype.register = function (cxOutlet, context) {
        this.outletRefs.next(this.outletRefs.value.set(cxOutlet, context));
    };
    /**
     * Returns map of outlets
     *
     */
    OutletRendererService.prototype.getOutletRef = function (outlet) {
        return this.outletRefs.asObservable().pipe(map(function (val) { return val.get(outlet); }), filter(function (val) { return Boolean(val); }));
    };
    OutletRendererService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OutletRendererService_Factory() { return new OutletRendererService(); }, token: OutletRendererService, providedIn: "root" });
    OutletRendererService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], OutletRendererService);
    return OutletRendererService;
}());
export { OutletRendererService };
