// @spartacus/core/esm5/src/occ/services/occ-requests-optimizer.service.js
import { __assign, __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { map, shareReplay } from 'rxjs/operators';
import { extractFields } from '../utils/occ-fields';
import { OccFieldsModel, OccFieldsService, ScopedDataWithUrl, } from './occ-fields.service';
import { HttpClient } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./occ-fields.service";
var OccRequestsOptimizerService = /** @class */ (function () {
    function OccRequestsOptimizerService(http, occFields) {
        this.http = http;
        this.occFields = occFields;
    }
    /**
     * Optimize occ endpoint calls merging requests to the same url by merging field parameters
     *
     * @param scopedDataWithUrls
     * @param dataFactory
     */
    OccRequestsOptimizerService.prototype.scopedDataLoad = function (scopedDataWithUrls, dataFactory) {
        var _this = this;
        var result = [];
        if (!dataFactory) {
            dataFactory = function (url) { return _this.http.get(url); };
        }
        var mergedUrls = this.occFields.getOptimalUrlGroups(scopedDataWithUrls);
        Object.entries(mergedUrls).forEach(function (_a) {
            var _b = __read(_a, 2), url = _b[0], groupedModelsSet = _b[1];
            var groupedModels = Object.values(groupedModelsSet);
            if (groupedModels.length === 1) {
                // only one scope for url, we can pass the data straightaway
                result.push(__assign(__assign({}, groupedModels[0].scopedData), { data$: dataFactory(url) }));
            }
            else {
                // multiple scopes per url
                // we have to split the model per each scope
                var data$_1 = dataFactory(url).pipe(shareReplay(1));
                groupedModels.forEach(function (modelData) {
                    result.push(__assign(__assign({}, modelData.scopedData), { data$: data$_1.pipe(map(function (data) { return extractFields(data, modelData.fields); })) }));
                });
            }
        });
        return result;
    };
    OccRequestsOptimizerService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccFieldsService }
    ]; };
    OccRequestsOptimizerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccRequestsOptimizerService_Factory() { return new OccRequestsOptimizerService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.OccFieldsService)); }, token: OccRequestsOptimizerService, providedIn: "root" });
    OccRequestsOptimizerService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], OccRequestsOptimizerService);
    return OccRequestsOptimizerService;
}());
export { OccRequestsOptimizerService };
