// @spartacus/core/esm5/src/cart/services/multi-cart-state-persistence.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { distinctUntilKeyChanged, filter, map } from 'rxjs/operators';
import { BASE_SITE_CONTEXT_ID } from '../../site-context';
import { SiteContextParamsService } from '../../site-context/services/site-context-params.service';
import { StatePersistenceService } from '../../state/services/state-persistence.service';
import { CartActions, MultiCartSelectors } from '../store';
import * as i0 from "@angular/core";
import * as i1 from "../../state/services/state-persistence.service";
import * as i2 from "@ngrx/store";
import * as i3 from "../../site-context/services/site-context-params.service";
var MultiCartStatePersistenceService = /** @class */ (function () {
    function MultiCartStatePersistenceService(statePersistenceService, store, siteContextParamsService) {
        this.statePersistenceService = statePersistenceService;
        this.store = store;
        this.siteContextParamsService = siteContextParamsService;
    }
    MultiCartStatePersistenceService.prototype.sync = function () {
        var _this = this;
        this.statePersistenceService.syncWithStorage({
            key: 'cart',
            state$: this.getCartState(),
            context$: this.siteContextParamsService.getValues([BASE_SITE_CONTEXT_ID]),
            onRead: function (state) { return _this.onRead(state); },
        });
    };
    MultiCartStatePersistenceService.prototype.getCartState = function () {
        return this.store.pipe(select(MultiCartSelectors.getMultiCartState), filter(function (state) { return !!state; }), distinctUntilKeyChanged('active'), map(function (state) {
            return {
                active: state.active,
            };
        }));
    };
    MultiCartStatePersistenceService.prototype.onRead = function (state) {
        this.store.dispatch(new CartActions.ClearCartState());
        if (state) {
            this.store.dispatch(new CartActions.SetActiveCartId(state.active));
        }
    };
    MultiCartStatePersistenceService.ctorParameters = function () { return [
        { type: StatePersistenceService },
        { type: Store },
        { type: SiteContextParamsService }
    ]; };
    MultiCartStatePersistenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MultiCartStatePersistenceService_Factory() { return new MultiCartStatePersistenceService(i0.ɵɵinject(i1.StatePersistenceService), i0.ɵɵinject(i2.Store), i0.ɵɵinject(i3.SiteContextParamsService)); }, token: MultiCartStatePersistenceService, providedIn: "root" });
    MultiCartStatePersistenceService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], MultiCartStatePersistenceService);
    return MultiCartStatePersistenceService;
}());
export { MultiCartStatePersistenceService };
