// @spartacus/core/esm5/src/occ/adapters/cart/occ-cart-voucher.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { CART_VOUCHER_NORMALIZER } from '../../../cart/connectors/voucher/converters';
import { OCC_USER_ID_ANONYMOUS } from '../../../occ/utils/occ-constants';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
import { InterceptorUtil, USE_CLIENT_TOKEN, } from '../../utils/interceptor-util';
var OccCartVoucherAdapter = /** @class */ (function () {
    function OccCartVoucherAdapter(http, occEndpoints, converter) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
    }
    OccCartVoucherAdapter.prototype.getCartVoucherEndpoint = function (userId, cartId) {
        return this.occEndpoints.getUrl('cartVoucher', { userId: userId, cartId: cartId });
    };
    OccCartVoucherAdapter.prototype.getHeaders = function (userId) {
        var headers = new HttpHeaders({
            'Content-Type': 'application/json',
        });
        if (userId === OCC_USER_ID_ANONYMOUS) {
            headers = InterceptorUtil.createHeader(USE_CLIENT_TOKEN, true, headers);
        }
        return headers;
    };
    OccCartVoucherAdapter.prototype.add = function (userId, cartId, voucherId) {
        var url = this.getCartVoucherEndpoint(userId, cartId);
        var toAdd = JSON.stringify({});
        var params = new HttpParams().set('voucherId', voucherId);
        var headers = this.getHeaders(userId);
        return this.http.post(url, toAdd, { headers: headers, params: params }).pipe(catchError(function (error) { return throwError(error); }), this.converter.pipeable(CART_VOUCHER_NORMALIZER));
    };
    OccCartVoucherAdapter.prototype.remove = function (userId, cartId, voucherId) {
        var url = this.getCartVoucherEndpoint(userId, cartId) +
            '/' +
            encodeURIComponent(voucherId);
        var headers = this.getHeaders(userId);
        return this.http
            .delete(url, { headers: headers })
            .pipe(catchError(function (error) { return throwError(error); }));
    };
    OccCartVoucherAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService }
    ]; };
    OccCartVoucherAdapter = __decorate([
        Injectable()
    ], OccCartVoucherAdapter);
    return OccCartVoucherAdapter;
}());
export { OccCartVoucherAdapter };
