// @spartacus/storefront/esm5/cms-components/user/login-form/login-form.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { LoginFormComponent } from './login-form.component';
import { FormErrorsModule } from '../../../shared/index';
var LoginFormModule = /** @class */ (function () {
    function LoginFormModule() {
    }
    LoginFormModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
                RouterModule,
                UrlModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ReturningCustomerLoginComponent: {
                            component: LoginFormComponent,
                            guards: [NotAuthGuard],
                        },
                    },
                }),
            ],
            declarations: [LoginFormComponent],
            exports: [LoginFormComponent],
            entryComponents: [LoginFormComponent],
        })
    ], LoginFormModule);
    return LoginFormModule;
}());
export { LoginFormModule };
