// @spartacus/storefront/esm5/cms-components/user/login-form/login-form.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ActivatedRoute } from '@angular/router';
import { AuthRedirectService, AuthService, GlobalMessageService, GlobalMessageType, WindowRef, } from '@spartacus/core';
import { CheckoutConfigService } from '../../checkout/services/checkout-config.service';
import { CustomFormValidators } from '../../../shared/index';
var LoginFormComponent = /** @class */ (function () {
    function LoginFormComponent(auth, globalMessageService, fb, authRedirectService, winRef, activatedRoute, checkoutConfigService) {
        this.auth = auth;
        this.globalMessageService = globalMessageService;
        this.fb = fb;
        this.authRedirectService = authRedirectService;
        this.winRef = winRef;
        this.activatedRoute = activatedRoute;
        this.checkoutConfigService = checkoutConfigService;
        this.loginAsGuest = false;
    }
    LoginFormComponent.prototype.ngOnInit = function () {
        var _a, _b, _c, _d, _e;
        var routeState = (_b = (_a = this.winRef.nativeWindow) === null || _a === void 0 ? void 0 : _a.history) === null || _b === void 0 ? void 0 : _b.state;
        var prefilledEmail = routeState === null || routeState === void 0 ? void 0 : routeState['newUid'];
        this.loginForm = this.fb.group({
            userId: [
                (prefilledEmail === null || prefilledEmail === void 0 ? void 0 : prefilledEmail.length) ? prefilledEmail : '',
                [Validators.required, CustomFormValidators.emailValidator],
            ],
            password: ['', Validators.required],
        });
        if (this.checkoutConfigService.isGuestCheckout()) {
            this.loginAsGuest = (_e = (_d = (_c = this.activatedRoute) === null || _c === void 0 ? void 0 : _c.snapshot) === null || _d === void 0 ? void 0 : _d.queryParams) === null || _e === void 0 ? void 0 : _e['forced'];
        }
    };
    LoginFormComponent.prototype.submitForm = function () {
        if (this.loginForm.valid) {
            this.loginUser();
        }
        else {
            this.loginForm.markAllAsTouched();
        }
    };
    LoginFormComponent.prototype.ngOnDestroy = function () {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    };
    LoginFormComponent.prototype.loginUser = function () {
        var _this = this;
        var _a = this.loginForm.controls, userId = _a.userId, password = _a.password;
        this.auth.authorize(userId.value.toLowerCase(), // backend accepts lowercase emails only
        password.value);
        if (!this.sub) {
            this.sub = this.auth.getUserToken().subscribe(function (data) {
                if (data && data.access_token) {
                    _this.globalMessageService.remove(GlobalMessageType.MSG_TYPE_ERROR);
                    _this.authRedirectService.redirect();
                }
            });
        }
    };
    LoginFormComponent.ctorParameters = function () { return [
        { type: AuthService },
        { type: GlobalMessageService },
        { type: FormBuilder },
        { type: AuthRedirectService },
        { type: WindowRef },
        { type: ActivatedRoute },
        { type: CheckoutConfigService }
    ]; };
    LoginFormComponent = __decorate([
        Component({
            selector: 'cx-login-form',
            template: "<form (ngSubmit)=\"submitForm()\" [formGroup]=\"loginForm\">\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'loginForm.emailAddress.label' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        type=\"email\"\r\n        class=\"form-control\"\r\n        formControlName=\"userId\"\r\n        placeholder=\"{{ 'loginForm.emailAddress.placeholder' | cxTranslate }}\"\r\n      />\r\n      <cx-form-errors [control]=\"loginForm.get('userId')\"></cx-form-errors>\r\n    </label>\r\n  </div>\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'loginForm.password.label' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        type=\"password\"\r\n        class=\"form-control\"\r\n        placeholder=\"{{ 'loginForm.password.placeholder' | cxTranslate }}\"\r\n        formControlName=\"password\"\r\n      />\r\n      <cx-form-errors [control]=\"loginForm.get('password')\"></cx-form-errors>\r\n    </label>\r\n  </div>\r\n  <p>\r\n    <a\r\n      [routerLink]=\"{ cxRoute: 'forgotPassword' } | cxUrl\"\r\n      aria-controls=\"reset-password\"\r\n      class=\"btn-link\"\r\n      >{{ 'loginForm.forgotPassword' | cxTranslate }}</a\r\n    >\r\n  </p>\r\n\r\n  <button type=\"submit\" class=\"btn btn-block btn-primary\">\r\n    {{ 'loginForm.signIn' | cxTranslate }}\r\n  </button>\r\n</form>\r\n\r\n<div class=\"register\">\r\n  <p class=\"cx-section-title\">\r\n    {{ 'loginForm.dontHaveAccount' | cxTranslate }}\r\n  </p>\r\n\r\n  <ng-container *ngIf=\"!loginAsGuest\">\r\n    <a\r\n      [routerLink]=\"{ cxRoute: 'register' } | cxUrl\"\r\n      class=\"btn btn-block btn-secondary btn-register\"\r\n      >{{ 'loginForm.register' | cxTranslate }}</a\r\n    >\r\n  </ng-container>\r\n\r\n  <ng-container *ngIf=\"loginAsGuest\">\r\n    <a\r\n      [routerLink]=\"{ cxRoute: 'checkoutLogin' } | cxUrl\"\r\n      class=\"btn btn-block btn-secondary btn-guest\"\r\n      >{{ 'loginForm.guestCheckout' | cxTranslate }}</a\r\n    >\r\n  </ng-container>\r\n</div>\r\n"
        })
    ], LoginFormComponent);
    return LoginFormComponent;
}());
export { LoginFormComponent };
