// @spartacus/storefront/esm2015/cms-components/checkout/components/delivery-mode/delivery-mode.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { SpinnerModule } from '../../../../shared/components/spinner/spinner.module';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { ShippingAddressSetGuard } from '../../guards/shipping-address-set.guard';
import { CartNotEmptyGuard } from './../../../../cms-components/cart/cart-not-empty.guard';
import { DeliveryModeComponent } from './delivery-mode.component';
let DeliveryModeModule = class DeliveryModeModule {
};
DeliveryModeModule = __decorate([
    NgModule({
        imports: [CommonModule, ReactiveFormsModule, I18nModule, SpinnerModule],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    CheckoutDeliveryMode: {
                        component: DeliveryModeComponent,
                        guards: [
                            CheckoutAuthGuard,
                            CartNotEmptyGuard,
                            ShippingAddressSetGuard,
                        ],
                    },
                },
            }),
        ],
        declarations: [DeliveryModeComponent],
        entryComponents: [DeliveryModeComponent],
        exports: [DeliveryModeComponent],
    })
], DeliveryModeModule);
export { DeliveryModeModule };
