// @spartacus/storefront/esm5/shared/components/card/card.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ICON_TYPE } from '../../../cms-components/misc/icon/icon.model';
var CardComponent = /** @class */ (function () {
    function CardComponent() {
        this.iconTypes = ICON_TYPE;
        this.deleteCard = new EventEmitter();
        this.setDefaultCard = new EventEmitter();
        this.sendCard = new EventEmitter();
        this.editCard = new EventEmitter();
        this.cancelCard = new EventEmitter();
        this.border = false;
        this.editMode = false;
        this.isDefault = false;
        this.fitToContainer = false;
    }
    // ACTIONS
    CardComponent.prototype.setEditMode = function () {
        this.editMode = true;
    };
    CardComponent.prototype.cancelEdit = function () {
        this.editMode = false;
        this.cancelCard.emit(5);
    };
    CardComponent.prototype.delete = function () {
        this.deleteCard.emit(1);
    };
    CardComponent.prototype.setDefault = function () {
        this.isDefault = true;
        this.setDefaultCard.emit(2);
    };
    CardComponent.prototype.send = function () {
        this.sendCard.emit(3);
    };
    CardComponent.prototype.edit = function () {
        this.editCard.emit(4);
    };
    CardComponent.prototype.ngOnInit = function () { };
    __decorate([
        Output()
    ], CardComponent.prototype, "deleteCard", void 0);
    __decorate([
        Output()
    ], CardComponent.prototype, "setDefaultCard", void 0);
    __decorate([
        Output()
    ], CardComponent.prototype, "sendCard", void 0);
    __decorate([
        Output()
    ], CardComponent.prototype, "editCard", void 0);
    __decorate([
        Output()
    ], CardComponent.prototype, "cancelCard", void 0);
    __decorate([
        Input()
    ], CardComponent.prototype, "border", void 0);
    __decorate([
        Input()
    ], CardComponent.prototype, "editMode", void 0);
    __decorate([
        Input()
    ], CardComponent.prototype, "isDefault", void 0);
    __decorate([
        Input()
    ], CardComponent.prototype, "content", void 0);
    __decorate([
        Input()
    ], CardComponent.prototype, "fitToContainer", void 0);
    CardComponent = __decorate([
        Component({
            selector: 'cx-card',
            template: "<div\r\n  *ngIf=\"content\"\r\n  class=\"cx-card\"\r\n  [class.cx-card-border]=\"border\"\r\n  [class.cx-card-fit-to-container]=\"fitToContainer\"\r\n>\r\n  <!-- Card Header -->\r\n  <div *ngIf=\"content.header && !editMode\" class=\"card-header\">\r\n    {{ content.header }}\r\n  </div>\r\n  <!-- Card Body -->\r\n  <div class=\"card-body cx-card-body\" [class.cx-card-delete]=\"editMode\">\r\n    <!-- Edit message -->\r\n    <div *ngIf=\"editMode\" class=\"cx-card-delete-msg\">\r\n      {{ content.deleteMsg }}\r\n    </div>\r\n    <!-- Card title -->\r\n    <h4 *ngIf=\"content.title\" class=\"cx-card-title\">\r\n      {{ content.title }}\r\n    </h4>\r\n    <!-- Card Content -->\r\n    <div class=\"cx-card-container\">\r\n      <!-- Card Label -->\r\n      <div class=\"cx-card-label-container\">\r\n        <div *ngIf=\"content.textBold\" class=\"cx-card-label-bold\">\r\n          {{ content.textBold }}\r\n        </div>\r\n        <div *ngFor=\"let line of content.text\">\r\n          <div class=\"cx-card-label\">{{ line }}</div>\r\n        </div>\r\n      </div>\r\n      <!-- Image -->\r\n      <div *ngIf=\"content.img\" class=\"cx-card-img-container\">\r\n        <cx-icon [type]=\"content.img\"></cx-icon>\r\n      </div>\r\n    </div>\r\n    <!-- Edit Mode Actions -->\r\n    <div *ngIf=\"editMode\" class=\"row cx-card-body-delete\">\r\n      <div class=\"col-md-6\">\r\n        <button class=\"btn btn-block btn-secondary\" (click)=\"cancelEdit()\">\r\n          {{ 'common.cancel' | cxTranslate }}\r\n        </button>\r\n      </div>\r\n      <div class=\"col-md-6\">\r\n        <button class=\"btn btn-block btn-primary\" (click)=\"delete()\">\r\n          {{ 'common.delete' | cxTranslate }}\r\n        </button>\r\n      </div>\r\n    </div>\r\n    <!-- Actions -->\r\n    <div *ngIf=\"content.actions && !editMode\" class=\"cx-card-actions\">\r\n      <div *ngFor=\"let action of content.actions\">\r\n        <div [ngSwitch]=\"action.event\">\r\n          <a\r\n            *ngSwitchCase=\"'delete'\"\r\n            class=\"cx-card-link card-link btn-link cx-action-link\"\r\n            (click)=\"delete()\"\r\n            (keydown.enter)=\"delete()\"\r\n            tabindex=\"0\"\r\n            >{{ action.name }}</a\r\n          >\r\n          <a\r\n            *ngSwitchCase=\"'default'\"\r\n            class=\"cx-card-link card-link btn-link cx-action-link\"\r\n            (click)=\"setDefault()\"\r\n            (keydown.enter)=\"setDefault()\"\r\n            tabindex=\"0\"\r\n            >{{ action.name }}</a\r\n          >\r\n          <a\r\n            *ngSwitchCase=\"'send'\"\r\n            class=\"cx-card-link card-link btn-link cx-action-link\"\r\n            (click)=\"send()\"\r\n            (keydown.enter)=\"send()\"\r\n            tabindex=\"0\"\r\n            >{{ action.name }}</a\r\n          >\r\n          <a\r\n            *ngSwitchCase=\"'edit'\"\r\n            class=\"cx-card-link card-link btn-link cx-action-link\"\r\n            (click)=\"edit()\"\r\n            (keydown.enter)=\"edit()\"\r\n            tabindex=\"0\"\r\n            >{{ action.name }}</a\r\n          >\r\n          <a\r\n            *ngSwitchDefault\r\n            href=\"{{ action.link }}\"\r\n            class=\"card-link btn-link cx-action-link\"\r\n            tabindex=\"0\"\r\n            >{{ action.name }}</a\r\n          >\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n"
        })
    ], CardComponent);
    return CardComponent;
}());
export { CardComponent };
