// @spartacus/storefront/esm5/cms-components/myaccount/update-password/components/update-password-form/update-password-form.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { CustomFormValidators } from '../../../../../shared/utils/validators/custom-form-validators';
var UpdatePasswordFormComponent = /** @class */ (function () {
    function UpdatePasswordFormComponent(fb) {
        this.fb = fb;
        this.submitted = new EventEmitter();
        this.cancelled = new EventEmitter();
    }
    UpdatePasswordFormComponent.prototype.ngOnInit = function () {
        this.updatePasswordForm = this.fb.group({
            oldPassword: ['', [Validators.required]],
            newPassword: [
                '',
                [Validators.required, CustomFormValidators.passwordValidator],
            ],
            newPasswordConfirm: ['', [Validators.required]],
        }, {
            validators: CustomFormValidators.passwordsMustMatch('newPassword', 'newPasswordConfirm'),
        });
    };
    UpdatePasswordFormComponent.prototype.onSubmit = function () {
        if (this.updatePasswordForm.valid) {
            this.submitted.emit({
                oldPassword: this.updatePasswordForm.value.oldPassword,
                newPassword: this.updatePasswordForm.value.newPassword,
            });
        }
        else {
            this.updatePasswordForm.markAllAsTouched();
        }
    };
    UpdatePasswordFormComponent.prototype.onCancel = function () {
        this.cancelled.emit();
    };
    UpdatePasswordFormComponent.ctorParameters = function () { return [
        { type: FormBuilder }
    ]; };
    __decorate([
        Output()
    ], UpdatePasswordFormComponent.prototype, "submitted", void 0);
    __decorate([
        Output()
    ], UpdatePasswordFormComponent.prototype, "cancelled", void 0);
    UpdatePasswordFormComponent = __decorate([
        Component({
            selector: 'cx-update-password-form',
            template: "<form\r\n  (ngSubmit)=\"onSubmit()\"\r\n  [formGroup]=\"updatePasswordForm\"\r\n  class=\"cx-update-password-component\"\r\n>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updatePasswordForm.oldPassword.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          class=\"form-control\"\r\n          type=\"password\"\r\n          name=\"oldPassword\"\r\n          placeholder=\"{{\r\n            'updatePasswordForm.oldPassword.placeholder' | cxTranslate\r\n          }}\"\r\n          formControlName=\"oldPassword\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updatePasswordForm.get('oldPassword')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updatePasswordForm.newPassword.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          class=\"form-control\"\r\n          type=\"password\"\r\n          name=\"newPassword\"\r\n          placeholder=\"{{\r\n            'updatePasswordForm.newPassword.placeholder' | cxTranslate\r\n          }}\"\r\n          formControlName=\"newPassword\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updatePasswordForm.get('newPassword')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updatePasswordForm.confirmPassword.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          class=\"form-control\"\r\n          type=\"password\"\r\n          name=\"newPasswordConfirm\"\r\n          placeholder=\"{{\r\n            'updatePasswordForm.confirmPassword.placeholder' | cxTranslate\r\n          }}\"\r\n          formControlName=\"newPasswordConfirm\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updatePasswordForm.get('newPasswordConfirm')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-lg-6 col-md-12\">\r\n      <button\r\n        class=\"btn btn-block btn-secondary\"\r\n        type=\"button\"\r\n        (click)=\"onCancel()\"\r\n      >\r\n        {{ 'common.cancel' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n    <div class=\"col-lg-6 col-md-12\">\r\n      <button class=\"btn btn-block btn-primary\" type=\"submit\">\r\n        {{ 'common.save' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n  </div>\r\n</form>\r\n"
        })
    ], UpdatePasswordFormComponent);
    return UpdatePasswordFormComponent;
}());
export { UpdatePasswordFormComponent };
