// @spartacus/core/esm5/src/global-message/store/effects/global-message.effect.js
import { __decorate, __param, __read } from "tslib";
import { isPlatformBrowser } from '@angular/common';
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { select, Store } from '@ngrx/store';
import { EMPTY, of } from 'rxjs';
import { concatMap, delay, filter, map, pluck, switchMap, take, withLatestFrom, } from 'rxjs/operators';
import { countOfDeepEqualObjects, indexOfFirstOccurrence, } from '../../../util/compare-equal-objects';
import { GlobalMessageConfig } from '../../config/global-message-config';
import { GlobalMessageActions } from '../actions/index';
import { GlobalMessageSelectors } from '../selectors/index';
var GlobalMessageEffect = /** @class */ (function () {
    function GlobalMessageEffect(actions$, store, config, platformId) {
        var _this = this;
        this.actions$ = actions$;
        this.store = store;
        this.config = config;
        this.platformId = platformId;
        this.removeDuplicated$ = this.actions$.pipe(ofType(GlobalMessageActions.ADD_MESSAGE), pluck('payload'), switchMap(function (message) {
            return of(message.text).pipe(withLatestFrom(_this.store.pipe(select(GlobalMessageSelectors.getGlobalMessageEntitiesByType(message.type)))), filter(function (_a) {
                var _b = __read(_a, 2), text = _b[0], messages = _b[1];
                return countOfDeepEqualObjects(text, messages) > 1;
            }), map(function (_a) {
                var _b = __read(_a, 2), text = _b[0], messages = _b[1];
                return new GlobalMessageActions.RemoveMessage({
                    type: message.type,
                    index: indexOfFirstOccurrence(text, messages),
                });
            }));
        }));
        this.hideAfterDelay$ = isPlatformBrowser(this.platformId) // we don't want to run this logic when doing SSR
            ? this.actions$.pipe(ofType(GlobalMessageActions.ADD_MESSAGE), pluck('payload'), concatMap(function (message) {
                var config = _this.config.globalMessages[message.type];
                return _this.store.pipe(select(GlobalMessageSelectors.getGlobalMessageCountByType(message.type)), take(1), filter(function (count) {
                    return ((config && config.timeout !== undefined) || message.timeout) &&
                        count &&
                        count > 0;
                }), delay(message.timeout || config.timeout), switchMap(function () {
                    return of(new GlobalMessageActions.RemoveMessage({
                        type: message.type,
                        index: 0,
                    }));
                }));
            }))
            : EMPTY;
    }
    GlobalMessageEffect.ctorParameters = function () { return [
        { type: Actions },
        { type: Store },
        { type: GlobalMessageConfig },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    __decorate([
        Effect()
    ], GlobalMessageEffect.prototype, "removeDuplicated$", void 0);
    __decorate([
        Effect()
    ], GlobalMessageEffect.prototype, "hideAfterDelay$", void 0);
    GlobalMessageEffect = __decorate([
        Injectable(),
        __param(3, Inject(PLATFORM_ID))
    ], GlobalMessageEffect);
    return GlobalMessageEffect;
}());
export { GlobalMessageEffect };
