// @spartacus/storefront/esm5/cms-components/asm/asm-loader.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { PageComponentModule } from '../../cms-structure/page/component/page-component.module';
import { AsmEnablerService } from './services/asm-enabler.service';
/**
 * The ASM loader module takes care of loading the ASM UI
 * only in case there's a reason to do so.
 */
var AsmLoaderModule = /** @class */ (function () {
    function AsmLoaderModule() {
    }
    AsmLoaderModule = __decorate([
        NgModule({
            imports: [CommonModule, PageComponentModule],
            providers: [
                {
                    provide: APP_INITIALIZER,
                    useFactory: asmFactory,
                    deps: [AsmEnablerService],
                    multi: true,
                },
            ],
        })
    ], AsmLoaderModule);
    return AsmLoaderModule;
}());
export { AsmLoaderModule };
/**
 *
 * We do not like to block the UI, which is why we delgate loading of ASM
 * to a real component; the router and state aren't available in an optimized
 * way during the APP_INITIALIZER.
 */
export function asmFactory(asmEnablerService) {
    var isReady = function () {
        asmEnablerService.load();
    };
    return isReady;
}
