// @spartacus/core/esm5/src/routing/protected-routes/protected-routes.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { RoutingConfig } from '../configurable-routes/config/routing-config';
import * as i0 from "@angular/core";
import * as i1 from "../configurable-routes/config/routing-config";
var ProtectedRoutesService = /** @class */ (function () {
    function ProtectedRoutesService(config) {
        var _this = this;
        this.config = config;
        this.nonProtectedPaths = []; // arrays of paths' segments list
        if (this.shouldProtect) {
            // pre-process config for performance:
            this.nonProtectedPaths = this.getNonProtectedPaths().map(function (path) {
                return _this.getSegments(path);
            });
        }
    }
    Object.defineProperty(ProtectedRoutesService.prototype, "routingConfig", {
        get: function () {
            return this.config && this.config.routing;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProtectedRoutesService.prototype, "shouldProtect", {
        /**
         * Returns 'protected' property (boolean) from routing config
         *
         * @returns boolean
         */
        get: function () {
            return this.routingConfig.protected;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Tells if the url is protected
     */
    ProtectedRoutesService.prototype.isUrlProtected = function (urlSegments) {
        return (this.shouldProtect &&
            !this.matchAnyPath(urlSegments, this.nonProtectedPaths));
    };
    /**
     * Tells whether the url matches at least one of the paths
     */
    ProtectedRoutesService.prototype.matchAnyPath = function (urlSegments, pathsSegments) {
        var _this = this;
        return pathsSegments.some(function (pathSegments) {
            return _this.matchPath(urlSegments, pathSegments);
        });
    };
    /**
     * Tells whether the url matches the path
     */
    ProtectedRoutesService.prototype.matchPath = function (urlSegments, pathSegments) {
        if (urlSegments.length !== pathSegments.length) {
            return false;
        }
        for (var i = 0; i < pathSegments.length; i++) {
            var pathSeg = pathSegments[i];
            var urlSeg = urlSegments[i];
            // compare only static segments:
            if (!pathSeg.startsWith(':') && pathSeg !== urlSeg) {
                return false;
            }
        }
        return true;
    };
    /**
     * Returns a list of paths that are not protected
     */
    ProtectedRoutesService.prototype.getNonProtectedPaths = function () {
        return Object.values(this.routingConfig.routes).reduce(function (acc, routeConfig) {
            return routeConfig.protected === false && // must be explicitly false, ignore undefined
                routeConfig.paths &&
                routeConfig.paths.length
                ? acc.concat(routeConfig.paths)
                : acc;
        }, []);
    };
    /**
     * Splits the url by slashes
     */
    ProtectedRoutesService.prototype.getSegments = function (url) {
        return (url || '').split('/');
    };
    ProtectedRoutesService.ctorParameters = function () { return [
        { type: RoutingConfig }
    ]; };
    ProtectedRoutesService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProtectedRoutesService_Factory() { return new ProtectedRoutesService(i0.ɵɵinject(i1.RoutingConfig)); }, token: ProtectedRoutesService, providedIn: "root" });
    ProtectedRoutesService = __decorate([
        Injectable({ providedIn: 'root' })
    ], ProtectedRoutesService);
    return ProtectedRoutesService;
}());
export { ProtectedRoutesService };
