// @spartacus/storefront/esm5/cms-structure/page/component/handlers/default-component.handler.js
import { __decorate } from "tslib";
import { ComponentFactoryResolver, Injectable, } from '@angular/core';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Default component handler used for dynamically launching cms components implemented
 * as native Angular components.
 */
var DefaultComponentHandler = /** @class */ (function () {
    function DefaultComponentHandler() {
    }
    DefaultComponentHandler.prototype.hasMatch = function (componentMapping) {
        return typeof componentMapping.component === 'function';
    };
    DefaultComponentHandler.prototype.getPriority = function () {
        return -50 /* FALLBACK */;
    };
    DefaultComponentHandler.prototype.launcher = function (componentMapping, viewContainerRef, elementInjector) {
        var _this = this;
        return new Observable(function (subscriber) {
            var componentRef;
            var injector = elementInjector !== null && elementInjector !== void 0 ? elementInjector : viewContainerRef.injector;
            var dispose = function () {
                if (componentRef) {
                    componentRef.destroy();
                }
            };
            var factory = _this.getComponentFactory(injector, componentMapping.component);
            if (factory) {
                componentRef = viewContainerRef.createComponent(factory, undefined, injector);
                subscriber.next({ elementRef: componentRef.location, componentRef: componentRef });
            }
            return dispose;
        });
    };
    DefaultComponentHandler.prototype.getComponentFactory = function (injector, component) {
        if (!component) {
            return null;
        }
        var factory = injector
            .get(ComponentFactoryResolver)
            .resolveComponentFactory(component);
        return factory;
    };
    DefaultComponentHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function DefaultComponentHandler_Factory() { return new DefaultComponentHandler(); }, token: DefaultComponentHandler, providedIn: "root" });
    DefaultComponentHandler = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], DefaultComponentHandler);
    return DefaultComponentHandler;
}());
export { DefaultComponentHandler };
