// @spartacus/storefront/esm5/cms-structure/page/component/handlers/lazy-component.handler.js
import { __assign, __decorate } from "tslib";
import { Injectable, } from '@angular/core';
import { from } from 'rxjs';
import { DefaultComponentHandler } from './default-component.handler';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./default-component.handler";
/**
 * Lazy component handler used for launching lazy loaded cms components implemented
 * as native Angular components.
 */
var LazyComponentHandler = /** @class */ (function () {
    function LazyComponentHandler(defaultHandler) {
        this.defaultHandler = defaultHandler;
    }
    /**
     * We want to mach dynamic import signature () => import('')
     */
    LazyComponentHandler.prototype.hasMatch = function (componentMapping) {
        return (typeof componentMapping.component === 'function' &&
            this.isNotClass(componentMapping.component));
    };
    LazyComponentHandler.prototype.isNotClass = function (symbol) {
        var signature = symbol.toString().substr(0, 20).replace(' ', '');
        return signature.startsWith('function()') || signature.startsWith('()=>');
    };
    LazyComponentHandler.prototype.getPriority = function () {
        return -10 /* LOW */;
    };
    LazyComponentHandler.prototype.launcher = function (componentMapping, viewContainerRef, elementInjector) {
        var _this = this;
        return from(componentMapping.component()).pipe(switchMap(function (component) {
            return _this.defaultHandler.launcher(__assign(__assign({}, componentMapping), { component: component }), viewContainerRef, elementInjector);
        }));
    };
    LazyComponentHandler.ctorParameters = function () { return [
        { type: DefaultComponentHandler }
    ]; };
    LazyComponentHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function LazyComponentHandler_Factory() { return new LazyComponentHandler(i0.ɵɵinject(i1.DefaultComponentHandler)); }, token: LazyComponentHandler, providedIn: "root" });
    LazyComponentHandler = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], LazyComponentHandler);
    return LazyComponentHandler;
}());
export { LazyComponentHandler };
