// @spartacus/storefront/esm5/cms-components/cart/cart-not-empty.guard.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { ActiveCartService, Cart, RoutingService } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var CartNotEmptyGuard = /** @class */ (function () {
    function CartNotEmptyGuard(routingService, activeCartService) {
        this.routingService = routingService;
        this.activeCartService = activeCartService;
    }
    CartNotEmptyGuard.prototype.canActivate = function () {
        var _this = this;
        return combineLatest([
            this.activeCartService.getActive(),
            this.activeCartService.isStable(),
        ]).pipe(filter(function (_a) {
            var _b = __read(_a, 2), _ = _b[0], loaded = _b[1];
            return loaded;
        }), map(function (_a) {
            var _b = __read(_a, 1), cart = _b[0];
            if (_this.isEmpty(cart)) {
                _this.routingService.go({ cxRoute: 'home' });
                return false;
            }
            return true;
        }));
    };
    CartNotEmptyGuard.prototype.isEmpty = function (cart) {
        return cart && !cart.totalItems;
    };
    CartNotEmptyGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: ActiveCartService }
    ]; };
    CartNotEmptyGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartNotEmptyGuard_Factory() { return new CartNotEmptyGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.ActiveCartService)); }, token: CartNotEmptyGuard, providedIn: "root" });
    CartNotEmptyGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CartNotEmptyGuard);
    return CartNotEmptyGuard;
}());
export { CartNotEmptyGuard };
