// @spartacus/storefront/esm5/cms-components/product/product-variants/product-variants.component.js
import { __decorate } from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { VariantType } from '@spartacus/core';
import { CurrentProductService } from '../current-product.service';
import { tap, filter, distinctUntilChanged } from 'rxjs/operators';
var ProductVariantsComponent = /** @class */ (function () {
    function ProductVariantsComponent(currentProductService) {
        this.currentProductService = currentProductService;
        this.variants = [];
        this.variantType = VariantType;
    }
    ProductVariantsComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.product$ = this.currentProductService.getProduct().pipe(filter(function (product) { return !!(product && product.baseOptions); }), distinctUntilChanged(), tap(function (product) {
            product.baseOptions.forEach(function (option) {
                if (option && option.variantType) {
                    _this.variants[option.variantType] = option;
                }
            });
        }));
    };
    ProductVariantsComponent.ctorParameters = function () { return [
        { type: CurrentProductService }
    ]; };
    ProductVariantsComponent = __decorate([
        Component({
            selector: 'cx-product-variants',
            template: "<ng-container *ngIf=\"product$ | async as product\">\n  <div class=\"variant-section\" *ngIf=\"product.baseOptions?.length\">\n    <cx-variant-style-selector\n      *ngIf=\"variants[variantType.STYLE]\"\n      [variants]=\"variants[variantType.STYLE]\"\n    ></cx-variant-style-selector>\n    <cx-variant-size-selector\n      *ngIf=\"variants[variantType.SIZE]\"\n      [product]=\"product\"\n      [variants]=\"variants[variantType.SIZE]\"\n    ></cx-variant-size-selector>\n    <cx-variant-color-selector\n      *ngIf=\"variants[variantType.COLOR]\"\n      [product]=\"product\"\n      [variants]=\"variants[variantType.COLOR]\"\n    ></cx-variant-color-selector>\n  </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ProductVariantsComponent);
    return ProductVariantsComponent;
}());
export { ProductVariantsComponent };
