// @spartacus/storefront/esm2015/cms-components/cart/mini-cart/mini-cart.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActiveCartService } from '@spartacus/core';
import { filter, map, startWith } from 'rxjs/operators';
import { ICON_TYPE } from '../../misc/icon/icon.model';
let MiniCartComponent = class MiniCartComponent {
    constructor(activeCartService) {
        this.activeCartService = activeCartService;
        this.iconTypes = ICON_TYPE;
        this.quantity$ = this.activeCartService.getActive().pipe(startWith({ deliveryItemsQuantity: 0 }), map((cart) => cart.deliveryItemsQuantity || 0));
        this.total$ = this.activeCartService.getActive().pipe(filter((cart) => !!cart.totalPrice), map((cart) => cart.totalPrice.formattedValue));
    }
};
MiniCartComponent.ctorParameters = () => [
    { type: ActiveCartService }
];
MiniCartComponent = __decorate([
    Component({
        selector: 'cx-mini-cart',
        template: "<a\r\n  [attr.aria-label]=\"\r\n    'miniCart.item' | cxTranslate: { count: quantity$ | async }\r\n  \"\r\n  [routerLink]=\"{ cxRoute: 'cart' } | cxUrl\"\r\n>\r\n  <cx-icon [type]=\"iconTypes.CART\"></cx-icon>\r\n\r\n  <span class=\"total\">\r\n    {{ 'miniCart.total' | cxTranslate: { total: total$ | async } }}\r\n  </span>\r\n\r\n  <span class=\"count\">\r\n    {{ 'miniCart.count' | cxTranslate: { count: quantity$ | async } }}\r\n  </span>\r\n</a>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], MiniCartComponent);
export { MiniCartComponent };
