// @spartacus/storefront/esm2015/cms-components/product/product-list/product-view/product-view.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { ICON_TYPE } from '../../../misc/icon/icon.model';
export var ViewModes;
(function (ViewModes) {
    ViewModes["Grid"] = "grid";
    ViewModes["List"] = "list";
})(ViewModes || (ViewModes = {}));
let ProductViewComponent = class ProductViewComponent {
    constructor() {
        this.iconTypes = ICON_TYPE;
        this.modeChange = new EventEmitter();
    }
    get buttonClass() {
        const viewName = this.viewMode.toLowerCase();
        return `cx-product-${viewName}`;
    }
    /**
     *   Display icons inversely to allow users
     *   to see the view they will navigate to
     */
    get viewMode() {
        if (this.mode === 'list') {
            return this.iconTypes.GRID;
        }
        else if (this.mode === 'grid') {
            return this.iconTypes.LIST;
        }
    }
    changeMode() {
        const newMode = this.mode === ViewModes.Grid ? ViewModes.List : ViewModes.Grid;
        this.modeChange.emit(newMode);
    }
};
__decorate([
    Input()
], ProductViewComponent.prototype, "mode", void 0);
__decorate([
    Output()
], ProductViewComponent.prototype, "modeChange", void 0);
ProductViewComponent = __decorate([
    Component({
        selector: 'cx-product-view',
        template: "<button\r\n  class=\"btn cx-product-layout\"\r\n  [ngClass]=\"buttonClass\"\r\n  (click)=\"changeMode()\"\r\n>\r\n  <cx-icon\r\n    *ngIf=\"viewMode === iconTypes.GRID\"\r\n    [type]=\"iconTypes.GRID\"\r\n  ></cx-icon>\r\n  <cx-icon\r\n    *ngIf=\"viewMode === iconTypes.LIST\"\r\n    [type]=\"iconTypes.LIST\"\r\n  ></cx-icon>\r\n</button>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], ProductViewComponent);
export { ProductViewComponent };
