// @spartacus/storefront/esm5/cms-components/myaccount/order/amend-order/cancellations/cancel-order-confirmation/cancel-order-confirmation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { tap } from 'rxjs/operators';
import { OrderAmendService } from '../../amend-order.service';
var CancelOrderConfirmationComponent = /** @class */ (function () {
    function CancelOrderConfirmationComponent(orderAmendService) {
        var _this = this;
        this.orderAmendService = orderAmendService;
        this.form$ = this.orderAmendService
            .getForm()
            .pipe(tap(function (form) { return (_this.orderCode = form.value.orderCode); }));
        this.entries$ = this.orderAmendService.getAmendedEntries();
    }
    CancelOrderConfirmationComponent.prototype.submit = function (form) {
        if (form.valid) {
            this.orderAmendService.save();
        }
        else {
            form.markAllAsTouched();
        }
    };
    CancelOrderConfirmationComponent.ctorParameters = function () { return [
        { type: OrderAmendService }
    ]; };
    CancelOrderConfirmationComponent = __decorate([
        Component({
            selector: 'cx-cancel-order-confirmation',
            template: "<form\r\n  *ngIf=\"form$ | async as form\"\r\n  [formGroup]=\"form\"\r\n  (ngSubmit)=\"submit(form)\"\r\n>\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <cx-amend-order-items\r\n    *ngIf=\"entries$ | async as entries\"\r\n    [entries]=\"entries\"\r\n    [isConfirmation]=\"true\"\r\n  >\r\n  </cx-amend-order-items>\r\n\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <ng-template #actions>\r\n    <cx-amend-order-actions\r\n      *ngIf=\"orderCode\"\r\n      [orderCode]=\"orderCode\"\r\n      [amendOrderForm]=\"form\"\r\n      backRoute=\"orderCancel\"\r\n    ></cx-amend-order-actions>\r\n  </ng-template>\r\n</form>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CancelOrderConfirmationComponent);
    return CancelOrderConfirmationComponent;
}());
export { CancelOrderConfirmationComponent };
