// @spartacus/storefront/esm5/cms-components/asm/services/asm-component.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { AsmAuthService, AuthService, RoutingService, WindowRef, } from '@spartacus/core';
import { of } from 'rxjs';
import { mergeMap, take } from 'rxjs/operators';
import { ASM_ENABLED_LOCAL_STORAGE_KEY } from '../asm-constants';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var AsmComponentService = /** @class */ (function () {
    function AsmComponentService(authService, asmAuthService, routingService, winRef) {
        this.authService = authService;
        this.asmAuthService = asmAuthService;
        this.routingService = routingService;
        this.winRef = winRef;
    }
    AsmComponentService.prototype.logoutCustomerSupportAgentAndCustomer = function () {
        var _this = this;
        this.authService
            .getUserToken()
            .pipe(take(1))
            .subscribe(function (token) {
            if (_this.asmAuthService.isCustomerEmulationToken(token)) {
                _this.logoutCustomer();
            }
            _this.asmAuthService.logoutCustomerSupportAgent();
        });
    };
    AsmComponentService.prototype.logoutCustomer = function () {
        this.authService.logout();
        this.routingService.go({ cxRoute: 'home' });
    };
    AsmComponentService.prototype.isCustomerEmulationSessionInProgress = function () {
        var _this = this;
        return this.authService
            .getUserToken()
            .pipe(mergeMap(function (userToken) {
            return of(_this.asmAuthService.isCustomerEmulationToken(userToken));
        }));
    };
    /**
     * We're currently only removing the persisted storage in the browser
     * to ensure the ASM experience isn't loaded on the next visit. There are a few
     * optimsiations we could think of:
     * - drop the `asm` parameter from the URL, in case it's still there
     * - remove the generated UI from the DOM (outlets currently do not support this)
     */
    AsmComponentService.prototype.unload = function () {
        if (this.winRef.localStorage) {
            this.winRef.localStorage.removeItem(ASM_ENABLED_LOCAL_STORAGE_KEY);
        }
    };
    AsmComponentService.ctorParameters = function () { return [
        { type: AuthService },
        { type: AsmAuthService },
        { type: RoutingService },
        { type: WindowRef }
    ]; };
    AsmComponentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AsmComponentService_Factory() { return new AsmComponentService(i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.AsmAuthService), i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.WindowRef)); }, token: AsmComponentService, providedIn: "root" });
    AsmComponentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], AsmComponentService);
    return AsmComponentService;
}());
export { AsmComponentService };
