// @spartacus/storefront/esm2015/layout/a11y/keyboard-focus/trap/trap-focus.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, HostListener, OnInit } from '@angular/core';
import { TabFocusDirective } from '../tab/tab-focus.directive';
import { TrapFocusService } from './trap-focus.service';
/**
 * Directive that keeps the focus inside the focussable child elements,
 * also known as a _focus trap_.
 */
let TrapFocusDirective = class TrapFocusDirective extends TabFocusDirective {
    constructor(elementRef, service) {
        super(elementRef, service);
        this.elementRef = elementRef;
        this.service = service;
        this.defaultConfig = { trap: true };
        // @Input('cxTrapFocus')
        this.config = {};
        this.handleTrapDown = (event) => {
            if (!!this.config.trap) {
                this.moveFocus(event, 1 /* NEXT */);
            }
        };
        this.handleTrapUp = (event) => {
            if (!!this.config.trap) {
                this.moveFocus(event, -1 /* PREV */);
            }
        };
    }
    /**
     * Moves the focus of the element reference up or down, depending on the increment.
     * The focus of the element is trapped to avoid it from going out of the group.
     *
     * @param event UIEvent that is used to get the target element. The event is blocked
     *   from standard execution and further bubbling.
     * @param increment indicates whether the next or previous is focussed.
     */
    moveFocus(event, increment) {
        if (this.service.hasFocusableChildren(this.host)) {
            this.service.moveFocus(this.host, this.config, increment, event);
        }
    }
};
TrapFocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: TrapFocusService }
];
__decorate([
    HostListener('keydown.arrowdown', ['$event']),
    HostListener('keydown.tab', ['$event'])
], TrapFocusDirective.prototype, "handleTrapDown", void 0);
__decorate([
    HostListener('keydown.arrowup', ['$event']),
    HostListener('keydown.shift.tab', ['$event'])
], TrapFocusDirective.prototype, "handleTrapUp", void 0);
TrapFocusDirective = __decorate([
    Directive() // selector: '[cxTrapFocus]'
], TrapFocusDirective);
export { TrapFocusDirective };
