// @spartacus/storefront/esm5/cms-components/myaccount/reset-password/reset-password.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, } from '@spartacus/core';
import { ResetPasswordFormComponent } from './reset-password-form.component';
import { FormErrorsModule } from '../../../shared/index';
var ResetPasswordModule = /** @class */ (function () {
    function ResetPasswordModule() {
    }
    ResetPasswordModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
                RouterModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ResetPasswordComponent: {
                            component: ResetPasswordFormComponent,
                            guards: [NotAuthGuard],
                        },
                    },
                }),
            ],
            declarations: [ResetPasswordFormComponent],
            exports: [ResetPasswordFormComponent],
            entryComponents: [ResetPasswordFormComponent],
        })
    ], ResetPasswordModule);
    return ResetPasswordModule;
}());
export { ResetPasswordModule };
