// @spartacus/core/esm5/src/product/services/coupon-search-page-meta.resolver.js
import { __decorate, __extends, __read } from "tslib";
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { combineLatest } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
import { AuthService } from '../../auth/facade/auth.service';
import { PageMetaResolver, } from '../../cms/page';
import { TranslationService } from '../../i18n/translation.service';
import { PageType } from '../../model/cms.model';
import { ProductSearchService } from '../../product/facade/product-search.service';
import { SemanticPathService } from '../../routing/configurable-routes/url-translation/semantic-path.service';
import * as i0 from "@angular/core";
import * as i1 from "../facade/product-search.service";
import * as i2 from "../../i18n/translation.service";
import * as i3 from "../../auth/facade/auth.service";
import * as i4 from "@angular/router";
import * as i5 from "../../routing/configurable-routes/url-translation/semantic-path.service";
/**
 * Resolves page meta data for the search result page, in case it's used
 * to query coupons. This is done by adding a `couponcode` query parameter
 * to the search page route.
 *
 * The page resolves an alternative page title and breadcrumb.
 */
var CouponSearchPageResolver = /** @class */ (function (_super) {
    __extends(CouponSearchPageResolver, _super);
    function CouponSearchPageResolver(productSearchService, translation, authService, route, semanticPathService) {
        var _this = _super.call(this) || this;
        _this.productSearchService = productSearchService;
        _this.translation = translation;
        _this.authService = authService;
        _this.route = route;
        _this.semanticPathService = semanticPathService;
        _this.total$ = _this.productSearchService.getResults().pipe(filter(function (data) { return !!(data === null || data === void 0 ? void 0 : data.pagination); }), map(function (results) { return results.pagination.totalResults; }));
        _this.pageType = PageType.CONTENT_PAGE;
        _this.pageTemplate = 'SearchResultsListPageTemplate';
        return _this;
    }
    CouponSearchPageResolver.prototype.resolveBreadcrumbs = function () {
        var _this = this;
        return combineLatest([
            this.translation.translate('common.home'),
            this.translation.translate('myCoupons.myCoupons'),
            this.authService.isUserLoggedIn(),
        ]).pipe(map(function (_a) {
            var _b = __read(_a, 3), homeLabel = _b[0], couponLabel = _b[1], isLoggedIn = _b[2];
            var breadcrumbs = [];
            breadcrumbs.push({ label: homeLabel, link: '/' });
            if (isLoggedIn) {
                breadcrumbs.push({
                    label: couponLabel,
                    link: _this.semanticPathService.transform({
                        cxRoute: 'coupons',
                    }),
                });
            }
            return breadcrumbs;
        }));
    };
    CouponSearchPageResolver.prototype.resolveTitle = function () {
        var _this = this;
        return this.total$.pipe(switchMap(function (total) {
            return _this.translation.translate('pageMetaResolver.search.findProductTitle', {
                count: total,
                coupon: _this.couponCode,
            });
        }));
    };
    CouponSearchPageResolver.prototype.getScore = function (page) {
        return _super.prototype.getScore.call(this, page) + (this.couponCode ? 1 : -1);
    };
    Object.defineProperty(CouponSearchPageResolver.prototype, "couponCode", {
        get: function () {
            var _a, _b;
            return (_b = (_a = this.route.snapshot) === null || _a === void 0 ? void 0 : _a.queryParams) === null || _b === void 0 ? void 0 : _b.couponcode;
        },
        enumerable: true,
        configurable: true
    });
    CouponSearchPageResolver.ctorParameters = function () { return [
        { type: ProductSearchService },
        { type: TranslationService },
        { type: AuthService },
        { type: ActivatedRoute },
        { type: SemanticPathService }
    ]; };
    CouponSearchPageResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function CouponSearchPageResolver_Factory() { return new CouponSearchPageResolver(i0.ɵɵinject(i1.ProductSearchService), i0.ɵɵinject(i2.TranslationService), i0.ɵɵinject(i3.AuthService), i0.ɵɵinject(i4.ActivatedRoute), i0.ɵɵinject(i5.SemanticPathService)); }, token: CouponSearchPageResolver, providedIn: "root" });
    CouponSearchPageResolver = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CouponSearchPageResolver);
    return CouponSearchPageResolver;
}(PageMetaResolver));
export { CouponSearchPageResolver };
