// @spartacus/core/esm5/src/product/services/search-page-meta.resolver.js
import { __decorate, __extends, __read } from "tslib";
import { Injectable } from '@angular/core';
import { combineLatest } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
import { PageMetaResolver } from '../../cms';
import { TranslationService } from '../../i18n/translation.service';
import { PageType } from '../../model/cms.model';
import { RoutingService } from '../../routing/facade/routing.service';
import { ProductSearchService } from '../facade/product-search.service';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "../facade/product-search.service";
import * as i3 from "../../i18n/translation.service";
/**
 * Resolves the page data for the Search Result Page based on the
 * `PageType.CATEGORY_PAGE` and the `SearchResultsListPageTemplate` template.
 *
 * Only the page title is resolved in the standard implemenation.
 */
var SearchPageMetaResolver = /** @class */ (function (_super) {
    __extends(SearchPageMetaResolver, _super);
    function SearchPageMetaResolver(routingService, productSearchService, translation) {
        var _this = _super.call(this) || this;
        _this.routingService = routingService;
        _this.productSearchService = productSearchService;
        _this.translation = translation;
        _this.total$ = _this.productSearchService.getResults().pipe(filter(function (data) { return !!(data === null || data === void 0 ? void 0 : data.pagination); }), map(function (results) { return results.pagination.totalResults; }));
        _this.query$ = _this.routingService
            .getRouterState()
            .pipe(map(function (state) { return state.state.params['query']; }));
        _this.pageType = PageType.CONTENT_PAGE;
        _this.pageTemplate = 'SearchResultsListPageTemplate';
        return _this;
    }
    SearchPageMetaResolver.prototype.resolveTitle = function () {
        var _this = this;
        var sources = [this.total$, this.query$];
        return combineLatest(sources).pipe(switchMap(function (_a) {
            var _b = __read(_a, 2), t = _b[0], q = _b[1];
            return _this.translation.translate('pageMetaResolver.search.title', {
                count: t,
                query: q,
            });
        }));
    };
    SearchPageMetaResolver.ctorParameters = function () { return [
        { type: RoutingService },
        { type: ProductSearchService },
        { type: TranslationService }
    ]; };
    SearchPageMetaResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function SearchPageMetaResolver_Factory() { return new SearchPageMetaResolver(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.ProductSearchService), i0.ɵɵinject(i3.TranslationService)); }, token: SearchPageMetaResolver, providedIn: "root" });
    SearchPageMetaResolver = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], SearchPageMetaResolver);
    return SearchPageMetaResolver;
}(PageMetaResolver));
export { SearchPageMetaResolver };
