// @spartacus/storefront/esm2015/layout/a11y/keyboard-focus/visible/visible-focus.directive.js
import { __decorate } from "tslib";
import { Directive, HostBinding, HostListener } from '@angular/core';
import { BaseFocusDirective } from '../base/base-focus.directive';
/**
 * Directive implementation that adds a CSS class to the host element
 * when the moused is used to focus an element. As soon as the keyboard
 * is used, the class is removed.
 */
let VisibleFocusDirective = class VisibleFocusDirective extends BaseFocusDirective {
    constructor() {
        super(...arguments);
        this.defaultConfig = { disableMouseFocus: true };
        /** controls a polyfill class for the lacking focus-visible feature */
        this.mouseFocus = false;
    }
    handleMousedown() {
        if (this.shouldFocusVisible) {
            this.mouseFocus = true;
        }
    }
    handleKeydown() {
        if (this.shouldFocusVisible) {
            this.mouseFocus = false;
        }
    }
    get shouldFocusVisible() {
        var _a;
        return (_a = this.config) === null || _a === void 0 ? void 0 : _a.disableMouseFocus;
    }
};
__decorate([
    HostBinding('class.mouse-focus')
], VisibleFocusDirective.prototype, "mouseFocus", void 0);
__decorate([
    HostListener('mousedown')
], VisibleFocusDirective.prototype, "handleMousedown", null);
__decorate([
    HostListener('keydown')
], VisibleFocusDirective.prototype, "handleKeydown", null);
VisibleFocusDirective = __decorate([
    Directive() // selector: '[cxVisibleFocus]'
], VisibleFocusDirective);
export { VisibleFocusDirective };
