// @spartacus/core/esm5/src/site-context/store/effects/currencies.effect.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { NEVER, of } from 'rxjs';
import { bufferCount, catchError, exhaustMap, filter, map, switchMapTo, tap, } from 'rxjs/operators';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { WindowRef } from '../../../window/window-ref';
import { SiteConnector } from '../../connectors/site.connector';
import { SiteContextActions } from '../actions/index';
import { getActiveCurrency } from '../selectors/currencies.selectors';
var CurrenciesEffects = /** @class */ (function () {
    function CurrenciesEffects(actions$, siteConnector, winRef, state) {
        var _this = this;
        this.actions$ = actions$;
        this.siteConnector = siteConnector;
        this.winRef = winRef;
        this.state = state;
        this.loadCurrencies$ = this.actions$.pipe(ofType(SiteContextActions.LOAD_CURRENCIES), exhaustMap(function () {
            return _this.siteConnector.getCurrencies().pipe(map(function (currencies) {
                return new SiteContextActions.LoadCurrenciesSuccess(currencies);
            }), catchError(function (error) {
                return of(new SiteContextActions.LoadCurrenciesFail(makeErrorSerializable(error)));
            }));
        }));
        this.persist$ = this.actions$.pipe(ofType(SiteContextActions.SET_ACTIVE_CURRENCY), tap(function (action) {
            if (_this.winRef.sessionStorage) {
                _this.winRef.sessionStorage.setItem('currency', action.payload);
            }
        }), switchMapTo(NEVER));
        this.activateCurrency$ = this.state.select(getActiveCurrency).pipe(bufferCount(2, 1), 
        // avoid dispatching `change` action when we're just setting the initial value:
        filter(function (_a) {
            var _b = __read(_a, 1), previous = _b[0];
            return !!previous;
        }), map(function (_a) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            return new SiteContextActions.CurrencyChange({ previous: previous, current: current });
        }));
    }
    CurrenciesEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: SiteConnector },
        { type: WindowRef },
        { type: Store }
    ]; };
    __decorate([
        Effect()
    ], CurrenciesEffects.prototype, "loadCurrencies$", void 0);
    __decorate([
        Effect()
    ], CurrenciesEffects.prototype, "persist$", void 0);
    __decorate([
        Effect()
    ], CurrenciesEffects.prototype, "activateCurrency$", void 0);
    CurrenciesEffects = __decorate([
        Injectable()
    ], CurrenciesEffects);
    return CurrenciesEffects;
}());
export { CurrenciesEffects };
