// @spartacus/core/esm5/src/site-context/store/effects/languages.effect.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { NEVER, of } from 'rxjs';
import { bufferCount, catchError, exhaustMap, filter, map, switchMapTo, tap, } from 'rxjs/operators';
import { makeErrorSerializable } from '../../../util/serialization-utils';
import { WindowRef } from '../../../window/window-ref';
import { SiteConnector } from '../../connectors/site.connector';
import { SiteContextActions } from '../actions/index';
import { getActiveLanguage } from '../selectors/languages.selectors';
var LanguagesEffects = /** @class */ (function () {
    function LanguagesEffects(actions$, siteConnector, winRef, state) {
        var _this = this;
        this.actions$ = actions$;
        this.siteConnector = siteConnector;
        this.winRef = winRef;
        this.state = state;
        this.loadLanguages$ = this.actions$.pipe(ofType(SiteContextActions.LOAD_LANGUAGES), exhaustMap(function () {
            return _this.siteConnector.getLanguages().pipe(map(function (languages) { return new SiteContextActions.LoadLanguagesSuccess(languages); }), catchError(function (error) {
                return of(new SiteContextActions.LoadLanguagesFail(makeErrorSerializable(error)));
            }));
        }));
        this.persist$ = this.actions$.pipe(ofType(SiteContextActions.SET_ACTIVE_LANGUAGE), tap(function (action) {
            if (_this.winRef.sessionStorage) {
                _this.winRef.sessionStorage.setItem('language', action.payload);
            }
        }), switchMapTo(NEVER));
        this.activateLanguage$ = this.state.select(getActiveLanguage).pipe(bufferCount(2, 1), 
        // avoid dispatching `change` action when we're just setting the initial value:
        filter(function (_a) {
            var _b = __read(_a, 1), previous = _b[0];
            return !!previous;
        }), map(function (_a) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            return new SiteContextActions.LanguageChange({ previous: previous, current: current });
        }));
    }
    LanguagesEffects.ctorParameters = function () { return [
        { type: Actions },
        { type: SiteConnector },
        { type: WindowRef },
        { type: Store }
    ]; };
    __decorate([
        Effect()
    ], LanguagesEffects.prototype, "loadLanguages$", void 0);
    __decorate([
        Effect()
    ], LanguagesEffects.prototype, "persist$", void 0);
    __decorate([
        Effect()
    ], LanguagesEffects.prototype, "activateLanguage$", void 0);
    LanguagesEffects = __decorate([
        Injectable()
    ], LanguagesEffects);
    return LanguagesEffects;
}());
export { LanguagesEffects };
