// @spartacus/storefront/esm2015/cms-components/checkout/components/place-order/place-order.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { CartNotEmptyGuard } from '../../../cart/cart-not-empty.guard';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { PlaceOrderComponent } from './place-order.component';
import { FormErrorsModule } from '../../../../shared/index';
import { ReactiveFormsModule } from '@angular/forms';
let PlaceOrderModule = class PlaceOrderModule {
};
PlaceOrderModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            RouterModule,
            UrlModule,
            I18nModule,
            ReactiveFormsModule,
            FormErrorsModule,
        ],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    CheckoutPlaceOrder: {
                        component: PlaceOrderComponent,
                        guards: [CheckoutAuthGuard, CartNotEmptyGuard],
                    },
                },
            }),
        ],
        declarations: [PlaceOrderComponent],
        entryComponents: [PlaceOrderComponent],
        exports: [PlaceOrderComponent],
    })
], PlaceOrderModule);
export { PlaceOrderModule };
