// @spartacus/storefront/esm2015/cms-components/checkout/components/place-order/place-order.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { CheckoutService, RoutingService } from '@spartacus/core';
import { filter } from 'rxjs/operators';
let PlaceOrderComponent = class PlaceOrderComponent {
    constructor(checkoutService, routingService, fb) {
        this.checkoutService = checkoutService;
        this.routingService = routingService;
        this.fb = fb;
        this.checkoutSubmitForm = this.fb.group({
            termsAndConditions: [false, Validators.requiredTrue],
        });
    }
    submitForm() {
        if (this.checkoutSubmitForm.valid) {
            this.checkoutService.placeOrder();
        }
        else {
            this.checkoutSubmitForm.markAllAsTouched();
        }
    }
    ngOnInit() {
        this.placeOrderSubscription = this.checkoutService
            .getOrderDetails()
            .pipe(filter((order) => Object.keys(order).length !== 0))
            .subscribe(() => {
            this.routingService.go({ cxRoute: 'orderConfirmation' });
        });
    }
    ngOnDestroy() {
        if (this.placeOrderSubscription) {
            this.placeOrderSubscription.unsubscribe();
        }
    }
};
PlaceOrderComponent.ctorParameters = () => [
    { type: CheckoutService },
    { type: RoutingService },
    { type: FormBuilder }
];
PlaceOrderComponent = __decorate([
    Component({
        selector: 'cx-place-order',
        template: "<form\r\n  (ngSubmit)=\"submitForm()\"\r\n  class=\"cx-place-order-form form-check\"\r\n  [formGroup]=\"checkoutSubmitForm\"\r\n>\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <input\r\n        formControlName=\"termsAndConditions\"\r\n        class=\"form-check-input\"\r\n        type=\"checkbox\"\r\n      />\r\n      <span class=\"form-check-label\">\r\n        {{ 'checkoutReview.confirmThatRead' | cxTranslate }}\r\n        <a\r\n          [routerLink]=\"{ cxRoute: 'termsAndConditions' } | cxUrl\"\r\n          class=\"cx-tc-link\"\r\n          target=\"_blank\"\r\n        >\r\n          {{ 'checkoutReview.termsAndConditions' | cxTranslate }}\r\n        </a>\r\n      </span>\r\n      <cx-form-errors\r\n        [control]=\"checkoutSubmitForm.get('termsAndConditions')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <button type=\"submit\" class=\"btn btn-primary btn-block\">\r\n    {{ 'checkoutReview.placeOrder' | cxTranslate }}\r\n  </button>\r\n</form>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], PlaceOrderComponent);
export { PlaceOrderComponent };
