// @spartacus/storefront/esm2015/cms-components/myaccount/notification-preference/notification-preference.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { NotificationPreferenceComponent } from './notification-preference.component';
let NotificationPreferenceModule = class NotificationPreferenceModule {
};
NotificationPreferenceModule = __decorate([
    NgModule({
        declarations: [NotificationPreferenceComponent],
        imports: [CommonModule, SpinnerModule, I18nModule],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    NotificationPreferenceComponent: {
                        component: NotificationPreferenceComponent,
                        guards: [AuthGuard],
                    },
                },
            }),
        ],
        exports: [NotificationPreferenceComponent],
        entryComponents: [NotificationPreferenceComponent],
    })
], NotificationPreferenceModule);
export { NotificationPreferenceModule };
