// @spartacus/storefront/esm5/cms-components/product/product-tabs/product-attributes/product-attributes.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { ProductAttributesComponent } from './product-attributes.component';
var ProductAttributesModule = /** @class */ (function () {
    function ProductAttributesModule() {
    }
    ProductAttributesModule = __decorate([
        NgModule({
            imports: [CommonModule, I18nModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ProductSpecsTabComponent: {
                            component: ProductAttributesComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductAttributesComponent],
            entryComponents: [ProductAttributesComponent],
            exports: [ProductAttributesComponent],
        })
    ], ProductAttributesModule);
    return ProductAttributesModule;
}());
export { ProductAttributesModule };
