// @spartacus/storefront/esm5/layout/a11y/keyboard-focus/visible/visible-focus.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, HostBinding, HostListener } from '@angular/core';
import { BaseFocusDirective } from '../base/base-focus.directive';
/**
 * Directive implementation that adds a CSS class to the host element
 * when the moused is used to focus an element. As soon as the keyboard
 * is used, the class is removed.
 */
var VisibleFocusDirective = /** @class */ (function (_super) {
    __extends(VisibleFocusDirective, _super);
    function VisibleFocusDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultConfig = { disableMouseFocus: true };
        /** controls a polyfill class for the lacking focus-visible feature */
        _this.mouseFocus = false;
        return _this;
    }
    VisibleFocusDirective.prototype.handleMousedown = function () {
        if (this.shouldFocusVisible) {
            this.mouseFocus = true;
        }
    };
    VisibleFocusDirective.prototype.handleKeydown = function () {
        if (this.shouldFocusVisible) {
            this.mouseFocus = false;
        }
    };
    Object.defineProperty(VisibleFocusDirective.prototype, "shouldFocusVisible", {
        get: function () {
            var _a;
            return (_a = this.config) === null || _a === void 0 ? void 0 : _a.disableMouseFocus;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        HostBinding('class.mouse-focus')
    ], VisibleFocusDirective.prototype, "mouseFocus", void 0);
    __decorate([
        HostListener('mousedown')
    ], VisibleFocusDirective.prototype, "handleMousedown", null);
    __decorate([
        HostListener('keydown')
    ], VisibleFocusDirective.prototype, "handleKeydown", null);
    VisibleFocusDirective = __decorate([
        Directive() // selector: '[cxVisibleFocus]'
    ], VisibleFocusDirective);
    return VisibleFocusDirective;
}(BaseFocusDirective));
export { VisibleFocusDirective };
