// @spartacus/storefront/esm5/cms-components/myaccount/order/return-request-detail/return-request-overview/return-request-overview.component.js
import { __decorate } from "tslib";
import { Component, ChangeDetectionStrategy, } from '@angular/core';
import { tap } from 'rxjs/operators';
import { ReturnRequestService } from '../return-request.service';
var ReturnRequestOverviewComponent = /** @class */ (function () {
    function ReturnRequestOverviewComponent(returnRequestService) {
        var _this = this;
        this.returnRequestService = returnRequestService;
        this.returnRequest$ = this.returnRequestService
            .getReturnRequest()
            .pipe(tap(function (returnRequest) { return (_this.rma = returnRequest.rma); }));
        this.isCancelling$ = this.returnRequestService.isCancelling$;
    }
    ReturnRequestOverviewComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.returnRequestService.isCancelSuccess$.subscribe(function (success) {
            if (success) {
                _this.returnRequestService.cancelSuccess(_this.rma);
            }
        });
    };
    ReturnRequestOverviewComponent.prototype.cancelReturn = function (returnRequestCode) {
        this.returnRequestService.cancelReturnRequest(returnRequestCode);
    };
    ReturnRequestOverviewComponent.prototype.back = function () {
        this.returnRequestService.backToList();
    };
    ReturnRequestOverviewComponent.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    ReturnRequestOverviewComponent.ctorParameters = function () { return [
        { type: ReturnRequestService }
    ]; };
    ReturnRequestOverviewComponent = __decorate([
        Component({
            selector: 'cx-return-request-overview',
            template: "<ng-container *ngIf=\"returnRequest$ | async as returnRequest\">\n  <div class=\"cx-nav row\">\n    <div class=\"col-xs-12 col-md-4 col-lg-3\">\n      <button (click)=\"back()\" class=\"btn btn-block btn-action\">\n        {{ 'common.back' | cxTranslate }}\n      </button>\n    </div>\n\n    <div class=\"col-xs-12 col-md-4 col-lg-3\">\n      <button\n        *ngIf=\"returnRequest.cancellable\"\n        class=\"btn btn-block btn-primary\"\n        (click)=\"cancelReturn(returnRequest.rma)\"\n        [disabled]=\"isCancelling$ | async\"\n      >\n        {{ 'returnRequest.cancel' | cxTranslate }}\n      </button>\n    </div>\n  </div>\n  <div class=\"cx-header row\">\n    <div class=\"cx-detail col-sm-12 col-md-4\">\n      <div class=\"cx-detail-label\">\n        {{ 'returnRequest.returnRequestId' | cxTranslate }}\n      </div>\n      <div class=\"cx-detail-value\">{{ returnRequest.rma }}</div>\n    </div>\n    <div class=\"cx-detail col-sm-12 col-md-4\">\n      <div class=\"cx-detail-label\">\n        {{ 'returnRequest.orderCode' | cxTranslate }}\n      </div>\n      <div class=\"cx-detail-value\">{{ returnRequest.order?.code }}</div>\n    </div>\n    <div class=\"cx-detail col-sm-12 col-md-4\">\n      <div class=\"cx-detail-label\">\n        {{ 'returnRequest.status' | cxTranslate }}\n      </div>\n      <div class=\"cx-detail-value\">\n        {{\n          'returnRequestList.statusDisplay'\n            | cxTranslate: { context: returnRequest.status }\n        }}\n      </div>\n    </div>\n  </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ReturnRequestOverviewComponent);
    return ReturnRequestOverviewComponent;
}());
export { ReturnRequestOverviewComponent };
