// @spartacus/storefront/esm2015/shared/components/list-navigation/pagination/pagination.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { ActivatedRoute, Params } from '@angular/router';
import { PaginationBuilder } from './pagination.builder';
import { PaginationItemType } from './pagination.model';
/**
 * The `PaginationComponent` is a generic component that is used for
 * all lists in Spartacus that require pagination. The component supports
 * all common features, which can be configured or hidden by CSS.
 */
let PaginationComponent = class PaginationComponent {
    constructor(paginationBuilder, activatedRoute) {
        this.paginationBuilder = paginationBuilder;
        this.activatedRoute = activatedRoute;
        this.viewPageEvent = new EventEmitter();
        this.pages = [];
    }
    get pagination() {
        return this._pagination;
    }
    set pagination(value) {
        this._pagination = value;
        this.render(value);
    }
    render(pagination) {
        this.pages = this.paginationBuilder.paginate(pagination.totalPages, pagination.currentPage);
    }
    /**
     * Inidicates whether the given item is the current item.
     *
     * @param item PaginationItem
     * @returns boolean
     */
    isCurrent(item) {
        return (item.type === PaginationItemType.PAGE &&
            item.number === this.pagination.currentPage);
    }
    /**
     * Indicates whether the pagination item is inactive. This is used
     * to disabled a link or set the tabindex to `-1`.
     *
     * Defaults to true
     *
     * @param item PaginationItem
     * @returns returns -1 in case of a disabled
     */
    isInactive(item) {
        return (!item.hasOwnProperty('number') ||
            item.number === this.pagination.currentPage);
    }
    getQueryParams(item) {
        const queryParams = Object.assign({}, this.activatedRoute.snapshot.queryParams);
        if (this.queryParam &&
            item.number < this.pagination.totalPages &&
            !this.isCurrent(item)) {
            queryParams[this.queryParam] = item.number;
        }
        // omit the page number from the query parameters in case it's the default
        // to clean up the experience and avoid unnecessary polluting of the URL
        if (queryParams[this.queryParam] === this.defaultPage) {
            delete queryParams[this.queryParam];
        }
        return queryParams;
    }
    pageChange(page) {
        this.viewPageEvent.emit(page.number);
    }
};
PaginationComponent.ctorParameters = () => [
    { type: PaginationBuilder },
    { type: ActivatedRoute }
];
__decorate([
    Input()
], PaginationComponent.prototype, "pageRoute", void 0);
__decorate([
    Input()
], PaginationComponent.prototype, "queryParam", void 0);
__decorate([
    Input()
], PaginationComponent.prototype, "defaultPage", void 0);
__decorate([
    Input()
], PaginationComponent.prototype, "pagination", null);
__decorate([
    Output()
], PaginationComponent.prototype, "viewPageEvent", void 0);
PaginationComponent = __decorate([
    Component({
        selector: 'cx-pagination',
        template: "<a\r\n  *ngFor=\"let item of pages\"\r\n  [class]=\"item.type\"\r\n  [class.disabled]=\"isInactive(item)\"\r\n  [class.current]=\"isCurrent(item)\"\r\n  [routerLink]=\"pageRoute\"\r\n  [queryParams]=\"getQueryParams(item)\"\r\n  [tabIndex]=\"isInactive(item) ? -1 : 0\"\r\n  (click)=\"pageChange(item)\"\r\n>\r\n  {{ item.label }}\r\n</a>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], PaginationComponent);
export { PaginationComponent };
