// @spartacus/storefront/esm2015/cms-components/myaccount/address-book/address-book.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AuthGuard, I18nModule, provideDefaultConfig, UserAddressService, } from '@spartacus/core';
import { CardModule } from '../../../shared/components/card/card.module';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { AddressFormModule } from '../../checkout/components/shipping-address/address-form/address-form.module';
import { AddressBookComponent } from './address-book.component';
import { AddressBookComponentService } from './address-book.component.service';
let AddressBookModule = class AddressBookModule {
};
AddressBookModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            CardModule,
            AddressFormModule,
            SpinnerModule,
            I18nModule,
        ],
        declarations: [AddressBookComponent],
        exports: [AddressBookComponent],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    AccountAddressBookComponent: {
                        component: AddressBookComponent,
                        providers: [
                            {
                                provide: AddressBookComponentService,
                                useClass: AddressBookComponentService,
                                deps: [UserAddressService],
                            },
                        ],
                        guards: [AuthGuard],
                    },
                },
            }),
            UserAddressService,
        ],
        entryComponents: [AddressBookComponent],
    })
], AddressBookModule);
export { AddressBookModule };
