// @spartacus/core/esm5/src/kyma/services/open-id-token/open-id-token.service.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { KymaConfig } from '../../config/kyma-config';
import * as i0 from "@angular/core";
import * as i1 from "../../config/kyma-config";
import * as i2 from "@angular/common/http";
var OAUTH_ENDPOINT = '/authorizationserver/oauth/token';
var OpenIdAuthenticationTokenService = /** @class */ (function () {
    function OpenIdAuthenticationTokenService(config, http) {
        this.config = config;
        this.http = http;
    }
    OpenIdAuthenticationTokenService.prototype.loadOpenIdAuthenticationToken = function (username, password) {
        var url = this.getOAuthEndpoint();
        var params = new HttpParams()
            .set('client_id', encodeURIComponent(this.config.authentication.kyma_client_id))
            .set('client_secret', encodeURIComponent(this.config.authentication.kyma_client_secret))
            .set('grant_type', 'password') // authorization_code, client_credentials, password
            .set('username', username)
            .set('password', password)
            .set('scope', 'openid');
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        return this.http
            .post(url, params, { headers: headers })
            .pipe(catchError(function (error) { return throwError(error); }));
    };
    OpenIdAuthenticationTokenService.prototype.getOAuthEndpoint = function () {
        return (this.config.backend.occ.baseUrl || '') + OAUTH_ENDPOINT;
    };
    OpenIdAuthenticationTokenService.ctorParameters = function () { return [
        { type: KymaConfig },
        { type: HttpClient }
    ]; };
    OpenIdAuthenticationTokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OpenIdAuthenticationTokenService_Factory() { return new OpenIdAuthenticationTokenService(i0.ɵɵinject(i1.KymaConfig), i0.ɵɵinject(i2.HttpClient)); }, token: OpenIdAuthenticationTokenService, providedIn: "root" });
    OpenIdAuthenticationTokenService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], OpenIdAuthenticationTokenService);
    return OpenIdAuthenticationTokenService;
}());
export { OpenIdAuthenticationTokenService };
