// @spartacus/core/esm5/src/cart/event/cart-event.builder.js
import { __decorate, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { ofType } from '@ngrx/effects';
import { ActionsSubject } from '@ngrx/store';
import { filter, map, withLatestFrom } from 'rxjs/operators';
import { EventService } from '../../event/event.service';
import { createFrom } from '../../util/create-from';
import { ActiveCartService } from '../facade/active-cart.service';
import { CartActions } from '../store';
import { CartAddEntryEvent, CartAddEntryFailEvent, CartAddEntrySuccessEvent, } from './cart.events';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../event/event.service";
import * as i3 from "../facade/active-cart.service";
/**
 * Registers events for the active cart
 */
var CartEventBuilder = /** @class */ (function () {
    function CartEventBuilder(actionsSubject, event, activeCartService) {
        this.actionsSubject = actionsSubject;
        this.event = event;
        this.activeCartService = activeCartService;
        this.register();
    }
    /**
     * Registers events for the active cart
     */
    CartEventBuilder.prototype.register = function () {
        this.registerAddEntry();
    };
    /**
     * Register events for adding entry to the active cart
     */
    CartEventBuilder.prototype.registerAddEntry = function () {
        this.registerMapped({
            action: CartActions.CART_ADD_ENTRY,
            event: CartAddEntryEvent,
        });
        this.registerMapped({
            action: CartActions.CART_ADD_ENTRY_SUCCESS,
            event: CartAddEntrySuccessEvent,
        });
        this.registerMapped({
            action: CartActions.CART_ADD_ENTRY_FAIL,
            event: CartAddEntryFailEvent,
        });
    };
    /**
     * Registers a stream of target events mapped from the source actions that contain the cart id equal to the active cart id.
     *
     * @param mapping mapping declaration - from `action` string type to `event` class type
     *   (an with optional `factory` function - by default `action.payload` will be assigned to the properties of the event instance).
     */
    CartEventBuilder.prototype.registerMapped = function (mapping) {
        var eventStream$ = this.getAction(mapping.action).pipe(withLatestFrom(this.activeCartService.getActiveCartId()), filter(function (_a) {
            var _b = __read(_a, 2), action = _b[0], activeCartId = _b[1];
            return action.payload['cartId'] === activeCartId;
        } // assuming that action's payload contains the cart id
        ), map(function (_a) {
            var _b = __read(_a, 1), action = _b[0];
            return createFrom(mapping.event, action.payload);
        }));
        return this.event.register(mapping.event, eventStream$);
    };
    /**
     * Returns a stream of actions only of a given type(s)
     *
     * @param actionType type(s) of actions
     */
    CartEventBuilder.prototype.getAction = function (actionType) {
        return this.actionsSubject.pipe(ofType.apply(void 0, __spread([].concat(actionType))));
    };
    CartEventBuilder.ctorParameters = function () { return [
        { type: ActionsSubject },
        { type: EventService },
        { type: ActiveCartService }
    ]; };
    CartEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartEventBuilder_Factory() { return new CartEventBuilder(i0.ɵɵinject(i1.ActionsSubject), i0.ɵɵinject(i2.EventService), i0.ɵɵinject(i3.ActiveCartService)); }, token: CartEventBuilder, providedIn: "root" });
    CartEventBuilder = __decorate([
        Injectable({ providedIn: 'root' })
    ], CartEventBuilder);
    return CartEventBuilder;
}());
export { CartEventBuilder };
