// @spartacus/storefront/esm5/cms-components/myaccount/order/amend-order/returns/order-return.guard.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { RoutingService } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { OrderReturnService } from './order-return.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "./order-return.service";
var OrderReturnGuard = /** @class */ (function () {
    function OrderReturnGuard(routing, orderAmendService) {
        this.routing = routing;
        this.orderAmendService = orderAmendService;
    }
    OrderReturnGuard.prototype.canActivate = function () {
        var _this = this;
        return this.orderAmendService.getForm().pipe(map(function (form) {
            if (!form.valid) {
                // the order code is not available in the route
                // as long as we're inside a guard, hence we redirect
                // to the common orders page.
                _this.routing.go({ cxRoute: 'orders' });
                return false;
            }
            else {
                return true;
            }
        }));
    };
    OrderReturnGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: OrderReturnService }
    ]; };
    OrderReturnGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function OrderReturnGuard_Factory() { return new OrderReturnGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.OrderReturnService)); }, token: OrderReturnGuard, providedIn: "root" });
    OrderReturnGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], OrderReturnGuard);
    return OrderReturnGuard;
}());
export { OrderReturnGuard };
