// @spartacus/storefront/esm5/cms-components/myaccount/notification-preference/notification-preference.component.js
import { __assign, __decorate, __read } from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { NotificationPreference, UserNotificationPreferenceService, } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { map, tap } from 'rxjs/operators';
var NotificationPreferenceComponent = /** @class */ (function () {
    function NotificationPreferenceComponent(notificationPreferenceService) {
        this.notificationPreferenceService = notificationPreferenceService;
        this.preferences = [];
    }
    NotificationPreferenceComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.notificationPreferenceService.resetNotificationPreferences();
        this.preferences$ = this.notificationPreferenceService
            .getPreferences()
            .pipe(tap(function (preferences) { return (_this.preferences = preferences); }));
        this.notificationPreferenceService.loadPreferences();
        this.isLoading$ = combineLatest([
            this.notificationPreferenceService.getPreferencesLoading(),
            this.notificationPreferenceService.getUpdatePreferencesResultLoading(),
        ]).pipe(map(function (_a) {
            var _b = __read(_a, 2), prefsLoading = _b[0], updateLoading = _b[1];
            return prefsLoading || updateLoading;
        }));
    };
    NotificationPreferenceComponent.prototype.updatePreference = function (preference) {
        var updatedPreferences = [];
        this.preferences.forEach(function (p) {
            if (p.channel === preference.channel) {
                updatedPreferences.push(__assign(__assign({}, p), { enabled: !p.enabled }));
            }
            else {
                updatedPreferences.push(p);
            }
        });
        this.notificationPreferenceService.updatePreferences(updatedPreferences);
    };
    NotificationPreferenceComponent.ctorParameters = function () { return [
        { type: UserNotificationPreferenceService }
    ]; };
    NotificationPreferenceComponent = __decorate([
        Component({
            selector: 'cx-notification-preference',
            template: "<ng-container *ngIf=\"preferences$ | async as preferences\">\n  <div *ngIf=\"preferences.length > 0; else loading\">\n    <div class=\"row d-flex justify-content-center\">\n      <div class=\"col-md-8\">\n        <div class=\"pref-header\">\n          {{ 'notificationPreference.message' | cxTranslate }}\n        </div>\n        <div class=\"form-check cx-notification-channels\">\n          <ng-container *ngFor=\"let preference of preferences\">\n            <label *ngIf=\"preference.visible\" class=\"pref-channel\">\n              <input\n                class=\"form-check-input cx-np-checkbox\"\n                role=\"checkbox\"\n                type=\"checkbox\"\n                [checked]=\"preference.enabled\"\n                (change)=\"updatePreference(preference)\"\n                [disabled]=\"isLoading$ | async\"\n              />\n              <span class=\"form-check-label\">\n                {{\n                  'notificationPreference.' + preference.channel | cxTranslate\n                }}\n                {{ preference.value }}\n              </span>\n            </label>\n          </ng-container>\n        </div>\n        <label class=\"pref-note\"\n          ><strong>{{ 'notificationPreference.note' | cxTranslate }}</strong\n          >{{ 'notificationPreference.noteMessage' | cxTranslate }}\n        </label>\n      </div>\n    </div>\n  </div>\n\n  <ng-template #loading>\n    <div class=\"cx-spinner\">\n      <cx-spinner></cx-spinner>\n    </div>\n  </ng-template>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NotificationPreferenceComponent);
    return NotificationPreferenceComponent;
}());
export { NotificationPreferenceComponent };
