// @spartacus/storefront/esm2015/shared/components/form/form-errors/form-errors.component.js
import { __decorate } from "tslib";
import { Component, Input, ChangeDetectionStrategy, HostBinding, } from '@angular/core';
import { startWith, map } from 'rxjs/operators';
/**
 * This component renders form errors.
 */
let FormErrorsComponent = class FormErrorsComponent {
    set control(control) {
        this._control = control;
        this.errors$ = control === null || control === void 0 ? void 0 : control.statusChanges.pipe(startWith({}), map(() => control.errors || {}), map((errors) => Object.entries(errors)
            .filter((error) => error[1])
            .map((error) => error[0])));
    }
    get control() {
        return this._control;
    }
    get invalid() {
        return this.control.invalid;
    }
    get dirty() {
        return this.control.dirty;
    }
    get touched() {
        return this.control.touched;
    }
};
__decorate([
    Input()
], FormErrorsComponent.prototype, "control", null);
__decorate([
    HostBinding('class.control-invalid')
], FormErrorsComponent.prototype, "invalid", null);
__decorate([
    HostBinding('class.control-dirty')
], FormErrorsComponent.prototype, "dirty", null);
__decorate([
    HostBinding('class.control-touched')
], FormErrorsComponent.prototype, "touched", null);
FormErrorsComponent = __decorate([
    Component({
        selector: 'cx-form-errors',
        template: "<p *ngFor=\"let errorName of errors$ | async\">\r\n  {{ 'formErrors.' + errorName | cxTranslate }}\r\n</p>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], FormErrorsComponent);
export { FormErrorsComponent };
