// @spartacus/storefront/esm2015/cms-components/cart/add-to-wishlist/add-to-wish-list.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { AuthService, OrderEntry, Product, WishListService, } from '@spartacus/core';
import { filter, map, tap } from 'rxjs/operators';
import { ICON_TYPE } from '../../../cms-components/misc/icon/icon.model';
import { CurrentProductService } from '../../product/current-product.service';
let AddToWishListComponent = class AddToWishListComponent {
    constructor(wishListService, currentProductService, authService) {
        this.wishListService = wishListService;
        this.currentProductService = currentProductService;
        this.authService = authService;
        this.product$ = this.currentProductService.getProduct().pipe(filter((product) => Boolean(product)), tap((product) => this.setStockInfo(product)));
        this.wishListEntries$ = this.wishListService.getWishList().pipe(filter((wishlist) => Boolean(wishlist)), map((wishList) => wishList.entries));
        this.userLoggedIn$ = this.authService.isUserLoggedIn();
        this.loading$ = this.wishListService.getWishListLoading();
        this.hasStock = false;
        this.iconTypes = ICON_TYPE;
    }
    add(product) {
        this.wishListService.addEntry(product.code);
    }
    remove(entry) {
        this.wishListService.removeEntry(entry);
    }
    getProductInWishList(product, entries) {
        const item = entries.find((entry) => entry.product.code === product.code);
        return item;
    }
    setStockInfo(product) {
        this.hasStock =
            product.stock && product.stock.stockLevelStatus !== 'outOfStock';
    }
};
AddToWishListComponent.ctorParameters = () => [
    { type: WishListService },
    { type: CurrentProductService },
    { type: AuthService }
];
AddToWishListComponent = __decorate([
    Component({
        selector: 'cx-add-to-wishlist',
        template: "<ng-container *ngIf=\"product$ | async as product\">\r\n  <ng-container *ngIf=\"userLoggedIn$ | async; else loginPrompt\">\r\n    <ng-container *ngIf=\"wishListEntries$ | async as entries\">\r\n      <ng-container *ngIf=\"hasStock\">\r\n        <div\r\n          *ngIf=\"getProductInWishList(product, entries) as entry; else addItem\"\r\n        >\r\n          <button\r\n            class=\"btn btn-link button-remove cx-action-link\"\r\n            (click)=\"remove(entry)\"\r\n            [disabled]=\"loading$ | async\"\r\n          >\r\n            <cx-icon [type]=\"iconTypes.HEART\"></cx-icon>\r\n            <span class=\"button-text\">{{\r\n              'addToWishList.remove' | cxTranslate\r\n            }}</span>\r\n          </button>\r\n        </div>\r\n        <ng-template #addItem>\r\n          <button\r\n            class=\"btn btn-link button-add cx-action-link\"\r\n            (click)=\"add(product)\"\r\n            [disabled]=\"loading$ | async\"\r\n          >\r\n            <cx-icon [type]=\"iconTypes.EMPTY_HEART\"></cx-icon>\r\n            <span class=\"button-text\">{{\r\n              'addToWishList.add' | cxTranslate\r\n            }}</span>\r\n          </button>\r\n        </ng-template>\r\n      </ng-container>\r\n    </ng-container>\r\n  </ng-container>\r\n</ng-container>\r\n\r\n<ng-template #loginPrompt>\r\n  <ng-container *ngIf=\"hasStock\">\r\n    <a\r\n      class=\"btn btn-link button-add-link cx-action-link\"\r\n      [routerLink]=\"{ cxRoute: 'login' } | cxUrl\"\r\n    >\r\n      <cx-icon [type]=\"iconTypes.EMPTY_HEART\"></cx-icon>\r\n      <span class=\"button-text\">{{\r\n        'addToWishList.anonymous' | cxTranslate\r\n      }}</span>\r\n    </a>\r\n  </ng-container>\r\n</ng-template>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], AddToWishListComponent);
export { AddToWishListComponent };
