// @spartacus/core/esm5/src/auth/store/reducers/index.js
import { __assign } from "tslib";
import { InjectionToken } from '@angular/core';
import { combineReducers, } from '@ngrx/store';
import { loaderReducer } from '../../../state/utils/loader/loader.reducer';
import { AuthActions } from '../actions/index';
import { CLIENT_TOKEN_DATA } from '../auth-state';
import * as fromUserTokenReducer from './user-token.reducer';
export function getReducers() {
    return {
        userToken: combineReducers({ token: fromUserTokenReducer.reducer }),
        clientToken: loaderReducer(CLIENT_TOKEN_DATA),
    };
}
export var reducerToken = new InjectionToken('AuthReducers');
export var reducerProvider = {
    provide: reducerToken,
    useFactory: getReducers,
};
export function clearAuthState(reducer) {
    return function (state, action) {
        if (action.type === AuthActions.LOGOUT) {
            state = __assign(__assign({}, state), { userToken: undefined });
        }
        return reducer(state, action);
    };
}
export var metaReducers = [clearAuthState];
