// @spartacus/storefront/esm2015/cms-components/myaccount/order/amend-order/returns/return-order-confirmation/return-order-confirmation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { tap } from 'rxjs/operators';
import { OrderAmendService } from '../../amend-order.service';
let ReturnOrderConfirmationComponent = class ReturnOrderConfirmationComponent {
    constructor(orderAmendService) {
        this.orderAmendService = orderAmendService;
        this.form$ = this.orderAmendService
            .getForm()
            .pipe(tap((form) => (this.orderCode = form.value.orderCode)));
        this.entries$ = this.orderAmendService.getAmendedEntries();
    }
    submit(form) {
        form.disable();
        this.orderAmendService.save();
    }
};
ReturnOrderConfirmationComponent.ctorParameters = () => [
    { type: OrderAmendService }
];
ReturnOrderConfirmationComponent = __decorate([
    Component({
        selector: 'cx-return-order-confirmation',
        template: "<form\r\n  *ngIf=\"form$ | async as form\"\r\n  [formGroup]=\"form\"\r\n  (ngSubmit)=\"submit(form)\"\r\n>\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <cx-amend-order-items\r\n    *ngIf=\"entries$ | async as entries\"\r\n    [entries]=\"entries\"\r\n    [isConfirmation]=\"true\"\r\n  >\r\n  </cx-amend-order-items>\r\n\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <ng-template #actions>\r\n    <cx-amend-order-actions\r\n      *ngIf=\"orderCode\"\r\n      [orderCode]=\"orderCode\"\r\n      [amendOrderForm]=\"form\"\r\n      backRoute=\"orderReturn\"\r\n    ></cx-amend-order-actions>\r\n  </ng-template>\r\n</form>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], ReturnOrderConfirmationComponent);
export { ReturnOrderConfirmationComponent };
