// @spartacus/core/esm5/src/site-context/store/actions/currencies.action.js
export var LOAD_CURRENCIES = '[Site-context] Load Currencies';
export var LOAD_CURRENCIES_FAIL = '[Site-context] Load Currencies Fail';
export var LOAD_CURRENCIES_SUCCESS = '[Site-context] Load Currencies Success';
export var SET_ACTIVE_CURRENCY = '[Site-context] Set Active Currency';
export var CURRENCY_CHANGE = '[Site-context] Currency Change';
var LoadCurrencies = /** @class */ (function () {
    function LoadCurrencies() {
        this.type = LOAD_CURRENCIES;
    }
    return LoadCurrencies;
}());
export { LoadCurrencies };
var LoadCurrenciesFail = /** @class */ (function () {
    function LoadCurrenciesFail(payload) {
        this.payload = payload;
        this.type = LOAD_CURRENCIES_FAIL;
    }
    return LoadCurrenciesFail;
}());
export { LoadCurrenciesFail };
var LoadCurrenciesSuccess = /** @class */ (function () {
    function LoadCurrenciesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_CURRENCIES_SUCCESS;
    }
    return LoadCurrenciesSuccess;
}());
export { LoadCurrenciesSuccess };
var SetActiveCurrency = /** @class */ (function () {
    function SetActiveCurrency(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_CURRENCY;
    }
    return SetActiveCurrency;
}());
export { SetActiveCurrency };
var CurrencyChange = /** @class */ (function () {
    function CurrencyChange(payload) {
        this.payload = payload;
        this.type = CURRENCY_CHANGE;
    }
    return CurrencyChange;
}());
export { CurrencyChange };
