// @spartacus/core/esm5/src/site-context/store/actions/languages.action.js
export var LOAD_LANGUAGES = '[Site-context] Load Languages';
export var LOAD_LANGUAGES_FAIL = '[Site-context] Load Languages Fail';
export var LOAD_LANGUAGES_SUCCESS = '[Site-context] Load Languages Success';
export var SET_ACTIVE_LANGUAGE = '[Site-context] Set Active Language';
export var LANGUAGE_CHANGE = '[Site-context] Language Change';
var LoadLanguages = /** @class */ (function () {
    function LoadLanguages() {
        this.type = LOAD_LANGUAGES;
    }
    return LoadLanguages;
}());
export { LoadLanguages };
var LoadLanguagesFail = /** @class */ (function () {
    function LoadLanguagesFail(payload) {
        this.payload = payload;
        this.type = LOAD_LANGUAGES_FAIL;
    }
    return LoadLanguagesFail;
}());
export { LoadLanguagesFail };
var LoadLanguagesSuccess = /** @class */ (function () {
    function LoadLanguagesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_LANGUAGES_SUCCESS;
    }
    return LoadLanguagesSuccess;
}());
export { LoadLanguagesSuccess };
var SetActiveLanguage = /** @class */ (function () {
    function SetActiveLanguage(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_LANGUAGE;
    }
    return SetActiveLanguage;
}());
export { SetActiveLanguage };
var LanguageChange = /** @class */ (function () {
    function LanguageChange(payload) {
        this.payload = payload;
        this.type = LANGUAGE_CHANGE;
    }
    return LanguageChange;
}());
export { LanguageChange };
