// @spartacus/storefront/esm5/cms-components/product/product-list/product-facet-navigation/services/facet.service.js
import { __assign, __decorate } from "tslib";
import { HttpUrlEncodingCodec } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { tap } from 'rxjs/operators';
import { FacetGroupCollapsedState, } from '../facet.model';
import { ProductFacetService } from './product-facet.service';
import * as i0 from "@angular/core";
import * as i1 from "./product-facet.service";
/**
 * Provides access to the facets as well as their UI state. The UI state
 * represents user related changes on the facets, such as expanding or
 * collapsing a facet group or expanding the number of _visible_ facet values.
 */
var FacetService = /** @class */ (function () {
    function FacetService(productFacetService) {
        var _this = this;
        this.productFacetService = productFacetService;
        /**
         * An internal map where we keep the UI state of the facets.
         */
        this.facetState = new Map();
        /**
         * Observes the facets for the given page and configures the initial UI state.
         *
         * Facets are configured on each emission so that we keep the facet UI state.
         * This is mainly done to keep the state during usage of the facet, but also
         * benefitial when the facets are rebuild while using them.
         */
        this.facetList$ = this.productFacetService.facetList$.pipe(tap(function (facetList) {
            facetList.facets.forEach(function (facet) { return _this.initialize(facet); });
        }));
    }
    /**
     * Returns the observed UI state for the facet.
     *
     * The state is initialized using the `initialize` method.
     */
    FacetService.prototype.getState = function (facet) {
        this.initialize(facet);
        return this.facetState.get(facet.name);
    };
    /**
     * Returns the UI state for the facet.
     *
     * The state is initialized using the `initialize` method.
     */
    FacetService.prototype.getStateSnapshot = function (facet) {
        return this.getState(facet).value;
    };
    /**
     * Toggles the facet expanded state. If the expanded state becomes false,
     * the visible values will decrease to the top values only.
     *
     * If the optional value argument is provided the expanded state will be set
     * to this value, regardless of the current `expanded` state.
     */
    FacetService.prototype.toggle = function (facet, isExpanded) {
        var state = this.getStateSnapshot(facet);
        var toggledState = {
            toggled: isExpanded
                ? FacetGroupCollapsedState.COLLAPSED
                : FacetGroupCollapsedState.EXPANDED,
        };
        if (toggledState.toggled === FacetGroupCollapsedState.COLLAPSED) {
            toggledState.maxVisible = state.topVisible;
        }
        this.updateState(facet, toggledState);
    };
    /**
     * Increases the visible values to the maximum values of the facet.
     */
    FacetService.prototype.increaseVisibleValues = function (facet) {
        this.updateState(facet, { maxVisible: facet.values.length });
    };
    /**
     * Decreases the visible values to the topValueCount.
     *
     * The topValueCount defaults to 6, but can be controlled in
     * the backend as well.
     */
    FacetService.prototype.decreaseVisibleValues = function (facet) {
        this.updateState(facet, { maxVisible: facet.topValueCount });
    };
    /**
     * Persists the facet state and initializes the default values for the top
     * and max visible values.
     */
    FacetService.prototype.initialize = function (facet) {
        if (!this.hasState(facet)) {
            this.facetState.set(facet.name, new BehaviorSubject({
                topVisible: facet.topValueCount || 0,
                maxVisible: facet.topValueCount || 0,
            }));
        }
    };
    /**
     * Updates the state of the facet in the local facet map.
     */
    FacetService.prototype.updateState = function (facet, property) {
        var state = __assign(__assign({}, this.getStateSnapshot(facet)), property);
        this.facetState.get(facet.name).next(state);
    };
    FacetService.prototype.hasState = function (facet) {
        return this.facetState.has(facet.name);
    };
    FacetService.prototype.getLinkParams = function (query) {
        return {
            // to avoid encoding issues with facets that have space (' ') in their name,
            // we replace the decoded '+' back to empty space ' '.
            // For more, see https://github.com/SAP/spartacus/issues/7348
            query: new HttpUrlEncodingCodec().decodeValue(query).replace(/\+/g, ' '),
        };
    };
    FacetService.ctorParameters = function () { return [
        { type: ProductFacetService }
    ]; };
    FacetService.ɵprov = i0.ɵɵdefineInjectable({ factory: function FacetService_Factory() { return new FacetService(i0.ɵɵinject(i1.ProductFacetService)); }, token: FacetService, providedIn: "root" });
    FacetService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], FacetService);
    return FacetService;
}());
export { FacetService };
