// @spartacus/storefront/esm5/cms-components/product/product-list/product-facet-navigation/services/product-facet.service.js
import { __assign, __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ActivatedRouterStateSnapshot, Breadcrumb, PageType, ProductSearchPage, RoutingService, } from '@spartacus/core';
import { filter, map, pluck, switchMap } from 'rxjs/operators';
import { ProductListComponentService } from '../../container/product-list-component.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "../../container/product-list-component.service";
/**
 * Provides access to all the facets and active facets for the Product Listing Page.
 */
var ProductFacetService = /** @class */ (function () {
    function ProductFacetService(routing, productListComponentService) {
        var _this = this;
        this.routing = routing;
        this.productListComponentService = productListComponentService;
        this.routeState$ = this.routing
            .getRouterState()
            .pipe(pluck('state'));
        this.searchResult$ = this.routeState$.pipe(switchMap(function (state) {
            return _this.productListComponentService.model$.pipe(filter(function (page) { return _this.filterForPage(state, page); }), map(function (page) { return _this.mapResults(state, page); }));
        }));
        /**
         * Observes the facets and active facets for the given page. The facet data
         * is provided in a `FacetList`.
         */
        this.facetList$ = this.searchResult$.pipe(map(function (result) {
            return ({
                facets: result.facets,
                activeFacets: result.breadcrumbs,
            });
        }));
    }
    /**
     * Filters the current result by verifying if the result is related to the page.
     * This is done to avoid a combination of the next page and the current search results.
     */
    ProductFacetService.prototype.filterForPage = function (state, page) {
        var _a, _b, _c;
        if (state.context.type === PageType.CATEGORY_PAGE) {
            return (((_c = (_b = (_a = page.currentQuery) === null || _a === void 0 ? void 0 : _a.query) === null || _b === void 0 ? void 0 : _b.value) === null || _c === void 0 ? void 0 : _c.indexOf("allCategories:" + state.context.id)) > -1);
        }
        if (state.context.type === PageType.CONTENT_PAGE &&
            state.context.id === 'search') {
            return page.currentQuery.query.value.startsWith(state.params.query + ":");
        }
        return false;
    };
    ProductFacetService.prototype.mapResults = function (state, page) {
        return __assign(__assign({}, page), { breadcrumbs: this.filterBreadcrumbs(page.breadcrumbs, state.params) });
    };
    /**
     * filter breadcrumbs which are not actively selected
     * but coming from the route navigation
     */
    ProductFacetService.prototype.filterBreadcrumbs = function (breadcrumbs, params) {
        var _this = this;
        return breadcrumbs
            ? breadcrumbs.filter(function (breadcrumb) { return !_this.hasBreadcrumb(breadcrumb, params); })
            : [];
    };
    /**
     * Indicates whether the breadcrumb is related to navigation parameters,
     * since either the category or brand code should match those codes.
     */
    ProductFacetService.prototype.hasBreadcrumb = function (breadcrumb, params) {
        return (breadcrumb.facetCode === 'allCategories' &&
            breadcrumb.facetValueCode === params.categoryCode);
    };
    ProductFacetService.ctorParameters = function () { return [
        { type: RoutingService },
        { type: ProductListComponentService }
    ]; };
    ProductFacetService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductFacetService_Factory() { return new ProductFacetService(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.ProductListComponentService)); }, token: ProductFacetService, providedIn: "root" });
    ProductFacetService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ProductFacetService);
    return ProductFacetService;
}());
export { ProductFacetService };
