// @spartacus/storefront/esm5/cms-components/product/product-list/product-facet-navigation/facet/facet.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, HostBinding, Input, QueryList, ViewChild, ViewChildren, } from '@angular/core';
import { ICON_TYPE } from '../../../../../cms-components/misc/icon/icon.model';
import { FocusDirective } from '../../../../../layout/a11y/keyboard-focus/focus.directive';
import { FacetService } from '../services/facet.service';
var FacetComponent = /** @class */ (function () {
    function FacetComponent(facetService, elementRef, cd) {
        this.facetService = facetService;
        this.elementRef = elementRef;
        this.cd = cd;
        /** configurable icon that is used to collapse the facet group  */
        this.expandIcon = ICON_TYPE.EXPAND;
        this.collapseIcon = ICON_TYPE.COLLAPSE;
    }
    Object.defineProperty(FacetComponent.prototype, "facet", {
        get: function () {
            return this._facet;
        },
        set: function (value) {
            this._facet = value;
            this.isMultiSelect = !!value.multiSelect;
            this.state$ = this.facetService.getState(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Handles clicking the heading of the facet group, which means toggling
     * the visibility of the group (collapse / expand) and optionally focusing
     * the group.
     */
    FacetComponent.prototype.toggleGroup = function (event) {
        var _a;
        var host = this.elementRef.nativeElement;
        var isLocked = (_a = this.keyboardFocus) === null || _a === void 0 ? void 0 : _a.isLocked;
        this.facetService.toggle(this.facet, this.isExpanded);
        if (!isLocked || this.isExpanded) {
            host.focus();
            // we stop propagating the event as otherwise the focus on the host will trigger
            // an unlock event from the LockFocus directive.
            event.stopPropagation();
        }
    };
    Object.defineProperty(FacetComponent.prototype, "isExpanded", {
        get: function () {
            return this.values.first.nativeElement.offsetParent !== null;
        },
        enumerable: true,
        configurable: true
    });
    FacetComponent.prototype.openLink = function (event) {
        event.target.click();
        event.preventDefault();
    };
    /**
     * Increases the number of visible values for the facet. This is delegated
     * to `facetService.increaseVisibleValues`.
     */
    FacetComponent.prototype.increaseVisibleValues = function () {
        this.facetService.increaseVisibleValues(this.facet);
    };
    /**
     * Decreases the number of visible values for the facet. This is delegated
     * to `facetService.decreaseVisibleValues`.
     */
    FacetComponent.prototype.decreaseVisibleValues = function () {
        this.facetService.decreaseVisibleValues(this.facet);
    };
    FacetComponent.prototype.getLinkParams = function (value) {
        var _a;
        return this.facetService.getLinkParams((_a = value.query) === null || _a === void 0 ? void 0 : _a.query.value);
    };
    FacetComponent.ctorParameters = function () { return [
        { type: FacetService },
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input()
    ], FacetComponent.prototype, "expandIcon", void 0);
    __decorate([
        Input()
    ], FacetComponent.prototype, "collapseIcon", void 0);
    __decorate([
        HostBinding('class.multi-select')
    ], FacetComponent.prototype, "isMultiSelect", void 0);
    __decorate([
        ViewChildren('facetValue')
    ], FacetComponent.prototype, "values", void 0);
    __decorate([
        ViewChild(FocusDirective)
    ], FacetComponent.prototype, "keyboardFocus", void 0);
    __decorate([
        Input()
    ], FacetComponent.prototype, "facet", null);
    FacetComponent = __decorate([
        Component({
            selector: 'cx-facet',
            template: "<ng-container *ngIf=\"state$ | async as state\">\r\n  <button class=\"heading\" (click)=\"toggleGroup($event)\">\r\n    {{ facet.name }}\r\n    <cx-icon class=\"collapse-icon\" [type]=\"collapseIcon\"></cx-icon>\r\n    <cx-icon class=\"expand-icon\" [type]=\"expandIcon\"></cx-icon>\r\n  </button>\r\n\r\n  <a\r\n    *ngFor=\"let value of facet.values | slice: 0:state.topVisible\"\r\n    #facetValue\r\n    routerLink=\"./\"\r\n    [queryParams]=\"getLinkParams(value)\"\r\n    class=\"value\"\r\n    [class.selected]=\"value.selected\"\r\n    [cxFocus]=\"value.name\"\r\n    (keydown.space)=\"openLink($event)\"\r\n  >\r\n    <span>\r\n      <span class=\"label\">{{ value.name }}</span>\r\n      <span class=\"count\">{{ value.count }}</span>\r\n    </span>\r\n  </a>\r\n\r\n  <div class=\"more\">\r\n    <a\r\n      *ngFor=\"\r\n        let value of facet.values | slice: state.topVisible:state.maxVisible\r\n      \"\r\n      #facetValue\r\n      routerLink=\"./\"\r\n      [queryParams]=\"getLinkParams(value)\"\r\n      class=\"value\"\r\n      [class.selected]=\"value.selected\"\r\n      [cxFocus]=\"value.name\"\r\n      (keydown.space)=\"openLink($event)\"\r\n    >\r\n      <span\r\n        >{{ value.name }}<span class=\"count\">{{ value.count }}</span></span\r\n      >\r\n    </a>\r\n\r\n    <button\r\n      *ngIf=\"state.maxVisible > state.topVisible\"\r\n      (click)=\"decreaseVisibleValues()\"\r\n      class=\"cx-action-link\"\r\n      cxFocus=\"moreorless\"\r\n    >\r\n      {{ 'productList.showLess' | cxTranslate }}\r\n    </button>\r\n\r\n    <button\r\n      *ngIf=\"state.maxVisible > 0 && state.maxVisible < facet.values.length\"\r\n      (click)=\"increaseVisibleValues()\"\r\n      class=\"cx-action-link\"\r\n      cxFocus=\"moreorless\"\r\n    >\r\n      {{ 'productList.showMore' | cxTranslate }}\r\n    </button>\r\n  </div>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], FacetComponent);
    return FacetComponent;
}());
export { FacetComponent };
