// @spartacus/storefront/esm2015/cms-components/content/banner/banner.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig } from '@spartacus/core';
import { GenericLinkModule } from '../../../shared/components/generic-link/generic-link.module';
import { MediaModule } from '../../../shared/components/media/media.module';
import { BannerComponent } from './banner.component';
let BannerModule = class BannerModule {
};
BannerModule = __decorate([
    NgModule({
        imports: [CommonModule, RouterModule, GenericLinkModule, MediaModule],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    SimpleResponsiveBannerComponent: {
                        component: BannerComponent,
                    },
                    BannerComponent: {
                        component: BannerComponent,
                    },
                    SimpleBannerComponent: {
                        component: BannerComponent,
                    },
                },
            }),
        ],
        declarations: [BannerComponent],
        entryComponents: [BannerComponent],
        exports: [BannerComponent],
    })
], BannerModule);
export { BannerModule };
