// @spartacus/storefront/esm5/cms-components/navigation/navigation/navigation.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { GenericLinkModule } from '../../../shared/components/generic-link/generic-link.module';
import { IconModule } from '../../misc/icon/icon.module';
import { NavigationUIComponent } from './navigation-ui.component';
import { NavigationComponent } from './navigation.component';
var NavigationModule = /** @class */ (function () {
    function NavigationModule() {
    }
    NavigationModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule,
                IconModule,
                GenericLinkModule,
                I18nModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        NavigationComponent: {
                            component: NavigationComponent,
                        },
                    },
                }),
            ],
            declarations: [NavigationComponent, NavigationUIComponent],
            entryComponents: [NavigationComponent],
            exports: [NavigationComponent, NavigationUIComponent],
        })
    ], NavigationModule);
    return NavigationModule;
}());
export { NavigationModule };
