// @spartacus/storefront/esm5/cms-components/navigation/navigation/navigation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { map } from 'rxjs/operators';
import { CmsComponentData } from '../../../cms-structure/page/model/cms-component-data';
import { NavigationService } from './navigation.service';
var NavigationComponent = /** @class */ (function () {
    function NavigationComponent(componentData, service) {
        this.componentData = componentData;
        this.service = service;
        this.node$ = this.service.createNavigation(this.componentData.data$);
        this.styleClass$ = this.componentData.data$.pipe(map(function (d) { return d === null || d === void 0 ? void 0 : d.styleClass; }));
    }
    NavigationComponent.ctorParameters = function () { return [
        { type: CmsComponentData },
        { type: NavigationService }
    ]; };
    NavigationComponent = __decorate([
        Component({
            selector: 'cx-navigation',
            template: "<cx-navigation-ui [node]=\"node$ | async\" [ngClass]=\"styleClass$ | async\">\n</cx-navigation-ui>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NavigationComponent);
    return NavigationComponent;
}());
export { NavigationComponent };
