// @spartacus/storefront/esm5/cms-components/navigation/breadcrumb/breadcrumb.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { CmsPageTitleModule, provideDefaultConfig, } from '@spartacus/core';
import { BreadcrumbComponent } from './breadcrumb.component';
var BreadcrumbModule = /** @class */ (function () {
    function BreadcrumbModule() {
    }
    BreadcrumbModule = __decorate([
        NgModule({
            imports: [CommonModule, RouterModule, CmsPageTitleModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        BreadcrumbComponent: {
                            component: BreadcrumbComponent,
                        },
                    },
                }),
            ],
            declarations: [BreadcrumbComponent],
            exports: [BreadcrumbComponent],
            entryComponents: [BreadcrumbComponent],
        })
    ], BreadcrumbModule);
    return BreadcrumbModule;
}());
export { BreadcrumbModule };
