// @spartacus/core/esm2015/src/user/store/selectors/order-return-request.selectors.js
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
import { StateUtils } from '../../../state/utils/index';
const ɵ0 = (state) => state.orderReturn;
export const getOrderReturnRequestState = createSelector(getUserState, ɵ0);
const ɵ1 = (state) => StateUtils.loaderValueSelector(state);
export const getOrderReturnRequest = createSelector(getOrderReturnRequestState, ɵ1);
const ɵ2 = (state) => StateUtils.loaderLoadingSelector(state);
export const getOrderReturnRequestLoading = createSelector(getOrderReturnRequestState, ɵ2);
const ɵ3 = (state) => StateUtils.loaderSuccessSelector(state) &&
    !StateUtils.loaderLoadingSelector(state);
export const getOrderReturnRequestSuccess = createSelector(getOrderReturnRequestState, ɵ3);
const ɵ4 = (state) => state.orderReturnList;
export const getOrderReturnRequestListState = createSelector(getUserState, ɵ4);
const ɵ5 = (state) => StateUtils.loaderValueSelector(state);
export const getOrderReturnRequestList = createSelector(getOrderReturnRequestListState, ɵ5);
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5 };
