// @spartacus/core/esm2015/src/user/store/selectors/regions.selectors.js
import { createSelector } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
const ɵ0 = (state) => state.regions;
export const getRegionsLoaderState = createSelector(getUserState, ɵ0);
const ɵ1 = (state) => {
    return StateUtils.loaderValueSelector(state).entities;
};
export const getAllRegions = createSelector(getRegionsLoaderState, ɵ1);
const ɵ2 = (state) => ({
    loaded: StateUtils.loaderSuccessSelector(state),
    loading: StateUtils.loaderLoadingSelector(state),
    regions: StateUtils.loaderValueSelector(state).entities,
    country: StateUtils.loaderValueSelector(state).country,
});
export const getRegionsDataAndLoading = createSelector(getRegionsLoaderState, ɵ2);
const ɵ3 = (state) => StateUtils.loaderValueSelector(state).country;
export const getRegionsCountry = createSelector(getRegionsLoaderState, ɵ3);
const ɵ4 = (state) => StateUtils.loaderLoadingSelector(state);
export const getRegionsLoading = createSelector(getRegionsLoaderState, ɵ4);
const ɵ5 = (state) => StateUtils.loaderSuccessSelector(state);
export const getRegionsLoaded = createSelector(getRegionsLoaderState, ɵ5);
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5 };
