// @spartacus/core/esm2015/src/anonymous-consents/store/selectors/anonymous-consent-templates.selectors.js
import { createSelector } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { getAnonymousConsentState } from './feature.selector';
const ɵ0 = (state) => state.templates;
export const getAnonymousConsentTemplatesState = createSelector(getAnonymousConsentState, ɵ0);
export const getAnonymousConsentTemplatesValue = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderValueSelector);
export const getAnonymousConsentTemplatesLoading = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderLoadingSelector);
export const getAnonymousConsentTemplatesSuccess = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderSuccessSelector);
export const getAnonymousConsentTemplatesError = createSelector(getAnonymousConsentTemplatesState, StateUtils.loaderErrorSelector);
export const getAnonymousConsentTemplate = (templateCode) => {
    return createSelector(getAnonymousConsentTemplatesValue, (templates) => {
        return templates
            ? templates.find((template) => template.id === templateCode)
            : null;
    });
};
export { ɵ0 };
