// @spartacus/storefront/esm5/layout/a11y/keyboard-focus/persist/persist-focus.directive.js
import { __decorate, __extends } from "tslib";
import { AfterViewInit, Directive, ElementRef, HostBinding, HostListener, OnInit, } from '@angular/core';
import { BlockFocusDirective } from '../block/block-focus.directive';
import { FOCUS_ATTR } from '../keyboard-focus.model';
import { PersistFocusService } from './persist-focus.service';
/**
 * Directive that provides persistence of the focused state. This is useful
 * when a group of focusable elements got refocused or even recreated. That
 * happens often when the DOM is constructed with an `*ngIf` or `*ngFor`.
 *
 * The focus state is based on a configured _key_, which can be passed in the
 * config input, either by using a string primitive or `PersistFocusConfig.key`:
 *
 * ```html
 * <button cxPersistFocus="myKey"></button>
 * <button cxFocus="myKey"></button>
 * <button [cxFocus]="{{key:'myKey'}"></button>
 * ```
 *
 * The focus state can be part of a focus _group_, so that the state is shared
 * and remember for the given group. In order to detect the persistence for a
 * given element, we store the persistence key as a data attribute (`data-cx-focus`):
 *
 * ```html
 * <button data-cx-focus="myKey"></button>
 * ```
 *
 * Other keyboard focus directives can read the key to understand whether the element
 * should retrieve focus.
 *
 */
var PersistFocusDirective = /** @class */ (function (_super) {
    __extends(PersistFocusDirective, _super);
    function PersistFocusDirective(elementRef, service) {
        var _this = _super.call(this, elementRef, service) || this;
        _this.elementRef = elementRef;
        _this.service = service;
        _this.defaultConfig = {};
        /**
         * The persistence key can be passed directly or through the `FocusConfig.key`.
         * While this could be considered a global key, the likeliness of conflicts
         * is very small since the key is cleared when the focus is changed.
         */
        // @Input('cxPersistFocus')
        _this.config = {};
        return _this;
    }
    PersistFocusDirective.prototype.handleFocus = function (event) {
        this.service.set(this.key, this.group);
        event === null || event === void 0 ? void 0 : event.preventDefault();
        event === null || event === void 0 ? void 0 : event.stopPropagation();
    };
    PersistFocusDirective.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
        this.attr = this.key ? this.key : undefined;
    };
    PersistFocusDirective.prototype.setDefaultConfiguration = function () {
        if (typeof this.config === 'string' && this.config !== '') {
            this.config = { key: this.config };
        }
        _super.prototype.setDefaultConfiguration.call(this);
    };
    /**
     * Focus the element explicitly if it was focused before.
     */
    PersistFocusDirective.prototype.ngAfterViewInit = function () {
        if (this.isPersisted) {
            this.host.focus({ preventScroll: true });
        }
    };
    Object.defineProperty(PersistFocusDirective.prototype, "isPersisted", {
        get: function () {
            return !!this.key && this.service.get(this.group) === this.key;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PersistFocusDirective.prototype, "key", {
        /**
         * Returns the key for the host element, which is used to persist the
         * focus state. This is useful in cases where the DOM is rebuild.
         */
        get: function () {
            var _a;
            return (_a = this.config) === null || _a === void 0 ? void 0 : _a.key;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PersistFocusDirective.prototype, "group", {
        /**
         * returns the persistence group (if any) for the focusable elements.
         */
        get: function () {
            return this.service.getPersistenceGroup(this.host, this.config);
        },
        enumerable: true,
        configurable: true
    });
    PersistFocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: PersistFocusService }
    ]; };
    __decorate([
        HostBinding("attr." + FOCUS_ATTR)
    ], PersistFocusDirective.prototype, "attr", void 0);
    __decorate([
        HostListener('focus', ['$event'])
    ], PersistFocusDirective.prototype, "handleFocus", null);
    PersistFocusDirective = __decorate([
        Directive() // selector: '[cxPersistFocus]',
    ], PersistFocusDirective);
    return PersistFocusDirective;
}(BlockFocusDirective));
export { PersistFocusDirective };
