// @spartacus/storefront/esm5/cms-components/product/product-list/product-facet-navigation/product-facet-navigation.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ConfigModule, I18nModule } from '@spartacus/core';
import { IconModule } from '../../../misc/icon/icon.module';
import { ActiveFacetsModule } from './active-facets/active-facets.module';
import { FacetListModule } from './facet-list/facet-list.module';
import { ProductFacetNavigationComponent } from './product-facet-navigation.component';
var ProductFacetNavigationModule = /** @class */ (function () {
    function ProductFacetNavigationModule() {
    }
    ProductFacetNavigationModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FacetListModule,
                ActiveFacetsModule,
                IconModule,
                I18nModule,
                ConfigModule.withConfig({
                    cmsComponents: {
                        ProductRefinementComponent: {
                            component: ProductFacetNavigationComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductFacetNavigationComponent],
            exports: [ProductFacetNavigationComponent],
        })
    ], ProductFacetNavigationModule);
    return ProductFacetNavigationModule;
}());
export { ProductFacetNavigationModule };
