// @spartacus/core/esm5/src/config/config-validator/config-validator.module.js
import { __decorate } from "tslib";
import { APP_INITIALIZER, isDevMode, NgModule, Optional, } from '@angular/core';
import { ConfigInitializerService } from '../config-initializer/config-initializer.service';
import { ConfigValidatorToken, validateConfig, } from './config-validator';
export function configValidatorFactory(configInitializer, validators) {
    var validate = function () {
        if (isDevMode()) {
            configInitializer
                .getStableConfig()
                .then(function (config) { return validateConfig(config, validators || []); });
        }
    };
    return validate;
}
/**
 * Should stay private in 1.x
 * as forRoot() is used internally by ConfigInitializerModule
 *
 * issue: #5279
 */
var ConfigValidatorModule = /** @class */ (function () {
    function ConfigValidatorModule() {
    }
    ConfigValidatorModule_1 = ConfigValidatorModule;
    ConfigValidatorModule.forRoot = function () {
        return {
            ngModule: ConfigValidatorModule_1,
            providers: [
                {
                    provide: APP_INITIALIZER,
                    multi: true,
                    useFactory: configValidatorFactory,
                    deps: [
                        ConfigInitializerService,
                        [new Optional(), ConfigValidatorToken],
                    ],
                },
            ],
        };
    };
    var ConfigValidatorModule_1;
    ConfigValidatorModule = ConfigValidatorModule_1 = __decorate([
        NgModule()
    ], ConfigValidatorModule);
    return ConfigValidatorModule;
}());
export { ConfigValidatorModule };
