// @spartacus/core/esm5/src/global-message/http-interceptors/handlers/bad-request/bad-request.handler.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { isCartNotFoundError } from '../../../../cart/utils/utils';
import { GlobalMessageType } from '../../../models/global-message.model';
import { HttpResponseStatus } from '../../../models/response-status.model';
import { HttpErrorHandler } from '../http-error.handler';
import * as i0 from "@angular/core";
import * as i1 from "../../../facade/global-message.service";
var OAUTH_ENDPOINT = '/authorizationserver/oauth/token';
var BadRequestHandler = /** @class */ (function (_super) {
    __extends(BadRequestHandler, _super);
    function BadRequestHandler() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.responseStatus = HttpResponseStatus.BAD_REQUEST;
        return _this;
    }
    BadRequestHandler.prototype.handleError = function (request, response) {
        this.handleBadPassword(request, response);
        this.handleBadLoginResponse(request, response);
        this.handleBadCartRequest(request, response);
        this.handleValidationError(request, response);
        this.handleVoucherOperationError(request, response);
    };
    BadRequestHandler.prototype.handleBadPassword = function (request, response) {
        var _a, _b, _c;
        if (((_a = response.url) === null || _a === void 0 ? void 0 : _a.includes(OAUTH_ENDPOINT)) &&
            ((_b = response.error) === null || _b === void 0 ? void 0 : _b.error) === 'invalid_grant' &&
            ((_c = request.body) === null || _c === void 0 ? void 0 : _c.get('grant_type')) === 'password') {
            this.globalMessageService.add({
                key: 'httpHandlers.badRequestPleaseLoginAgain',
                params: {
                    errorMessage: response.error.error_description || response.message || '',
                },
            }, GlobalMessageType.MSG_TYPE_ERROR);
            this.globalMessageService.remove(GlobalMessageType.MSG_TYPE_CONFIRMATION);
        }
    };
    BadRequestHandler.prototype.handleBadLoginResponse = function (_request, response) {
        var _this = this;
        this.getErrors(response)
            .filter(function (error) { return error.type === 'PasswordMismatchError'; })
            .forEach(function () {
            _this.globalMessageService.add({ key: 'httpHandlers.badRequestOldPasswordIncorrect' }, GlobalMessageType.MSG_TYPE_ERROR);
        });
    };
    BadRequestHandler.prototype.handleValidationError = function (_request, response) {
        var _this = this;
        this.getErrors(response)
            .filter(function (e) { return e.type === 'ValidationError'; })
            .forEach(function (error) {
            _this.globalMessageService.add({
                key: "httpHandlers.validationErrors." + error.reason + "." + error.subject,
            }, GlobalMessageType.MSG_TYPE_ERROR);
        });
    };
    BadRequestHandler.prototype.handleBadCartRequest = function (_request, response) {
        var _this = this;
        this.getErrors(response)
            .filter(function (e) { return isCartNotFoundError(e); })
            .forEach(function () {
            _this.globalMessageService.add({ key: 'httpHandlers.cartNotFound' }, GlobalMessageType.MSG_TYPE_ERROR);
        });
    };
    BadRequestHandler.prototype.handleVoucherOperationError = function (_request, response) {
        var _this = this;
        this.getErrors(response)
            .filter(function (e) {
            return e.message === 'coupon.invalid.code.provided' &&
                e.type === 'VoucherOperationError';
        })
            .forEach(function () {
            _this.globalMessageService.add({ key: 'httpHandlers.invalidCodeProvided' }, GlobalMessageType.MSG_TYPE_ERROR);
        });
    };
    BadRequestHandler.prototype.getErrors = function (response) {
        var _a;
        return (((_a = response.error) === null || _a === void 0 ? void 0 : _a.errors) || []).filter(function (error) { return error.type !== 'JaloObjectNoLongerValidError'; });
    };
    BadRequestHandler.prototype.getPriority = function () {
        return -10 /* LOW */;
    };
    BadRequestHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function BadRequestHandler_Factory() { return new BadRequestHandler(i0.ɵɵinject(i1.GlobalMessageService)); }, token: BadRequestHandler, providedIn: "root" });
    BadRequestHandler = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], BadRequestHandler);
    return BadRequestHandler;
}(HttpErrorHandler));
export { BadRequestHandler };
