// @spartacus/core/esm5/src/state/event/state-event.service.js
import { __decorate, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { ofType } from '@ngrx/effects';
import { ActionsSubject } from '@ngrx/store';
import { map } from 'rxjs/operators';
import { EventService } from '../../event/event.service';
import { createFrom } from '../../util/create-from';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../event/event.service";
/**
 * Registers streams of ngrx actions as events source streams
 */
var StateEventService = /** @class */ (function () {
    function StateEventService(actionsSubject, eventService) {
        this.actionsSubject = actionsSubject;
        this.eventService = eventService;
    }
    /**
     * Registers an event source stream of specific events
     * mapped from a given action type.
     *
     * @param mapping mapping from action to event
     *
     * @returns a teardown function that unregisters the event source
     */
    StateEventService.prototype.register = function (mapping) {
        return this.eventService.register(mapping.event, this.getFromAction(mapping));
    };
    /**
     * Returns a stream of specific events mapped from a specific action.
     * @param mapping mapping from action to event
     */
    StateEventService.prototype.getFromAction = function (mapping) {
        var _this = this;
        return this.actionsSubject
            .pipe(ofType.apply(void 0, __spread([].concat(mapping.action))))
            .pipe(map(function (action) {
            return _this.createEvent(action, mapping.event, mapping.factory);
        }));
    };
    /**
     * Creates an event instance for given class out from the action object.
     * Unless the `factory` parameter is given, the action's `payload` is used
     * as the argument for the event's constructor.
     *
     * @param action instance of an Action
     * @param mapping mapping from action to event
     * @param factory optional function getting an action instance and returning an event instance
     *
     * @returns instance of an Event
     */
    StateEventService.prototype.createEvent = function (action, eventType, factory) {
        var _a;
        return factory
            ? factory(action)
            : createFrom(eventType, (_a = action.payload) !== null && _a !== void 0 ? _a : {});
    };
    StateEventService.ctorParameters = function () { return [
        { type: ActionsSubject },
        { type: EventService }
    ]; };
    StateEventService.ɵprov = i0.ɵɵdefineInjectable({ factory: function StateEventService_Factory() { return new StateEventService(i0.ɵɵinject(i1.ActionsSubject), i0.ɵɵinject(i2.EventService)); }, token: StateEventService, providedIn: "root" });
    StateEventService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], StateEventService);
    return StateEventService;
}());
export { StateEventService };
