// @spartacus/storefront/esm2015/cms-components/order-confirmation/components/order-confirmation-items/order-confirmation-items.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, } from '@angular/core';
import { CheckoutService, Order, PromotionResult, PromotionLocation, } from '@spartacus/core';
import { PromotionService } from '../../../../shared/services/promotion/promotion.service';
let OrderConfirmationItemsComponent = class OrderConfirmationItemsComponent {
    constructor(checkoutService, promotionService) {
        this.checkoutService = checkoutService;
        this.promotionService = promotionService;
        this.promotionLocation = PromotionLocation.Checkout;
    }
    ngOnInit() {
        this.order$ = this.checkoutService.getOrderDetails();
        this.orderPromotions$ = this.promotionService.getOrderPromotions(this.promotionLocation);
    }
    ngOnDestroy() {
        this.checkoutService.clearCheckoutData();
    }
};
OrderConfirmationItemsComponent.ctorParameters = () => [
    { type: CheckoutService },
    { type: PromotionService }
];
OrderConfirmationItemsComponent = __decorate([
    Component({
        selector: 'cx-order-confirmation-items',
        template: "<div class=\"cx-order-items container\" *ngIf=\"order$ | async as order\">\r\n  <h4 class=\"cx-order-items-header\">\r\n    {{ 'checkoutOrderConfirmation.orderItems' | cxTranslate }}\r\n  </h4>\r\n\r\n  <ng-container *ngIf=\"orderPromotions$ | async as orderPromotions\">\r\n    <cx-promotions [promotions]=\"orderPromotions\"></cx-promotions>\r\n  </ng-container>\r\n\r\n  <cx-cart-item-list\r\n    [items]=\"order.entries\"\r\n    [readonly]=\"true\"\r\n    [promotionLocation]=\"promotionLocation\"\r\n  ></cx-cart-item-list>\r\n</div>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], OrderConfirmationItemsComponent);
export { OrderConfirmationItemsComponent };
