// @spartacus/storefront/esm5/layout/a11y/keyboard-focus/base/base-focus.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, HostBinding, Input, OnInit, } from '@angular/core';
import { BaseFocusService } from './base-focus.service';
/**
 * Abstract directive that provides a common interface for all focus directives:
 * - Block Focus
 * - Persist Focus
 * - Escape Focus
 * - Auto Focus
 * - Tab Focus
 * - Trap Focus
 * - Lock Focus
 */
var BaseFocusDirective = /** @class */ (function () {
    function BaseFocusDirective(elementRef, service) {
        this.elementRef = elementRef;
        this.service = service;
        /**
         * A default config can be provided for each directive if a specific focus directive
         * is used directly. i.e. `<div cxAutoFocus></div>`
         */
        this.defaultConfig = {};
    }
    BaseFocusDirective.prototype.ngOnInit = function () {
        this.setDefaultConfiguration();
        this.requiredTabindex = -1;
    };
    /**
     * Override the (input) config if it undefined or an empty string, with the
     * `defaultConfig`. The `defaultConfig` might be specified for each directive
     * differently. If a specific directive is used (i.e. `cxAutoFocus`), the
     * specific (inherited) defaultConfig will be used.
     */
    BaseFocusDirective.prototype.setDefaultConfiguration = function () {
        if ((!this.config || this.config === '') && this.defaultConfig) {
            this.config = this.defaultConfig;
        }
    };
    Object.defineProperty(BaseFocusDirective.prototype, "host", {
        /**
         * Helper method to return the host element for the directive
         * given by the `elementRef`.
         */
        get: function () {
            return this.elementRef.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFocusDirective.prototype, "requiredTabindex", {
        /**
         * Force a tabindex on the host element if it is _requried_ to make the element
         * focusable. If the element is focusable by nature or by a given tabindex, the
         * `tabindex` is not applied.
         *
         * Buttons, active links, etc. do no need an explicit tabindex to receive focus.
         */
        set: function (tabindex) {
            if (this.requiresExplicitTabIndex) {
                this.tabindex = tabindex;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFocusDirective.prototype, "requiresExplicitTabIndex", {
        /**
         * Returns true if the host element does not have a tabindex defined
         * and it also doesn't get focus by browsers nature (i.e. button or
         * active link).
         */
        get: function () {
            return (this.tabindex === undefined &&
                ['button', 'input', 'select', 'textarea'].indexOf(this.host.tagName.toLowerCase()) === -1 &&
                !(this.host.tagName === 'A' &&
                    (this.host.hasAttribute('href') || this.host.hasAttribute('routerlink'))));
        },
        enumerable: true,
        configurable: true
    });
    BaseFocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: BaseFocusService }
    ]; };
    __decorate([
        Input(), HostBinding('attr.tabindex')
    ], BaseFocusDirective.prototype, "tabindex", void 0);
    BaseFocusDirective = __decorate([
        Directive()
    ], BaseFocusDirective);
    return BaseFocusDirective;
}());
export { BaseFocusDirective };
