// @spartacus/storefront/esm5/cms-components/myaccount/order/amend-order/returns/return-order-confirmation/return-order-confirmation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { tap } from 'rxjs/operators';
import { OrderAmendService } from '../../amend-order.service';
var ReturnOrderConfirmationComponent = /** @class */ (function () {
    function ReturnOrderConfirmationComponent(orderAmendService) {
        var _this = this;
        this.orderAmendService = orderAmendService;
        this.form$ = this.orderAmendService
            .getForm()
            .pipe(tap(function (form) { return (_this.orderCode = form.value.orderCode); }));
        this.entries$ = this.orderAmendService.getAmendedEntries();
    }
    ReturnOrderConfirmationComponent.prototype.submit = function (form) {
        form.disable();
        this.orderAmendService.save();
    };
    ReturnOrderConfirmationComponent.ctorParameters = function () { return [
        { type: OrderAmendService }
    ]; };
    ReturnOrderConfirmationComponent = __decorate([
        Component({
            selector: 'cx-return-order-confirmation',
            template: "<form\r\n  *ngIf=\"form$ | async as form\"\r\n  [formGroup]=\"form\"\r\n  (ngSubmit)=\"submit(form)\"\r\n>\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <cx-amend-order-items\r\n    *ngIf=\"entries$ | async as entries\"\r\n    [entries]=\"entries\"\r\n    [isConfirmation]=\"true\"\r\n  >\r\n  </cx-amend-order-items>\r\n\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <ng-template #actions>\r\n    <cx-amend-order-actions\r\n      *ngIf=\"orderCode\"\r\n      [orderCode]=\"orderCode\"\r\n      [amendOrderForm]=\"form\"\r\n      backRoute=\"orderReturn\"\r\n    ></cx-amend-order-actions>\r\n  </ng-template>\r\n</form>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ReturnOrderConfirmationComponent);
    return ReturnOrderConfirmationComponent;
}());
export { ReturnOrderConfirmationComponent };
