// @spartacus/core/esm5/src/product/store/reducers/index.js
import { InjectionToken } from '@angular/core';
import { SiteContextActions } from '../../../site-context/store/actions/index';
import { PRODUCT_DETAIL_ENTITY } from '../product-state';
import * as fromProductReferences from './product-references.reducer';
import * as fromProductReviews from './product-reviews.reducer';
import * as fromProductsSearch from './product-search.reducer';
import { entityScopedLoaderReducer } from '../../../state/utils/scoped-loader/entity-scoped-loader.reducer';
export function getReducers() {
    return {
        search: fromProductsSearch.reducer,
        details: entityScopedLoaderReducer(PRODUCT_DETAIL_ENTITY),
        reviews: fromProductReviews.reducer,
        references: fromProductReferences.reducer,
    };
}
export var reducerToken = new InjectionToken('ProductReducers');
export var reducerProvider = {
    provide: reducerToken,
    useFactory: getReducers,
};
export function clearProductsState(reducer) {
    return function (state, action) {
        if (action.type === SiteContextActions.CURRENCY_CHANGE ||
            action.type === SiteContextActions.LANGUAGE_CHANGE) {
            state = undefined;
        }
        return reducer(state, action);
    };
}
export var metaReducers = [clearProductsState];
