// @spartacus/core/esm5/src/product/store/reducers/product-references.reducer.js
import { __assign, __read, __spread } from "tslib";
import { ProductActions } from '../actions/index';
export var initialState = {
    productCode: '',
    list: [],
};
export function reducer(state, action) {
    if (state === void 0) { state = initialState; }
    switch (action.type) {
        case ProductActions.LOAD_PRODUCT_REFERENCES_SUCCESS: {
            var productCode = action.payload.productCode;
            var list = action.payload.list;
            return __assign(__assign({}, state), { list: __spread(state.list, (list ? list : [])).reduce(function (productReferences, productReference) {
                    if (!productReferences.some(function (obj) {
                        return obj.referenceType === productReference.referenceType &&
                            obj.target.code === productReference.target.code;
                    })) {
                        productReferences.push(productReference);
                    }
                    return productReferences;
                }, []), productCode: productCode });
        }
        case ProductActions.CLEAN_PRODUCT_REFERENCES: {
            return initialState;
        }
    }
    return state;
}
export var getProductReferenceList = function (state) { return state.list; };
export var getProductReferenceProductCode = function (state) { return state.productCode; };
