// @spartacus/core/esm5/src/product/store/selectors/product.selectors.js
import { createSelector } from '@ngrx/store';
import { StateUtils } from '../../../state/utils/index';
import { getProductsState } from './feature.selector';
var ɵ0 = function (state) { return state.details; };
export var getProductState = createSelector(getProductsState, ɵ0);
export var getSelectedProductStateFactory = function (code, scope) {
    if (scope === void 0) { scope = ''; }
    return createSelector(getProductState, function (details) {
        return StateUtils.entityLoaderStateSelector(details, code)[scope] ||
            StateUtils.initialLoaderState;
    });
};
export var getSelectedProductFactory = function (code, scope) {
    if (scope === void 0) { scope = ''; }
    return createSelector(getSelectedProductStateFactory(code, scope), function (productState) { return StateUtils.loaderValueSelector(productState); });
};
export var getSelectedProductLoadingFactory = function (code, scope) {
    if (scope === void 0) { scope = ''; }
    return createSelector(getSelectedProductStateFactory(code, scope), function (productState) { return StateUtils.loaderLoadingSelector(productState); });
};
export var getSelectedProductSuccessFactory = function (code, scope) {
    if (scope === void 0) { scope = ''; }
    return createSelector(getSelectedProductStateFactory(code, scope), function (productState) { return StateUtils.loaderSuccessSelector(productState); });
};
export var getSelectedProductErrorFactory = function (code, scope) {
    if (scope === void 0) { scope = ''; }
    return createSelector(getSelectedProductStateFactory(code, scope), function (productState) { return StateUtils.loaderErrorSelector(productState); });
};
var ɵ1 = function (details) {
    return Object.keys(details.entities);
};
export var getAllProductCodes = createSelector(getProductState, ɵ1);
export { ɵ0, ɵ1 };
