// @spartacus/storefront/esm5/cms-structure/page/component/page-component.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentWrapperDirective } from './component-wrapper.directive';
import { ComponentHandler } from './handlers/component-handler';
import { DefaultComponentHandler } from './handlers/default-component.handler';
import { LazyComponentHandler } from './handlers/lazy-component.handler';
var PageComponentModule = /** @class */ (function () {
    function PageComponentModule() {
    }
    PageComponentModule = __decorate([
        NgModule({
            imports: [CommonModule],
            providers: [
                {
                    provide: ComponentHandler,
                    useExisting: DefaultComponentHandler,
                    multi: true,
                },
                {
                    provide: ComponentHandler,
                    useExisting: LazyComponentHandler,
                    multi: true,
                },
            ],
            declarations: [ComponentWrapperDirective],
            exports: [ComponentWrapperDirective],
        })
    ], PageComponentModule);
    return PageComponentModule;
}());
export { PageComponentModule };
