// @spartacus/storefront/esm5/cms-components/product/product-variants/variant-size-selector/variant-size-selector.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Product, RoutingService, BaseOption, VariantQualifier, VariantOptionQualifier, ProductService, ProductScope, } from '@spartacus/core';
import { filter, take } from 'rxjs/operators';
var VariantSizeSelectorComponent = /** @class */ (function () {
    function VariantSizeSelectorComponent(productService, routingService) {
        this.productService = productService;
        this.routingService = routingService;
    }
    VariantSizeSelectorComponent.prototype.changeSize = function (code) {
        var _this = this;
        if (code) {
            this.productService
                .get(code, ProductScope.LIST)
                .pipe(
            // below call might looks redundant but in fact this data is going to be loaded anyways
            // we're just calling it earlier and storing
            filter(Boolean), take(1))
                .subscribe(function (product) {
                _this.routingService.go({
                    cxRoute: 'product',
                    params: product,
                });
            });
        }
        return null;
    };
    VariantSizeSelectorComponent.prototype.getVariantOptionValue = function (qualifiers) {
        var obj = qualifiers.find(function (q) { return q.qualifier === VariantQualifier.SIZE; });
        return obj ? obj.value : '';
    };
    VariantSizeSelectorComponent.ctorParameters = function () { return [
        { type: ProductService },
        { type: RoutingService }
    ]; };
    __decorate([
        Input()
    ], VariantSizeSelectorComponent.prototype, "product", void 0);
    __decorate([
        Input()
    ], VariantSizeSelectorComponent.prototype, "variants", void 0);
    VariantSizeSelectorComponent = __decorate([
        Component({
            selector: 'cx-variant-size-selector',
            template: "<ng-container>\n  <div class=\"variant-selector\">\n    <div class=\"variant-name\">{{ 'variant.size' | cxTranslate }}:</div>\n    <select\n      (change)=\"changeSize($event.target.value)\"\n      class=\"form-control variant-select\"\n    >\n      <option\n        *ngFor=\"let v of variants?.options\"\n        value=\"{{ v.code }}\"\n        [selected]=\"v.code === product?.code\"\n        >{{ getVariantOptionValue(v.variantOptionQualifiers) }}</option\n      >\n    </select>\n    <a\n      href=\"#\"\n      class=\"size-guide\"\n      title=\"{{ 'variant.sizeGuideLabel' | cxTranslate }}\"\n    >\n      {{ 'variant.sizeGuideLabel' | cxTranslate }}\n    </a>\n  </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], VariantSizeSelectorComponent);
    return VariantSizeSelectorComponent;
}());
export { VariantSizeSelectorComponent };
