// @spartacus/storefront/esm2015/layout/a11y/keyboard-focus/tab/tab-focus.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, HostListener } from '@angular/core';
import { AutoFocusDirective } from '../autofocus/auto-focus.directive';
import { TabFocusService } from './tab-focus.service';
/**
 * Directive to move the focus of ("locked") child elements. This is useful
 * for a nested list of tabs, carousel slides or any group of elements that
 * requires horizontal navigation.
 */
let TabFocusDirective = class TabFocusDirective extends AutoFocusDirective {
    constructor(elementRef, service) {
        super(elementRef, service);
        this.elementRef = elementRef;
        this.service = service;
        /** `tab` defaults to true if the directive `cxTabFocus` is used. */
        this.defaultConfig = { tab: true };
        // @Input('cxTabFocus')
        this.config = {};
    }
    handleNextTab(event) {
        var _a;
        if ((_a = this.config) === null || _a === void 0 ? void 0 : _a.tab) {
            this.service.moveTab(this.host, this.config, 1 /* NEXT */, event);
        }
    }
    handlePreviousTab(event) {
        var _a;
        if ((_a = this.config) === null || _a === void 0 ? void 0 : _a.tab) {
            this.service.moveTab(this.host, this.config, -1 /* PREV */, event);
        }
    }
};
TabFocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: TabFocusService }
];
__decorate([
    HostListener('keydown.arrowRight', ['$event'])
], TabFocusDirective.prototype, "handleNextTab", null);
__decorate([
    HostListener('keydown.arrowLeft', ['$event'])
], TabFocusDirective.prototype, "handlePreviousTab", null);
TabFocusDirective = __decorate([
    Directive() // selector: '[cxTabFocus]'
], TabFocusDirective);
export { TabFocusDirective };
