// @spartacus/storefront/esm5/cms-components/anonymous-consent-management/open-dialog/anonymous-consent-open-dialog.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, ViewChild, ViewContainerRef, } from '@angular/core';
import { take } from 'rxjs/operators';
import { AnonymousConsentLaunchDialogService } from '../anonymous-consent-launch-dialog.service';
var AnonymousConsentOpenDialogComponent = /** @class */ (function () {
    function AnonymousConsentOpenDialogComponent(vcr, anonymousConsentLaunchDialogService) {
        this.vcr = vcr;
        this.anonymousConsentLaunchDialogService = anonymousConsentLaunchDialogService;
    }
    AnonymousConsentOpenDialogComponent.prototype.openDialog = function () {
        var dialog = this.anonymousConsentLaunchDialogService.openDialog(this.openElement, this.vcr);
        if (dialog) {
            dialog.pipe(take(1)).subscribe();
        }
    };
    AnonymousConsentOpenDialogComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: AnonymousConsentLaunchDialogService }
    ]; };
    __decorate([
        ViewChild('open')
    ], AnonymousConsentOpenDialogComponent.prototype, "openElement", void 0);
    AnonymousConsentOpenDialogComponent = __decorate([
        Component({
            selector: 'cx-anonymous-consent-open-dialog',
            template: "<button #open class=\"btn btn-link\" (click)=\"openDialog()\">\r\n  {{ 'anonymousConsents.preferences' | cxTranslate }}\r\n</button>\r\n"
        })
    ], AnonymousConsentOpenDialogComponent);
    return AnonymousConsentOpenDialogComponent;
}());
export { AnonymousConsentOpenDialogComponent };
