// @spartacus/core/esm5/src/store-finder/store/actions/find-stores.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { STORE_FINDER_DATA } from '../store-finder-state';
export var FIND_STORES_ON_HOLD = '[StoreFinder] On Hold';
export var FIND_STORES = '[StoreFinder] Find Stores';
export var FIND_STORES_FAIL = '[StoreFinder] Find Stores Fail';
export var FIND_STORES_SUCCESS = '[StoreFinder] Find Stores Success';
export var FIND_STORE_BY_ID = '[StoreFinder] Find a Store by Id';
export var FIND_STORE_BY_ID_FAIL = '[StoreFinder] Find a Store by Id Fail';
export var FIND_STORE_BY_ID_SUCCESS = '[StoreFinder] Find a Store by Id Success';
var FindStoresOnHold = /** @class */ (function (_super) {
    __extends(FindStoresOnHold, _super);
    function FindStoresOnHold() {
        var _this = _super.call(this, STORE_FINDER_DATA) || this;
        _this.type = FIND_STORES_ON_HOLD;
        return _this;
    }
    return FindStoresOnHold;
}(StateUtils.LoaderLoadAction));
export { FindStoresOnHold };
var FindStores = /** @class */ (function (_super) {
    __extends(FindStores, _super);
    function FindStores(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA) || this;
        _this.payload = payload;
        _this.type = FIND_STORES;
        return _this;
    }
    return FindStores;
}(StateUtils.LoaderLoadAction));
export { FindStores };
var FindStoresFail = /** @class */ (function (_super) {
    __extends(FindStoresFail, _super);
    function FindStoresFail(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA, payload) || this;
        _this.payload = payload;
        _this.type = FIND_STORES_FAIL;
        return _this;
    }
    return FindStoresFail;
}(StateUtils.LoaderFailAction));
export { FindStoresFail };
var FindStoresSuccess = /** @class */ (function (_super) {
    __extends(FindStoresSuccess, _super);
    function FindStoresSuccess(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA) || this;
        _this.payload = payload;
        _this.type = FIND_STORES_SUCCESS;
        return _this;
    }
    return FindStoresSuccess;
}(StateUtils.LoaderSuccessAction));
export { FindStoresSuccess };
var FindStoreById = /** @class */ (function (_super) {
    __extends(FindStoreById, _super);
    function FindStoreById(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA) || this;
        _this.payload = payload;
        _this.type = FIND_STORE_BY_ID;
        return _this;
    }
    return FindStoreById;
}(StateUtils.LoaderLoadAction));
export { FindStoreById };
var FindStoreByIdFail = /** @class */ (function (_super) {
    __extends(FindStoreByIdFail, _super);
    function FindStoreByIdFail(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA, payload) || this;
        _this.payload = payload;
        _this.type = FIND_STORE_BY_ID_FAIL;
        return _this;
    }
    return FindStoreByIdFail;
}(StateUtils.LoaderFailAction));
export { FindStoreByIdFail };
var FindStoreByIdSuccess = /** @class */ (function (_super) {
    __extends(FindStoreByIdSuccess, _super);
    function FindStoreByIdSuccess(payload) {
        var _this = _super.call(this, STORE_FINDER_DATA) || this;
        _this.payload = payload;
        _this.type = FIND_STORE_BY_ID_SUCCESS;
        return _this;
    }
    return FindStoreByIdSuccess;
}(StateUtils.LoaderSuccessAction));
export { FindStoreByIdSuccess };
