// @spartacus/storefront/esm5/cms-components/product/product-variants/variant-color-selector/variant-color-selector.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BaseOption, Product, RoutingService, VariantOptionQualifier, VariantQualifier, } from '@spartacus/core';
var VariantColorSelectorComponent = /** @class */ (function () {
    function VariantColorSelectorComponent(routingService) {
        this.routingService = routingService;
    }
    VariantColorSelectorComponent.prototype.changeColor = function (code, name) {
        if (code) {
            this.routingService.go({
                cxRoute: 'product',
                params: { code: code, name: name },
            });
        }
        return null;
    };
    VariantColorSelectorComponent.prototype.getVariantOptionValue = function (qualifiers) {
        var obj = qualifiers.find(function (q) { return q.qualifier === VariantQualifier.COLOR; });
        return obj ? obj.value : '';
    };
    VariantColorSelectorComponent.ctorParameters = function () { return [
        { type: RoutingService }
    ]; };
    __decorate([
        Input()
    ], VariantColorSelectorComponent.prototype, "product", void 0);
    __decorate([
        Input()
    ], VariantColorSelectorComponent.prototype, "variants", void 0);
    VariantColorSelectorComponent = __decorate([
        Component({
            selector: 'cx-variant-color-selector',
            template: "<ng-container>\n  <div class=\"variant-selector\">\n    <div class=\"variant-name\">{{ 'variant.color' | cxTranslate }}:</div>\n\n    <select\n      (change)=\"changeColor($event.target.value, product?.name)\"\n      class=\"form-control variant-select\"\n    >\n      <option\n        *ngFor=\"let v of variants?.options\"\n        value=\"{{ v.code }}\"\n        [selected]=\"v.code === product?.code\"\n        >{{ getVariantOptionValue(v.variantOptionQualifiers) }}</option\n      >\n    </select>\n  </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], VariantColorSelectorComponent);
    return VariantColorSelectorComponent;
}());
export { VariantColorSelectorComponent };
