// @spartacus/storefront/esm5/layout/a11y/skip-link/directive/skip-link.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Input, OnDestroy, OnInit } from '@angular/core';
import { SkipLinkService } from '../service/skip-link.service';
var SkipLinkDirective = /** @class */ (function () {
    function SkipLinkDirective(elementRef, skipLinkService) {
        this.elementRef = elementRef;
        this.skipLinkService = skipLinkService;
    }
    SkipLinkDirective.prototype.ngOnInit = function () {
        this.skipLinkService.add(this.cxSkipLink, this.elementRef.nativeElement);
    };
    SkipLinkDirective.prototype.ngOnDestroy = function () {
        this.skipLinkService.remove(this.cxSkipLink);
    };
    SkipLinkDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: SkipLinkService }
    ]; };
    __decorate([
        Input()
    ], SkipLinkDirective.prototype, "cxSkipLink", void 0);
    SkipLinkDirective = __decorate([
        Directive({
            selector: '[cxSkipLink]',
        })
    ], SkipLinkDirective);
    return SkipLinkDirective;
}());
export { SkipLinkDirective };
