// @spartacus/storefront/esm5/cms-components/cart/cart-coupon/applied-coupons/applied-coupons.component.js
import { __decorate } from "tslib";
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { Voucher, CartVoucherService } from '@spartacus/core';
import { ICON_TYPE } from '../../../../cms-components/misc/icon/icon.model';
var AppliedCouponsComponent = /** @class */ (function () {
    function AppliedCouponsComponent(cartVoucherService) {
        this.cartVoucherService = cartVoucherService;
        this.cartIsLoading = false;
        this.isReadOnly = false;
        this.iconTypes = ICON_TYPE;
    }
    Object.defineProperty(AppliedCouponsComponent.prototype, "sortedVouchers", {
        get: function () {
            this.vouchers = this.vouchers || [];
            return this.vouchers.slice().sort(function (a, b) {
                return a.code.localeCompare(b.code);
            });
        },
        enumerable: true,
        configurable: true
    });
    AppliedCouponsComponent.prototype.removeVoucher = function (voucherId) {
        this.cartVoucherService.removeVoucher(voucherId);
    };
    AppliedCouponsComponent.ctorParameters = function () { return [
        { type: CartVoucherService }
    ]; };
    __decorate([
        Input()
    ], AppliedCouponsComponent.prototype, "vouchers", void 0);
    __decorate([
        Input()
    ], AppliedCouponsComponent.prototype, "cartIsLoading", void 0);
    __decorate([
        Input()
    ], AppliedCouponsComponent.prototype, "isReadOnly", void 0);
    AppliedCouponsComponent = __decorate([
        Component({
            selector: 'cx-applied-coupons',
            template: "<div *ngIf=\"isReadOnly; else editableCoupons\">\r\n  <div *ngIf=\"sortedVouchers.length > 0\">\r\n    <div class=\"cx-applied-coupon-title\">\r\n      {{ 'voucher.vouchersApplied' | cxTranslate }}\r\n    </div>\r\n  </div>\r\n  <div\r\n    *ngFor=\"let voucher of sortedVouchers\"\r\n    class=\"coupon-summary cx-coupon-card textonly\"\r\n    role=\"filter\"\r\n  >\r\n    <span class=\"cx-applied-coupon-code\">{{ voucher.voucherCode }}</span>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #editableCoupons>\r\n  <div class=\"row\">\r\n    <div\r\n      *ngFor=\"let voucher of sortedVouchers\"\r\n      class=\"col-sm-12 col-md-6 col-lg-12 cx-coupon-card-grid\"\r\n      role=\"filter\"\r\n    >\r\n      <div class=\"cx-coupon-apply cx-coupon-card cx-coupon-list-wrap\">\r\n        <span class=\"cx-cart-coupon-code\">{{ voucher.voucherCode }}</span>\r\n        <button\r\n          type=\"button\"\r\n          class=\"close\"\r\n          aria-label=\"Close\"\r\n          (click)=\"removeVoucher(voucher.voucherCode)\"\r\n          [disabled]=\"cartIsLoading\"\r\n          [class.disabled]=\"cartIsLoading\"\r\n        >\r\n          <span aria-hidden=\"true\">\r\n            <cx-icon [type]=\"iconTypes.CLOSE\"></cx-icon>\r\n          </span>\r\n        </button>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</ng-template>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], AppliedCouponsComponent);
    return AppliedCouponsComponent;
}());
export { AppliedCouponsComponent };
