// @spartacus/storefront/esm5/cms-components/content/paragraph/paragraph.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ParagraphComponent } from './paragraph.component';
import { provideDefaultConfig } from '@spartacus/core';
var CmsParagraphModule = /** @class */ (function () {
    function CmsParagraphModule() {
    }
    CmsParagraphModule = __decorate([
        NgModule({
            imports: [CommonModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CMSParagraphComponent: {
                            component: ParagraphComponent,
                        },
                        CMSTabParagraphComponent: {
                            component: ParagraphComponent,
                        },
                    },
                }),
            ],
            declarations: [ParagraphComponent],
            exports: [ParagraphComponent],
            entryComponents: [ParagraphComponent],
        })
    ], CmsParagraphModule);
    return CmsParagraphModule;
}());
export { CmsParagraphModule };
