// @spartacus/core/esm5/src/user/store/user-state.js
export var USER_FEATURE = 'user';
export var UPDATE_EMAIL_PROCESS_ID = 'updateEmail';
export var UPDATE_PASSWORD_PROCESS_ID = 'updatePassword';
export var UPDATE_USER_DETAILS_PROCESS_ID = 'updateUserDetails';
export var REGISTER_USER_PROCESS_ID = 'registerUser';
export var REMOVE_USER_PROCESS_ID = 'removeUser';
export var GIVE_CONSENT_PROCESS_ID = 'giveConsent';
export var WITHDRAW_CONSENT_PROCESS_ID = 'withdrawConsent';
export var UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID = 'updateNotificationPreferences';
export var ADD_PRODUCT_INTEREST_PROCESS_ID = 'addProductInterests';
export var REMOVE_PRODUCT_INTERESTS_PROCESS_ID = 'removeProductInterests';
export var CANCEL_ORDER_PROCESS_ID = 'cancelOrder';
export var CANCEL_RETURN_PROCESS_ID = 'cancelReturn';
export var USER_CONSENTS = '[User] User Consents';
export var USER_PAYMENT_METHODS = '[User] User Payment Methods';
export var USER_ORDERS = '[User] User Orders';
export var USER_ADDRESSES = '[User] User Addresses';
export var USER_RETURN_REQUESTS = '[User] Order Return Requests';
export var USER_RETURN_REQUEST_DETAILS = '[User] Return Request Details';
export var USER_ORDER_DETAILS = '[User] User Order Details';
export var REGIONS = '[User] Regions';
export var CUSTOMER_COUPONS = '[User] Customer Coupons';
export var SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID = 'subscribeCustomerCoupon';
export var UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID = 'unsubscribeCustomerCoupon';
export var CLAIM_CUSTOMER_COUPON_PROCESS_ID = 'claimCustomerCoupon';
export var NOTIFICATION_PREFERENCES = '[User] Notification Preferences';
export var PRODUCT_INTERESTS = '[User] Product Interests';
