// @spartacus/core/esm5/src/global-message/http-interceptors/handlers/unknown-error/unknown-error.handler.js
import { __decorate, __extends } from "tslib";
import { Injectable, isDevMode } from '@angular/core';
import { HttpResponseStatus } from '../../../models/response-status.model';
import { HttpErrorHandler } from '../http-error.handler';
import * as i0 from "@angular/core";
import * as i1 from "../../../facade/global-message.service";
/**
 * Unknown Error Handler works as an fallback, to handle errors that were
 * not handled by any other error handlers
 */
var UnknownErrorHandler = /** @class */ (function (_super) {
    __extends(UnknownErrorHandler, _super);
    function UnknownErrorHandler() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.responseStatus = HttpResponseStatus.UNKNOWN;
        return _this;
    }
    /**
     * hasMatch always returns true, to mach all errors
     */
    UnknownErrorHandler.prototype.hasMatch = function (_errorResponse) {
        return true;
    };
    UnknownErrorHandler.prototype.handleError = function () {
        if (isDevMode()) {
            console.warn("Unknown http response error: " + this.responseStatus);
        }
    };
    /**
     * Fallback priority assures that the handler is used as a last resort
     */
    UnknownErrorHandler.prototype.getPriority = function () {
        return -50 /* FALLBACK */;
    };
    UnknownErrorHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function UnknownErrorHandler_Factory() { return new UnknownErrorHandler(i0.ɵɵinject(i1.GlobalMessageService)); }, token: UnknownErrorHandler, providedIn: "root" });
    UnknownErrorHandler = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], UnknownErrorHandler);
    return UnknownErrorHandler;
}(HttpErrorHandler));
export { UnknownErrorHandler };
