// @spartacus/storefront/esm2015/cms-components/anonymous-consent-management/open-dialog/anonymous-consent-open-dialog.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, ViewChild, ViewContainerRef, } from '@angular/core';
import { take } from 'rxjs/operators';
import { AnonymousConsentLaunchDialogService } from '../anonymous-consent-launch-dialog.service';
let AnonymousConsentOpenDialogComponent = class AnonymousConsentOpenDialogComponent {
    constructor(vcr, anonymousConsentLaunchDialogService) {
        this.vcr = vcr;
        this.anonymousConsentLaunchDialogService = anonymousConsentLaunchDialogService;
    }
    openDialog() {
        const dialog = this.anonymousConsentLaunchDialogService.openDialog(this.openElement, this.vcr);
        if (dialog) {
            dialog.pipe(take(1)).subscribe();
        }
    }
};
AnonymousConsentOpenDialogComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: AnonymousConsentLaunchDialogService }
];
__decorate([
    ViewChild('open')
], AnonymousConsentOpenDialogComponent.prototype, "openElement", void 0);
AnonymousConsentOpenDialogComponent = __decorate([
    Component({
        selector: 'cx-anonymous-consent-open-dialog',
        template: "<button #open class=\"btn btn-link\" (click)=\"openDialog()\">\r\n  {{ 'anonymousConsents.preferences' | cxTranslate }}\r\n</button>\r\n"
    })
], AnonymousConsentOpenDialogComponent);
export { AnonymousConsentOpenDialogComponent };
