// @spartacus/core/esm5/src/cms/store/actions/page.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
export var LOAD_CMS_PAGE_DATA = '[Cms] Load Page Data';
export var LOAD_CMS_PAGE_DATA_FAIL = '[Cms] Load Page Data Fail';
export var LOAD_CMS_PAGE_DATA_SUCCESS = '[Cms] Load Page Data Success';
export var CMS_SET_PAGE_SUCCESS_INDEX = '[Cms] Set Page Success Index';
export var CMS_SET_PAGE_FAIL_INDEX = '[Cms] Set Page Fail Index';
var LoadCmsPageData = /** @class */ (function (_super) {
    __extends(LoadCmsPageData, _super);
    function LoadCmsPageData(payload) {
        var _this = _super.call(this, payload.type, payload.id) || this;
        _this.payload = payload;
        _this.type = LOAD_CMS_PAGE_DATA;
        return _this;
    }
    return LoadCmsPageData;
}(StateUtils.EntityLoadAction));
export { LoadCmsPageData };
var LoadCmsPageDataFail = /** @class */ (function (_super) {
    __extends(LoadCmsPageDataFail, _super);
    function LoadCmsPageDataFail(pageContext, error) {
        var _this = _super.call(this, pageContext.type, pageContext.id, error) || this;
        _this.type = LOAD_CMS_PAGE_DATA_FAIL;
        return _this;
    }
    return LoadCmsPageDataFail;
}(StateUtils.EntityFailAction));
export { LoadCmsPageDataFail };
var LoadCmsPageDataSuccess = /** @class */ (function (_super) {
    __extends(LoadCmsPageDataSuccess, _super);
    function LoadCmsPageDataSuccess(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id, payload) || this;
        _this.type = LOAD_CMS_PAGE_DATA_SUCCESS;
        return _this;
    }
    return LoadCmsPageDataSuccess;
}(StateUtils.EntitySuccessAction));
export { LoadCmsPageDataSuccess };
var CmsSetPageSuccessIndex = /** @class */ (function (_super) {
    __extends(CmsSetPageSuccessIndex, _super);
    function CmsSetPageSuccessIndex(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id, payload) || this;
        _this.type = CMS_SET_PAGE_SUCCESS_INDEX;
        return _this;
    }
    return CmsSetPageSuccessIndex;
}(StateUtils.EntitySuccessAction));
export { CmsSetPageSuccessIndex };
var CmsSetPageFailIndex = /** @class */ (function (_super) {
    __extends(CmsSetPageFailIndex, _super);
    function CmsSetPageFailIndex(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id) || this;
        _this.payload = payload;
        _this.type = CMS_SET_PAGE_FAIL_INDEX;
        return _this;
    }
    return CmsSetPageFailIndex;
}(StateUtils.EntityFailAction));
export { CmsSetPageFailIndex };
