// @spartacus/storefront/esm5/cms-components/checkout/guards/checkout-auth.guard.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { ActiveCartService, AuthRedirectService, AuthService, RoutingService, User, UserToken, } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { map } from 'rxjs/operators';
import { CheckoutConfigService } from '../services/checkout-config.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "../services/checkout-config.service";
var CheckoutAuthGuard = /** @class */ (function () {
    function CheckoutAuthGuard(routingService, authService, authRedirectService, checkoutConfigService, activeCartService) {
        this.routingService = routingService;
        this.authService = authService;
        this.authRedirectService = authRedirectService;
        this.checkoutConfigService = checkoutConfigService;
        this.activeCartService = activeCartService;
    }
    CheckoutAuthGuard.prototype.canActivate = function () {
        var _this = this;
        return combineLatest([
            this.authService.getUserToken(),
            this.activeCartService.getAssignedUser(),
        ]).pipe(map(function (_a) {
            var _b = __read(_a, 2), token = _b[0], user = _b[1];
            if (!token.access_token) {
                if (_this.activeCartService.isGuestCart()) {
                    return Boolean(user);
                }
                if (_this.checkoutConfigService.isGuestCheckout()) {
                    _this.routingService.go({ cxRoute: 'login' }, { forced: true });
                }
                else {
                    _this.routingService.go({ cxRoute: 'login' });
                }
                _this.authRedirectService.reportAuthGuard();
            }
            return !!token.access_token;
        }));
    };
    CheckoutAuthGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: AuthService },
        { type: AuthRedirectService },
        { type: CheckoutConfigService },
        { type: ActiveCartService }
    ]; };
    CheckoutAuthGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function CheckoutAuthGuard_Factory() { return new CheckoutAuthGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.AuthRedirectService), i0.ɵɵinject(i2.CheckoutConfigService), i0.ɵɵinject(i1.ActiveCartService)); }, token: CheckoutAuthGuard, providedIn: "root" });
    CheckoutAuthGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CheckoutAuthGuard);
    return CheckoutAuthGuard;
}());
export { CheckoutAuthGuard };
