// @spartacus/storefront/esm5/cms-components/checkout/guards/not-checkout-auth.guard.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ActiveCartService, AuthService, RoutingService, } from '@spartacus/core';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var NotCheckoutAuthGuard = /** @class */ (function () {
    function NotCheckoutAuthGuard(routingService, authService, activeCartService) {
        this.routingService = routingService;
        this.authService = authService;
        this.activeCartService = activeCartService;
    }
    NotCheckoutAuthGuard.prototype.canActivate = function () {
        var _this = this;
        return this.authService.getUserToken().pipe(map(function (token) {
            if (token.access_token) {
                _this.routingService.go({ cxRoute: 'home' });
            }
            else if (_this.activeCartService.isGuestCart()) {
                _this.routingService.go({ cxRoute: 'cart' });
                return false;
            }
            return !token.access_token;
        }));
    };
    NotCheckoutAuthGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: AuthService },
        { type: ActiveCartService }
    ]; };
    NotCheckoutAuthGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotCheckoutAuthGuard_Factory() { return new NotCheckoutAuthGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.ActiveCartService)); }, token: NotCheckoutAuthGuard, providedIn: "root" });
    NotCheckoutAuthGuard = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], NotCheckoutAuthGuard);
    return NotCheckoutAuthGuard;
}());
export { NotCheckoutAuthGuard };
