// @spartacus/core/esm5/src/product/facade/product-review.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { tap } from 'rxjs/operators';
import { ProductActions } from '../store/actions/index';
import { ProductSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var ProductReviewService = /** @class */ (function () {
    function ProductReviewService(store) {
        this.store = store;
    }
    ProductReviewService.prototype.getByProductCode = function (productCode) {
        var _this = this;
        return this.store.pipe(select(ProductSelectors.getSelectedProductReviewsFactory(productCode)), tap(function (reviews) {
            if (reviews === undefined && productCode !== undefined) {
                _this.store.dispatch(new ProductActions.LoadProductReviews(productCode));
            }
        }));
    };
    ProductReviewService.prototype.add = function (productCode, review) {
        this.store.dispatch(new ProductActions.PostProductReview({
            productCode: productCode,
            review: review,
        }));
    };
    ProductReviewService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    ProductReviewService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductReviewService_Factory() { return new ProductReviewService(i0.ɵɵinject(i1.Store)); }, token: ProductReviewService, providedIn: "root" });
    ProductReviewService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ProductReviewService);
    return ProductReviewService;
}());
export { ProductReviewService };
