// @spartacus/core/esm5/src/product/facade/product-reference.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { tap } from 'rxjs/operators';
import { ProductActions } from '../store/actions/index';
import { ProductSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var ProductReferenceService = /** @class */ (function () {
    function ProductReferenceService(store) {
        this.store = store;
    }
    ProductReferenceService.prototype.get = function (productCode, referenceType, pageSize) {
        var _this = this;
        return this.store.pipe(select(ProductSelectors.getSelectedProductReferencesFactory(productCode, referenceType)), tap(function (references) {
            if (references === undefined && productCode !== undefined) {
                _this.store.dispatch(new ProductActions.LoadProductReferences({
                    productCode: productCode,
                    referenceType: referenceType,
                    pageSize: pageSize,
                }));
            }
        }));
    };
    ProductReferenceService.prototype.cleanReferences = function () {
        this.store.dispatch(new ProductActions.CleanProductReferences());
    };
    ProductReferenceService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    ProductReferenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductReferenceService_Factory() { return new ProductReferenceService(i0.ɵɵinject(i1.Store)); }, token: ProductReferenceService, providedIn: "root" });
    ProductReferenceService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ProductReferenceService);
    return ProductReferenceService;
}());
export { ProductReferenceService };
