// @spartacus/core/esm5/src/product/facade/product-search.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { ProductActions } from '../store/actions/index';
import { ProductSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var ProductSearchService = /** @class */ (function () {
    function ProductSearchService(store) {
        this.store = store;
    }
    ProductSearchService.prototype.search = function (query, searchConfig) {
        this.store.dispatch(new ProductActions.SearchProducts({
            queryText: query,
            searchConfig: searchConfig,
        }));
    };
    ProductSearchService.prototype.getResults = function () {
        return this.store.pipe(select(ProductSelectors.getSearchResults));
    };
    ProductSearchService.prototype.clearResults = function () {
        this.store.dispatch(new ProductActions.ClearProductSearchResult({
            clearPageResults: true,
        }));
    };
    ProductSearchService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    ProductSearchService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductSearchService_Factory() { return new ProductSearchService(i0.ɵɵinject(i1.Store)); }, token: ProductSearchService, providedIn: "root" });
    ProductSearchService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ProductSearchService);
    return ProductSearchService;
}());
export { ProductSearchService };
