// @spartacus/storefront/esm2015/cms-components/wish-list/wish-list.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ItemCounterModule, MediaModule, StarRatingModule, } from '../../shared/index';
import { AddToCartModule } from '../cart/index';
import { WishListItemComponent } from './components/wish-list-item/wish-list-item.component';
import { WishListComponent } from './components/wish-list/wish-list.component';
let WishListModule = class WishListModule {
};
WishListModule = __decorate([
    NgModule({
        imports: [
            AddToCartModule,
            CommonModule,
            I18nModule,
            MediaModule,
            RouterModule,
            StarRatingModule,
            UrlModule,
            ItemCounterModule,
        ],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    WishListComponent: {
                        component: WishListComponent,
                        guards: [AuthGuard],
                    },
                },
            }),
        ],
        declarations: [WishListComponent, WishListItemComponent],
        entryComponents: [WishListComponent],
        exports: [WishListComponent, WishListItemComponent],
    })
], WishListModule);
export { WishListModule };
