// @spartacus/storefront/esm5/cms-components/checkout/components/shipping-address/address-form/address-form.component.js
import { __decorate, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { Address, AddressValidation, CheckoutDeliveryService, Country, ErrorModel, GlobalMessageService, GlobalMessageType, Region, Title, UserAddressService, UserService, } from '@spartacus/core';
import { BehaviorSubject } from 'rxjs';
import { map, switchMap, take, tap } from 'rxjs/operators';
import { ModalRef, ModalService, } from '../../../../../shared/components/modal/index';
import { sortTitles } from '../../../../../shared/utils/forms/title-utils';
import { SuggestedAddressDialogComponent } from './suggested-addresses-dialog/suggested-addresses-dialog.component';
var AddressFormComponent = /** @class */ (function () {
    function AddressFormComponent(fb, checkoutDeliveryService, userService, userAddressService, globalMessageService, modalService) {
        this.fb = fb;
        this.checkoutDeliveryService = checkoutDeliveryService;
        this.userService = userService;
        this.userAddressService = userAddressService;
        this.globalMessageService = globalMessageService;
        this.modalService = modalService;
        this.selectedCountry$ = new BehaviorSubject('');
        this.setAsDefaultField = true;
        this.showCancelBtn = true;
        this.submitAddress = new EventEmitter();
        this.backToAddress = new EventEmitter();
        this.addressForm = this.fb.group({
            country: this.fb.group({
                isocode: [null, Validators.required],
            }),
            titleCode: [''],
            firstName: ['', Validators.required],
            lastName: ['', Validators.required],
            line1: ['', Validators.required],
            line2: [''],
            town: ['', Validators.required],
            region: this.fb.group({
                isocode: [null, Validators.required],
            }),
            postalCode: ['', Validators.required],
            phone: '',
            defaultAddress: [false],
        });
    }
    AddressFormComponent.prototype.ngOnInit = function () {
        var _this = this;
        // Fetching countries
        this.countries$ = this.userAddressService.getDeliveryCountries().pipe(tap(function (countries) {
            if (Object.keys(countries).length === 0) {
                _this.userAddressService.loadDeliveryCountries();
            }
        }));
        // Fetching titles
        this.titles$ = this.userService.getTitles().pipe(tap(function (titles) {
            if (Object.keys(titles).length === 0) {
                _this.userService.loadTitles();
            }
        }), map(function (titles) {
            titles.sort(sortTitles);
            var noneTitle = { code: '', name: 'Title' };
            return __spread([noneTitle], titles);
        }));
        // Fetching regions
        this.regions$ = this.selectedCountry$.pipe(switchMap(function (country) { return _this.userAddressService.getRegions(country); }), tap(function (regions) {
            var regionControl = _this.addressForm.get('region.isocode');
            if (regions && regions.length > 0) {
                regionControl.enable();
            }
            else {
                regionControl.disable();
            }
        }));
        // verify the new added address
        this.addressVerifySub = this.checkoutDeliveryService
            .getAddressVerificationResults()
            .subscribe(function (results) {
            if (results.decision === 'FAIL') {
                _this.checkoutDeliveryService.clearAddressVerificationResults();
            }
            else if (results.decision === 'ACCEPT') {
                _this.submitAddress.emit(_this.addressForm.value);
            }
            else if (results.decision === 'REJECT') {
                // TODO: Workaround: allow server for decide is titleCode mandatory (if yes, provide personalized message)
                if (results.errors.errors.some(function (error) { return error.subject === 'titleCode'; })) {
                    _this.globalMessageService.add({ key: 'addressForm.titleRequired' }, GlobalMessageType.MSG_TYPE_ERROR);
                }
                else {
                    _this.globalMessageService.add({ key: 'addressForm.invalidAddress' }, GlobalMessageType.MSG_TYPE_ERROR);
                }
                _this.checkoutDeliveryService.clearAddressVerificationResults();
            }
            else if (results.decision === 'REVIEW') {
                _this.openSuggestedAddress(results);
            }
        });
        if (this.addressData && Object.keys(this.addressData).length !== 0) {
            this.addressForm.patchValue(this.addressData);
            this.countrySelected(this.addressData.country);
            if (this.addressData.region) {
                this.regionSelected(this.addressData.region);
            }
        }
    };
    AddressFormComponent.prototype.countrySelected = function (country) {
        this.addressForm['controls'].country['controls'].isocode.setValue(country.isocode);
        this.selectedCountry$.next(country.isocode);
    };
    AddressFormComponent.prototype.regionSelected = function (region) {
        this.addressForm['controls'].region['controls'].isocode.setValue(region.isocode);
    };
    AddressFormComponent.prototype.toggleDefaultAddress = function () {
        this.addressForm['controls'].defaultAddress.setValue(this.addressForm.value.defaultAddress);
    };
    AddressFormComponent.prototype.back = function () {
        this.backToAddress.emit();
    };
    AddressFormComponent.prototype.verifyAddress = function () {
        var _this = this;
        if (this.addressForm.valid) {
            if (this.addressForm.get('region').value.isocode) {
                this.regionsSub = this.regions$.pipe(take(1)).subscribe(function (regions) {
                    var obj = regions.find(function (region) {
                        return region.isocode ===
                            _this.addressForm.controls['region'].value.isocode;
                    });
                    Object.assign(_this.addressForm.value.region, {
                        isocodeShort: obj.isocodeShort,
                    });
                });
            }
            if (this.addressForm.dirty) {
                this.checkoutDeliveryService.verifyAddress(this.addressForm.value);
            }
            else {
                // address form value not changed
                // ignore duplicate address
                this.submitAddress.emit(undefined);
            }
        }
        else {
            this.addressForm.markAllAsTouched();
        }
    };
    AddressFormComponent.prototype.openSuggestedAddress = function (results) {
        var _this = this;
        if (!this.suggestedAddressModalRef) {
            this.suggestedAddressModalRef = this.modalService.open(SuggestedAddressDialogComponent, { centered: true, size: 'lg' });
            this.suggestedAddressModalRef.componentInstance.enteredAddress = this.addressForm.value;
            this.suggestedAddressModalRef.componentInstance.suggestedAddresses =
                results.suggestedAddresses;
            this.suggestedAddressModalRef.result
                .then(function (address) {
                _this.checkoutDeliveryService.clearAddressVerificationResults();
                if (address) {
                    address = Object.assign({
                        titleCode: _this.addressForm.value.titleCode,
                        phone: _this.addressForm.value.phone,
                        selected: true,
                    }, address);
                    _this.submitAddress.emit(address);
                }
                _this.suggestedAddressModalRef = null;
            })
                .catch(function () {
                // this  callback is called when modal is closed with Esc key or clicking backdrop
                _this.checkoutDeliveryService.clearAddressVerificationResults();
                var address = Object.assign({
                    selected: true,
                }, _this.addressForm.value);
                _this.submitAddress.emit(address);
                _this.suggestedAddressModalRef = null;
            });
        }
    };
    AddressFormComponent.prototype.ngOnDestroy = function () {
        this.checkoutDeliveryService.clearAddressVerificationResults();
        if (this.addressVerifySub) {
            this.addressVerifySub.unsubscribe();
        }
        if (this.regionsSub) {
            this.regionsSub.unsubscribe();
        }
    };
    AddressFormComponent.ctorParameters = function () { return [
        { type: FormBuilder },
        { type: CheckoutDeliveryService },
        { type: UserService },
        { type: UserAddressService },
        { type: GlobalMessageService },
        { type: ModalService }
    ]; };
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "addressData", void 0);
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "actionBtnLabel", void 0);
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "cancelBtnLabel", void 0);
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "setAsDefaultField", void 0);
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "showTitleCode", void 0);
    __decorate([
        Input()
    ], AddressFormComponent.prototype, "showCancelBtn", void 0);
    __decorate([
        Output()
    ], AddressFormComponent.prototype, "submitAddress", void 0);
    __decorate([
        Output()
    ], AddressFormComponent.prototype, "backToAddress", void 0);
    AddressFormComponent = __decorate([
        Component({
            selector: 'cx-address-form',
            template: "<form (ngSubmit)=\"verifyAddress()\" [formGroup]=\"addressForm\">\r\n  <div class=\"row\">\r\n    <div class=\"col-md-12 col-lg-9\">\r\n      <div class=\"form-group\" formGroupName=\"country\">\r\n        <ng-container *ngIf=\"countries$ | async as countries\">\r\n          <div *ngIf=\"countries.length !== 0\">\r\n            <label aria-required=\"true\">\r\n              <span class=\"label-content required\">{{\r\n                'addressForm.country' | cxTranslate\r\n              }}</span>\r\n              <ng-select\r\n                class=\"country-select\"\r\n                formControlName=\"isocode\"\r\n                [searchable]=\"true\"\r\n                [clearable]=\"false\"\r\n                [items]=\"countries\"\r\n                bindLabel=\"name\"\r\n                bindValue=\"isocode\"\r\n                placeholder=\"{{ 'addressForm.selectOne' | cxTranslate }}\"\r\n                (change)=\"countrySelected($event)\"\r\n              >\r\n              </ng-select>\r\n              <cx-form-errors\r\n                [control]=\"addressForm.get('country.isocode')\"\r\n              ></cx-form-errors>\r\n            </label>\r\n          </div>\r\n        </ng-container>\r\n      </div>\r\n      <div class=\"form-group\" *ngIf=\"showTitleCode\">\r\n        <ng-container *ngIf=\"titles$ | async as titles\">\r\n          <div *ngIf=\"titles.length !== 0\">\r\n            <label aria-required=\"true\">\r\n              <span class=\"label-content required\">{{\r\n                'addressForm.title' | cxTranslate\r\n              }}</span>\r\n              <ng-select\r\n                formControlName=\"titleCode\"\r\n                [searchable]=\"false\"\r\n                [clearable]=\"false\"\r\n                [items]=\"titles\"\r\n                bindLabel=\"name\"\r\n                bindValue=\"code\"\r\n              >\r\n              </ng-select>\r\n            </label>\r\n          </div>\r\n        </ng-container>\r\n      </div>\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content required\">{{\r\n            'addressForm.firstName.label' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            class=\"form-control\"\r\n            type=\"text\"\r\n            placeholder=\"{{\r\n              'addressForm.firstName.placeholder' | cxTranslate\r\n            }}\"\r\n            formControlName=\"firstName\"\r\n          />\r\n          <cx-form-errors\r\n            [control]=\"addressForm.get('firstName')\"\r\n          ></cx-form-errors>\r\n        </label>\r\n      </div>\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content required\">{{\r\n            'addressForm.lastName.label' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            type=\"text\"\r\n            class=\"form-control\"\r\n            placeholder=\"{{ 'addressForm.lastName.placeholder' | cxTranslate }}\"\r\n            formControlName=\"lastName\"\r\n          />\r\n          <cx-form-errors\r\n            [control]=\"addressForm.get('lastName')\"\r\n          ></cx-form-errors>\r\n        </label>\r\n      </div>\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content required\">{{\r\n            'addressForm.address1' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            type=\"text\"\r\n            class=\"form-control\"\r\n            placeholder=\"{{ 'addressForm.streetAddress' | cxTranslate }}\"\r\n            formControlName=\"line1\"\r\n          />\r\n          <cx-form-errors [control]=\"addressForm.get('line1')\"></cx-form-errors>\r\n        </label>\r\n      </div>\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content\">{{\r\n            'addressForm.address2' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            type=\"text\"\r\n            class=\"form-control\"\r\n            placeholder=\"{{ 'addressForm.aptSuite' | cxTranslate }}\"\r\n            formControlName=\"line2\"\r\n          />\r\n        </label>\r\n      </div>\r\n      <div class=\"row\">\r\n        <div class=\"form-group col-md-6\">\r\n          <label>\r\n            <span class=\"label-content required\">{{\r\n              'addressForm.city.label' | cxTranslate\r\n            }}</span>\r\n            <input\r\n              type=\"text\"\r\n              class=\"form-control\"\r\n              placeholder=\"{{ 'addressForm.city.placeholder' | cxTranslate }}\"\r\n              formControlName=\"town\"\r\n            />\r\n            <cx-form-errors\r\n              [control]=\"addressForm.get('town')\"\r\n            ></cx-form-errors>\r\n          </label>\r\n        </div>\r\n        <div class=\"form-group col-md-6\">\r\n          <label>\r\n            <span class=\"label-content required\">{{\r\n              'addressForm.zipCode.label' | cxTranslate\r\n            }}</span>\r\n            <input\r\n              type=\"text\"\r\n              class=\"form-control\"\r\n              placeholder=\"{{\r\n                'addressForm.zipCode.placeholder' | cxTranslate\r\n              }}\"\r\n              formControlName=\"postalCode\"\r\n            />\r\n            <cx-form-errors\r\n              [control]=\"addressForm.get('postalCode')\"\r\n            ></cx-form-errors>\r\n          </label>\r\n        </div>\r\n        <ng-container\r\n          *ngIf=\"regions$ | async as regions\"\r\n          formGroupName=\"region\"\r\n        >\r\n          <ng-container *ngIf=\"regions.length !== 0\">\r\n            <div class=\"form-group col-md-6\">\r\n              <label aria-required=\"true\">\r\n                <span class=\"label-content required\">{{\r\n                  'addressForm.state' | cxTranslate\r\n                }}</span>\r\n                <ng-select\r\n                  class=\"region-select\"\r\n                  formControlName=\"isocode\"\r\n                  [searchable]=\"true\"\r\n                  [clearable]=\"false\"\r\n                  [items]=\"regions\"\r\n                  bindLabel=\"{{ regions[0].name ? 'name' : 'isocode' }}\"\r\n                  bindValue=\"{{ regions[0].name ? 'isocode' : 'region' }}\"\r\n                  placeholder=\"{{ 'addressForm.selectOne' | cxTranslate }}\"\r\n                >\r\n                </ng-select>\r\n                <cx-form-errors\r\n                  [control]=\"addressForm.get('region.isocode')\"\r\n                ></cx-form-errors>\r\n              </label>\r\n            </div>\r\n          </ng-container>\r\n        </ng-container>\r\n      </div>\r\n      <div class=\"form-group\">\r\n        <label>\r\n          <span class=\"label-content\">{{\r\n            'addressForm.phoneNumber.label' | cxTranslate\r\n          }}</span>\r\n          <input\r\n            type=\"tel\"\r\n            class=\"form-control\"\r\n            placeholder=\"{{\r\n              'addressForm.phoneNumber.placeholder' | cxTranslate\r\n            }}\"\r\n            formControlName=\"phone\"\r\n          />\r\n        </label>\r\n      </div>\r\n      <div class=\"form-group\" *ngIf=\"setAsDefaultField\">\r\n        <div class=\"form-check\">\r\n          <label>\r\n            <input\r\n              type=\"checkbox\"\r\n              class=\"form-check-input\"\r\n              formControlName=\"defaultAddress\"\r\n              (change)=\"toggleDefaultAddress()\"\r\n            />\r\n            <span class=\"form-check-label\">{{\r\n              'addressForm.setAsDefault' | cxTranslate\r\n            }}</span>\r\n          </label>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n  <div class=\"cx-checkout-btns row\">\r\n    <div class=\"col-md-12 col-lg-6\" *ngIf=\"showCancelBtn\">\r\n      <button class=\"btn btn-block btn-action\" (click)=\"back()\">\r\n        {{ cancelBtnLabel || ('addressForm.chooseAddress' | cxTranslate) }}\r\n      </button>\r\n    </div>\r\n    <div class=\"col-md-12 col-lg-6\">\r\n      <button class=\"btn btn-block btn-primary\" type=\"submit\">\r\n        {{ actionBtnLabel || ('common.continue' | cxTranslate) }}\r\n      </button>\r\n    </div>\r\n  </div>\r\n</form>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], AddressFormComponent);
    return AddressFormComponent;
}());
export { AddressFormComponent };
