// @spartacus/core/esm5/src/auth/services/client-authentication/client-authentication-token.service.js
import { __decorate } from "tslib";
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { OccEndpointsService } from '../../../occ/services/occ-endpoints.service';
import { AuthConfig } from '../../config/auth-config';
import * as i0 from "@angular/core";
import * as i1 from "../../config/auth-config";
import * as i2 from "@angular/common/http";
import * as i3 from "../../../occ/services/occ-endpoints.service";
var ClientAuthenticationTokenService = /** @class */ (function () {
    function ClientAuthenticationTokenService(config, http, occEndpointsService) {
        this.config = config;
        this.http = http;
        this.occEndpointsService = occEndpointsService;
    }
    ClientAuthenticationTokenService.prototype.loadClientAuthenticationToken = function () {
        var url = this.occEndpointsService.getRawEndpoint('login');
        var params = new HttpParams()
            .set('client_id', encodeURIComponent(this.config.authentication.client_id))
            .set('client_secret', encodeURIComponent(this.config.authentication.client_secret))
            .set('grant_type', 'client_credentials');
        var headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        return this.http.post(url, params, { headers: headers });
    };
    ClientAuthenticationTokenService.ctorParameters = function () { return [
        { type: AuthConfig },
        { type: HttpClient },
        { type: OccEndpointsService }
    ]; };
    ClientAuthenticationTokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ClientAuthenticationTokenService_Factory() { return new ClientAuthenticationTokenService(i0.ɵɵinject(i1.AuthConfig), i0.ɵɵinject(i2.HttpClient), i0.ɵɵinject(i3.OccEndpointsService)); }, token: ClientAuthenticationTokenService, providedIn: "root" });
    ClientAuthenticationTokenService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], ClientAuthenticationTokenService);
    return ClientAuthenticationTokenService;
}());
export { ClientAuthenticationTokenService };
