// @spartacus/storefront/esm2015/cms-components/product/carousel/product-carousel/product-carousel.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { CmsProductCarouselComponent as model, Product, ProductScope, ProductService, } from '@spartacus/core';
import { filter, map } from 'rxjs/operators';
import { CmsComponentData } from '../../../../cms-structure/page/model/cms-component-data';
let ProductCarouselComponent = class ProductCarouselComponent {
    constructor(componentData, productService) {
        this.componentData = componentData;
        this.productService = productService;
        this.PRODUCT_SCOPE = ProductScope.LIST;
        this.componentData$ = this.componentData.data$.pipe(filter(Boolean));
        /**
         * returns an Obervable string for the title.
         */
        this.title$ = this.componentData$.pipe(map((data) => data.title));
        /**
         * Obervable that holds an Array of Observables. This is done, so that
         * the component UI could consider to lazy load the UI components when they're
         * in the viewpoint.
         */
        this.items$ = this.componentData$.pipe(map((data) => data.productCodes.trim().split(' ')), map((codes) => codes.map((code) => this.productService.get(code, this.PRODUCT_SCOPE))));
    }
};
ProductCarouselComponent.ctorParameters = () => [
    { type: CmsComponentData },
    { type: ProductService }
];
ProductCarouselComponent = __decorate([
    Component({
        selector: 'cx-product-carousel',
        template: "<cx-carousel\r\n  [items]=\"items$ | async\"\r\n  [title]=\"title$ | async\"\r\n  [template]=\"carouselItem\"\r\n  itemWidth=\"285px\"\r\n>\r\n</cx-carousel>\r\n\r\n<ng-template #carouselItem let-item=\"item\">\r\n  <a tabindex=\"0\" [routerLink]=\"{ cxRoute: 'product', params: item } | cxUrl\">\r\n    <cx-media [container]=\"item.images?.PRIMARY\"></cx-media>\r\n    <h4>\r\n      {{ item.name }}\r\n    </h4>\r\n    <div class=\"price\">\r\n      {{ item.price?.formattedValue }}\r\n    </div>\r\n  </a>\r\n</ng-template>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], ProductCarouselComponent);
export { ProductCarouselComponent };
