// @spartacus/storefront/esm2015/cms-components/product/carousel/product-carousel/product-carousel.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig, UrlModule } from '@spartacus/core';
import { CarouselModule, MediaModule, } from '../../../../shared/components/index';
import { ProductCarouselComponent } from './product-carousel.component';
let ProductCarouselModule = class ProductCarouselModule {
};
ProductCarouselModule = __decorate([
    NgModule({
        imports: [CommonModule, CarouselModule, MediaModule, RouterModule, UrlModule],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    ProductCarouselComponent: {
                        component: ProductCarouselComponent,
                    },
                },
            }),
        ],
        declarations: [ProductCarouselComponent],
        entryComponents: [ProductCarouselComponent],
        exports: [ProductCarouselComponent],
    })
], ProductCarouselModule);
export { ProductCarouselModule };
