// @spartacus/storefront/esm5/cms-components/cart/save-for-later/save-for-later.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { CartSharedModule } from '../cart-shared/cart-shared.module';
import { SaveForLaterComponent } from './save-for-later.component';
var SaveForLaterModule = /** @class */ (function () {
    function SaveForLaterModule() {
    }
    SaveForLaterModule = __decorate([
        NgModule({
            imports: [CommonModule, I18nModule, CartSharedModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        SaveForLaterComponent: {
                            component: SaveForLaterComponent,
                        },
                    },
                }),
            ],
            declarations: [SaveForLaterComponent],
            exports: [SaveForLaterComponent],
            entryComponents: [SaveForLaterComponent],
        })
    ], SaveForLaterModule);
    return SaveForLaterModule;
}());
export { SaveForLaterModule };
