// @spartacus/core/esm5/src/occ/adapters/product/occ-product.adapter.js
import { __assign, __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
import { ConverterService } from '../../../util/converter.service';
import { PRODUCT_NORMALIZER } from '../../../product/connectors/product/converters';
import { OccRequestsOptimizerService } from '../../services/occ-requests-optimizer.service';
var OccProductAdapter = /** @class */ (function () {
    function OccProductAdapter(http, occEndpoints, converter, requestsOptimizer) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
        this.requestsOptimizer = requestsOptimizer;
    }
    OccProductAdapter.prototype.load = function (productCode, scope) {
        return this.http
            .get(this.getEndpoint(productCode, scope))
            .pipe(this.converter.pipeable(PRODUCT_NORMALIZER));
    };
    OccProductAdapter.prototype.loadMany = function (products) {
        var _this = this;
        var scopedDataWithUrls = products.map(function (model) { return ({
            scopedData: model,
            url: _this.getEndpoint(model.code, model.scope),
        }); });
        return this.requestsOptimizer
            .scopedDataLoad(scopedDataWithUrls)
            .map(function (scopedProduct) {
            return (__assign(__assign({}, scopedProduct), { data$: scopedProduct.data$.pipe(_this.converter.pipeable(PRODUCT_NORMALIZER)) }));
        });
    };
    OccProductAdapter.prototype.getEndpoint = function (code, scope) {
        return this.occEndpoints.getUrl('product', {
            productCode: code,
        }, undefined, scope);
    };
    OccProductAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService },
        { type: OccRequestsOptimizerService }
    ]; };
    OccProductAdapter = __decorate([
        Injectable()
    ], OccProductAdapter);
    return OccProductAdapter;
}());
export { OccProductAdapter };
