// @spartacus/storefront/esm5/cms-components/asm/asm-toggle-ui/asm-toggle-ui.component.js
import { __decorate } from "tslib";
import { Component, ViewEncapsulation } from '@angular/core';
import { AsmService } from '@spartacus/core';
import { Subscription } from 'rxjs';
var AsmToggleUiComponent = /** @class */ (function () {
    function AsmToggleUiComponent(asmService) {
        this.asmService = asmService;
        this.subscription = new Subscription();
    }
    AsmToggleUiComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription.add(this.asmService.getAsmUiState().subscribe(function (uiState) {
            _this.isCollapsed = uiState.collapsed;
        }));
    };
    AsmToggleUiComponent.prototype.toggleUi = function () {
        this.asmService.updateAsmUiState({ collapsed: !this.isCollapsed });
    };
    AsmToggleUiComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    AsmToggleUiComponent.ctorParameters = function () { return [
        { type: AsmService }
    ]; };
    AsmToggleUiComponent = __decorate([
        Component({
            selector: 'cx-asm-toggle-ui',
            template: "<a class=\"toggleUi\" (click)=\"toggleUi()\">\r\n  <span [ngClass]=\"!isCollapsed ? 'collapseIcon' : 'expandIcon'\"></span>\r\n  <span *ngIf=\"!isCollapsed\" class=\"label\">\r\n    {{ 'asm.toggleUi.collapse' | cxTranslate }}\r\n  </span>\r\n  <span *ngIf=\"isCollapsed\" class=\"label\">\r\n    {{ 'asm.toggleUi.expand' | cxTranslate }}\r\n  </span>\r\n</a>\r\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["cx-asm-toggle-ui{cursor:pointer;display:flex;align-items:center;height:16px;margin:0 15px}cx-asm-toggle-ui .toggleUi{display:inherit;align-items:inherit}cx-asm-toggle-ui .toggleUi .label{margin-left:5px}@media (max-width:575px){cx-asm-toggle-ui .toggleUi .label{display:none}}cx-asm-toggle-ui .toggleUi .collapseIcon,cx-asm-toggle-ui .toggleUi .expandIcon{width:16px;height:16px}cx-asm-toggle-ui .toggleUi .collapseIcon{background:url(\"data:image/svg+xml,%3Csvg aria-hidden='true' focusable='false' data-prefix='fas' data-icon='chevron-circle-up' class='svg-inline--fa fa-chevron-circle-up fa-w-16' role='img' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'%3E%3Cpath fill='%23d1e3ff' d='M8 256C8 119 119 8 256 8s248 111 248 248-111 248-248 248S8 393 8 256zm231-113.9L103.5 277.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L256 226.9l101.6 101.6c9.4 9.4 24.6 9.4 33.9 0l17-17c9.4-9.4 9.4-24.6 0-33.9L273 142.1c-9.4-9.4-24.6-9.4-34 0z'%3E%3C/path%3E%3C/svg%3E\") center center no-repeat}cx-asm-toggle-ui .toggleUi .expandIcon{background:url(\"data:image/svg+xml,%3Csvg aria-hidden='true' focusable='false' data-prefix='fas' data-icon='chevron-circle-down' class='svg-inline--fa fa-chevron-circle-down fa-w-16' role='img' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'%3E%3Cpath fill='%23d1e3ff' d='M504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zM273 369.9l135.5-135.5c9.4-9.4 9.4-24.6 0-33.9l-17-17c-9.4-9.4-24.6-9.4-33.9 0L256 285.1 154.4 183.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L239 369.9c9.4 9.4 24.6 9.4 34 0z'%3E%3C/path%3E%3C/svg%3E\") center center no-repeat}"]
        })
    ], AsmToggleUiComponent);
    return AsmToggleUiComponent;
}());
export { AsmToggleUiComponent };
