// @spartacus/core/esm5/src/site-context/facade/base-site.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { filter, map, take, tap } from 'rxjs/operators';
import { getContextParameterDefault } from '../config/context-config-utils';
import { SiteContextConfig } from '../config/site-context-config';
import { BASE_SITE_CONTEXT_ID } from '../providers/context-ids';
import { SiteContextActions } from '../store/actions/index';
import { SiteContextSelectors } from '../store/selectors/index';
var BaseSiteService = /** @class */ (function () {
    function BaseSiteService(store, config) {
        this.store = store;
        this.config = config;
    }
    /**
     * Represents the current baseSite uid.
     */
    BaseSiteService.prototype.getActive = function () {
        return this.store.pipe(select(SiteContextSelectors.getActiveBaseSite), filter(function (active) { return Boolean(active); }));
    };
    /**
     * We currently don't support switching baseSite at run time
     */
    BaseSiteService.prototype.getAll = function () {
        return this.getActive().pipe(map(function (baseSite) { return [baseSite]; }));
    };
    BaseSiteService.prototype.setActive = function (baseSite) {
        var _this = this;
        return this.store
            .pipe(select(SiteContextSelectors.getActiveBaseSite), take(1))
            .subscribe(function (activeBaseSite) {
            if (baseSite && activeBaseSite !== baseSite) {
                _this.store.dispatch(new SiteContextActions.SetActiveBaseSite(baseSite));
            }
        });
    };
    /**
     * Initializes the active baseSite.
     */
    BaseSiteService.prototype.initialize = function () {
        var value;
        this.getActive()
            .subscribe(function (val) { return (value = val); })
            .unsubscribe();
        if (value) {
            // don't initialize, if there is already a value (i.e. retrieved from route or transferred from SSR)
            return;
        }
        this.setActive(getContextParameterDefault(this.config, BASE_SITE_CONTEXT_ID));
    };
    /**
     * Get the base site details data
     */
    BaseSiteService.prototype.getBaseSiteData = function () {
        var _this = this;
        return this.store.pipe(select(SiteContextSelectors.getBaseSiteData), tap(function (baseSite) {
            if (Object.keys(baseSite).length === 0) {
                _this.store.dispatch(new SiteContextActions.LoadBaseSite());
            }
        }));
    };
    BaseSiteService.ctorParameters = function () { return [
        { type: Store },
        { type: SiteContextConfig }
    ]; };
    BaseSiteService = __decorate([
        Injectable()
    ], BaseSiteService);
    return BaseSiteService;
}());
export { BaseSiteService };
