// @spartacus/core/esm5/src/features-config/features-config.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '../config/config.module';
import { FeatureLevelDirective } from './directives/feature-level.directive';
import { FeatureDirective } from './directives/feature.directive';
var FeaturesConfigModule = /** @class */ (function () {
    function FeaturesConfigModule() {
    }
    FeaturesConfigModule_1 = FeaturesConfigModule;
    FeaturesConfigModule.forRoot = function (defaultLevel) {
        return {
            ngModule: FeaturesConfigModule_1,
            providers: [
                provideDefaultConfig({
                    features: {
                        level: defaultLevel || '*',
                    },
                }),
            ],
        };
    };
    var FeaturesConfigModule_1;
    FeaturesConfigModule = FeaturesConfigModule_1 = __decorate([
        NgModule({
            declarations: [FeatureLevelDirective, FeatureDirective],
            exports: [FeatureLevelDirective, FeatureDirective],
        })
    ], FeaturesConfigModule);
    return FeaturesConfigModule;
}());
export { FeaturesConfigModule };
