// @spartacus/core/esm5/src/occ/adapters/site-context/occ-site.adapter.js
import { __decorate } from "tslib";
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { COUNTRY_NORMALIZER, CURRENCY_NORMALIZER, LANGUAGE_NORMALIZER, REGION_NORMALIZER, } from '../../../site-context/connectors/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
var OccSiteAdapter = /** @class */ (function () {
    function OccSiteAdapter(http, occEndpointsService, converterService) {
        this.http = http;
        this.occEndpointsService = occEndpointsService;
        this.converterService = converterService;
    }
    OccSiteAdapter.prototype.loadLanguages = function () {
        return this.http
            .get(this.occEndpointsService.getUrl('languages'))
            .pipe(map(function (languageList) { return languageList.languages; }), this.converterService.pipeableMany(LANGUAGE_NORMALIZER));
    };
    OccSiteAdapter.prototype.loadCurrencies = function () {
        return this.http
            .get(this.occEndpointsService.getUrl('currencies'))
            .pipe(map(function (currencyList) { return currencyList.currencies; }), this.converterService.pipeableMany(CURRENCY_NORMALIZER));
    };
    OccSiteAdapter.prototype.loadCountries = function (type) {
        return this.http
            .get(this.occEndpointsService.getUrl('countries', undefined, type ? { type: type } : undefined))
            .pipe(map(function (countryList) { return countryList.countries; }), this.converterService.pipeableMany(COUNTRY_NORMALIZER));
    };
    OccSiteAdapter.prototype.loadRegions = function (countryIsoCode) {
        return this.http
            .get(this.occEndpointsService.getUrl('regions', { isoCode: countryIsoCode }))
            .pipe(map(function (regionList) { return regionList.regions; }), this.converterService.pipeableMany(REGION_NORMALIZER));
    };
    OccSiteAdapter.prototype.loadBaseSite = function () {
        var baseUrl = this.occEndpointsService.getBaseEndpoint();
        var urlSplits = baseUrl.split('/');
        var activeSite = urlSplits.pop();
        var url = urlSplits.join('/') + '/basesites';
        var params = new HttpParams({
            fromString: 'fields=FULL',
        });
        return this.http
            .get(url, { params: params })
            .pipe(map(function (siteList) {
            return siteList.baseSites.find(function (site) { return site.uid === activeSite; });
        }));
    };
    OccSiteAdapter.ctorParameters = function () { return [
        { type: HttpClient },
        { type: OccEndpointsService },
        { type: ConverterService }
    ]; };
    OccSiteAdapter = __decorate([
        Injectable()
    ], OccSiteAdapter);
    return OccSiteAdapter;
}());
export { OccSiteAdapter };
