// @spartacus/core/esm5/src/occ/adapters/site-context/site-context-occ.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '../../../config/config.module';
import { SiteAdapter } from '../../../site-context/connectors/site.adapter';
import { defaultOccSiteContextConfig } from './default-occ-site-context-config';
import { OccSiteAdapter } from './occ-site.adapter';
import { SiteContextInterceptor } from './site-context.interceptor';
var SiteContextOccModule = /** @class */ (function () {
    function SiteContextOccModule() {
    }
    SiteContextOccModule = __decorate([
        NgModule({
            imports: [CommonModule, HttpClientModule],
            providers: [
                provideDefaultConfig(defaultOccSiteContextConfig),
                {
                    provide: SiteAdapter,
                    useClass: OccSiteAdapter,
                },
                {
                    provide: HTTP_INTERCEPTORS,
                    useExisting: SiteContextInterceptor,
                    multi: true,
                },
            ],
        })
    ], SiteContextOccModule);
    return SiteContextOccModule;
}());
export { SiteContextOccModule };
