// @spartacus/storefront/esm5/cms-components/myaccount/update-profile/update-profile.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { UpdateProfileFormComponent } from './components/update-profile-form.component';
import { UpdateProfileComponent } from './update-profile.component';
import { FormErrorsModule } from '../../../shared/index';
var UpdateProfileModule = /** @class */ (function () {
    function UpdateProfileModule() {
    }
    UpdateProfileModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
                SpinnerModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        UpdateProfileComponent: {
                            component: UpdateProfileComponent,
                            guards: [AuthGuard],
                        },
                    },
                }),
            ],
            declarations: [UpdateProfileComponent, UpdateProfileFormComponent],
            exports: [UpdateProfileComponent, UpdateProfileFormComponent],
            entryComponents: [UpdateProfileComponent],
        })
    ], UpdateProfileModule);
    return UpdateProfileModule;
}());
export { UpdateProfileModule };
