// @spartacus/storefront/esm5/cms-components/myaccount/update-profile/update-profile.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { GlobalMessageService, GlobalMessageType, RoutingService, Title, User, UserService, } from '@spartacus/core';
import { Subscription } from 'rxjs';
import { tap } from 'rxjs/operators';
var UpdateProfileComponent = /** @class */ (function () {
    function UpdateProfileComponent(routingService, userService, globalMessageService) {
        this.routingService = routingService;
        this.userService = userService;
        this.globalMessageService = globalMessageService;
        this.subscription = new Subscription();
    }
    UpdateProfileComponent.prototype.ngOnInit = function () {
        var _this = this;
        // reset the previous form processing state
        this.userService.resetUpdatePersonalDetailsProcessingState();
        this.user$ = this.userService.get();
        this.titles$ = this.userService.getTitles().pipe(tap(function (titles) {
            if (Object.keys(titles).length === 0) {
                _this.userService.loadTitles();
            }
        }));
        this.loading$ = this.userService.getUpdatePersonalDetailsResultLoading();
        this.subscription.add(this.userService
            .getUpdatePersonalDetailsResultSuccess()
            .subscribe(function (success) { return _this.onSuccess(success); }));
    };
    UpdateProfileComponent.prototype.onSuccess = function (success) {
        if (success) {
            this.globalMessageService.add({ key: 'updateProfileForm.profileUpdateSuccess' }, GlobalMessageType.MSG_TYPE_CONFIRMATION);
            this.routingService.go({ cxRoute: 'home' });
        }
    };
    UpdateProfileComponent.prototype.onCancel = function () {
        this.routingService.go({ cxRoute: 'home' });
    };
    UpdateProfileComponent.prototype.onSubmit = function (_a) {
        var userUpdates = _a.userUpdates;
        this.userService.updatePersonalDetails(userUpdates);
    };
    UpdateProfileComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
        // clean up the state
        this.userService.resetUpdatePersonalDetailsProcessingState();
    };
    UpdateProfileComponent.ctorParameters = function () { return [
        { type: RoutingService },
        { type: UserService },
        { type: GlobalMessageService }
    ]; };
    UpdateProfileComponent = __decorate([
        Component({
            selector: 'cx-update-profile',
            template: "<ng-container>\r\n  <div *ngIf=\"loading$ | async; else updateForm\">\r\n    <div class=\"cx-spinner\">\r\n      <cx-spinner></cx-spinner>\r\n    </div>\r\n  </div>\r\n\r\n  <ng-template #updateForm>\r\n    <div class=\"container\">\r\n      <div class=\"row d-flex justify-content-center\">\r\n        <cx-update-profile-form\r\n          *ngIf=\"(user$ | async)?.uid\"\r\n          class=\"col-md-6\"\r\n          [user]=\"user$ | async\"\r\n          [titles]=\"titles$ | async\"\r\n          (cancelled)=\"onCancel()\"\r\n          (submitted)=\"onSubmit($event)\"\r\n        ></cx-update-profile-form>\r\n      </div>\r\n    </div>\r\n  </ng-template>\r\n</ng-container>\r\n"
        })
    ], UpdateProfileComponent);
    return UpdateProfileComponent;
}());
export { UpdateProfileComponent };
