// @spartacus/storefront/esm2015/cms-components/myaccount/order/amend-order/amend-order-actions/amend-order-actions.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input, } from '@angular/core';
import { RoutingService } from '@spartacus/core';
let AmendOrderActionsComponent = class AmendOrderActionsComponent {
    constructor(routingService) {
        this.routingService = routingService;
        this.styles = 'row';
    }
    continue(event) {
        if (this.amendOrderForm.valid) {
            this.routingService.go({
                cxRoute: this.forwardRoute,
                params: { code: this.orderCode },
            });
        }
        else {
            this.amendOrderForm.markAllAsTouched();
            event.stopPropagation();
        }
    }
};
AmendOrderActionsComponent.ctorParameters = () => [
    { type: RoutingService }
];
__decorate([
    Input()
], AmendOrderActionsComponent.prototype, "orderCode", void 0);
__decorate([
    Input()
], AmendOrderActionsComponent.prototype, "amendOrderForm", void 0);
__decorate([
    Input()
], AmendOrderActionsComponent.prototype, "backRoute", void 0);
__decorate([
    Input()
], AmendOrderActionsComponent.prototype, "forwardRoute", void 0);
__decorate([
    HostBinding('class')
], AmendOrderActionsComponent.prototype, "styles", void 0);
AmendOrderActionsComponent = __decorate([
    Component({
        selector: 'cx-amend-order-actions',
        template: "<div class=\"col-xs-12 col-md-4 col-lg-3\">\r\n  <a\r\n    [routerLink]=\"\r\n      {\r\n        cxRoute: backRoute,\r\n        params: { code: orderCode }\r\n      } | cxUrl\r\n    \"\r\n    class=\"btn btn-block btn-action\"\r\n  >\r\n    {{ 'common.back' | cxTranslate }}\r\n  </a>\r\n</div>\r\n<div class=\"col-xs-12 col-md-4 col-lg-3\">\r\n  <button\r\n    *ngIf=\"forwardRoute\"\r\n    class=\"btn btn-block btn-primary\"\r\n    (click)=\"continue($event)\"\r\n  >\r\n    {{ 'common.continue' | cxTranslate }}\r\n  </button>\r\n\r\n  <button *ngIf=\"!forwardRoute\" class=\"btn btn-block btn-primary\" type=\"submit\">\r\n    {{ 'orderDetails.cancellationAndReturn.submit' | cxTranslate }}\r\n  </button>\r\n</div>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], AmendOrderActionsComponent);
export { AmendOrderActionsComponent };
