// @spartacus/storefront/esm5/cms-components/anonymous-consent-management/anonymous-consent-management.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DeferLoadingStrategy, FeaturesConfigModule, I18nModule, provideConfig, provideDefaultConfig, } from '@spartacus/core';
import { KeyboardFocusModule } from '../../layout/a11y/keyboard-focus/index';
import { AnonymousConsentManagementBannerComponent } from './banner/anonymous-consent-management-banner.component';
import { defaultAnonymousConsentLayoutConfig } from './default-anonymous-consent-layout.config';
import { AnonymousConsentOpenDialogComponent } from './open-dialog/anonymous-consent-open-dialog.component';
var AnonymousConsentManagementBannerModule = /** @class */ (function () {
    function AnonymousConsentManagementBannerModule() {
    }
    AnonymousConsentManagementBannerModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                I18nModule,
                FeaturesConfigModule,
                KeyboardFocusModule,
            ],
            providers: [
                provideConfig(defaultAnonymousConsentLayoutConfig),
                provideDefaultConfig({
                    cmsComponents: {
                        AnonymousConsentManagementBannerComponent: {
                            component: AnonymousConsentManagementBannerComponent,
                            deferLoading: DeferLoadingStrategy.INSTANT,
                        },
                        AnonymousConsentOpenDialogComponent: {
                            component: AnonymousConsentOpenDialogComponent,
                        },
                    },
                }),
            ],
            declarations: [
                AnonymousConsentManagementBannerComponent,
                AnonymousConsentOpenDialogComponent,
            ],
            exports: [
                AnonymousConsentManagementBannerComponent,
                AnonymousConsentOpenDialogComponent,
            ],
            entryComponents: [
                AnonymousConsentManagementBannerComponent,
                AnonymousConsentOpenDialogComponent,
            ],
        })
    ], AnonymousConsentManagementBannerModule);
    return AnonymousConsentManagementBannerModule;
}());
export { AnonymousConsentManagementBannerModule };
