// @spartacus/storefront/esm2015/cms-components/product/product-tabs/product-reviews/product-reviews.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { ProductReviewsComponent } from './product-reviews.component';
import { StarRatingModule, FormErrorsModule } from '../../../../shared/index';
let ProductReviewsModule = class ProductReviewsModule {
};
ProductReviewsModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            ReactiveFormsModule,
            FormsModule,
            I18nModule,
            StarRatingModule,
            FormErrorsModule,
        ],
        providers: [
            provideDefaultConfig({
                cmsComponents: {
                    ProductReviewsTabComponent: {
                        component: ProductReviewsComponent,
                    },
                },
            }),
        ],
        declarations: [ProductReviewsComponent],
        entryComponents: [ProductReviewsComponent],
        exports: [ProductReviewsComponent],
    })
], ProductReviewsModule);
export { ProductReviewsModule };
