// @spartacus/storefront/esm5/cms-components/checkout/components/checkout-orchestrator/checkout-orchestrator.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { defaultCheckoutConfig } from '../../config/default-checkout-config';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { CheckoutGuard } from '../../guards/checkout.guard';
import { CartNotEmptyGuard } from './../../../../cms-components/cart/cart-not-empty.guard';
import { CheckoutOrchestratorComponent } from './checkout-orchestrator.component';
var CheckoutOrchestratorModule = /** @class */ (function () {
    function CheckoutOrchestratorModule() {
    }
    CheckoutOrchestratorModule = __decorate([
        NgModule({
            imports: [CommonModule],
            providers: [
                provideDefaultConfig(defaultCheckoutConfig),
                provideDefaultConfig({
                    cmsComponents: {
                        CheckoutOrchestrator: {
                            component: CheckoutOrchestratorComponent,
                            guards: [CheckoutAuthGuard, CartNotEmptyGuard, CheckoutGuard],
                        },
                    },
                }),
            ],
            declarations: [CheckoutOrchestratorComponent],
            entryComponents: [CheckoutOrchestratorComponent],
            exports: [CheckoutOrchestratorComponent],
        })
    ], CheckoutOrchestratorModule);
    return CheckoutOrchestratorModule;
}());
export { CheckoutOrchestratorModule };
