// @spartacus/storefront/esm2015/cms-components/myaccount/reset-password/reset-password-form.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { RoutingService, UserService } from '@spartacus/core';
import { Subscription } from 'rxjs';
import { CustomFormValidators } from '../../../shared/utils/validators/custom-form-validators';
let ResetPasswordFormComponent = class ResetPasswordFormComponent {
    constructor(fb, routingService, userService) {
        this.fb = fb;
        this.routingService = routingService;
        this.userService = userService;
        this.subscription = new Subscription();
        this.resetPasswordForm = this.fb.group({
            password: [
                '',
                [Validators.required, CustomFormValidators.passwordValidator],
            ],
            repassword: ['', [Validators.required]],
        }, {
            validators: CustomFormValidators.passwordsMustMatch('password', 'repassword'),
        });
    }
    ngOnInit() {
        this.subscription.add(this.routingService
            .getRouterState()
            .subscribe((state) => (this.token = state.state.queryParams['token'])));
        this.subscription.add(this.userService.isPasswordReset().subscribe((reset) => {
            if (reset) {
                this.routingService.go({ cxRoute: 'login' });
            }
        }));
    }
    resetPassword() {
        if (this.resetPasswordForm.valid) {
            const password = this.resetPasswordForm.get('password').value;
            this.userService.resetPassword(this.token, password);
        }
        else {
            this.resetPasswordForm.markAllAsTouched();
        }
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
};
ResetPasswordFormComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: RoutingService },
    { type: UserService }
];
ResetPasswordFormComponent = __decorate([
    Component({
        selector: 'cx-reset-password-form',
        template: "<form\r\n  (ngSubmit)=\"resetPassword()\"\r\n  [formGroup]=\"resetPasswordForm\"\r\n  class=\"cx-reset-password-form-component\"\r\n>\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'register.newPassword' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        class=\"form-control\"\r\n        type=\"password\"\r\n        name=\"password\"\r\n        placeholder=\"{{ 'register.password.placeholder' | cxTranslate }}\"\r\n        formControlName=\"password\"\r\n      />\r\n      <cx-form-errors\r\n        [control]=\"resetPasswordForm.get('password')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <div class=\"form-group\">\r\n    <label>\r\n      <span class=\"label-content\">{{\r\n        'register.passwordMinRequirements' | cxTranslate\r\n      }}</span>\r\n      <input\r\n        class=\"form-control\"\r\n        type=\"password\"\r\n        name=\"confirmpassword\"\r\n        placeholder=\"{{ 'register.confirmPassword.placeholder' | cxTranslate }}\"\r\n        formControlName=\"repassword\"\r\n      />\r\n      <cx-form-errors\r\n        [control]=\"resetPasswordForm.get('repassword')\"\r\n      ></cx-form-errors>\r\n    </label>\r\n  </div>\r\n\r\n  <div class=\"form-group\">\r\n    <button class=\"btn btn-block btn-primary\" type=\"submit\">\r\n      {{ 'register.resetPassword' | cxTranslate }}\r\n    </button>\r\n  </div>\r\n</form>\r\n"
    })
], ResetPasswordFormComponent);
export { ResetPasswordFormComponent };
