// @spartacus/core/esm5/src/asm/store/actions/csagent-token.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { CSAGENT_TOKEN_DATA } from '../asm-state';
import { LOGOUT_CUSTOMER_SUPPORT_AGENT } from '../../../auth/store/actions/login-logout.action';
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN = '[Auth] Load Customer Service Agent Token';
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_FAIL = '[Auth] Load Customer Service Agent Token Fail';
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_SUCCESS = '[Auth] Load Customer Service Agent Token Success';
var LoadCustomerSupportAgentToken = /** @class */ (function (_super) {
    __extends(LoadCustomerSupportAgentToken, _super);
    function LoadCustomerSupportAgentToken(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN;
        return _this;
    }
    return LoadCustomerSupportAgentToken;
}(StateUtils.LoaderLoadAction));
export { LoadCustomerSupportAgentToken };
var LoadCustomerSupportAgentTokenFail = /** @class */ (function (_super) {
    __extends(LoadCustomerSupportAgentTokenFail, _super);
    function LoadCustomerSupportAgentTokenFail(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_FAIL;
        return _this;
    }
    return LoadCustomerSupportAgentTokenFail;
}(StateUtils.LoaderFailAction));
export { LoadCustomerSupportAgentTokenFail };
var LoadCustomerSupportAgentTokenSuccess = /** @class */ (function (_super) {
    __extends(LoadCustomerSupportAgentTokenSuccess, _super);
    function LoadCustomerSupportAgentTokenSuccess(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_SUCCESS;
        return _this;
    }
    return LoadCustomerSupportAgentTokenSuccess;
}(StateUtils.LoaderSuccessAction));
export { LoadCustomerSupportAgentTokenSuccess };
var LogoutCustomerSupportAgent = /** @class */ (function () {
    function LogoutCustomerSupportAgent() {
        this.type = LOGOUT_CUSTOMER_SUPPORT_AGENT;
    }
    return LogoutCustomerSupportAgent;
}());
export { LogoutCustomerSupportAgent };
