// @spartacus/core/esm5/src/asm/store/actions/customer.action.js
import { __extends } from "tslib";
import { StateUtils } from '../../../state/utils/index';
import { CUSTOMER_SEARCH_DATA } from '../asm-state';
export var CUSTOMER_SEARCH = '[Asm] Customer Search';
export var CUSTOMER_SEARCH_FAIL = '[Asm] Customer Search Fail';
export var CUSTOMER_SEARCH_SUCCESS = '[Asm] Customer Search Success';
export var CUSTOMER_SEARCH_RESET = '[Asm] Customer Search Reset';
var CustomerSearch = /** @class */ (function (_super) {
    __extends(CustomerSearch, _super);
    function CustomerSearch(payload) {
        var _this = _super.call(this, CUSTOMER_SEARCH_DATA) || this;
        _this.payload = payload;
        _this.type = CUSTOMER_SEARCH;
        return _this;
    }
    return CustomerSearch;
}(StateUtils.LoaderLoadAction));
export { CustomerSearch };
var CustomerSearchFail = /** @class */ (function (_super) {
    __extends(CustomerSearchFail, _super);
    function CustomerSearchFail(payload) {
        var _this = _super.call(this, CUSTOMER_SEARCH_DATA) || this;
        _this.payload = payload;
        _this.type = CUSTOMER_SEARCH_FAIL;
        return _this;
    }
    return CustomerSearchFail;
}(StateUtils.LoaderFailAction));
export { CustomerSearchFail };
var CustomerSearchSuccess = /** @class */ (function (_super) {
    __extends(CustomerSearchSuccess, _super);
    function CustomerSearchSuccess(payload) {
        var _this = _super.call(this, CUSTOMER_SEARCH_DATA) || this;
        _this.payload = payload;
        _this.type = CUSTOMER_SEARCH_SUCCESS;
        return _this;
    }
    return CustomerSearchSuccess;
}(StateUtils.LoaderSuccessAction));
export { CustomerSearchSuccess };
var CustomerSearchReset = /** @class */ (function (_super) {
    __extends(CustomerSearchReset, _super);
    function CustomerSearchReset() {
        var _this = _super.call(this, CUSTOMER_SEARCH_DATA) || this;
        _this.type = CUSTOMER_SEARCH_RESET;
        return _this;
    }
    return CustomerSearchReset;
}(StateUtils.LoaderResetAction));
export { CustomerSearchReset };
