// @spartacus/core/esm5/src/state/utils/entity-loader/entity-loader.action.js
import { __assign } from "tslib";
import { entityMeta } from '../entity/entity.action';
import { failMeta, loadMeta, resetMeta, successMeta, } from '../loader/loader.action';
export var ENTITY_LOAD_ACTION = '[ENTITY] LOAD';
export var ENTITY_FAIL_ACTION = '[ENTITY] LOAD FAIL';
export var ENTITY_SUCCESS_ACTION = '[ENTITY] LOAD SUCCESS';
export var ENTITY_RESET_ACTION = '[ENTITY] RESET';
export function entityLoadMeta(entityType, id) {
    return __assign(__assign({}, loadMeta(entityType)), entityMeta(entityType, id));
}
export function entityFailMeta(entityType, id, error) {
    return __assign(__assign({}, failMeta(entityType, error)), entityMeta(entityType, id));
}
export function entitySuccessMeta(entityType, id) {
    return __assign(__assign({}, successMeta(entityType)), entityMeta(entityType, id));
}
export function entityResetMeta(entityType, id) {
    return __assign(__assign({}, resetMeta(entityType)), entityMeta(entityType, id));
}
var EntityLoadAction = /** @class */ (function () {
    function EntityLoadAction(entityType, id) {
        this.type = ENTITY_LOAD_ACTION;
        this.meta = entityLoadMeta(entityType, id);
    }
    return EntityLoadAction;
}());
export { EntityLoadAction };
var EntityFailAction = /** @class */ (function () {
    function EntityFailAction(entityType, id, error) {
        this.type = ENTITY_FAIL_ACTION;
        this.meta = entityFailMeta(entityType, id, error);
    }
    return EntityFailAction;
}());
export { EntityFailAction };
var EntitySuccessAction = /** @class */ (function () {
    function EntitySuccessAction(entityType, id, payload) {
        this.payload = payload;
        this.type = ENTITY_SUCCESS_ACTION;
        this.meta = entitySuccessMeta(entityType, id);
    }
    return EntitySuccessAction;
}());
export { EntitySuccessAction };
var EntityLoaderResetAction = /** @class */ (function () {
    function EntityLoaderResetAction(entityType, id) {
        this.type = ENTITY_RESET_ACTION;
        this.meta = entityResetMeta(entityType, id);
    }
    return EntityLoaderResetAction;
}());
export { EntityLoaderResetAction };
