// @spartacus/storefront/esm5/cms-pages/product-details-page/product-details-page.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig } from '@spartacus/core';
import { CmsPageGuard } from '../../cms-structure/guards/cms-page.guard';
import { PageLayoutComponent } from '../../cms-structure/page/page-layout/page-layout.component';
import { PRODUCT_DETAILS_URL_MATCHER } from './product-details-url-matcher';
var ɵ0 = { cxRoute: 'product' };
var ProductDetailsPageModule = /** @class */ (function () {
    function ProductDetailsPageModule() {
    }
    ProductDetailsPageModule = __decorate([
        NgModule({
            imports: [
                RouterModule.forChild([
                    {
                        path: null,
                        canActivate: [CmsPageGuard],
                        component: PageLayoutComponent,
                        data: ɵ0,
                    },
                ]),
            ],
            providers: [
                provideDefaultConfig({
                    routing: {
                        routes: {
                            product: {
                                matchers: [PRODUCT_DETAILS_URL_MATCHER],
                            },
                        },
                    },
                }),
            ],
        })
    ], ProductDetailsPageModule);
    return ProductDetailsPageModule;
}());
export { ProductDetailsPageModule };
export { ɵ0 };
