// @spartacus/storefront/esm2015/cms-components/cart/save-for-later/save-for-later.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { ActiveCartService, Cart, CmsParagraphComponent, CmsService, OrderEntry, SelectiveCartService, } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { filter, map } from 'rxjs/operators';
let SaveForLaterComponent = class SaveForLaterComponent {
    constructor(cmsService, cartService, selectiveCartService) {
        this.cmsService = cmsService;
        this.cartService = cartService;
        this.selectiveCartService = selectiveCartService;
    }
    ngOnInit() {
        this.isCartEmpty$ = this.cartService
            .getActive()
            .pipe(map((cart) => !(cart && cart.totalItems && cart.totalItems > 0)));
        this.saveForLater$ = this.selectiveCartService.getCart();
        this.entries$ = this.selectiveCartService
            .getEntries()
            .pipe(filter((entries) => entries.length > 0));
        this.cartLoaded$ = combineLatest([
            this.cartService.isStable(),
            this.selectiveCartService.getLoaded(),
        ]).pipe(map(([cartLoaded, sflLoaded]) => cartLoaded && sflLoaded));
        this.data$ = this.cmsService.getComponentData('EmptyCartParagraphComponent');
    }
    moveToCart(item) {
        this.selectiveCartService.removeEntry(item);
        this.cartService.addEntry(item.product.code, item.quantity);
    }
};
SaveForLaterComponent.ctorParameters = () => [
    { type: CmsService },
    { type: ActiveCartService },
    { type: SelectiveCartService }
];
SaveForLaterComponent = __decorate([
    Component({
        selector: 'cx-save-for-later',
        template: "<ng-container *ngIf=\"isCartEmpty$ | async\">\r\n  <p\r\n    *ngIf=\"data$ | async as data\"\r\n    [innerHTML]=\"data.content\"\r\n    class=\"cx-empty-cart-info\"\r\n  ></p>\r\n</ng-container>\r\n\r\n<ng-container *ngIf=\"saveForLater$ | async as saveForLater\">\r\n  <ng-container *ngIf=\"entries$ | async as entries\">\r\n    <div *ngIf=\"saveForLater.totalItems > 0\" class=\"cart-details-wrapper\">\r\n      <div class=\"cx-total\">\r\n        {{\r\n          'saveForLaterItems.itemTotal'\r\n            | cxTranslate: { count: saveForLater.totalItems }\r\n        }}\r\n      </div>\r\n      <cx-cart-item-list\r\n        [items]=\"entries\"\r\n        [readonly]=\"false\"\r\n        [cartIsLoading]=\"!(cartLoaded$ | async)\"\r\n        [options]=\"{\r\n          isSaveForLater: true,\r\n          optionalBtn: moveToCartBtn\r\n        }\"\r\n      ></cx-cart-item-list>\r\n    </div>\r\n  </ng-container>\r\n</ng-container>\r\n\r\n<ng-template let-ctx #moveToCartBtn>\r\n  <div class=\"col-md-3 col-lg-3 col-xl-3 cx-sfl-btn\">\r\n    <button\r\n      class=\"link cx-action-link\"\r\n      [disabled]=\"ctx.loading\"\r\n      (click)=\"moveToCart(ctx.item)\"\r\n      type=\"button\"\r\n    >\r\n      {{ 'saveForLaterItems.moveToCart' | cxTranslate }}\r\n    </button>\r\n  </div>\r\n</ng-template>\r\n"
    })
], SaveForLaterComponent);
export { SaveForLaterComponent };
