// @spartacus/storefront/esm5/cms-components/product/carousel/product-references/product-references.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig, UrlModule } from '@spartacus/core';
import { CarouselModule } from '../../../../shared/components/carousel/carousel.module';
import { MediaModule } from '../../../../shared/components/media/media.module';
import { ProductReferencesComponent } from './product-references.component';
var ProductReferencesModule = /** @class */ (function () {
    function ProductReferencesModule() {
    }
    ProductReferencesModule = __decorate([
        NgModule({
            imports: [CommonModule, CarouselModule, MediaModule, RouterModule, UrlModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        ProductReferencesComponent: {
                            component: ProductReferencesComponent,
                        },
                    },
                }),
            ],
            declarations: [ProductReferencesComponent],
            entryComponents: [ProductReferencesComponent],
            exports: [ProductReferencesComponent],
        })
    ], ProductReferencesModule);
    return ProductReferencesModule;
}());
export { ProductReferencesModule };
