// @spartacus/storefront/esm5/cms-components/myaccount/my-coupons/my-coupons.component.service.js
import { __decorate, __read } from "tslib";
import { Injectable } from '@angular/core';
import { CustomerCoupon, RoutingService, TranslationService, } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { combineLatest } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
var MyCouponsComponentService = /** @class */ (function () {
    function MyCouponsComponentService(routingService, translation) {
        this.routingService = routingService;
        this.translation = translation;
        this.RELEVANCE = ':relevance';
        this.CUSTOMER_COUPON_CODE = ':customerCouponCode:';
    }
    MyCouponsComponentService.prototype.launchSearchPage = function (coupon) {
        this.routingService.go({
            cxRoute: 'search',
            params: { query: this.buildSearchParam(coupon) },
        }, { couponcode: coupon.couponId });
    };
    MyCouponsComponentService.prototype.buildSearchParam = function (coupon) {
        return coupon.allProductsApplicable
            ? this.RELEVANCE
            : this.RELEVANCE + this.CUSTOMER_COUPON_CODE + coupon.couponId;
    };
    MyCouponsComponentService.prototype.getSortLabels = function () {
        return combineLatest([
            this.translation.translate('myCoupons.startDateAsc'),
            this.translation.translate('myCoupons.startDateDesc'),
            this.translation.translate('myCoupons.endDateAsc'),
            this.translation.translate('myCoupons.endDateDesc'),
        ]).pipe(map(function (_a) {
            var _b = __read(_a, 4), textByStartDateAsc = _b[0], textByStartDateDesc = _b[1], textByEndDateAsc = _b[2], textByEndDateDesc = _b[3];
            return {
                byStartDateAsc: textByStartDateAsc,
                byStartDateDesc: textByStartDateDesc,
                byEndDateAsc: textByEndDateAsc,
                byEndDateDesc: textByEndDateDesc,
            };
        }));
    };
    MyCouponsComponentService.ctorParameters = function () { return [
        { type: RoutingService },
        { type: TranslationService }
    ]; };
    MyCouponsComponentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MyCouponsComponentService_Factory() { return new MyCouponsComponentService(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i1.TranslationService)); }, token: MyCouponsComponentService, providedIn: "root" });
    MyCouponsComponentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], MyCouponsComponentService);
    return MyCouponsComponentService;
}());
export { MyCouponsComponentService };
