// @spartacus/storefront/esm5/cms-components/myaccount/update-password/update-password.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { UpdatePasswordFormComponent } from './components/update-password-form/update-password-form.component';
import { UpdatePasswordComponent } from './components/update-password/update-password.component';
import { FormErrorsModule } from '../../../shared/index';
var UpdatePasswordModule = /** @class */ (function () {
    function UpdatePasswordModule() {
    }
    UpdatePasswordModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
                SpinnerModule,
                I18nModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        UpdatePasswordComponent: {
                            component: UpdatePasswordComponent,
                            guards: [AuthGuard],
                        },
                    },
                }),
            ],
            declarations: [UpdatePasswordComponent, UpdatePasswordFormComponent],
            exports: [UpdatePasswordComponent, UpdatePasswordFormComponent],
            entryComponents: [UpdatePasswordComponent],
        })
    ], UpdatePasswordModule);
    return UpdatePasswordModule;
}());
export { UpdatePasswordModule };
