// @spartacus/storefront/esm2015/cms-components/myaccount/update-profile/components/update-profile-form.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
let UpdateProfileFormComponent = class UpdateProfileFormComponent {
    constructor(fb) {
        this.fb = fb;
        this.submitted = new EventEmitter();
        this.cancelled = new EventEmitter();
        this.updateProfileForm = this.fb.group({
            titleCode: [''],
            firstName: ['', Validators.required],
            lastName: ['', Validators.required],
        });
    }
    ngOnInit() {
        if (this.user) {
            this.updateProfileForm.patchValue(this.user);
        }
    }
    onSubmit() {
        if (this.updateProfileForm.valid) {
            this.submitted.emit({
                userUpdates: Object.assign({}, this.updateProfileForm.value),
            });
        }
        else {
            this.updateProfileForm.markAllAsTouched();
        }
    }
    onCancel() {
        this.cancelled.emit();
    }
};
UpdateProfileFormComponent.ctorParameters = () => [
    { type: FormBuilder }
];
__decorate([
    Input()
], UpdateProfileFormComponent.prototype, "user", void 0);
__decorate([
    Input()
], UpdateProfileFormComponent.prototype, "titles", void 0);
__decorate([
    Output()
], UpdateProfileFormComponent.prototype, "submitted", void 0);
__decorate([
    Output()
], UpdateProfileFormComponent.prototype, "cancelled", void 0);
UpdateProfileFormComponent = __decorate([
    Component({
        selector: 'cx-update-profile-form',
        template: "<form (ngSubmit)=\"onSubmit()\" [formGroup]=\"updateProfileForm\">\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateProfileForm.title' | cxTranslate\r\n        }}</span>\r\n        <select formControlName=\"titleCode\" class=\"form-control\">\r\n          <option value=\"\">{{ 'updateProfileForm.none' | cxTranslate }}</option>\r\n          <option *ngFor=\"let title of titles\" [value]=\"title.code\">{{\r\n            title.name\r\n          }}</option>\r\n        </select>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateProfileForm.firstName.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          type=\"text\"\r\n          class=\"form-control\"\r\n          name=\"firstName\"\r\n          placeholder=\"{{\r\n            'updateProfileForm.firstName.placeholder' | cxTranslate\r\n          }}\"\r\n          formControlName=\"firstName\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updateProfileForm.get('firstName')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-md-12\">\r\n      <label>\r\n        <span class=\"label-content\">{{\r\n          'updateProfileForm.lastName.label' | cxTranslate\r\n        }}</span>\r\n        <input\r\n          type=\"text\"\r\n          class=\"form-control\"\r\n          name=\"lastName\"\r\n          placeholder=\"{{\r\n            'updateProfileForm.lastName.placeholder' | cxTranslate\r\n          }}\"\r\n          formControlName=\"lastName\"\r\n        />\r\n        <cx-form-errors\r\n          [control]=\"updateProfileForm.get('lastName')\"\r\n        ></cx-form-errors>\r\n      </label>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"form-group row\">\r\n    <div class=\"col-lg-6 col-md-12\">\r\n      <button\r\n        class=\"btn btn-block btn-secondary\"\r\n        type=\"button\"\r\n        (click)=\"onCancel()\"\r\n      >\r\n        {{ 'common.cancel' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n    <div class=\"col-lg-6 col-md-12\">\r\n      <button class=\"btn btn-block btn-primary\" type=\"submit\">\r\n        {{ 'common.save' | cxTranslate }}\r\n      </button>\r\n    </div>\r\n  </div>\r\n</form>\r\n"
    })
], UpdateProfileFormComponent);
export { UpdateProfileFormComponent };
