// @spartacus/storefront/esm5/cms-components/myaccount/order/order-history/order-history.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { NgSelectModule } from '@ng-select/ng-select';
import { AuthGuard, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ListNavigationModule } from '../../../../shared/components/list-navigation/list-navigation.module';
import { CmsPageGuard } from '../../../../cms-structure/guards/cms-page.guard';
import { PageLayoutComponent } from '../../../../cms-structure/page/page-layout/page-layout.component';
import { OrderHistoryComponent } from './order-history.component';
var ɵ0 = { cxRoute: 'orders' };
var OrderHistoryModule = /** @class */ (function () {
    function OrderHistoryModule() {
    }
    OrderHistoryModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule.forChild([
                    {
                        path: null,
                        canActivate: [AuthGuard, CmsPageGuard],
                        component: PageLayoutComponent,
                        data: ɵ0,
                    },
                ]),
                RouterModule,
                FormsModule,
                NgSelectModule,
                ListNavigationModule,
                UrlModule,
                I18nModule,
            ],
            declarations: [OrderHistoryComponent],
            exports: [OrderHistoryComponent],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        AccountOrderHistoryComponent: {
                            component: OrderHistoryComponent,
                            guards: [AuthGuard],
                        },
                    },
                }),
            ],
            entryComponents: [OrderHistoryComponent],
        })
    ], OrderHistoryModule);
    return OrderHistoryModule;
}());
export { OrderHistoryModule };
export { ɵ0 };
