// @spartacus/storefront/esm5/cms-components/checkout/components/checkout-order-summary/checkout-order-summary.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { CartSharedModule } from '../../../../cms-components/cart/cart-shared/cart-shared.module';
import { CheckoutOrderSummaryComponent } from './checkout-order-summary.component';
var CheckoutOrderSummaryModule = /** @class */ (function () {
    function CheckoutOrderSummaryModule() {
    }
    CheckoutOrderSummaryModule = __decorate([
        NgModule({
            imports: [CommonModule, CartSharedModule],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        CheckoutOrderSummary: {
                            component: CheckoutOrderSummaryComponent,
                        },
                    },
                }),
            ],
            declarations: [CheckoutOrderSummaryComponent],
            entryComponents: [CheckoutOrderSummaryComponent],
            exports: [CheckoutOrderSummaryComponent],
        })
    ], CheckoutOrderSummaryModule);
    return CheckoutOrderSummaryModule;
}());
export { CheckoutOrderSummaryModule };
