// @spartacus/storefront/esm5/cms-components/myaccount/order/amend-order/returns/return-order/return-order.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { tap } from 'rxjs/operators';
import { OrderAmendService } from '../../amend-order.service';
var ReturnOrderComponent = /** @class */ (function () {
    function ReturnOrderComponent(orderAmendService) {
        var _this = this;
        this.orderAmendService = orderAmendService;
        this.form$ = this.orderAmendService
            .getForm()
            .pipe(tap(function (form) { return (_this.orderCode = form.value.orderCode); }));
        this.entries$ = this.orderAmendService.getEntries();
    }
    ReturnOrderComponent.ctorParameters = function () { return [
        { type: OrderAmendService }
    ]; };
    ReturnOrderComponent = __decorate([
        Component({
            selector: 'cx-return-order',
            template: "<ng-container *ngIf=\"form$ | async as form\">\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <cx-amend-order-items *ngIf=\"entries$ | async as entries\" [entries]=\"entries\">\r\n  </cx-amend-order-items>\r\n\r\n  <ng-container *ngTemplateOutlet=\"actions\"></ng-container>\r\n\r\n  <ng-template #actions>\r\n    <cx-amend-order-actions\r\n      *ngIf=\"orderCode\"\r\n      [orderCode]=\"orderCode\"\r\n      [amendOrderForm]=\"form\"\r\n      backRoute=\"orderDetails\"\r\n      forwardRoute=\"orderReturnConfirmation\"\r\n    ></cx-amend-order-actions>\r\n  </ng-template>\r\n</ng-container>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], ReturnOrderComponent);
    return ReturnOrderComponent;
}());
export { ReturnOrderComponent };
