// @spartacus/storefront/esm2015/cms-components/checkout/components/checkout-progress/checkout-progress-mobile-top/checkout-progress-mobile-top.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { ActiveCartService, Cart, RoutingConfigService, RoutingService, } from '@spartacus/core';
import { tap } from 'rxjs/operators';
import { CheckoutConfig } from '../../../config/checkout-config';
let CheckoutProgressMobileTopComponent = class CheckoutProgressMobileTopComponent {
    constructor(config, routingService, routingConfigService, activeCartService) {
        this.config = config;
        this.routingService = routingService;
        this.routingConfigService = routingConfigService;
        this.activeCartService = activeCartService;
    }
    ngOnInit() {
        this.steps = this.config.checkout.steps;
        this.cart$ = this.activeCartService.getActive();
        this.routerState$ = this.routingService.getRouterState().pipe(tap((router) => {
            this.activeStepUrl = router.state.context.id;
            this.steps.forEach((step, index) => {
                const routeUrl = `/${this.routingConfigService.getRouteConfig(step.routeName).paths[0]}`;
                if (routeUrl === this.activeStepUrl) {
                    this.activeStepIndex = index;
                }
            });
        }));
    }
};
CheckoutProgressMobileTopComponent.ctorParameters = () => [
    { type: CheckoutConfig },
    { type: RoutingService },
    { type: RoutingConfigService },
    { type: ActiveCartService }
];
CheckoutProgressMobileTopComponent = __decorate([
    Component({
        selector: 'cx-checkout-progress-mobile-top',
        template: "<div *ngIf=\"routerState$ | async as routerState\">\n  <div *ngIf=\"cart$ | async as cart\">\n    <div class=\"cx-media\">\n      <div class=\"cx-list-media\" *ngIf=\"cart?.totalItems && cart?.subTotal\">\n        {{ 'cartItems.cartTotal' | cxTranslate: { count: cart.totalItems } }}:\n        {{ cart.subTotal.formattedValue }}\n      </div>\n      <div *ngFor=\"let step of steps; let i = index\">\n        <div class=\"cx-list-media\" *ngIf=\"i < activeStepIndex\">\n          <div>{{ i + 1 }}. {{ step.name | cxTranslate }}</div>\n          <button\n            class=\"btn btn-link\"\n            [routerLink]=\"{ cxRoute: step.routeName } | cxUrl\"\n          >\n            {{ 'common.edit' | cxTranslate }}\n          </button>\n        </div>\n        <div class=\"cx-list-media is-active\" *ngIf=\"i === activeStepIndex\">\n          <div>{{ i + 1 }}. {{ step.name | cxTranslate }}</div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n"
    })
], CheckoutProgressMobileTopComponent);
export { CheckoutProgressMobileTopComponent };
