// @spartacus/storefront/esm2015/cms-components/navigation/footer-navigation/footer-navigation.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { map } from 'rxjs/operators';
import { CmsComponentData } from '../../../cms-structure/page/model/cms-component-data';
import { NavigationService } from '../navigation/navigation.service';
let FooterNavigationComponent = class FooterNavigationComponent {
    constructor(componentData, service) {
        this.componentData = componentData;
        this.service = service;
        this.node$ = this.service.getNavigationNode(this.componentData.data$);
        this.styleClass$ = this.componentData.data$.pipe(map((d) => d === null || d === void 0 ? void 0 : d.styleClass));
    }
};
FooterNavigationComponent.ctorParameters = () => [
    { type: CmsComponentData },
    { type: NavigationService }
];
FooterNavigationComponent = __decorate([
    Component({
        selector: 'cx-footer-navigation',
        template: "<cx-navigation-ui\r\n  [node]=\"node$ | async\"\r\n  [flyout]=\"false\"\r\n  [ngClass]=\"styleClass$ | async\"\r\n></cx-navigation-ui>\r\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], FooterNavigationComponent);
export { FooterNavigationComponent };
