// @spartacus/storefront/esm5/cms-components/user/checkout-login/checkout-login.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { CheckoutLoginComponent } from './checkout-login.component';
import { NotCheckoutAuthGuard } from '../../checkout/guards/not-checkout-auth.guard';
import { FormErrorsModule } from '../../../shared/index';
var CheckoutLoginModule = /** @class */ (function () {
    function CheckoutLoginModule() {
    }
    CheckoutLoginModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                I18nModule,
                FormsModule,
                ReactiveFormsModule,
                FormsModule,
                ReactiveFormsModule,
                FormErrorsModule,
            ],
            providers: [
                provideDefaultConfig({
                    cmsComponents: {
                        GuestCheckoutLoginComponent: {
                            component: CheckoutLoginComponent,
                            guards: [NotCheckoutAuthGuard],
                        },
                    },
                }),
            ],
            declarations: [CheckoutLoginComponent],
            exports: [CheckoutLoginComponent],
            entryComponents: [CheckoutLoginComponent],
        })
    ], CheckoutLoginModule);
    return CheckoutLoginModule;
}());
export { CheckoutLoginModule };
