// @spartacus/core/esm5/src/checkout/facade/checkout-payment.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { AuthService } from '../../auth/facade/auth.service';
import { ActiveCartService } from '../../cart/facade/active-cart.service';
import { OCC_USER_ID_ANONYMOUS } from '../../occ/utils/occ-constants';
import { getProcessStateFactory } from '../../process/store/selectors/process-group.selectors';
import { CheckoutActions } from '../store/actions/index';
import { SET_PAYMENT_DETAILS_PROCESS_ID, } from '../store/checkout-state';
import { CheckoutSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../auth/facade/auth.service";
import * as i3 from "../../cart/facade/active-cart.service";
var CheckoutPaymentService = /** @class */ (function () {
    function CheckoutPaymentService(checkoutStore, authService, activeCartService) {
        this.checkoutStore = checkoutStore;
        this.authService = authService;
        this.activeCartService = activeCartService;
    }
    /**
     * Get card types
     */
    CheckoutPaymentService.prototype.getCardTypes = function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getAllCardTypes));
    };
    /**
     * Get payment details
     */
    CheckoutPaymentService.prototype.getPaymentDetails = function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getPaymentDetails));
    };
    /**
     * Get status about set Payment Details process
     */
    CheckoutPaymentService.prototype.getSetPaymentDetailsResultProcess = function () {
        return this.checkoutStore.pipe(select(getProcessStateFactory(SET_PAYMENT_DETAILS_PROCESS_ID)));
    };
    /**
     * Clear info about process of setting Payment Details
     */
    CheckoutPaymentService.prototype.resetSetPaymentDetailsProcess = function () {
        this.checkoutStore.dispatch(new CheckoutActions.ResetSetPaymentDetailsProcess());
    };
    /**
     * Load the supported card types
     */
    CheckoutPaymentService.prototype.loadSupportedCardTypes = function () {
        this.checkoutStore.dispatch(new CheckoutActions.LoadCardTypes());
    };
    /**
     * Create payment details using the given paymentDetails param
     * @param paymentDetails: the PaymentDetails to be created
     */
    CheckoutPaymentService.prototype.createPaymentDetails = function (paymentDetails) {
        if (this.actionAllowed()) {
            var userId_1;
            this.authService
                .getOccUserId()
                .subscribe(function (occUserId) { return (userId_1 = occUserId); })
                .unsubscribe();
            var cartId_1;
            this.activeCartService
                .getActiveCartId()
                .subscribe(function (activeCartId) { return (cartId_1 = activeCartId); })
                .unsubscribe();
            if (userId_1 && cartId_1) {
                this.checkoutStore.dispatch(new CheckoutActions.CreatePaymentDetails({
                    userId: userId_1,
                    cartId: cartId_1,
                    paymentDetails: paymentDetails,
                }));
            }
        }
    };
    /**
     * Set payment details
     * @param paymentDetails : the PaymentDetails to be set
     */
    CheckoutPaymentService.prototype.setPaymentDetails = function (paymentDetails) {
        if (this.actionAllowed()) {
            var userId_2;
            this.authService
                .getOccUserId()
                .subscribe(function (occUserId) { return (userId_2 = occUserId); })
                .unsubscribe();
            var cart_1;
            this.activeCartService
                .getActive()
                .subscribe(function (activeCart) { return (cart_1 = activeCart); })
                .unsubscribe();
            if (userId_2 && cart_1) {
                this.checkoutStore.dispatch(new CheckoutActions.SetPaymentDetails({
                    userId: userId_2,
                    cartId: cart_1.code,
                    paymentDetails: paymentDetails,
                }));
            }
        }
    };
    /**
     * Sets payment loading to true without having the flicker issue (GH-3102)
     */
    CheckoutPaymentService.prototype.paymentProcessSuccess = function () {
        this.checkoutStore.dispatch(new CheckoutActions.PaymentProcessSuccess());
    };
    CheckoutPaymentService.prototype.actionAllowed = function () {
        var userId;
        this.authService
            .getOccUserId()
            .subscribe(function (occUserId) { return (userId = occUserId); })
            .unsubscribe();
        return ((userId && userId !== OCC_USER_ID_ANONYMOUS) ||
            this.activeCartService.isGuestCart());
    };
    CheckoutPaymentService.ctorParameters = function () { return [
        { type: Store },
        { type: AuthService },
        { type: ActiveCartService }
    ]; };
    CheckoutPaymentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CheckoutPaymentService_Factory() { return new CheckoutPaymentService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.AuthService), i0.ɵɵinject(i3.ActiveCartService)); }, token: CheckoutPaymentService, providedIn: "root" });
    CheckoutPaymentService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], CheckoutPaymentService);
    return CheckoutPaymentService;
}());
export { CheckoutPaymentService };
